---
title: "Tp53 SNV hotspots"
author: "K S Gaonkar (D3B)"
output: html_notebook

---

In this PR we will add TP53 alteration status as discussed in [#837](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/837):

**TP53 altered - loss**, if :

 - A sample contains a TP53 hotspot SNV mutation. (Cancer hotspot database and downloadable file available). Please also crosscheck that all mutations from this table are included.
 - A sample contains two TP53 alterations, suggesting (but not confirming) that both alleles are affected (SNV+SNV, CNV+SNV).
 - A sample contains one alteration (SNV or CNV) + has cancer_predispositions == "Li-Fraumeni syndrome", suggesting there is a germline variant in addition to the somatic variant we observe.
 - A sample does not have a TP53 alterations, but has cancer_predispositions == "Li-Fraumeni syndrome" and TP53 classifier score for matched RNA-Seq > 0.5 (or higher cutoff we decide upon later).
 
**TP53 altered - activated**, if:

- A sample contains one of the two TP53 activating mutations R273C and R248W. Reference and reference.





## Setup
```{r}
library("ggpubr")
library("ggthemes")
library("tidyverse")

# rootdir
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")

input_dir <- file.path(root_dir,
                         "analyses",
                         "tp53_nf1_score",
                         "input")
# histology
 histology <- read_tsv(file.path(data_dir,"pbta-histologies.tsv")) %>%
   select("Kids_First_Biospecimen_ID",
          "sample_id",
          "cancer_predispositions",
          "sample_type",
          "composition")

# Cancer database
hotspot_database_2017_tp53_snv <- readxl::read_xls(file.path(input_dir,"hotspots_v2.xls"),sheet = 1) %>%
  filter(Hugo_Symbol == "TP53")
hotspot_database_2017_tp53_indel <- readxl::read_xls(file.path(input_dir,"hotspots_v2.xls"),sheet = 2) %>%
  filter(Hugo_Symbol == "TP53")

# p53, p63 and p73 functional sites
functional_sites_tp53 <- read_tsv(file.path(input_dir,"hotspot_Chen_2006.tsv")) 

results_dir <- file.path(root_dir,
                         "analyses",
                         "tp53_nf1_score",
                         "results")

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# classifier scores
classifier_score_stranded <- read_tsv(
  file.path(
    results_dir,
    "pbta-gene-expression-rsem-fpkm-collapsed.stranded_classifier_scores.tsv"))%>%
  select(sample_id,tp53_score)
classifier_score_polya <-read_tsv(
  file.path(
    results_dir,
    "pbta-gene-expression-rsem-fpkm-collapsed.polya_classifier_scores.tsv")) %>%
  select(sample_id,tp53_score)

classifier_score<- rbind(classifier_score_polya,classifier_score_stranded) %>%
  rename(biospecimen_id=sample_id) %>%
  left_join(histology[,c("Kids_First_Biospecimen_ID","sample_id")] ,by=c("biospecimen_id"="Kids_First_Biospecimen_ID")) 

# Copy number per sample 
# overlapping functional domain
cnv_domain_overlap <- read_tsv(
  file.path(
    results_dir,
    "loss_overlap_domains_tp53.tsv"))


```

### Check if all function sites are in hotspots

All functional sites are just 1 base so checking in `hotspot_database_2017_snv`

```{r}

functional_sites %>%
  filter(!gsub("[A-Z|a-z]","",p53) %in% hotspot_database_2017_tp53_snv$Amino_Acid_Position )

```
2 functional sites are missing in hotspots database.

## SNV for TP53

Removing Silent or Intron classified variants to capture only putative damaging mutations

```{r}
consensus_tp53_snv_indel <- data.table::fread(
  file.path(data_dir,"pbta-snv-consensus-mutation.maf.tsv.gz"),
                                   select = c("Chromosome",
                                              "Start_Position",
                                              "End_Position",
                                              "Strand",
                                              "Variant_Classification",
                                              "Tumor_Sample_Barcode",
                                              "Hugo_Symbol",
                                              "HGVSp_Short"),
                                   data.table = FALSE) %>%
  filter(Hugo_Symbol == "TP53") %>%
  filter(!(Variant_Classification %in% c("Silent", "Intron")))
```

###  Gather annotation for SNV 

hotspots : overlaps AA position which are statistically significant SNV/Indels
activating : overlaps AA position which are found to act as gain-of-function according to literature

```{r}

consensus_tp53_snv_indel <- consensus_tp53_snv_indel %>%
  mutate(
    hotspot = case_when(
      # strip REF and Variant AA to get Amino_Acid_Position in consensus maf
      (gsub("[A-Z|a-z]|[.]","",HGVSp_Short) %in% 
         # if overlaps the hotspot Amino_Acid_Position 
         hotspot_database_2017_tp53_snv$Amino_Acid_Position)  ~ "Yes"),
    activating = case_when(
      # strip REF and Variant AA to get Amino_Acid_Position in consensus maf
      (gsub("[A-Z|a-z]|[.]","",HGVSp_Short) %in% 
         # if overlaps the activating  Amino_Acid_Position 
         c("273","248"))  ~ "Yes"
    )
  )

```

## Gather SNV,CNV, classifier and cancer_predisposition values

We will gather values per sample_id since DNA and RNA samples can be only matched by `sample_id`


```{r}

tp53_alterations_score <- histology %>%
  # filter for Tumors
  filter(sample_type=="Tumor") %>%
  # join consensus calls overlapping hotspots
  left_join(consensus_tp53_snv_indel, 
            by=c("Kids_First_Biospecimen_ID"="Tumor_Sample_Barcode")) %>%
  # join filtered cnv losses
  left_join(cnv_domain_overlap,by=c("Kids_First_Biospecimen_ID"="biospecimen_id",
                                    "sample_id"))  %>%
  # add classifier score
  left_join(classifier_score,by=c("sample_id")) %>%
  # select useful columns
  select(sample_id,tp53_score,cancer_predispositions,HGVSp_Short,copy_number,hotspot,activating) %>%
  arrange(sample_id) %>%
  unique() 

tp53_alterations_score 

```

Convert the above to wide format

```{r}

tp53_alterations_score_wide <- tp53_alterations_score %>%
  # group 
  group_by(sample_id,tp53_score,cancer_predispositions) %>%
  # 
  # sample_id as rows add SNV counts,CNV loss counts 
  # and hotspot/activating mutation annotation  
  summarise(
    # summarize length of SNV and CNv alterations 
    SNV_indel_counts= length(unique(HGVSp_Short[!is.na(HGVSp_Short)])), 
    CNV_loss_counts = length(unique(copy_number[!is.na(copy_number)])),
    # summarize unique hotspot values per sample_id
    hotspot = toString(unique(hotspot[!is.na(hotspot) ])),
    activating = toString(unique(activating[!is.na(activating)]))) 

tp53_alterations_score_wide

```

### Add annotation

As discussed above we want to annotate TP53 mutants that are putative loss-of-function OR  gain-of-function.

```{r}

tp53_alterations_score_wide <- tp53_alterations_score_wide %>%
  mutate(
    # add tp53_altered annotation column
    tp53_altered = 
      case_when(
        # when activating == ""
        activating == "" &
          # check if mutated variant AA position overlaps hotspot database 
          ( hotspot == "Yes"  |
              # check if sample_id has SNV+CNV mutation
              # suggesting both alleles are mutated
              (SNV_indel_counts >= 1 & CNV_loss_counts >=1) |
              # check if more than 1 SNV is present
              # suggesting both alleles are mutated
              SNV_indel_counts >= 1 |
              # check if SNV mutant and has
              # Li-Fraumeni syndrome suggesting
              # germline TP53 mutant as cancer predisposition 
              (SNV_indel_counts >= 1 & 
                             grepl("Li-Fraumeni syndrome",cancer_predispositions)) |
              # check if CNV mutant  and has
              # Li-Fraumeni syndrome suggesting
              # germline TP53 mutant as cancer predisposition 
              (CNV_loss_counts >= 1 & 
                             grepl("Li-Fraumeni syndrome",cancer_predispositions)) |
              # check if tp53 inactivating score for RNA_Seq is greater that 0.5
              # and has Li-Fraumeni syndrome suggesting
              # germline TP53 mutant as cancer predisposition 
              (tp53_score > 0.5 & 
                             grepl("Li-Fraumeni syndrome",cancer_predispositions))
          ) ~ "loss",
        # when activating == "Yes"
        activating == "Yes" ~ "activated",
        # if no evidence supports TP53 inativation or activation 
        TRUE ~ "No evidence"
      )
  )

```

## Explore distribution of tp53_altered status vs tp53 inactivation scores

```{r}
ggplot(tp53_alterations_score_wide, aes(x = factor(tp53_altered), y = tp53_score)) +
  geom_violin()+
  geom_jitter(alpha = 0.5, width = 0.2) +
  stat_compare_means() +
  theme_bw() +
  ggtitle("Distribution of scores across tp53 altered status") +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  xlab("tp53 altered status") 

```
TP53 mutants with gain-of-function (activating) mutants promote tumorigenesis according to literature. [Reference](https://pubmed.ncbi.nlm.nih.gov/17417627/) and [reference](https://pubmed.ncbi.nlm.nih.gov/24677579/) have similar distribution of classifier scores to that of potential loss-of-function (multi-allelic) Tp53 mutations. 

"No evidence" annotated samples either don't have any loss/gain TP53 SNVs nor CNV losses OR DNA sample is not available.




### Check if other cancer predisposition have high TP53 inactivation scores  

```{r}

ggplot(tp53_alterations_score_wide, aes(x = factor(tp53_altered), y = tp53_score)) +
  geom_violin()+
  geom_jitter(alpha = 0.5, width = 0.2) +
  theme_bw() +
  ggtitle("Distribution of scores across tp53 altered status") +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  xlab("tp53 altered status") +
  facet_wrap(.~cancer_predispositions)

```
 Some NF-1 and/or Other inherited conditions NOS have high scoring TP53 mutants as well.

## Save file 

Adding DNA and RNA biospecimen ids to `tp53_alterations_score_wide` and saving

```{r}

tp53_alterations_score_wide %>% 
  left_join(histology[,c("Kids_First_Biospecimen_ID","sample_id")]) %>%
  write_tsv(file.path(results_dir,"tp53_altered_status.tsv"))

```


