---
title: "Find CNV losses that overlap with TP53 domains"
author: "K S Gaonkar (D3B)"
output: html_notebook
params:
  base_run:
    label: "1/0 to run with base histology"
    value: 0
    input: integer
---
    
In this script we will find if there are a Structural Variant breakpoints 

1)  within TP53 or covering the gene locus such that it is deleted and concurrently, there are either one or both of:
  - 0 or 1 copy of TP53 (we would already have captured them as loss)
  - expression less than 1 FPKM

2) within TP53 resulting in a fusion

We want to subset SV calls which will possibly	lead to loss of function to use for evaluation of TP53 inactivation score at a later step.

### Setup
```{r}
suppressPackageStartupMessages({
  library(ggpubr)
  library(ggthemes)
  library(tidyverse)
})

# rootdir
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")
results_dir <- file.path(
  root_dir,
  "analyses",
  "tp53_nf1_score",
  "results"
)

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

### Input files
```{r}

manta_sv <- read_tsv(file.path(data_dir, "pbta-sv-manta.tsv.gz"))
putative_onco_fusion <- read_tsv(file.path(data_dir, "pbta-fusion-putative-oncogenic.tsv"))
stranded_expr <- readRDS(file.path(data_dir, "pbta-gene-expression-rsem-fpkm-collapsed.stranded.rds"))
polya_expr <- readRDS(file.path(data_dir, "pbta-gene-expression-rsem-fpkm-collapsed.polya.rds"))
if ( params$base_run ==0 ){
  clinical<-read.delim(file.path(data_dir,"pbta-histologies.tsv"), stringsAsFactors = FALSE)
} else{
  clinical<-read.delim(file.path(data_dir,"pbta-histologies-base.tsv"), stringsAsFactors = FALSE)  
}

histology <- dplyr::select(clinical,c("Kids_First_Biospecimen_ID", "sample_id"))

# high confidence cnv losses from `03-tp53-cnv-loss-domain.Rmd`
cnv_tp53_loss <- read_tsv(file.path(results_dir, "loss_overlap_domains_tp53.tsv"))

# generate txdb oblect from gtf file to gather exon and intron locations
annotation_directory <- file.path(root_dir,
                                  "analyses",
                                  "focal-cn-file-preparation",
                                  "annotation_files")
annotation_file <- file.path(annotation_directory,
                             "txdb_from_gencode.v27.gtf.db")

if (!file.exists(annotation_file)) {
  # Define the annotations for the hg38 genome
  txdb <- GenomicFeatures::makeTxDbFromGFF(
    file = file.path(data_dir,"gencode.v27.primary_assembly.annotation.gtf.gz"),
    format = "gtf"
  )
  # can do this even if the directory exists
  dir.create(annotation_directory, showWarnings = FALSE)
  # write this to file to save time next time
  AnnotationDbi::saveDb(txdb, annotation_file)
} else {
  txdb <- AnnotationDbi::loadDb(annotation_file)
}

# exon regions
exon <- GenomicFeatures::exonicParts(txdb, linked.to.single.gene.only=FALSE)
# intron regions
intron <- GenomicFeatures::intronicParts(txdb, linked.to.single.gene.only=FALSE)

```

## SV breakpoints associated with TP53 loss 
Since we want to capture TP53 loss phenotypes we will only keep breakpoints that are either DEL: deletion, INV: inversion and BND: translocations which are breakpoint notations with with 2 breakpoints in different chromosomes
 
```{r}
manta_sv_tp53 <- manta_sv %>%
  dplyr::select(SV.chrom, SV.start, SV.end, SV.length, Kids.First.Biospecimen.ID.Tumor, SV.type, Gene.name, ALT) %>%
  # Distribution of TP53 SVs
  filter(
    # only overlaps TP53 gene
    grepl("^TP53$", Gene.name) |
      # part of larger region also encompasing TP53
      grepl("/TP53/", Gene.name),
    SV.type %in% c("BND", "INV", "DEL")
  ) %>%
  left_join(histology,
    by = c("Kids.First.Biospecimen.ID.Tumor" = "Kids_First_Biospecimen_ID")
  ) 

manta_sv_tp53 %>%
  arrange(Kids.First.Biospecimen.ID.Tumor)
```

## Fusion breakpoints within TP53 
We will also add exon and intron information to this fusion call.

```{r}
putative_onco_tp53_fusion <- putative_onco_fusion %>%
  filter(str_detect(FusionName,"^TP53--|--TP53$")) %>%
  left_join(histology,
            by = c("Sample" = "Kids_First_Biospecimen_ID")
  ) %>%
  dplyr::select("Sample",
                "sample_id",
                "FusionName",
                "RightBreakpoint",
                "LeftBreakpoint") %>%
  reshape2::melt(c("Sample","sample_id","FusionName"))

```

### Generate Granges and overlaps
```{r}

if(nrow(putative_onco_tp53_fusion) > 0) {
  putative_onco_tp53_fusion_gr <- putative_onco_tp53_fusion %>%
    mutate(
      BreakpointChr = paste0("chr",gsub(":.*$", "", value)),
      Breakpoint = gsub("^.*:", "", value)) %>%
    GenomicRanges::makeGRangesFromDataFrame(seqnames.field = "BreakpointChr",
                                            start.field = "Breakpoint",
                                            end.field = "Breakpoint",
                                            
                                            keep.extra.columns = TRUE,
                                            starts.in.df.are.0based = FALSE)
  
  # Check for exon overlap
  bk_overlaps_exon <- IRanges::mergeByOverlaps(putative_onco_tp53_fusion_gr,exon) %>%
    as.data.frame() %>%
    mutate(exon_rank = unlist(exon_rank)) %>%
    dplyr::filter(grepl("ENSG00000141510",gene_id,
                        exon_rank==1)) %>%
    dplyr::select(starts_with("putative_onco")) %>%
    dplyr::rename_at(vars(starts_with("putative_onco")),~str_remove(.,"putative_onco_tp53_fusion_gr."))
  
  # Check for intron overlap
  bk_overlaps_intron <- IRanges::mergeByOverlaps(putative_onco_tp53_fusion_gr,intron) %>%
    as.data.frame() %>%
    dplyr::filter(grepl("ENSG00000141510",gene_id)) %>%
    dplyr::select(starts_with("putative_onco")) %>%
    dplyr::rename_at(vars(starts_with("putative_onco")),~str_remove(.,"putative_onco_tp53_fusion_gr.")) 
  
  # Save
  if (nrow(bk_overlaps_intron) == 0) {
    # only saving fusion calls overlapping exon 1
    bk_overlaps_exon %>% 
      write_tsv(file.path(results_dir, "fusion_bk_tp53_loss.tsv"))
  }else if (nrow(bk_overlaps_exon) == 0) {
    # only saving fusion calls overlapping intron 1
    bk_overlaps_intron %>% 
      write_tsv(file.path(results_dir, "fusion_bk_tp53_loss.tsv"))
  } else {
    # bind calls overlapping exon1 or intron1
    bind_rows(bk_overlaps_exon,
              bk_overlaps_intron) %>%
      write_tsv(file.path(results_dir, "fusion_bk_tp53_loss.tsv"))
  }
  
}
```


## Filter SV calls that are associated with sample with TP53 CNV loss
High confidence cnv losses from `03-tp53-cnv-loss-domain.Rmd` overlap TP53 domains:

 - TAD = trans-activating domain (essential for function)
 - DBD = DNA-binding domain (residues 102–292)
 - TD = tetramerization domain (residues 326–356) 
 
```{r}
sv_with_loss_cnv <- manta_sv_tp53 %>%
  # filter to samples with SV within or overlapping TP53
  filter(sample_id %in% cnv_tp53_loss$sample_id)

sv_with_loss_cnv
```
BS_3NX3RBCX and BS_EJV0N3BX have a CNV loss and multiple SV breakpoints overlapping TP53 locus!


## Filter SV calls that are associated with sample with low TP53 expression
We will use sample_id from `manta_sv_tp53` to filter stranded and polya expression file 

```{r}
stranded_expr_tp53 <- stranded_expr["TP53", ] %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Kids_First_Biospecimen_ID")

polya_expr_tp53 <- polya_expr["TP53", ] %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Kids_First_Biospecimen_ID")

lowexpr_tp53_with_sv <- bind_rows(stranded_expr_tp53, polya_expr_tp53) %>%
  # add sample_id to expression df
  left_join(histology,
    by = c("Kids_First_Biospecimen_ID")
  ) %>%
  # filter to keep only low expressed TP53 sample_ids
  filter(TP53 < 1)

sv_with_lowexpr <- manta_sv_tp53 %>%
  # filter to samples with SV within or overlapping TP53
  filter(sample_id %in% lowexpr_tp53_with_sv$sample_id)

sv_with_lowexpr
```
No sample_ids with a SV within or overlapping TP53 have low expression of TP53!

## Save high confidence SV associated with TP53 loss
```{r}
if (nrow(sv_with_lowexpr) == 0) {
  # only saving sv_with_loss_cnv here
  # because sv_with_lowexpr is empty
  sv_with_loss_cnv %>% 
    write_tsv(file.path(results_dir, "sv_overlap_tp53.tsv"))
} else if (nrow(sv_with_loss_cnv) == 0) {
  # only saving sv_with_lowexpr here
  # because sv_with_loss_cnv is empty
  sv_with_lowexpr %>% 
    write_tsv(file.path(results_dir, "sv_overlap_tp53.tsv"))
} else {
  # bind rows of both sv_with_lowexpr
  # and sv_with_loss_cnv and save
  bind_rows(
    sv_with_loss_cnv,
    sv_with_lowexpr
  ) %>%
    write_tsv(file.path(results_dir, "sv_overlap_tp53.tsv"))
}
```

