---
title: "Tp53 SNV hotspots"
author: "K S Gaonkar (D3B)"
output: html_notebook
params:
  base_run:
    label: "1/0 to run with base histology"
    value: 0
    input: integer
---

In this notebook we will add TP53 alteration status as discussed in [#837](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/837):

**TP53 altered - loss**, if :

 - A sample contains a TP53 hotspot SNV mutation. (Cancer hotspot database and downloadable file available). Please also crosscheck that all mutations from this table are included.
 - A sample contains two TP53 alterations, suggesting (but not confirming) that both alleles are affected (SNV+SNV, (CNV or SV) + SNV).
 - A sample contains one alteration (SNV or (CNV or SV)) + has cancer_predispositions == "Li-Fraumeni syndrome", suggesting there is a germline variant in addition to the somatic variant we observe.
 - A sample does not have a TP53 alterations, but has cancer_predispositions == "Li-Fraumeni syndrome" and TP53 classifier score for matched RNA-Seq > 0.5 (or higher cutoff we decide upon later).
 
 Note: CNV and SV will be considered as the same event, but we will be adding SV_counts and SV.type to the altered status output file
 
**TP53 altered - activated**, if:

- A sample contains one of the two TP53 activating mutations R273C and R248W. Reference and reference.





## Setup
```{r}
library("ggpubr")
library("ggthemes")
library("tidyverse")
library("ggfortify")
library("broom")

# rootdir
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")

input_dir <- file.path(root_dir,
                         "analyses",
                         "tp53_nf1_score",
                         "input")

plot_dir <- file.path(root_dir,
                         "analyses",
                         "tp53_nf1_score",
                         "plots")

if (!dir.exists(plot_dir)) {
  dir.create(plot_dir)
}

# cell composition
cell_line_composition <- read_tsv(file.path("..",
                                            "molecular-subtyping-HGG",
                                            "input",
                                            "cell-line-composition.tsv"),
                                  col_types = 
                                    readr::cols( aliquot_id  = readr::col_character())) 

# histology
if ( params$base_run ==0 ){
  clinical<-read.delim(file.path(data_dir,"pbta-histologies.tsv"), stringsAsFactors = FALSE)
} else{
  clinical<-read.delim(file.path(data_dir,"pbta-histologies-base.tsv"), stringsAsFactors = FALSE)  
}

histology <- clinical %>%
   dplyr::select("Kids_First_Biospecimen_ID",
          "sample_id",
          "cancer_predispositions",
          "sample_type",
          "experimental_strategy",
          "composition") %>%
   # merge cell line composition
  left_join(cell_line_composition)
 

# Cancer database
hotspot_database_2017_tp53_snv <- readxl::read_xls(file.path(input_dir,"hotspots_v2.xls"),sheet = 1) %>%
  filter(Hugo_Symbol == "TP53")
hotspot_database_2017_tp53_indel <- readxl::read_xls(file.path(input_dir,"hotspots_v2.xls"),sheet = 2) %>%
  filter(Hugo_Symbol == "TP53")

# p53, p63 and p73 functional sites
functional_sites_tp53 <- read_tsv(file.path(input_dir,"hotspot_Chen_2006.tsv")) 

results_dir <- file.path(root_dir,
                         "analyses",
                         "tp53_nf1_score",
                         "results")

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# classifier scores
classifier_score_stranded <- read_tsv(
  file.path(
    results_dir,
    "pbta-gene-expression-rsem-fpkm-collapsed.stranded_classifier_scores.tsv"))%>%
  dplyr::select(sample_id,tp53_score)

classifier_score_polya <-read_tsv(
  file.path(
    results_dir,
    "pbta-gene-expression-rsem-fpkm-collapsed.polya_classifier_scores.tsv")) %>%
  dplyr::select(sample_id,tp53_score)

classifier_score<- rbind(classifier_score_polya,classifier_score_stranded) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_RNA=sample_id) %>%
  left_join(histology ,by=c("Kids_First_Biospecimen_ID_RNA"="Kids_First_Biospecimen_ID")) 

# Copy number per sample 
# overlapping functional domain
cnv_domain_overlap <- read_tsv(
  file.path(
    results_dir,
    "loss_overlap_domains_tp53.tsv"))

# Structural variant overlapping 
# or within gene locus of TP53
sv_overlap <- read_tsv(
  file.path(
    results_dir,
    "sv_overlap_tp53.tsv"))

# Structural variant as a fusion in exon1/intron1
fusion_overlap <- read_tsv(
  file.path(
    results_dir,
    "fusion_bk_tp53_loss.tsv"))

```

### Check if all function sites are in hotspots

All functional sites are just 1 base so checking in `hotspot_database_2017_snv`

```{r}

functional_sites_tp53 %>%
  filter(!gsub("[A-Z|a-z]","",p53) %in% hotspot_database_2017_tp53_snv$Amino_Acid_Position )

```
2 functional sites are missing in hotspots database.

## SNV for TP53
 - We will use the `pbta-snv-consensus-mutation.maf.tsv.gz` from `snv-callers` module which gathers calls that are present in all 3 callers (strelka2,mutect2 and lancet)
 - In addition, we will also use `pbta-snv-scavenged-hotspots.maf.tsv.gz` from `hotspot-detection` module to gather calls that overlap MSKCC hotspots found in any caller (except if only vardict calls the site as variant, we remove these calls since we have a lot of calls unique to vardict which we consider as false positive as discussed [here](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/snv-callers#snv-caller-comparison-analysis))
 - We will also removing `Silent` or `Intron` classified variants to capture only putative damaging mutations

```{r}


tp53_consensus_maf <- data.table::fread(
  file.path(root_dir, "data" , "pbta-snv-consensus-mutation.maf.tsv.gz"),
                                   select = c("Chromosome",
                                              "Start_Position",
                                              "End_Position",
                                              "Strand",
                                              "Variant_Classification",
                                              "Tumor_Sample_Barcode",
                                              "Hugo_Symbol",
                                              "HGVSp_Short"),
                                   data.table = FALSE) %>%
  filter(Hugo_Symbol == "TP53")

tp53_hotspot_maf <- data.table::fread(
  file.path(root_dir, "analyses" , "hotspots-detection" , "results" , "pbta-snv-scavenged-hotspots.maf.tsv.gz"),
                                   select = c("Chromosome",
                                              "Start_Position",
                                              "End_Position",
                                              "Strand",
                                              "Variant_Classification",
                                              "Tumor_Sample_Barcode",
                                              "Hugo_Symbol",
                                              "HGVSp_Short"),
                                   data.table = FALSE) %>%
  filter(Hugo_Symbol == "TP53") %>%
  dplyr::select(colnames(tp53_consensus_maf))



tp53_consensus_hotspot_maf <- tp53_consensus_maf %>%
  bind_rows(tp53_hotspot_maf) %>%
  filter(!(Variant_Classification %in% c("Silent", "Intron",
                                         # remove other non-amino acid SNVs
                                         "3'Flank" ,"5'Flank",
                                         "3'UTR", "5'UTR"  ))) %>%
  unique()
```

###  Gather annotation for SNV 

hotspots : overlaps AA position which are statistically significant SNV/Indels
activating : overlaps AA position which are found to act as gain-of-function according to literature

```{r}

tp53_consensus_hotspot_maf  <- tp53_consensus_hotspot_maf  %>%
  mutate(
    hotspot = case_when(
      # strip REF and Variant AA to get Amino_Acid_Position in consensus maf
      (gsub("[A-Z|a-z]|[.]","",HGVSp_Short) %in% 
         # if overlaps the hotspot Amino_Acid_Position 
         hotspot_database_2017_tp53_snv$Amino_Acid_Position)  ~ 1,
      TRUE ~ 0),
    activating = case_when(
      # strip REF and Variant AA to get Amino_Acid_Position in consensus maf
      (gsub("[A-Z|a-z]|[.]","",HGVSp_Short) %in% 
         # if overlaps the activating  Amino_Acid_Position 
         c("273","248"))  ~ 1,
      TRUE ~ 0
    )
  )

```

## Gather SNV,CNV, classifier and cancer_predisposition values

We will gather values per sample_id since DNA and RNA samples can be only matched by `sample_id`


```{r}

tp53_alterations <- histology %>%
  filter(experimental_strategy != "RNA-Seq") %>%
  # merge cell line composition
  left_join(cell_line_composition) %>%
  # filter for Tumors
  filter(sample_type=="Tumor") %>%
  # join consensus calls overlapping hotspots
  left_join(tp53_consensus_hotspot_maf , 
            by=c("Kids_First_Biospecimen_ID"="Tumor_Sample_Barcode")) %>%
  # join filtered cnv losses
  left_join(cnv_domain_overlap,by=c("Kids_First_Biospecimen_ID"="biospecimen_id",
                                    "sample_id")) %>%
  # join SV 
  left_join(sv_overlap,by=c("Kids_First_Biospecimen_ID"="Kids.First.Biospecimen.ID.Tumor",
                            "sample_id")) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_DNA = Kids_First_Biospecimen_ID) %>%
  # join fusion
  left_join(fusion_overlap,by="sample_id") 

tp53_alterations_score <- tp53_alterations %>%
  # add classifier score
  full_join(classifier_score,by=c("sample_id",
                                  "cell_line_composition",
                                  "cancer_predispositions",
                                  "sample_type",
                                  "Kids_First_Participant_ID")) %>%
  # select useful columns
  dplyr::select(sample_id,Kids_First_Biospecimen_ID_RNA,Kids_First_Biospecimen_ID_DNA,
         tp53_score,cancer_predispositions,HGVSp_Short,copy_number,SV.type,FusionName,hotspot,activating) %>%
  arrange(sample_id) %>%
  unique() %>%
  replace_na(list(hotspot = 0, activating = 0))

tp53_alterations_score 

```

Convert the above to wide format

```{r}

tp53_alterations_score_wide <- tp53_alterations_score %>%
  # group 
  group_by(sample_id,
           Kids_First_Biospecimen_ID_DNA,
           Kids_First_Biospecimen_ID_RNA,
           cancer_predispositions,
           tp53_score) %>%
  # 
  # sample_id as rows add SNV counts,CNV loss counts 
  # and hotspot/activating mutation annotation  
  summarise(
    # summarize length of SNV and CNV alterations 
    SNV_indel_counts= length(unique(HGVSp_Short[!is.na(HGVSp_Short)])), 
    CNV_loss_counts = length(unique(copy_number[!is.na(copy_number)])),
    SV_counts = length(unique(SV.type[!is.na(SV.type)])),
    Fusion_counts = length(unique(FusionName[!is.na(FusionName)])),
    HGVSp_Short = toString(unique(HGVSp_Short)),
    CNV_loss_evidence = toString(unique(copy_number)),
    SV_type = toString(unique(SV.type)),
    Fusion_evidence = toString(unique(FusionName)),
    # summarize unique hotspot values per sample_id
    hotspot = max(unique(hotspot[!is.na(hotspot) ])),
    activating = max(unique(activating[!is.na(activating)]))) 

tp53_alterations_score_wide

```

### Add annotation

As discussed above we want to annotate TP53 mutants that are putative loss-of-function OR  gain-of-function.

```{r}

tp53_alterations_score_wide <- tp53_alterations_score_wide %>%
  mutate(
    # add tp53_altered annotation column
    tp53_altered = 
      case_when(
        # when activating == 0
        activating == 0 &
          # check if mutated variant AA position overlaps hotspot database 
          ( hotspot == 1  |
              # check if sample_id has SNV+(CNV|SV|Fusion) mutation
              # suggesting both alleles are mutated
              (SNV_indel_counts >= 1 & 
                 (CNV_loss_counts >=1 | SV_counts >=1 | Fusion_counts >=1 ) ) |
              # check if more than 1 SNV is present
              # suggesting both alleles are mutated
              SNV_indel_counts > 1 |
              # check if SNV mutant and has
              # Li-Fraumeni syndrome suggesting
              # germline TP53 mutant as cancer predisposition 
              (SNV_indel_counts >= 1 & 
                             grepl("Li-Fraumeni syndrome",cancer_predispositions)) |
              # check if CNV|SV mutant|Fusion  and has
              # Li-Fraumeni syndrome suggesting
              # germline TP53 mutant as cancer predisposition 
              ((CNV_loss_counts >= 1 | SV_counts >=1 | Fusion_counts >=1) & 
                             grepl("Li-Fraumeni syndrome",cancer_predispositions)) |
              # check if tp53 inactivating score for RNA_Seq is greater that 0.5
              # and has Li-Fraumeni syndrome suggesting
              # germline TP53 mutant as cancer predisposition 
              (tp53_score > 0.5 & 
                             grepl("Li-Fraumeni syndrome",cancer_predispositions)) |
              # check if tp53 inactivating score for RNA_Seq is greater that 0.5
              # and has 1 or more SNV| (CNV|SV|Fusion) loss in TP53
              (tp53_score > 0.5 & 
                             (SNV_indel_counts >= 1 | (CNV_loss_counts >=1 | SV_counts >=1 |Fusion_counts >=1)))
          ) ~ "loss",
        # when activating == 1
        activating == 1 ~ "activated",
        # if no evidence supports TP53 inativation or activation 
        TRUE ~ "other"
      )
  )

```

## Explore distribution of tp53_altered status vs tp53 inactivation scores

```{r, out.width="50%"}
# Default global test is kruskal-wallis for >2 groups
# Default is to perform multiple corrections, but not bonferroni - specifying bonferroni below
# Default p values on plots are uncorrected

# More on the functions: http://www.sthda.com/english/articles/24-ggpubr-publication-ready-plots/76-add-p-values-and-significance-levels-to-ggplots/#compare-more-than-two-groups

# From R help: 
# p.adjust.method: method for adjusting p values (see p.adjust). Has impact only in a situation, where multiple pairwise tests are performed; or when there are multiple grouping variables. Allowed values include "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none". If you don't want to adjust the p value (not recommended), use p.adjust.method = "none". Note that, when the formula contains multiple variables, the p-value adjustment is done independently for each variable.

# method: the type of test. Default for two samples is wilcox.test and for >2 is kruskal.test. 
# Allowed values for 2 samples include: "t.test" (parametric), "wilcox.test" (non-parametric) and >2: "anova" (parametric) and "kruskal.test" (non-parametric).

# Create table of statistics, including corrected p values
adj <- compare_means(tp53_score ~ tp53_altered, data = tp53_alterations_score_wide, p.adjust.method = "bonferroni")
adj

# Print table of n per group for legends
tp53_alterations_score_wide %>%
  filter(!is.na(tp53_score)) %>%
  group_by(tp53_altered) %>%
  tally()
```

```{r}
# Create plot
ggviolin(tp53_alterations_score_wide, x = "tp53_altered", y = "tp53_score", 
         color = "tp53_altered", 
         palette = "jco", 
         order = c("activated", "loss", "other"),
         add = "boxplot",  
   ggtheme = theme_pubr()) +
  # Default method = "kruskal.test" for multiple groups
  # Add adjusted p-value - has to be done manually and from a table - pulling from table above 
  stat_pvalue_manual(adj, label = "p.adj", y.position = c(1.5, 1.3, 1.1)) +
  # Add global p-value
  stat_compare_means(label.y = 1.7) +
  xlab("TP53 altered status") +
  ylab("TP53 score") +
  rremove("legend")

ggsave(file.path(plot_dir, "tp53_scores_by_altered_status.png"))

```
TP53 mutants with gain-of-function (activating) mutants promote tumorigenesis according to literature. [Reference](https://pubmed.ncbi.nlm.nih.gov/17417627/) and [reference](https://pubmed.ncbi.nlm.nih.gov/24677579/) have similar distribution of classifier scores to that of potential loss-of-function (multi-allelic) Tp53 mutations. 

"Other" annotated samples either don't have any high-confidence loss/gain TP53 SNVs nor CNV losses OR DNA sample is not available.

### (stranded) Expression profile for activating vs loss TP53 status

```{r}
stranded <- readRDS(file.path(data_dir,"pbta-gene-expression-rsem-fpkm-collapsed.stranded.rds"))

# subset to TP53
subset_stranded <- t(stranded)[,"TP53"]

stranded_tp53 <- as.data.frame(subset_stranded) %>%
  dplyr::rename(TP53=subset_stranded) %>%
  rownames_to_column() %>%
  left_join(tp53_alterations_score_wide,by=c("rowname"="Kids_First_Biospecimen_ID_RNA")) %>%
  # remove polya from matched scores
  filter(tp53_altered %in% c("activated","loss"))

# Print table of n per group for legends
stranded_tp53 %>%
  filter(!is.na(tp53_score)) %>%
  group_by(tp53_altered) %>%
  tally()
```

Plot distribution of TP53 gene expression (stranded)

```{r, out.width="50%"}
ggviolin(stranded_tp53, x = "tp53_altered", y = "TP53", 
         color = "tp53_altered", 
         palette = "jco",
         order = c("activated", "loss"),
         add = "boxplot",  
   ggtheme = theme_pubr()) +
  # Add pairwise comparisons p-value
  stat_compare_means(method = "wilcox.test", label.y = 120) +
  xlab("TP53 altered status") +
  ylab("TP53 RNA expression") +
  rremove("legend")

ggsave(file.path(plot_dir, "tp53_expression_by_altered_status_stranded.png"))

```

### (poly-a) Expression profile for activating vs loss TP53 status

```{r}
polya <- readRDS(file.path(data_dir,"pbta-gene-expression-rsem-fpkm-collapsed.polya.rds"))

# subset to TP53
subset_polya <- t(polya)[,"TP53"]

polya_tp53<- as.data.frame(subset_polya) %>%
  dplyr::rename(TP53=subset_polya) %>%
  rownames_to_column() %>%
  left_join(tp53_alterations_score_wide,by=c("rowname"="Kids_First_Biospecimen_ID_RNA")) %>%
  filter(tp53_altered %in% c("activated","loss"))

# Print table of n per group for legends
polya_tp53 %>%
  filter(!is.na(tp53_score)) %>%
  group_by(tp53_altered) %>%
  tally()
```

Plot distribution of TP53 gene expression (poly-a)

```{r, out.width="50%" }
ggviolin(polya_tp53, x = "tp53_altered", y = "TP53", 
         color = "tp53_altered", 
         palette = "jco",
         order = c("activated", "loss"),
         add = "boxplot",  
   ggtheme = theme_pubr()) +
  # Add pairwise comparisons p-value
  stat_compare_means(method = "wilcox.test", label.y = 80) +
  xlab("TP53 altered status") +
  ylab("TP53 RNA expression") +
  rremove("legend")

ggsave(file.path(plot_dir, "tp53_expression_by_altered_status_polya.png"))

```

### Check if other cancer predisposition have high TP53 inactivation scores  

```{r ,out.width="50%"}
ggplot(tp53_alterations_score_wide, aes(x = factor(tp53_altered), y = tp53_score)) +
  geom_violin()+
  geom_jitter(alpha = 0.5, width = 0.2) +
  theme_bw() +
  ggtitle("Distribution of scores across tp53 altered status") +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  xlab("tp53 altered status") +
  facet_wrap(.~cancer_predispositions)

# plot lfs for manuscript
lfs <- tp53_alterations_score_wide %>%
  filter(cancer_predispositions == "Li-Fraumeni syndrome")

lfs %>%
  filter(!is.na(tp53_score)) %>%
  group_by(tp53_altered) %>%
  tally()
```

```{r}
ggviolin(lfs, x = "tp53_altered", y = "tp53_score", 
         color = "tp53_altered", 
         palette = "jco",
         order = c("loss", "other"),
         add = c("boxplot", "jitter"),  
   ggtheme = theme_pubr()) +
  # Add pairwise comparisons p-value
  stat_compare_means(method = "wilcox.test", label.y = 1.2) +
  xlab("TP53 altered status") +
  ylab("TP53 score") +
  rremove("legend")
ggsave(file.path(plot_dir, "lfs_tp53_scores.png"))


```
Some NF-1 and/or Other inherited conditions NOS have high scoring TP53 mutants as well. 

Interesting 2 bs ids annotated with Li-Fraumeni synfrome pre-disposition have very low tp53 classifier scores.

```{r}

tp53_alterations_score_wide %>%
  filter(cancer_predispositions == "Li-Fraumeni syndrome",
         tp53_score < 0.5) 
  
```


## Save file 

Adding DNA and RNA biospecimen ids to `tp53_alterations_score_wide` and saving

```{r}

tp53_alterations_score_wide %>% 
  write_tsv(file.path(results_dir,"tp53_altered_status.tsv"))

```
