---
title: "Mutation Signatures"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---

**Purpose:**

Calculate and plot mutation signatures for all samples using COSMIC and 
Alexandrov et al, 2013 mutational signatures. 

### Summary of Findings:


#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/mutation-signatures/mutation_signatures.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

Read in set up script.

```{r}
if (!("deconstructSigs" %in% installed.packages())) {
  BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
  install.packages("deconstructSigs")
}

# Magrittr pipe
`%>%` <- dplyr::`%>%`

# Load this library
library(deconstructSigs)
```

Set up output directories. 

```{r}
results_dir <- "results"
plots_dir <- "plots"
cosmic_plots <- file.path(plots_dir, "cosmic")
nature_plots <- file.path(plots_dir, "nature")
```

Make new directories for the results. 

```{r}
if (!dir.exists(cosmic_plots)) {
  dir.create(cosmic_plots, recursive = TRUE)
}
if (!dir.exists(nature_plots)) {
  dir.create(nature_plots, recursive = TRUE)
}
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

Custom functions 

```{r}
sample_mut_sig_plot <- function(which_sig_list, label = "none", output_dir = getwd(), 
                                sample_ids) {
  # Given a list of `deconstructSigs::WhichSignature` output, make plots for each sample using 
  # deconstructSigs::plotSignatures
  #
  # Args:
  # 
  #   which_sig_list: a list of `whichSignature` output
  #   label: the label you would like associated with the plot as a character string.
  #   sample_ids: vector of the sample ids. 
  #   output_dir: where the plots should be saved. If this directory doesn' exist, 
  #               will create it. Default is current directory.
  # Returns:
  #   Saved png mutation signature plots to the specified directory. 
  
  if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
  }

  for(sample_num in 1:length(which_sig_list)) {
    png(file.path(output_dir, paste0(sample_ids[sample_num], "_", label, "_mutation_sig.png")))
    plotSignatures(which_sig_list[[sample_num]], sub = sample_ids[sample_num])
    dev.off()
  }
}
```

A function for calculating Mutation per Mb for all signatures. 

```{r}
calc_mut_per_sig <- function(which_sig_list, 
                             muts_per_sample, 
                             wgs_genome_size, 
                             wxs_exome_size, 
                             maf_df) {
  # Given a list of `deconstructSigs::WhichSignature` output, calculate the 
  # mutations per signature per Mb of the genome/exome.
  #
  # Args:
  # 
  #   which_sig_list: a list of `whichSignature` output
  #   muts_per_sample: a vector with the total mutation signature counts for each sample
  #   wgs_genome_size: size of the WGS genome in bp
  #   wxs_genome_size: size of the WXS exome in bp
  #   maf_df: a data.frame with `short_histology` and `experimental strategy` information columns
  #   
  # Returns:
  #   A data.frame that is samples x signatures and has the number of mutations 
  #   per mb for each signature x sample combo
  
  # Count the total number of signature mutations for each sample
  total_muts <- apply(sigs_input, 1, sum)

  # Pull out the signature weights and make into matrix
  sig_weights <- do.call("rbind.data.frame", 
          lapply(which_sig_list, function(sample_data) sample_data$weights)
          ) %>% 
    tibble::rownames_to_column("Tumor_Sample_Barcode") 

  # Calculate the number of mutations contributing to each signature
  sig_num_df <- apply(sig_weights[, -1], 2, function(row_weights) row_weights*total_muts) %>%
    as.data.frame() %>% 
    
    # Add back the sample ids
    dplyr::mutate(Tumor_Sample_Barcode = sig_weights$Tumor_Sample_Barcode) %>% 
    
    # Join the short_histology and experimental stategy information 
    dplyr::inner_join(dplyr::select(maf_df, 
                                    "Tumor_Sample_Barcode", 
                                    "short_histology", 
                                    "experimental_strategy"), 
                      by = "Tumor_Sample_Barcode")  %>% 
    # Get rid of Panel samples
    dplyr::filter(experimental_strategy != "Panel") %>%
    
    # Reformat for plotting
    reshape2::melt() %>% 
    
    # Only keep distinct
    dplyr::distinct(Tumor_Sample_Barcode, variable, .keep_all = TRUE) %>%
    
    # Add genome size and calculate the mutation per this column
    dplyr::mutate(genome_size = dplyr::recode(experimental_strategy,
        "WGS" = wgs_size,
        "WXS" = wxs_size
      ), 
      mut_per_mb = value/(genome_size/10^6))
  
  return(sig_num_df)
}
```
 
Customized plotting functions. 

```{r}
bubble_matrix_plot <- function(sig_num_df, label = "none") {
  # Given the data.frame output from `calc_mut_per_sig` where the number of mutations per Mb 
  # that belong to each sample x signature combo, make a bubble matrix plot that 
  # displays this information for all samples but summarized by histology. 
  #
  # Args:
  # 
  #   sig_num_df: a data.frame with number of mutations per Mb that belong to 
  #               each sample x signature combo
  #   label: a character string for the title of the plot to be passed to ggplot2::ggtitle
  #   
  # Returns:
  #   A bubble matrix plot with the number of mutations per Mb for all samples
  #   summarized by histology. 
  #   
  # Summarize the mut_per_mb column by histology
  grouped_sig_num <- sig_num_df %>% 
    dplyr::arrange(variable) %>%
    dplyr::group_by(short_histology, variable) %>%
    dplyr::summarize(num_tumors = sum(value > 0), 
                     mean_num = median(mut_per_mb)) 
 
  # Make the bubble matrix plot
  ggplot2::ggplot(grouped_sig_num , ggplot2::aes(x = short_histology, y = forcats::fct_rev(variable))) +      
    ggplot2::geom_point(ggplot2::aes(color = mean_num, size = num_tumors)) +
    ggplot2::scale_size("Number of Samples", range = c(0, 10)) +    
    ggplot2::scale_color_distiller("Median Number of Mutations per Mb", palette = "YlGnBu") +
    ggplot2::theme_classic() +
    ggplot2::scale_x_discrete(position = "top") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 55, hjust = -.01),
                   axis.text.y = ggplot2::element_text(size = ggplot2::rel(.75)),
                   legend.key.size = ggplot2::unit(.5, "cm")) +
    ggplot2::xlab("") +
    ggplot2::ylab("") +
    ggplot2::ggtitle(label) + 
    ggplot2::scale_x_discrete(position = "top") 
}
```

```{r}
grouped_sig_barplot <- function(hist_groups, sig_num_df, output_dir = getwd(), 
                                label) {
  # Given the data.frame output from `calc_mut_per_sig` where the number of 
  # mutations per Mb that belong to each sample x signature combo, plot the 
  # number of mutations per Mb for each sample as a grouped bar plot for all 
  # signatures. 
  #
  # Args:
  #   hist_groups: a vector of `short_histology` groups to each be plotted
  #                individually.
  #   sig_num_df: a data.frame with number of mutations per Mb that belong to 
  #               each sample x signature combo
  #   output_dir: where the plots should be saved. If this directory doesn' exist, 
  #               will create it. Default is current directory.
  #   label: a character string for the title of the plot to be passed to it's 
  #   png name and ggtitle.
  # 
  # Returns: 
  #  A grouped barplot saved as png with themutations per Mb for each sample 
  #  from each signature
  #  
  # Do this for each histology group provided in. the vector
  for (hist_group in hist_groups) {
    # Narrow the df down to just this histology's group
    histology_df <- sig_num_df %>% 
      dplyr::filter(short_histology == hist_group, mut_per_mb > 0) %>% 
      dplyr::mutate("Signatures" = droplevels(variable))
  
    # Make the grouped bar plot
    histology_df %>%
      ggplot2::ggplot(ggplot2::aes(x = reorder(Tumor_Sample_Barcode, -mut_per_mb), y = mut_per_mb)) +
      ggplot2::geom_bar(stat = "identity", ggplot2::aes(fill = Signatures)) + 
      ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 55, hjust = 1)) +
      ggplot2::ylab("Mutations per Mb") + 
      ggplot2::xlab("") + 
      ggplot2::theme_classic() + 
      ggplot2::ggtitle(paste(hist_group, label, "signatures"))

    # Save the plot
    ggplot2::ggsave(file.path(output_dir, 
                              paste0("barplot_", 
                                    hist_group, 
                                    "_", label, 
                                    "_mutation_sig.png")))
  }
}
```

## Set up data

```{r}
maf <- data.table::fread("consensus_mutation.maf.tsv")
```


```{r}
# Set up BED region files for TMB calculations
wgs_bed <- readr::read_tsv(file.path("..", "..", "data", "release-v8-20191104", "WGS.hg38.strelka2.unpadded.bed"),
                           col_names = FALSE)
wxs_bed <- readr::read_tsv(file.path("..", "..", "data", "release-v8-20191104", "WXS.hg38.100bp_padded.bed"), 
                           col_names = FALSE)

# Calculate size of genome surveyed
# These files are BED files where the third column is the End position and
# the second column is the Start position.
# So End - Start gives the size of each range. Sum the gives the total size in bp.
wgs_size <- sum(wgs_bed[, 3] - wgs_bed[, 2])
wxs_size <- sum(wxs_bed[, 3] - wxs_bed[, 2])
```

Determine how many mutations we have per sample.

```{r}
mut_per_sample <- maf %>% 
  dplyr::group_by(Tumor_Sample_Barcode) %>% 
  dplyr::tally() %>% 
  dplyr::arrange(n)

summary(mut_per_sample$n)
```

Graph this.

```{r}
  ggplot2::ggplot(mut_per_sample, ggplot2::aes(x = n, geom = "density")) + 
  ggplot2::geom_density() + 
  ggplot2::theme_classic() 
```

Make mutation data into `deconstructSigs` input format.

```{r}
# Convert to deconstructSigs input
sigs_input <- mut.to.sigs.input(mut.ref = maf, 
                                sample.id = "Tumor_Sample_Barcode", 
                                chr = "Chromosome", 
                                pos = "Start_Position", 
                                ref = "Reference_Allele", 
                                alt = "Allele", 
                                bsg = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38)
```

Add total mutations per sample. 

```{r}
# Count the total number of signature mutations for each sample
total_muts <- apply(sigs_input, 1, sum)
```

## Determine Signatures for COSMIC and Alexandrov et al, 2013

Get list of tumor sample ids. 

```{r}
tumor_sample_ids <- maf %>%
  dplyr::filter(Tumor_Sample_Barcode %in% rownames(sigs_input)) %>%
  dplyr::distinct(Tumor_Sample_Barcode) %>%
  dplyr::pull(Tumor_Sample_Barcode)
```

Get [COSMIC signatures](https://cancer.sanger.ac.uk/cosmic) for each sample. 

```{r}
sample_sigs_cosmic <- lapply(sample_ids, function(sample_id) {
  # Determine the signatures contributing to the sample
  whichSignatures(tumor.ref = sigs_input, 
                  signatures.ref = signatures.cosmic, 
                  sample.id = sample_id)
})
```

Get [Alexandrov et al, 2013](https://www.ncbi.nlm.nih.gov/pubmed/23945592) signatures for each sample. 

```{r}
sample_sigs_nature <- lapply(sample_ids, function(sample_id) {
  # Determine the signatures contributing to the sample
  whichSignatures(tumor.ref = sigs_input, 
                  signatures.ref = signatures.nature2013, 
                  sample.id = sample_id)
})
```

### COSMIC Signature Plots

```{r}
sample_mut_sig_plot(sample_sigs_cosmic, 
                    label = "cosmic", 
                    output_dir = file.path(cosmic_plots, "individ_mutation_sig"), 
                    sample_ids = tumor_sample_ids)
```

### Alexandrov et al, 2013 Signature Plots

```{r}
sample_mut_sig_plot(sample_sigs_nature, 
                    label = "nature", 
                    output_dir = file.path(nature_plots, "individ_mutation_sig"), 
                    sample_ids = tumor_sample_ids)
```

### Calculate the mutations per Mb for each signature

Do this for COSMIC mutation signatures.

```{r}
cosmic_sigs_df <- calc_mut_per_sig(sample_sigs_cosmic,
                                   muts_per_sample = total_muts, 
                                   wgs_genome_size = wgs_size, 
                                   wxs_exome_size = wxs_size, 
                                   maf_df = maf)
cosmic_sigs_df
```

Do this for Alexandrov et al, 2013 mutation signatures.

```{r}
nature_sigs_df <- calc_mut_per_sig(sample_sigs_nature,
                                   muts_per_sample = total_muts, 
                                   wgs_genome_size = wgs_size, 
                                   wxs_exome_size = wxs_size, 
                                   maf_df = maf)
nature_sigs_df
```

## Mutation signature bubble matrix by histology groups

```{r}
bubble_matrix_plot(cosmic_sigs_df, label = "cosmic")
```

```{r}
ggplot2::ggsave(file.path(cosmic_plots, paste0("bubble_matrix_cosmic_mutation_sig.png")))
```

```{r}
bubble_matrix_plot(nature_sigs_df, label = "nature")
```

```{r}
ggplot2::ggsave(file.path(nature_plots, paste0("bubble_matrix_nature_mutation_sig.png")))
```

## Mutation signature grouped bar plots for each histology group

```{r}
lapply(unique(cosmic_sigs_df$short_histology), 
       grouped_sig_barplot, 
       sig_num_df = cosmic_sigs_df, 
       output_dir = cosmic_plots, 
       label = "cosmic")
```


```{r}
lapply(unique(nature_sigs_df$short_histology), 
       grouped_sig_barplot, 
       sig_num_df = nature_sigs_df, 
       output_dir = nature_plots, 
       label = "nature")
```

## Session Info

```{r}
sessionInfo()
```
