---
title: "Create subset files for CI"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: J. Taroni for ALSF CCDL
date: 2019
---

In this notebook, we subset the files to include full features for some number 
of matched participants.
The output files are intended to be used for continuous integration.

#### USAGE

To run this from the command line, use:

```
Rscript -e "rmarkdown::render('analyses/create-subset-files/01-create_subset_files.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Number of participants and seed

This variable controls the number of participants that are included in the 
subset files.

```{r}
num_matched_participants <- 90
```

This is used to set the seed when selecting participant ids.

```{r}
seed <- 12345
```

#### Packages and functions

```{r}
# R.utils is required to gzip files here
if (!("R.utils" %in% installed.packages())) {
  install.packages("R.utils")
}
```

`magrittr` pipe.

```{r}
`%>%` <- dplyr::`%>%`
```

Custom function for writing gzipped MAF files.

```{r}
write_maf_file <- function(maf_df, file_name, version_string) {
  # Given a data.frame that contains the fields for a MAF file, write a gzipped
  # MAF file and include the version information provided in version_string.
  # 
  # Note: if file_name exists, it will be overwritten
  #
  # Args:
  #   maf_df: A data.frame that contains the MAF info.
  #   file_name: Output file name, including the full path.
  #   version_string: the version string that will be written to the first line
  #                   of the file at file_name
  #
  # Returns: intended to be used to write files only
  
  # if the file name supplied to this function ends in `.gz`, take it out for
  # the purposes of writeLines, etc.
  # we'll gzip it at the end with R.utils::gzip and this extension is not needed
  if (grepl(".gz", file_name)) {
    file_name <- sub(".gz", "", file_name)
  }
  
  # write the version string to the top of the file
  writeLines(version_string, con = file_name)
  
  # write the tabular data of maf_df
  readr::write_tsv(maf_df, path = file_name, append = TRUE, col_names = TRUE)
  
  # now gzip the file
  R.utils::gzip(file_name, overwrite = TRUE)
}
```

I can confirm that the file written by `write_maf_file` can be read into R via
`maftools::read.maf`.
See `02-subset_maf_concept` for a proof of concept.

### Directories and files

Path to the symlinked data obtained via `bash download-data.sh`.

```{r}
data_dir <- file.path("..", "..", "data")
```

We'll put the subset files in a new directory `../../data/testing`.

```{r}
testing_data_dir <- file.path(data_dir, "testing")
if (!dir.exists(testing_data_dir)) {
  dir.create(testing_data_dir, recursive = TRUE)
}
```

We want the file names to be consistent between the symlinked data and the
subset files that will be used for continuous integration, so we'll hardcode
them here.

```{r}
cnvkit_file <- "pbta-cnv-cnvkit.seg.gz"
controlfreec_file <- "pbta-cnv-controlfreec.seg.gz"
arriba_file <- "pbta-fusion-arriba.tsv.gz"
starfusion_file <- "pbta-fusion-starfusion.tsv.gz"
kallisto_file <- "pbta-gene-expression-kallisto.rds"
rsem_file <- "pbta-gene-expression-rsem.fpkm.rds"
histologies_file <- "pbta-histologies.tsv"
mutect2_file <- "pbta-snv-mutect2.vep.maf.gz"
strelka2_file <- "pbta-snv-strelka2.vep.maf.gz"
manta_file <- "pbta-sv-manta.tsv.gz"
lumpy_file <- "pbta-sv-lumpy.tsv.gz"
```

We'll use `Reduce` + `intersect` on a list that holds the sample identifiers 
from each file to find identifiers present in _all_ files.

```{r}
sample_id_list <- list()
```

## CNV

The CNV files are in [SEG format](https://software.broadinstitute.org/software/igv/SEG) 
and are gzipped.
These are tab-delimited files.

```{r}
cnvkit <- readr::read_tsv(file.path(data_dir, cnvkit_file))
controlfreec <- readr::read_tsv(file.path(data_dir, controlfreec_file))
```

Extract the identifiers from the SEG files.

```{r}
sample_id_list$cnvkit <- unique(cnvkit$ID)
sample_id_list$controlfreec <- unique(controlfreec$ID)
```

## Fusion

```{r}
arriba <- readr::read_tsv(file.path(data_dir, arriba_file))
starfusion <- readr::read_tsv(file.path(data_dir, starfusion_file))
```

```{r}
sample_id_list$arriba <- unique(arriba$tumor_id)
sample_id_list$starfusion <- unique(starfusion$tumor_id)
```

## Gene expression

```{r}
kallisto <- readr::read_rds(file.path(data_dir, kallisto_file))
rsem <- readr::read_rds(file.path(data_dir, rsem_file))
```

```{r}
# the first two columns are transcript and gene identifiers, respectively
sample_id_list$kallisto <- colnames(kallisto)[3:ncol(kallisto)]
# the first column contains gene identifiers
sample_id_list$rsem <- colnames(rsem)[2:ncol(kallisto)]
```

## SNV

SNV files are in [annotated MAF format](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/master/doc/format/vep-maf.md).
See also: https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/

These are a bit unwieldy to read in, so we'll split up the chunks below.

```{r}
mutect2 <- data.table::fread(file.path(data_dir, mutect2_file),
                             skip = 1,  # skip version string
                             data.table = FALSE)
```

```{r}
strelka2 <- data.table::fread(file.path(data_dir, strelka2_file),
                              skip = 1,  # skip version string
                              data.table = FALSE)
```

```{r}
sample_id_list$mutect2 <- unique(mutect2$Tumor_Sample_Barcode)
sample_id_list$strelka2 <- unique(strelka2$Tumor_Sample_Barcode)
```

## SV 

The Manta SV file is in [annotated Manta TSV format](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/master/doc/format/manta-tsv-header.md).

```{r}
manta <- readr::read_tsv(file.path(data_dir, manta_file))
lumpy <- readr::read_tsv(file.path(data_dir, lumpy_file))
```

```{r}
sample_id_list$manta <- unique(manta$Kids.First.Biospecimen.ID.Tumor)
sample_id_list$lumpy <- unique(lumpy$Kids.First.Biospecimen.ID.Tumor)
```

## Select samples and subset files

### Map to participant identifiers

These sample identifiers are all biospecimen identifiers.
The tools that use different experiment strategies will have no overlapping
biospecimen identifiers.
Thus, we'll need to map _biospecimen_ identifiers to _participant_ identifiers.
We'll use the histologies file to do this.

```{r}
histologies_df <- readr::read_tsv(file.path(data_dir, histologies_file))
```

```{r}
participant_id_list <- lapply(
  sample_id_list,
  function(sample_ids) {
    participant_ids <-  histologies_df %>%
      dplyr::filter(Kids_First_Biospecimen_ID %in% sample_ids) %>% 
      dplyr::pull(Kids_First_Participant_ID)
    unique(participant_ids)
  }
)
```

Now we can take the intersection of elements of `participant_id_list` to
identify participant identifiers represented in each of the files we're trying
to subset.

```{r}
participants_in_all <- base::Reduce(intersect, participant_id_list)
```

### Select samples

We'll select 5 participant identifiers for the purposes of producing subset
files (for continuous integration).

```{r}
# set seed for reproducibility
set.seed(seed)
selected_participant_ids <- sample(participants_in_all, 
                                   num_matched_participants)
```

Again, each of the tools use biospecimen identifiers. 
We'll have to map back to those to accomplish the subsetting task.

```{r}
selected_biospecimen_df <- histologies_df %>%
  dplyr::filter(Kids_First_Participant_ID %in% selected_participant_ids)
```

`selected_biospecimen_df` is our subset of `pbta-histologies.tsv`.

```{r}
readr::write_tsv(selected_biospecimen_df,
                 file.path(testing_data_dir, "pbta-histologies.tsv"))
```

To keep things less verbose below, we'll keep the biospecimen IDs as a vector.

```{r}
selected_biospecimen_ids <- selected_biospecimen_df %>%
  dplyr::pull(Kids_First_Biospecimen_ID)
```

### Subset files

#### CNV

```{r}
cnvkit_subset <- cnvkit %>%
  dplyr::filter(ID %in% selected_biospecimen_ids) %>%
  readr::write_tsv(file.path(testing_data_dir, cnvkit_file))
```

```{r}
controlfreec_subset <- controlfreec %>%
  dplyr::filter(ID %in% selected_biospecimen_ids) %>%
  readr::write_tsv(file.path(testing_data_dir, controlfreec_file))
```

#### Fusion

```{r}
arriba_subset <- arriba %>%
  dplyr::filter(tumor_id %in% selected_biospecimen_ids) %>%
  readr::write_tsv(file.path(testing_data_dir, arriba_file))
```

```{r}
starfusion_subset <- starfusion %>%
  dplyr::filter(tumor_id %in% selected_biospecimen_ids) %>%
  readr::write_tsv(file.path(testing_data_dir, starfusion_file))
```

#### Gene Expression

```{r}
# extract the RNA-seq biospecimen IDs, specifically to use with dplyr::select
rnaseq_selected_ids <- rlang::quos(selected_biospecimen_df %>%
  dplyr::filter(experimental_strategy == "RNA-Seq") %>%
  dplyr::pull(Kids_First_Biospecimen_ID))
```

```{r}
kallisto_subset <- kallisto %>%
  dplyr::select(transcript_id, gene_id, !!!rnaseq_selected_ids) %>%
  readr::write_rds(file.path(testing_data_dir, kallisto_file))
```

```{r}
rsem_subset <- rsem %>%
  dplyr::select(gene_id, !!!rnaseq_selected_ids) %>%
  readr::write_rds(file.path(testing_data_dir, rsem_file))
```

#### SNV

```{r}
mutect2_subset <- mutect2 %>%
  dplyr::filter(Tumor_Sample_Barcode %in% selected_biospecimen_ids)

# write compressed mutect2 subset file
write_maf_file(maf_df = mutect2_subset,
               file_name = file.path(testing_data_dir, mutect2_file),
               # we'll get the version string directly from the original 
               # mutect2 file
               version_string = readLines(file.path(data_dir, mutect2_file), 
                                          n = 1))
```

```{r}
strelka2_subset <- strelka2 %>%
  dplyr::filter(Tumor_Sample_Barcode %in% selected_biospecimen_ids)

# write compressed strelka2 subset file
write_maf_file(maf_df = strelka2_subset,
               file_name = file.path(testing_data_dir, strelka2_file),
               # we'll get the version string directly from the original 
               # strelka2 file
               version_string = readLines(file.path(data_dir, strelka2_file), 
                                          n = 1))
```

#### SV

We can use the participant identifiers directly to subset the Manta data.frame.

```{r}
manta_subset <- manta %>%
  dplyr::filter(Kids.First.Participant.ID %in% selected_participant_ids) %>%
  readr::write_tsv(file.path(testing_data_dir, manta_file))
```

```{r}
lumpy_subset <- lumpy %>%
  dplyr::filter(Kids.First.Participant.ID %in% selected_participant_ids) %>%
  readr::write_tsv(file.path(testing_data_dir, lumpy_file))
```

## Session Info

```{r}
sessionInfo()
```
