---
title: "Chromosomal Instability: By Histology Plots"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
params:
  min_samples: 5
---

This analysis evaluates chromosomal instability by using breakpoint SV and CNV 
data that was co-localized by histology in `01-localization-of-breakpoints.Rmd`. 
This notebook returns a CDF plot and genome map plots for each `short_histology` 
group in the `plots/tumor-type` directory.

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

```
Rscript -e "rmarkdown::render('analyses/chromosomal-instability/02b-plot-chr-instability-by-histology.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}
# Set seed so heatmaps turn out the same
set.seed(2020)

# Magrittr pipe
`%>%` <- dplyr::`%>%`

# Need the plotting functions
source(file.path("util", "chr-break-plot.R"))
```

Set up option for how many samples minimum we need to include in CDF plot. 
Borrowed this type of logic from @sjspielman in the [gene-set-enrichment analysis.](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/ee1283991d30237a6a04be8fc6fec7672844133b/analyses/gene-set-enrichment-analysis/02-model-gsea.Rmd#L55)

```{r}
# Let's check if this is numeric
if (!(is.numeric(params$min_samples))) {
  warning("The parameter min_samples should be numeric. Trying to coerce to numeric.")
  params$min_samples <- as.numeric(params$min_samples)
} 
```

### Directories and Files

```{r}
# Path to data directory
data_dir <- file.path("..", "..", "data")

# Path to output directory
plots_dir <- "plots"

# Path to tumor type plots output directory
hist_plots_dir <- file.path(plots_dir, "tumor-type")

# Create the hist_plots_dir  if it does not exist
if (!dir.exists(hist_plots_dir)) {
  dir.create(hist_plots_dir, recursive = TRUE)
}
```

Here's all the input files we will need: 

```{r}
metadata_file <- file.path(data_dir, "pbta-histologies.tsv")
hist_binned_breaks_file <- file.path("breakpoint-data", "histology_breakpoint_binned_counts.RDS")
total_densities_file <- file.path("breakpoint-data", "intersection_of_breaks_densities.tsv")
```

Output plot: 

```{r}
cdf_plot_file <- file.path(plots_dir, "breaks_cdf_plot.png")
```

### Read in data 

Set up metadata

```{r}
# Read in the metadata
metadata <- readr::read_tsv(metadata_file, guess_max = 10000)
```

Load in the localized by histology breakpoint data. 

```{r}
hist_break_local <- readr::read_rds(hist_binned_breaks_file)
```

Read in the total intersection breaks densities TSV that was made by the 
`00-setup-breakpoint-data.R` script. 

```{r}
total_break_densities_df <- readr::read_tsv(
  total_densities_file
) %>%
  # Add on the short histology column from the metadata for plotting purposes
  dplyr::inner_join(metadata %>% dplyr::select(Kids_First_Biospecimen_ID, short_histology),
    by = c("samples" = "Kids_First_Biospecimen_ID")
  )
```

## Plot CDF plot by histology

```{r}
breaks_cdf <- total_break_densities_df  %>%
  # Only plot histologies groups with more than `min_samples` number of samples
  dplyr::group_by(tools::toTitleCase(short_histology), add = TRUE) %>%
  dplyr::filter(dplyr::n() > params$min_samples) %>%
  # Calculate histology group mean
  dplyr::mutate(
    hist_mean = mean(breaks_count),
    hist_rank = rank(breaks_count) / dplyr::n()
  ) %>%
  dplyr::ungroup() %>%
  # Now we will plot these as cummulative distribution plots
  ggplot2::ggplot(ggplot2::aes(
    x = breaks_count,
    color = short_histology
  )) +
  ggplot2::stat_ecdf(pad = FALSE, geom = "point") + 
  ggplot2::coord_flip() +
  # Add summary line for mean
  ggplot2::geom_segment(y = 0, yend = 1, color = "grey", 
                        ggplot2::aes(x = hist_mean, xend = hist_mean)) +
  # Separate by histology
  ggplot2::facet_wrap(~short_histology, nrow = 1, strip.position = "bottom") +
  ggplot2::theme_classic() +
  ggplot2::ylab("") +
  ggplot2::xlab("Chromosomal Breaks per Genome") +
  # Transform to log10 make non-log y-axis labels
  ggplot2::scale_x_continuous(trans = "log1p", breaks = c(0, 1, 3, 10, 30)) +
  ggplot2::scale_y_continuous(limits = c(-0.1, 1.1), breaks = c()) +
  ggplot2::theme(legend.position = "none") +
  ggplot2::theme(
    axis.text.y = ggplot2::element_blank(),
    axis.ticks.y = ggplot2::element_blank(),
    strip.placement = "outside",
    strip.text = ggplot2::element_text(size = 10, angle = 90, hjust = 1), 
    strip.background = ggplot2::element_rect(fill = NA, color = NA)
  )

breaks_cdf
```


```{r}
# Save the plot to a png
ggplot2::ggsave(cdf_plot_file,
  plot = breaks_cdf, width = 40, height = 20, unit = "cm"
)
```

Print from png so rendering is smoother.
![Break Counts CDF Plot](./plots/breaks_cdf_plot.png)

### Plot the breakpoints for each tumor-type

Here we will plot median number of break points for the tumor-type group per 
each bin.

```{r}
purrr::imap(hist_break_local, function(.x, name = .y) {
  # Make the combo plot
  multipanel_break_plot(
    granges_list = .x,
    plot_name = name,
    y_val = "total_counts",
    y_lab = "Total Breaks per Mb",
    plot_dir = hist_plots_dir
  )
})
```

Zip up the PNG files into one file. 

```{r}
# Declare name of zip file
zip_file <- paste0(hist_plots_dir, ".zip")

# Remove any current zip_file of this name so we can overwrite it
if (file.exists(zip_file)) {
  file.remove(zip_file)
}
# Zip up the plots
zip(zip_file, hist_plots_dir, extras = "-j")
```

### Session Info

```{r}
sessionInfo()
```
