---
title: "Chromosomal Instability: Breakpoints"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
---

This analysis evaluates chromosomal instability by formatting SV and CNV data 
into chromosomal breaks and measuring chromosomal break density.
This notebook returns plots for each sample and tumor type group in the 
`plots` directory.

You can see the output notebook [here](https://cansavvy.github.io/openpbta-notebook-concept/chromosomal-instability/chromosomal-instability.nb.html).

Code adapted from [svcnvplus](https://github.com/gonzolgarcia/svcnvplus).

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

```
Rscript -e "rmarkdown::render('analyses/chromosomal-instability/chromosomal-instability.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}
# This threshold will be used to determine percent of genome changed
ch.pct <- 0.02

# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

Read in the custom functions needed for this analysis. 

```{r}
source(file.path("util", "chr-break-calculate.R"))
source(file.path("util", "chr-break-plot.R"))
```

### Directories and Files

```{r}
# Path to data directory
data_dir <- file.path("..", "..", "data")

# Path to output directory
plots_dir <- "plots"

# Path to tumor type plots output directory
hist_plots_dir <- file.path(plots_dir, "tumor-type")

# Create the hist_plots_dir  if it does not exist
if (!dir.exists(hist_plots_dir)) {
  dir.create(hist_plots_dir, recursive = TRUE)
}

# Path to tumor type plots output directory
sample_plots_dir <- file.path(plots_dir, "sample")

# Create the sample_plots_dir if it does not exist
if (!dir.exists(sample_plots_dir)) {
  dir.create(sample_plots_dir, recursive = TRUE)
}
```

### Read in data and format it 

#### Set up metadata

```{r}
# Read in the metadata
metadata <- readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv"))
```

#### Set up CNV data: 

Read in the CNV data and format it to make the chromosome and coordinates easier for
handling later. 

```{r}
# Read in the segment copy number data
cnv_df <- data.table::fread(file.path(data_dir, "pbta-cnv-cnvkit.seg.gz"),
  data.table = FALSE
) %>%
  dplyr::mutate(
    samples = as.factor(ID),
    # Calculate width
    width = loc.end - loc.start,
    # Make a binary variable that labels whether or not a sequence
    # is considered changed by the threshold set
    changed = as.factor(dplyr::case_when(
      seg.mean < log2(1 - ch.pct) ~ "change",
      seg.mean >= log2(1 - ch.pct) ~ "no_change"
    ))
  ) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(chrom = factor(gsub("chr", "", chrom),
    levels = c(1:22, "X", "Y")
  )) %>%
  # Changing these so they end up matching the SV data
  dplyr::rename(start = loc.start, end = loc.end)
```

Filter CNV data to only the changes that are larger than our cutoff `ch.pct`.

```{r}
cnv_filtered_df <- cnv_df %>%
  dplyr::filter(changed == "change")
```

#### Set up SV data

Read in the SV data.

```{r}
sv_df <- data.table::fread(file.path(data_dir, "pbta-sv-manta.tsv.gz"), 
                           data.table = FALSE) %>%
  # Reformat the 23 and 24 chromosomes so they are X and Y and also factors
  dplyr::mutate(
    chrom = factor(dplyr::recode(SV.chrom,
      `23` = "X", `24` = "Y"
    ),
    levels = c(1:22, "X", "Y")
    )
  )
```

#### Set up chromosome reference data

Set up chromosome size data 

```{r}
# Set up Chr sizes
chr_sizes <- readr::read_tsv(file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed"),
  col_names = FALSE
) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(X1 = factor(gsub("chr", "", X1),
    levels = c(1:22, "X", "Y", "M")
  )) %>%
  # Remove sex chromosomes
  dplyr::filter(!(X1 %in% c("X", "Y", "M")))

# Make chromosome size list
chr_sizes_list <- chr_sizes$X3
names(chr_sizes_list) <- chr_sizes$X1
```

## Format the data as chromosomes breaks

The SV and CNV data comes to us in the form of ranges, but for getting a look 
at chromosomal instability, we will want to convert this data into single break
points of the genome. 
We'll also remove the sex chromosomes. 

Make an CNV breaks data.frame. 

```{r}
# Make a CNV data.frame that has the breaks
cnv_breaks <- data.frame(
  samples = rep(cnv_filtered_df$ID, 2),
  chrom = c(cnv_filtered_df$chrom, cnv_filtered_df$chrom),
  start = c(cnv_filtered_df$start, cnv_filtered_df$end),
  end = c(cnv_filtered_df$start, cnv_filtered_df$end),
  seg.mean = rep(cnv_filtered_df$seg.mean, 2),
  copy.num = rep(cnv_filtered_df$copy.num, 2),
  stringsAsFactors = FALSE
) %>%
  # Remove sex chromosomes
  dplyr::filter(!(chrom %in% c("X", "Y")))
```

Make an SV breaks data.frame. 

```{r}
# Make a data.frame that has the breaks
sv_breaks <- data.frame(
  samples = rep(sv_df$Kids.First.Biospecimen.ID.Tumor, 2),
  chrom = c(sv_df$chrom, sv_df$chrom),
  start = c(sv_df$SV.start, sv_df$SV.end),
  end = c(sv_df$SV.start, sv_df$SV.end),
  svclass = rep(sv_df$SV.type, 2),
  stringsAsFactors = FALSE
) %>%
  # Remove sex chromosomes and NAs
  dplyr::filter(!(chrom %in% c("X", "Y", NA)))
```

### Co-localization of breakpoints for samples 

For each sample, we will bin the genome and count how many chromosome breaks 
occur for each bin, given the given `bin_size`.

We will run this for each sample and return a list of three `GenomicRanges` objects:
1) `common_density` contains the combined break counts for both SV and CNV break data.   
2) `cnv_density` contains the number of break counts for CNV.   
3) `sv_density` contains the number of break counts for SV.  

```{r, warning = FALSE, results='hide'}
# Change here and it will change the rest
bin_size <- 1e6

# Obtain a list of samples that are common to both datasets. 
common_samples <- intersect(sv_breaks$samples, cnv_breaks$samples)

# Get a big list of break densities for each sample.
sample_densities <- lapply(common_samples, function(sample_id) {

  # Calculate break densities with both datasets combined
  common_density <- break_density(
    sv_breaks,
    cnv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for CNV only
  cnv_density <- break_density(
    cnv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for SV only
  sv_density <- break_density(
    sv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Return this mega GRanges list
  return(list(
    common_density = common_density,
    cnv_density = cnv_density,
    sv_density = sv_density
  ))
})

# Bring along the sample IDs
names(sample_densities) <- common_samples
```

### Plot the breakpoints for each sample

Given the `GenomicRanges` objects for each sample, create a combined plot for 
each. 

```{r, results='hide'}
lapply(sample_densities, function(granges_list) {

  # Get the sample name
  sample_name <- common_samples[parent.frame()$i[]]

  # Make the combo plot
  multipanel_break_plot(
    granges_list = granges_list,
    plot_name = sample_name,
    y_val = "total_counts",
    y_lab = "Breaks per Mb",
    plot_dir = sample_plots_dir
  )
})
```

### Co-localization of breakpoints for tumor-type groups

Same as was done for each sample, now we will calculate densities for 
each tumor-type group. 

```{r, results='hide'}
# Change bin_size here and it will change the rest
bin_size <- 1e6

# Get a list of the tumor_types for which we have DNA-seq data
tumor_types <- metadata %>% 
  dplyr::filter(!is.na(short_histology), experimental_strategy != "RNA-Seq") %>% 
  dplyr::distinct(short_histology) %>%
  dplyr::pull(short_histology)

tumor_types[12]
```

Run the density calculations for the groups. 

```{r}
# Get a big list of break densities for each sample.
group_densities <- lapply(tumor_types, function(tumor_type) {

  # Obtain a list of sample_ids to subset by
  sample_ids <- metadata %>%
    dplyr::filter(metadata$short_histology == tumor_type) %>%
    dplyr::pull(Kids_First_Biospecimen_ID)
  
  # Double check these samples are in the list
  check_samples <- (sample_ids %in% common_samples)
  
  # If no samples, go to next
  if (sum(check_samples) > 1) {
  
  message(paste("Calculating breakpoint density for", tumor_type, "samples"))

    # Calculate break densities with both datasets combined
  common_density <- break_density(
    sv_breaks,
    cnv_breaks,
    sample_id = sample_ids,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for CNV only
  cnv_density <- break_density(
    cnv_breaks,
    sample_id = sample_ids,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for SV only
  sv_density <- break_density(
    sv_breaks,
    sample_id = sample_ids,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Return this mega GRanges list
  return(list(
    common_density = common_density,
    cnv_density = cnv_density,
    sv_density = sv_density
  ))
  }
})

# Bring along the tumor-type labels
names(group_densities) <- tumor_types

# Remove tumor_types data that didn't have at least two samples
group_densities <- group_densities[!sapply(group_densities, is.null)]
```

### Plot the breakpoints for each tumor-type

Here we will plot median number of break points for the tumor-type group per 
each bin.

```{r}
lapply(group_densities, function(granges_list) {
  # Get the tumor type label
  group_name <- tumor_types[parent.frame()$i[]]

  # Make the combo plot
  multipanel_break_plot(
    granges_list = granges_list,
    plot_name = group_name,
    y_val = "median_counts",
    y_lab = "Median Breaks per Mb",
    plot_dir = hist_plots_dir
  )
})
```

Zip up the PNG files into one file. 

```{r}
zip(paste0(hist_plots_dir, ".zip"), 
    hist_plots_dir
)
```

### Session Info

```{r}
sessionInfo()
```

