---
title: "Chromosomal Instability"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
---

Code adapted from [svcnvplus](https://github.com/gonzolgarcia/svcnvplus). 

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

- Generating a measure of chromosomal instability (CIN) burden that can be plotted by tumor type
- Identify recurrently altered genes to be included in the OncoPrint (#6)
- Amplification/tandem-duplication
- Deep deletion/deletion
- Other Structural variation: Inversion, translocations

```
Rscript -e "rmarkdown::render('analyses/cnv-chrom-plot/chromosomal-instability.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}
# This threshold will be used to determine percent of genome changed
ch.pct <- 0.02

# Path to scratch directory
sv_dir <- file.path("..", "..", "scratch", "sv-vcf")

if (!dir.exists(sv_dir)) {
  # Get SV file processed for the formatting we need
  source(file.path("..", "sv-analysis", "01-process-sv-file.R"))
}

# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

Read in the custom functions needed for this analysis. 

```{r}
source(file.path("util", "wrangling-functions.R"))
```

### Directories and Files

```{r}
# Path to data directory
data_dir <- file.path("..", "..", "data")

# Path to output directory
plots_dir <- "plots"

# Create the plots_dir if it does not exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir, recursive = TRUE)
}
```

### Read in data 

```{r}
# Read in metadata
metadata <-
  readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv"))
```

Read in the CNV data and format it to make the chromosome and coordinates easier for
handling later. 

```{r}
# Read in the segment copy number data
cnv_df <- data.table::fread(file.path(data_dir, "pbta-cnv-cnvkit.seg.gz"),
  data.table = FALSE
) %>% 
  dplyr::mutate(samples = as.factor(ID), 
                # Calculate width
                width = loc.end - loc.start, 
                # Make a binary variable that labels whether or not a sequence 
                # is considered changed by the threshold set
                changed = as.factor(dplyr::case_when(
                  seg.mean < log2(1 - ch.pct) ~ "change", 
                  seg.mean >= log2(1 - ch.pct) ~ "no_change"))) %>% 
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(chrom = factor(gsub("chr", "", chrom), 
                               levels = c(1:22, "X", "Y")))  %>% 
  # Changing these so they end up matching the SV data
  dplyr::rename(start = loc.start, end = loc.end)
```

Read in the SV data.

```{r}
# Get a list of the properly formatted SV files, we'll only take the ones without Y and M chromosomes
sv_files <- list.files(sv_dir, pattern = "_withoutYandM.tsv", full.names = TRUE)

# Read in each of the files
sv_data <- lapply(sv_files, function(sv_file) {
  data.table::fread(sv_file, data.table = FALSE) %>% 
    # The row binding step won't work unless all chromosome variables get 
    # converted to characters consistently
    dplyr::mutate(chrom1 = as.character(chrom1), 
                  chrom2 = as.character(chrom2))
})
```

Carry over the sample names

```{r}
# Extract and carry over the sample names to the list
sample_names <- gsub("_withoutYandM.tsv", "", sv_files)
names(sv_data) <- stringr::word(sample_names, sep = "/", 5)
```

Format this into a data.frame and make the chromosome and coordinates easier for
handling later. 

```{r}
# Bind all of them into one
sv_df <- dplyr::bind_rows(sv_data, .id = "sample_ids") %>% 
  # Reformat the 23 and 24 chromosomes so they are X and Y and also factors
  dplyr::mutate(chrom1 = factor(dplyr::recode(chrom1, 
                                       `23` = "X", `24` = "Y"), 
                                levels = c(1:22, "X", "Y")),
                chrom2 = factor(dplyr::recode(chrom2, 
                                       `23` = "X", `24` = "Y"),
                                levels = c(1:22, "X", "Y"))
                ) %>% 
  # Changing these so they end up matching the SV data
  dplyr::rename(samples = sample_ids, chrom = chrom1) %>% 
  # GenomicRanges is unhappy if end comes before start, so we need to switch these
  dplyr::mutate(start = dplyr::case_when(
    pos2 < pos1 ~ pos2,
    TRUE ~ pos1
  ), 
  end = dplyr::case_when(
    pos2 < pos1 ~ pos1,
    TRUE ~ pos2
  ))
```

### Percent genome change by CNV data

```{r}
perc_genome_change <- cnv_df %>% 
  # Group by biospecimen ID and whether the sequence is considered "changed"
  dplyr::group_by(ID, changed) %>% 
  # Add up total length of "changed" and "no_change" sequences for each sample
  dplyr::summarize(total_bp = sum(as.numeric(width))) %>% 
  # Reformat for easier calculation
  tidyr::spread(total_bp, key = changed) %>%
  # Determine ratio of changed to total
  dplyr::mutate(perc_change = change / sum(change + no_change))
```

Plot percent genome change

```{r}
dplyr::inner_join(perc_genome_change, metadata, 
                  by = c("ID" = "Kids_First_Biospecimen_ID")) %>% 
  ggplot2::ggplot(ggplot2::aes(x = short_histology, y = perc_change, color = short_histology)) + 
  ggforce::geom_sina() + 
  ggplot2::theme_classic() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1), 
                 legend.position = "none") + 
  ggplot2::ylab("Percent Genome Change (> 0.2 threshold)") + 
  ggplot2::xlab("")
```

### Find regions of CNV and SV overlap for each sample

```{r}
sv_cnv_common <- overlap_sv_cnv(cnv_df = cnv_df, 
                                sv_df = sv_df, 
                                by_sample = TRUE)
```

### Find regions of high break density

Make an CNV breaks data.frame. 

```{r}
# Make a CNV data.frame that has the breaks   
cnv_breaks <- data.frame(
  samples = rep(cnv_df$ID, 2), 
  chrom = c(cnv_df$chrom, cnv_df$chrom),
  start = c(cnv_df$start, cnv_df$end),
  end = c(cnv_df$start, cnv_df$end),
  seg.mean = rep(cnv_df$seg.mean, 2), 
  copy.num = rep(cnv_df$copy.num, 2), 
  stringsAsFactors = FALSE)
```

Make an SV breaks data.frame. 

```{r}
# Make a data.frame that has the breaks   
sv_breaks <- data.frame(
  samples = rep(sv_df$sample_ids, 2), 
  chrom = c(sv_df$chrom1, sv_df$chrom2),
  start = c(sv_df$pos1, sv_df$pos2),
  end = c(sv_df$pos1, sv_df$pos2),
  svclass = rep(sv_df$SVtype, 2), 
  stringsAsFactors = FALSE)
```

### Breakpoint burden per sample

Make a breakpoint burden summary table with both CNV and SV data. 

```{r}
# Combine SV and CNV breakpoint summaries into one data.frame
breakpoint_df <- data.frame(table(sv_breaks$samples)) %>% 
  dplyr::inner_join(data.frame(table(cnv_breaks$samples)), 
                    by = "Var1", 
                    suffix = c(".sv", ".cnv")) %>% 
  # Rename the sample column to reflect its info
  dplyr::rename(sample_id = Var1)

# Let's take a look at a preview of this
breakpoint_df
```

### Co-localization of breakpoints

```{r}
TBD
```

### Chromosome shattering

```{r}
TBD
```

### Recurrently shattered regions

```{r}
TBD
```

# Session Info

```{r}
sessionInfo()
```

