---
title: "Chromosomal Instability"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
---

Code adapted from [svcnvplus](https://github.com/gonzolgarcia/svcnvplus). 

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

- Generating a measure of chromosomal instability (CIN) burden that can be plotted by tumor type
- Identify recurrently altered genes to be included in the OncoPrint (#6)
- Amplification/tandem-duplication
- Deep deletion/deletion
- Other Structural variation: Inversion, translocations


```
Rscript -e "rmarkdown::render('analyses/cnv-chrom-plot/gistic_plot.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}
# This threshold will be used to determine percent of genome changed
ch.pct <- 0.02

# Path to scratch directory
sv_dir <- file.path("..", "..", "scratch", "sv-vcf")

if (!dir.exists(sv_dir)) {
  # Get SV file processed for the formatting we need
  source(file.path("..", "sv-analysis", "01-process-sv-file.R"))
}

# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

### Directories and Files

```{r}
# Path to data directory
data_dir <- file.path("..", "..", "data")

# Path to output directory
plots_dir <- "plots"

# Create the plots_dir if it does not exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir, recursive = TRUE)
}
```

### Read in data 

```{r}
# Read in metadata
metadata <-
  readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv"))
```

Read in the CNV data. 

```{r}
# Read in the segment copy number data
cnv_df <- data.table::fread(file.path(data_dir, "pbta-cnv-cnvkit.seg.gz"),
  data.table = FALSE
) %>% 
  dplyr::mutate(ID = as.factor(ID), 
                # Calculate width
                width = loc.end - loc.start, 
                # Make a binary variable that labels whether or not a sequence 
                # is considered changed by the threshold set
                changed = as.factor(dplyr::case_when(
                  seg.mean < log2(1 - ch.pct) ~ "change", 
                  seg.mean >= log2(1 - ch.pct) ~ "no_change"))) %>% 
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(chrom = factor(gsub("chr", "", chrom), 
                               levels = c(1:22, "X", "Y")))
```

Make a `GenomicRanges` object for CNV data. 

```{r}
cnv_ranges <- GenomicRanges::GRanges(
    seqnames = cnv_df$chrom,
    ranges = IRanges::IRanges(
      start = cnv_df$loc.start,
      end = cnv_df$loc.end
    ),
    mcols = cnv_df
  )
```

Read in the SV data.

```{r}
# Get a list of the properly formatted SV files, we'll only take the ones without Y and M chromosomes
sv_files <- list.files(sv_dir, pattern = "_withoutYandM.tsv", full.names = TRUE)

# Read in each of the files
sv_data <- lapply(sv_files, function(sv_file) {
  data.table::fread(sv_file, data.table = FALSE) %>% 
    # The row binding step won't work unless all chromosome variables get 
    # converted to characters consistently
    dplyr::mutate(chrom1 = as.character(chrom1), 
                  chrom2 = as.character(chrom2))
})

# Extract and carry over the sample names to the list
sample_names <- gsub("_withoutYandM.tsv", "", sv_files)
names(sv_data) <- stringr::word(sample_names, sep = "/", 5)

# Bind all of them into one
sv_df <- dplyr::bind_rows(sv_data, .id = "sample_ids")
```

Make an SV breaks data.frame. 

```{r}
# Make a data.frame that has the breaks   
sv_breaks <- data.frame(
  samples = rep(sv_df$sample_ids, 2), 
  chrom = c(sv_df$chrom1, sv_df$chrom2),
  start = c(sv_df$pos1, sv_df$pos2),
  end = c(sv_df$pos1, sv_df$pos2),
  svclass = rep(sv_df$SVtype, 2), 
  stringsAsFactors = FALSE)
```

Make a `GenomicRanges` object for SV breaks data.

```{r}
sv_ranges <- GenomicRanges::GRanges(
    seqnames = sv_breaks$chrom,
    ranges = IRanges::IRanges(
      start = as.numeric(sv_breaks$start),
      end = as.numeric(sv_breaks$end)
    ),
    mcols = sv_breaks
  )
```

### Percent genome change

```{r}
perc_genome_change <- cnv_df %>% 
  # Group by biospecimen ID and whether the sequence is considered "changed"
  dplyr::group_by(ID, changed) %>% 
  # Add up total length of "changed" and "no_change" sequences for each sample
  dplyr::summarize(total_bp = sum(as.numeric(width))) %>% 
  # Reformat for easier calculation
  tidyr::spread(total_bp, key = changed) %>%
  # Determine ratio of changed to total
  dplyr::mutate(perc_change = change / sum(change + no_change))
```

Plot percent genome change

```{r}
dplyr::inner_join(perc_genome_change, metadata, 
                  by = c("ID" = "Kids_First_Biospecimen_ID")) %>% 
  ggplot2::ggplot(ggplot2::aes(x = short_histology, y = perc_change, color = short_histology)) + 
  ggforce::geom_sina() + 
  ggplot2::theme_classic() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1), 
                 legend.position = "none") + 
  ggplot2::ylab("Percent Genome Change (> 0.2 threshold)") + 
  ggplot2::xlab("")
```


### Co-localization of breakpoints

```{r}
# identify the sample ids that are common to both SV and CNV data
common_sample_ids <- intersect(sv_df$sample_ids, cnv_df$ID)

# Find overlaps for each sample
lapply(common_sample_ids, function(sample_id) {
  
  sv_break_sample <- sv_breaks %>% 
    dplyr::filter(samples == sample_id)
  
  sv_sample_ranges <- GenomicRanges::GRanges(
      seqnames = sv_breaks$chrom,
    ranges = IRanges::IRanges(
      start = as.numeric(sv_breaks$start),
      end = as.numeric(sv_breaks$end)
    ),
    mcols = sv_breaks
  )
  cnv_sample <- cnv_df %>% 
    dplyr::filter(ID == sample_id)
  
  cnv_sample_ranges <- GenomicRanges::GRanges(
    seqnames = cnv_df$chrom,
    ranges = IRanges::IRanges(
      start = cnv_df$loc.start,
      end = cnv_df$loc.end
    ),
    mcols = cnv_df
  )
  
  overlaps <- GenomicAlignments::findOverlaps(sv_sample_ranges,
                                              cnv_sample_ranges)
  sv_match <- unique(queryHits(overlaps))
  cnv_match <- unique(subjectHits(overlaps))
  
}
)
```

### Breakpoint burden

Make a breakpoint burden summary table with both CNV and SV data. 

```{r}
# Combine SV and CNV breakpoint summaries into one data.frame
breakpoint_df <- data.frame(table(sv_df$sample_ids)) %>% 
  dplyr::inner_join(data.frame(table(cnv_df$ID)), 
                    by = "Var1", 
                    suffix = c(".sv", ".cnv")) %>% 
  # Rename the sample column to reflect its info
  dplyr::rename(sample_id = Var1)

# Let's take a look at a preview of this
breakpoint_df
```

```{r}
ggplot2::ggplot(breakpoint_df, ggplot2::aes(x = Freq.sv, y = Freq.cnv)) + 
  ggplot2::geom_point()
```



### Chromosome shattering

```{r}
shattered.regions(segdf, svdf, fc.pct = 0.2,  min.num.probes = 5, clean.brk = 8,
                                window.size = 10,slide.size = 2, num.seg.breaks = 6, 
                                num.seg.sd = 5, num.sv.breaks = 6, num.sv.sd = 5, 
                                num.common.breaks = 3, num.common.sd = 0, interleaved.cut = 0.33,
                                dist.iqm.cut = 100000,verbose=FALSE)
```


### Recurrently shattered regions

```{r}
freq.p.test
```


# Session Info

```{r}
sessionInfo()
```

