---
title: "Chromosomal Instability: Breakpoints"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
---

This analysis evaluates chromosomal instability by formatting SV and CNV data 
into chromsomal breaks. 

This notebook returns plots for each sample in the `plots` directory.

You can see the output notebook [here](https://cansavvy.github.io/openpbta-notebook-concept/chromosomal-instability/chromosomal-instability.nb.html).

Code adapted from [svcnvplus](https://github.com/gonzolgarcia/svcnvplus).

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

```
Rscript -e "rmarkdown::render('analyses/chromosomal-instability/chromosomal-instability.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}
# This threshold will be used to determine percent of genome changed
ch.pct <- 0.02

# Path to scratch directory
sv_dir <- file.path("..", "..", "scratch", "sv-vcf")

if (!dir.exists(sv_dir)) {
  # Get SV file processed for the formatting we need
  source(file.path("..", "sv-analysis", "01-process-sv-file.R"))
}

# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

Read in the custom functions needed for this analysis. 

```{r}
source(file.path("util", "chr-break-calculate.R"))
source(file.path("util", "chr-break-plot.R"))
```

### Directories and Files

```{r}
# Path to data directory
data_dir <- file.path("..", "..", "data")

# Path to output directory
plots_dir <- "plots"

# Create the plots_dir if it does not exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir, recursive = TRUE)
}
```

### Read in data and format it 

#### Set up CNV data: 

Read in the CNV data and format it to make the chromosome and coordinates easier for
handling later. 

```{r}
# Read in the segment copy number data
cnv_df <- data.table::fread(file.path(data_dir, "pbta-cnv-cnvkit.seg.gz"),
  data.table = FALSE
) %>%
  dplyr::mutate(
    samples = as.factor(ID),
    # Calculate width
    width = loc.end - loc.start,
    # Make a binary variable that labels whether or not a sequence
    # is considered changed by the threshold set
    changed = as.factor(dplyr::case_when(
      seg.mean < log2(1 - ch.pct) ~ "change",
      seg.mean >= log2(1 - ch.pct) ~ "no_change"
    ))
  ) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(chrom = factor(gsub("chr", "", chrom),
    levels = c(1:22, "X", "Y")
  )) %>%
  # Changing these so they end up matching the SV data
  dplyr::rename(start = loc.start, end = loc.end)
```

#### Set up SV data

Read in the SV data.

```{r}
# Get a list of the properly formatted SV files, we'll only take the ones without Y and M chromosomes
sv_files <- list.files(sv_dir, pattern = "_withoutYandM.tsv", full.names = TRUE)

# Read in each of the files
sv_data <- lapply(sv_files, function(sv_file) {
  data.table::fread(sv_file, data.table = FALSE) %>%
    # The row binding step won't work unless all chromosome variables get
    # converted to characters consistently
    dplyr::mutate(
      chrom1 = as.character(chrom1),
      chrom2 = as.character(chrom2)
    )
})
```

Carry over the sample names.

```{r}
# Extract and carry over the sample names to the list
sample_names <- gsub("_withoutYandM.tsv", "", sv_files)
names(sv_data) <- stringr::word(sample_names, sep = "/", 5)
```

Format this into a data.frame and make the chromosome and coordinates easier for
handling later. 

```{r}
# Bind all of them into one
sv_df <- dplyr::bind_rows(sv_data, .id = "sample_ids") %>%
  # Reformat the 23 and 24 chromosomes so they are X and Y and also factors
  dplyr::mutate(
    chrom1 = factor(dplyr::recode(chrom1,
      `23` = "X", `24` = "Y"
    ),
    levels = c(1:22, "X", "Y")
    ),
    chrom2 = factor(dplyr::recode(chrom2,
      `23` = "X", `24` = "Y"
    ),
    levels = c(1:22, "X", "Y")
    )
  ) %>%
  # Changing these so they end up matching the SV data
  dplyr::rename(samples = sample_ids, chrom = chrom1) %>%
  # GenomicRanges is unhappy if end comes before start, so we need to switch these
  dplyr::mutate(
    start = dplyr::case_when(
      pos2 < pos1 ~ pos2,
      TRUE ~ pos1
    ),
    end = dplyr::case_when(
      pos2 < pos1 ~ pos1,
      TRUE ~ pos2
    )
  )
```

### Set up other reference data

Obtain a list of samples that are common to both datasets. 

```{r}
common_samples <- intersect(sv_df$samples, cnv_df$samples)
```

Set up chromosome size data 

```{r}
# Set up Chr sizes
chr_sizes <- readr::read_tsv(file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed"),
  col_names = FALSE
) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(X1 = factor(gsub("chr", "", X1),
    levels = c(1:22, "X", "Y", "M")
  )) %>%
  # Remove sex chromosomes
  dplyr::filter(!(X1 %in% c("X", "Y", "M")))

# Make chromosome size list
chr_sizes_list <- chr_sizes$X3
names(chr_sizes_list) <- chr_sizes$X1
```

## Format the data as chromosomes breaks

The SV and CNV data comes to us in the form of ranges, but for getting a look 
at chromosomal instability, we will want to convert this data into single break
points of the genome. 
We'll also remove the sex chromosomes. 

Make an CNV breaks data.frame. 

```{r}
# Make a CNV data.frame that has the breaks
cnv_breaks <- data.frame(
  samples = rep(cnv_df$ID, 2),
  chrom = c(cnv_df$chrom, cnv_df$chrom),
  start = c(cnv_df$start, cnv_df$end),
  end = c(cnv_df$start, cnv_df$end),
  seg.mean = rep(cnv_df$seg.mean, 2),
  copy.num = rep(cnv_df$copy.num, 2),
  stringsAsFactors = FALSE
) %>%
  # Remove sex chromosomes
  dplyr::filter(!(chrom %in% c("X", "Y")))
```

Make an SV breaks data.frame. 

```{r}
# Make a data.frame that has the breaks
sv_breaks <- data.frame(
  samples = rep(sv_df$samples, 2),
  chrom = c(sv_df$chrom1, sv_df$chrom2),
  start = c(sv_df$start, sv_df$end),
  end = c(sv_df$start, sv_df$end),
  svclass = rep(sv_df$SVtype, 2),
  stringsAsFactors = FALSE
) %>%
  # Remove sex chromosomes
  dplyr::filter(!(chrom %in% c("X", "Y")))
```

### Co-localization of breakpoints

For each sample, we will bin the genome and count how many chromosome breaks 
occur for each bin, given the given bin_size.

We will run this for each sample and return a list of three `GenomicRanges` objects:
1) `common_density` contains the combined break counts for both SV and CNV break data.   
2) `cnv_density` contains the number of break counts for CNV.   
3) `sv_density` contains the number of break counts for SV.  

```{r, results='hide'}
# Change here and it will change the rest
bin_size <- 1e6

# Get a big list of break densities for each sample.
densities <- lapply(common_samples, function(sample_id) {

  # Calculate break densities with both datasets combined
  common_density <- break_density(sv_breaks,
    cnv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for CNV only
  cnv_density <- break_density(cnv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for SV only
  sv_density <- break_density(sv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Return this mega GRanges list
  return(list(
    common_density = common_density,
    cnv_density = cnv_density,
    sv_density = sv_density
  ))
})

# Bring along the sample IDs
names(densities) <- common_samples
```

## Plot the breakpoints for each sample

Given the `GenomicRanges` objects for each sample, create a combined plot for 
each. 

```{r, results='hide'}
lapply(densities, chr_break_plot)
```

# Session Info

```{r}
sessionInfo()
```

