---
title: "Chromosomal Instability: Breakpoints"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
---

This analysis evaluates chromosomal instability by formatting SV and CNV data 
into chromosomal breaks and measuring chromosomal break density.
This notebook returns chromosomal break heatmaps and plots for each tumor type 
group in the `plots` directory.

Code adapted from [svcnvplus](https://github.com/gonzolgarcia/svcnvplus).

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

```
Rscript -e "rmarkdown::render('analyses/chromosomal-instability/chromosomal-instability.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}
# Set seed so heatmaps turn out the same
set.seed(2020)

# This threshold will be used to determine percent of genome changed
# TODO: after updating to consensus CNV data, evaluate this threshold setting.
ch.pct <- 0.2

# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

Read in the custom functions needed for this analysis. 

```{r}
source(file.path("util", "chr-break-calculate.R"))
source(file.path("util", "chr-break-plot.R"))
```

### Directories and Files

```{r}
# Path to data directory
data_dir <- file.path("..", "..", "data")

# Path to output directory
plots_dir <- "plots"

# Path to tumor type plots output directory
hist_plots_dir <- file.path(plots_dir, "tumor-type")

# Create the hist_plots_dir  if it does not exist
if (!dir.exists(hist_plots_dir)) {
  dir.create(hist_plots_dir, recursive = TRUE)
}
```

### Read in data and format it 

#### Set up metadata

```{r}
# Read in the metadata
metadata <- readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv"))
```

#### Set up CNV data: 

Read in the CNV data and format it to make the chromosome and coordinates easier for
handling later. 
TODO: Update to consensus CNV data. 

```{r}
# Read in the segment copy number data
cnv_df <- data.table::fread(file.path(data_dir, "pbta-cnv-cnvkit.seg.gz"),
  data.table = FALSE
) %>%
  dplyr::mutate(
    samples = as.factor(ID),
    # Calculate width
    width = loc.end - loc.start,
    # Make a binary variable that labels whether or not a sequence
    # is considered changed by the threshold set
    # TODO: after updating to Consensus CNV data, see if sex chromosomes still
    # need to be removed and whether these thresholds should be changed.
    changed = as.factor(dplyr::case_when(
      abs(seg.mean) > log2(1 + ch.pct) ~ "change",
      TRUE ~ "no_change"
    ))
  ) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(chrom = factor(gsub("chr", "", chrom),
    levels = c(1:22, "X", "Y")
  )) %>%
  # Changing these so they end up matching the SV data
  dplyr::rename(start = loc.start, end = loc.end)
```

Filter CNV data to only the changes that are larger than our cutoff `ch.pct`.

```{r}
cnv_filtered_df <- cnv_df %>%
  dplyr::filter(changed == "change")
```

#### Set up SV data

Read in the SV data.

```{r}
sv_df <- data.table::fread(file.path(data_dir, "pbta-sv-manta.tsv.gz"),
  data.table = FALSE
) %>%
  # Reformat the 23 and 24 chromosomes so they are X and Y and also factors
  dplyr::mutate(
    chrom = dplyr::recode(SV.chrom,
      `23` = "X", `24` = "Y"
    )
  )
```

#### Set up chromosome reference data

Set up chromosome size data. 
It just so happens that this BED file: `WGS.hg38.strelka2.unpadded.bed` is actually 
just a list of the chromosome sizes so we are using that for now. 

```{r}
# Set up Chr sizes
chr_sizes <- readr::read_tsv(file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed"),
  col_names = FALSE
) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(X1 = factor(gsub("chr", "", X1),
    levels = c(1:22, "X", "Y", "M")
  )) %>%
  # Remove sex chromosomes
  dplyr::filter(!(X1 %in% c("X", "Y", "M")))

# Make chromosome size named vector
chr_sizes_vector <- chr_sizes$X3
names(chr_sizes_vector) <- chr_sizes$X1
```

## Format the data as chromosomes breaks

The SV and CNV data comes to us in the form of ranges, but for getting a look 
at chromosomal instability, we will want to convert this data into single break
points of the genome. 
We'll also remove the sex chromosomes. 
TODO: after updating to consensus CNV data, evaluate whether sex chromosomes should still be removed. 

Make an CNV breaks data.frame. 

```{r}
# Make a CNV data.frame that has the breaks
cnv_breaks <- data.frame(
  samples = rep(cnv_filtered_df$ID, 2),
  chrom = rep(cnv_filtered_df$chrom, 2),
  coord = c(cnv_filtered_df$start, cnv_filtered_df$end),
  seg.mean = rep(cnv_filtered_df$seg.mean, 2),
  copy.num = rep(cnv_filtered_df$copy.num, 2),
  stringsAsFactors = FALSE
) %>%
  # Remove sex chromosomes
  dplyr::filter(!(chrom %in% c("X", "Y")))
```

Make an SV breaks data.frame. 

```{r}
# Make a data.frame that has the breaks
sv_breaks <- data.frame(
  samples = rep(sv_df$Kids.First.Biospecimen.ID.Tumor, 2),
  chrom = rep(sv_df$chrom, 2),
  coord = c(sv_df$SV.start, sv_df$SV.end),
  svclass = rep(sv_df$SV.type, 2),
  stringsAsFactors = FALSE
) %>%
  # Remove sex chromosomes and NAs
  dplyr::filter(!(chrom %in% c("X", "Y", "M", NA)))
```

Make an common breaks data.frame. 

```{r}
common_breaks <- dplyr::bind_rows(sv_breaks, cnv_breaks) %>%
  dplyr::distinct(samples, chrom, coord, .keep_all = TRUE)
```

Put all the breaks into a list. 

```{r}
breaks_list <- list(
  common_breaks = common_breaks,
  cnv_breaks = cnv_breaks,
  sv_breaks = sv_breaks
)
```

### Co-localization of breakpoints for samples 

For each sample, we will bin the genome and count how many chromosome breaks 
occur for each bin, given the given `bin_size`.

We will run this for each sample and return a list of three `GenomicRanges` objects:
1) `common_breaks` contains the combined break counts for both SV and CNV break data.   
2) `cnv_breaks` contains the number of break counts for CNV.   
3) `sv_breaks` contains the number of break counts for SV.  

```{r, warning = FALSE, results='hide'}
# Change here and it will change the rest
bin_size <- 1e6

# Obtain a list of samples that are common to both datasets.
common_samples <- intersect(sv_breaks$samples, cnv_breaks$samples)

# Get a big list of break densities for each sample.
sample_densities <- lapply(common_samples, function(sample_id) {
  lapply(breaks_list, function(breaks_df) {
    break_density(breaks_df,
      sample_id = sample_id,
      start_col = "coord",
      end_col = "coord",
      window_size = bin_size,
      chr_sizes_vector = chr_sizes_vector
    )
  })
})

# Bring along the sample IDs
names(sample_densities) <- common_samples
```

### Set up for making heatmaps of the breakpoints

Given the `GenomicRanges` objects for each sample, create a combined plot for 
each. 

Make chromosome labeling `HeatmapAnnotation` object. 

```{r}
# Extract chromosome names from a GenomicRanges object
chr_labels <- sample_densities[[1]]$common_breaks@seqnames
chr_labels <- paste0("chr", as.factor(chr_labels))

# Make a key for assigning alternating colors to the chromosomes
chr_colors <- rep(c("grey", "lightblue"), length.out = length(unique(chr_labels)))
names(chr_colors) <- unique(chr_labels)

# Create the Heatmap annotation object
chr_annot <- ComplexHeatmap::HeatmapAnnotation(
  df = data.frame(chr_labels), col = list(chr_labels = c(chr_colors)),
  name = "",
  show_legend = FALSE,
  show_annotation_name = FALSE
)
```

Make histology labeling `HeatmapAnnotation` object.

```{r}
# Get the histologies for the samples in this set and order them by histology
histologies <-
  data.frame(Kids_First_Biospecimen_ID = common_samples) %>%
  dplyr::inner_join(dplyr::select(metadata, Kids_First_Biospecimen_ID, short_histology)) %>%
  dplyr::mutate(short_histology = as.factor(short_histology)) %>%
  dplyr::arrange(short_histology) %>%
  tibble::column_to_rownames("Kids_First_Biospecimen_ID")

# Create the Heatmap annotation object
hist_annot <- ComplexHeatmap::rowAnnotation(
  df = histologies,
  show_annotation_name = FALSE
)
```

Make a color function. 

```{r}
col_fun <- circlize::colorRamp2(
  c(0, 2, 5, 10, 20),
  c("#edf8fb", "#b2e2e2", "#66c2a4", "#2ca25f", "#006d2c")
)
```

Make a function for making the heatmaps. 

```{r}
breaks_heatmap <- function(data_name) {
  # A wrapper function for making a heatmap from the samples GenomicRanges list.
  #
  # Args:
  # data_name: a character string that matches a name in the list.

  # Returns:
  # A heatmap of the chromosomal breaks

  # Pull out the total_counts
  total_counts <- lapply(sample_densities, function(granges_list) {
    granges_list[[data_name]]@elementMetadata@listData$total_counts
  })
  # Make into a data.frame
  total_counts <- dplyr::bind_rows(total_counts) %>%
    dplyr::select(rownames(histologies)) %>%
    t()
  # Add chromosome labels
  colnames(total_counts) <- chr_labels
  # Plot on a heatmap
  heatmap <- ComplexHeatmap::Heatmap(total_counts,
    col = col_fun,
    heatmap_legend_param = list(title = "Count of chr breaks"),
    cluster_columns = FALSE,
    cluster_rows = FALSE,
    show_column_names = FALSE,
    show_row_names = FALSE,
    bottom_annotation = chr_annot,
    left_annotation = hist_annot
  )
  # Return plot
  return(heatmap)
}
```

## Common breaks heatmap

```{r}
common_heatmap <- breaks_heatmap(data_name = "common_breaks")

# Save plot as PNG
png(file.path(plots_dir, paste0("common_breaks_heatmap.png")),
  width = 1200, height = 900, units = "px"
)
common_heatmap
dev.off()

# Print out here
common_heatmap
```

## CNV breaks heatmap

```{r}
cnv_heatmap <- breaks_heatmap(data_name = "cnv_breaks")

# Save plot as PNG
png(file.path(plots_dir, paste0("cnv_breaks_heatmap.png")),
  width = 1200, height = 900, units = "px"
)
cnv_heatmap
dev.off()

# Print out here
cnv_heatmap
```

## SV breaks heatmap

```{r}
sv_heatmap <- breaks_heatmap(data_name = "sv_breaks")

# Save plot as PNG
png(file.path(plots_dir, paste0("sv_breaks_heatmap.png")),
  width = 1200, height = 900, units = "px"
)
sv_heatmap
dev.off()

# Print out here
sv_heatmap
```

### Co-localization of breakpoints for tumor-type groups

Same as was done for each sample, now we will calculate densities for 
each tumor-type group. 

```{r, results='hide'}
# Change bin_size here and it will change the rest
bin_size <- 1e6

# Get a list of the tumor_types for which we have DNA-seq data
tumor_types <- metadata %>%
  dplyr::filter(!is.na(short_histology), experimental_strategy != "RNA-Seq") %>%
  dplyr::distinct(short_histology) %>%
  dplyr::pull(short_histology)
```

Run the density calculations for the groups. 

```{r}
# Get a big list of break densities for each sample.
group_densities <- lapply(tumor_types, function(tumor_type) {

  # Obtain a list of sample_ids to subset by
  sample_ids <- metadata %>%
    dplyr::filter(metadata$short_histology == tumor_type) %>%
    dplyr::pull(Kids_First_Biospecimen_ID)

  # Double check these samples are in the list
  check_samples <- (sample_ids %in% common_samples)

  # If no samples, go to next
  if (sum(check_samples) > 1) {
    message(paste("Calculating breakpoint density for", tumor_type, "samples"))
    lapply(breaks_list, function(breaks_df) {
      break_density(breaks_df,
        sample_id = sample_ids,
        start_col = "coord",
        end_col = "coord",
        window_size = bin_size,
        chr_sizes_vector = chr_sizes_vector
      )
    })
  }
})

# Bring along the tumor-type labels
names(group_densities) <- tumor_types

# Remove tumor_types data that didn't have at least two samples
group_densities <- group_densities[!sapply(group_densities, is.null)]
```

### Plot the breakpoints for each tumor-type

Here we will plot median number of break points for the tumor-type group per 
each bin.

```{r}
purrr::imap(group_densities, function(.x, name = .y) {
  # Make the combo plot
  multipanel_break_plot(
    granges_list = .x,
    plot_name = name,
    y_val = "median_counts",
    y_lab = "Median Breaks per Mb",
    plot_dir = hist_plots_dir
  )
})
```

Zip up the PNG files into one file. 

```{r}
# Declare name of zip file
zip_file <- paste0(hist_plots_dir, ".zip")

# Remove any current zip_file of this name so we can overwrite it
if (file.exists(zip_file)) {
  file.remove(zip_file)
}
# Zip up the plots
zip(zip_file, hist_plots_dir)
```

### Session Info

```{r}
sessionInfo()
```

