---
title: "Subtyping of non-MB/non-ATRT Embryonal Tumors"
author: "Stephanie J. Spielman"
date: 2020
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Purpose

**THIS ANALYSIS IS A WORK IN PROGRESS.**

This RMarkdown file performs molecular subtyping on the unclassifed embryonal tumors. The majority of these samples are already classified as either Medulloblastoma or ATRT, but a subset are neither of these subtypes and require further classification.

We use information from fusions, CNV, and RNAseq expression to subtype these tumors into ETMR and CNS embryonal NOS, as follows:


+ **ETMR, C19MC-altered**
  + Embryonal tumor with multilayered rosettes (ETMR), C19MC-altered
  + These tumors have focal amplification of the miRNA cluster on chr19 (denoted C19MC) and often have gene fusions involving _TTYH1_ and chr19 miRNA cluster genes
  + These tumors have high expression of _LIN28A_ and serves as a biomarker for ETMRs

+ **ETMR, NOS**
  + Embryonal tumor with multilayered rosettes (ETMR)
  + These tumors have high expression of _LIN28A_ and serves as a biomarker for ETMRs, but do not show focal amplification of C19MC.

+ **CNS HGNET-MN1**
  + CNS high-grade neuroepithelial tumor with MN1 alteration
  + Likely previously diagnosed as PNET.
  + Contain gene fusions involving _MN1_. Fusion partners can include _BEND2_ and _CXXC5_. 
  + Predominantly female patients.

+ **CNS HGNET-BCOR**
  + CNS high-grade neuroepithelial tumor with BCOR alteration
  + Tumors have internal tandem duplication of _BCOR_
  + Median age of diagnosis less than 10 years

+ **CNS NB-FOXR2**
  + Central nervous system (CNS) neuroblastoma with FOXR2 activation
  + Over-expression and/or gene fusions in _FOXR2_

+ **CNS EFT-CIC**
  + CNS Ewing sarcoma family tumor with CIC alteration
  + Alterations in _ CIC_, commonly fused with _ NUTM1_

+ **CNS Embryonal, NOS**
  + CNS Embryonal tumor, not otherwise specified
  + Tumors previously called PNET that do not fit into other groups above.





## Setup

Load libraries and set paths to input, output directories and files.

```{r libs-paths}

####### Packages
library(readr)
library(tidyr)
library(dplyr)
library(stringr)
library(kableExtra)
`%>%` <- dplyr::`%>%`

####### Define directories
data_dir    <- file.path("..", "..", "data")
cnv_dir     <- file.path("..", "focal-cn-file-preparation", "results")
results_dir <- file.path("results")
if (!dir.exists(results_dir)) dir.create(results_dir)

######## Input file names
metadata_file <- file.path(data_dir, "pbta-histologies.tsv")
expression_file <- file.path(data_dir, "pbta-gene-expression-rsem-fpkm.stranded.rds")  ## TODO: Make this an input argument
fusion_file      <- file.path(data_dir, "pbta-fusion-putative-oncogenic.tsv")
cnv_file        <- file.path(cnv_dir, "controlfreec_annotated_cn_autosomes.tsv.gz") ## TODO: Is this the right file for now? And/or, make this an input argument
fusion_file_arriba     <- file.path(data_dir, "pbta-fusion-arriba.tsv.gz") ## TODO: Is this the right file for now? And/or, make this an input argument
fusion_arriba <- read_tsv(fusion_file_arriba)
fusion_file_star     <- file.path(data_dir, "pbta-fusion-starfusion.tsv.gz") ## TODO: Is this the right file for now? And/or, make this an input argument
fusion_star <- read_tsv(fusion_file_star)

###### Output file names
file_samples_with_tthy1_c19mc_fusions <- file.path(results_dir, "samples_with_TTYH1_C19MC_fusions.tsv")
```


#### Read in and prepare input files
```{r input-prep}
### Read input files
metadata   <- readr::read_tsv(metadata_file)
expression <- readr::read_rds(expression_file)
fusions    <- readr::read_tsv(fusion_file)
cnv        <- readr::read_tsv(cnv_file)

### Subset metadata to only samples annotated as embryonal but NOT MB, ATRT
# metadata %>% 
#   filter(str_detect(broad_histology, "Embryonal tumor"),
#          !(str_detect(short_histology, "ATRT")),
#          !(str_detect(short_histology, "Medulloblastoma"))) -> unclassified_embryonal
#metadata %>% 
#  filter(str_detect(broad_histology, "Embryonal tumor")) -> unclassified_embryonal

         
### Some useful definitions
## From UCSC Genome Brower, these are positions corresponding to 19q13.41 (C19MC region)
c19mc_region <- "19q13.41"
c19mc_start  <- 51400001
c19mc_end    <- 53600000
```


## Analysis

### Part One: ETMR Subtyping


#### Fusions

First, we identify samples with the following characteristics:

+ Fusions involving TTYH1
+ Fusions in C19MC region


```{r}
######### Samples with fusions involving TTYH1 
fusions %>%
  filter(str_detect(FusionName, "TTYH1")) %>%
  rename(Kids_First_Biospecimen_ID = Sample) %>% 
  dplyr::select(Kids_First_Biospecimen_ID) %>%
  distinct() %>%
  mutate(feature = "TTYH1 Fusion") -> fusions_involving_ttyh1


########### Samples with fusions in C19MC region, need to search by region here
#### Logic: Look for fusions where both breakpoints are on chr19, and at least ONE breakpoint lies within the C19MC, i.e. inclusive \in [c19mc_start,c19mc_end]
#### To do this we need to return to the original fusion files which contain the locations of breakpoints
fusion_star %>%
  dplyr::select(LeftBreakpoint, RightBreakpoint, tumor_id) %>%
  mutate(LeftBreakpoint = str_replace(LeftBreakpoint, "^chr", ""),
         LeftBreakpoint = str_replace(LeftBreakpoint, ":[\\+-]$",""),
         RightBreakpoint = str_replace(RightBreakpoint, "^chr", ""),
         RightBreakpoint = str_replace(RightBreakpoint, ":[\\+-]$","")) %>%
  rename(breakpoint1 = LeftBreakpoint, 
         breakpoint2 = RightBreakpoint) %>%
  mutate(caller = "star")-> star_columns

fusion_arriba %>%
  dplyr::select(breakpoint1, breakpoint2, tumor_id) %>%
  mutate(caller = "arriba") %>%
  bind_rows(star_columns) %>% 
  distinct() %>%
  filter(str_detect(breakpoint1, "^19") & str_detect(breakpoint2, "^19")) %>%
  mutate(breakpoint1 = as.numeric(str_replace(breakpoint1, "^19:","")),
         breakpoint2 = as.numeric(str_replace(breakpoint2, "^19:",""))) %>%
  filter( (breakpoint1 >= c19mc_start & breakpoint1 <= c19mc_end) | 
           (breakpoint2 >= c19mc_start & breakpoint2 <= c19mc_end) ) %>%
  dplyr::select(tumor_id) %>%
  rename(Kids_First_Biospecimen_ID = tumor_id) %>%
  distinct() %>%
  mutate(feature = "C19MC-region Fusion") -> fusions_involving_c19mc
```


```{r}
bind_rows(fusions_involving_ttyh1, fusions_involving_c19mc) %>%
  inner_join(metadata) %>%
  dplyr::select(Kids_First_Biospecimen_ID, sample_id, short_histology, broad_histology, disease_type_new, feature) %>%
  mutate(has_feature = TRUE) %>%
  spread(feature, has_feature) %>%
  replace(is.na(.), FALSE) %>%
  arrange(sample_id) -> samples_fusions_of_interest
write_tsv(samples_fusions_of_interest, file_samples_with_tthy1_c19mc_fusions)

kable(samples_fusions_of_interest) %>% kable_styling()
```


<!--
OLD CODE POSSIBLY TO MAKE A GRAND RETURN, BUT LEAVING HERE COMMENTED OUT UNTIL ITS FATE IS DECIDED.

#### CNV

Next, we identify samples with C19MC amplification. 

```{r, eval=FALSE, echo=FALSE}

########################### THIS CODE IS NOT CURRENTLY BEING RUN ###################################
########### Look for CNV of C19MC which is located at 19q13.41 (defined as `c19mc_region`)
cnv %>%
   rename(Kids_First_Biospecimen_ID = biospecimen_id) %>%
   filter(Kids_First_Biospecimen_ID %in% unclassified_embryonal$Kids_First_Biospecimen_ID,
          cytoband == c19mc_region, 
          status == "gain") %>%
  dplyr::select(Kids_First_Biospecimen_ID) %>% 
  distinct() %>%
  inner_join(metadata) %>%
  select(Kids_First_Biospecimen_ID, sample_id, short_histology, broad_histology) %>%
  mutate(feature = "C19MC Amplification") -> samples_with_c19mc_amp
head(samples_with_c19mc_amp)
```

#### Expression

Next, we obtain LIN28A expression levels.
```{r, eval=FALSE, echo=FALSE}
# RNA-seq samples only
unclassified_embryonal %>%
  filter(experimental_strategy == "RNA-Seq", 
         Kids_First_Biospecimen_ID %in% names(expression)) -> unclassified_embryonal_rnaseq

### Subset and normalize expression data to use only those embroynal samples in `all_embryonal`
expression %>% 
  dplyr::select(gene_id, unclassified_embryonal_rnaseq$Kids_First_Biospecimen_ID) %>%
  gather(Kids_First_Biospecimen_ID, expression_level, -gene_id) %>%
  mutate(norm_expression_level = log2(expression_level + 1)) %>%
  dplyr::select(-expression_level) -> expression_unclassified_embryonal

######### Isolate LIN28A expression levels (exclude pseudogenes P1, P2)
expression_unclassified_embryonal %>%
  filter(str_detect(gene_id, "LIN28A$")) %>%
  inner_join(unclassified_embryonal) %>%
  dplyr::select(sample_id, norm_expression_level) %>%
  rename(lin28a_norm_expression = norm_expression_level) -> samples_lin28a_expression
head(samples_lin28a_expression)
```

#### Merging

Finally, let's put it all together to see what we've found:
```{r, eval=FALSE, echo=FALSE}
fusion_cnv <- bind_rows(samples_with_ttyh1_fusion, samples_with_c19mc_amp, fusions_involving_c19mc_regions) 

left_join(fusion_cnv, samples_lin28a_expression, by = "sample_id") %>%
  arrange(sample_id)
```
We have **four samples** which may be potentially classified as ETMRs based on fusions and/or amplication:

+ `7316-3326` has TTYH1 and C19MC-region fusions, LIN28A log2 expression ~5.48
+ `7316-4034` has C19MC Amplification, LIN28A log2 expression ~0.14 (fairly low)
+ `7316-464` has C19MC Amplification, LIN28A log2 expression ~0.39 (still low, but getting higher!)
+ `7316-784` has TTYH1 fusion, LIN28A log2 expression ~5.5


We can also check for just LIN28A overexpression, revealing samples `7316-2658` and `7316-2987` has having some level of overexpression (_naively_ eyeballing here only). 
```{r, eval=FALSE, echo=FALSE}

samples_lin28a_expression %>% 
  arrange(desc(lin28a_norm_expression), sample_id) %>%
  head(20)
```

-->