---
title: "Update clinically reviewed subtype for PNOC003 samples"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: Krutika Gaonkar for D3b
---

As part of molecular-subtype-HGG analysis we assign a HGG or DMG subtype from looking for K28M histone variants, in this notebook we are identifying samples which have differences between the clinically reviewed subtype and subtype from molecular-subtype-HGG module in PNOC003 and update
those to the clinical subtypes

Adding information from the original [issue](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/751) added by @jharenza 
 
**UPDATE 2020-10-05** 
Below is a double-reviewed (by Cassie Kline and myself) table of histone mutations found in the specified patient tumor per the TGEN genomic reports. 

Kids_First_Participant_ID | H3 Status | tumor_descriptor
-- | -- | --
PT_M23Q0DC3 | H3F3A K28M | Initial CNS Tumor
PT_9GKVQ9QS | H3F3A K28M | Initial CNS Tumor
PT_KBFM551M | H3 WT | Initial CNS Tumor
PT_KBFM551M | no report | Progressive   Disease Post-Mortem
PT_V1HNAC2Q | H3F3A K28M | Initial CNS Tumor
PT_NK8A49X5 | H3F3A K28M | Initial CNS Tumor
PT_NK8A49X5 | H3F3A K28M | Progressive
PT_KZ56XHJT | H3F3A K28M | Initial CNS Tumor
PT_KZ56XHJT | H3F3A K28M | Progressive
PT_KZ56XHJT | no report | Progressive   Disease Post-Mortem
PT_QA9WJ679 | H3F3A K28M | Initial CNS Tumor
PT_WGVEF96B | H3F3A K28M | Initial CNS Tumor
PT_HGM20MW7 | H3F3A K28M | Initial CNS Tumor
PT_0MXPTTM3 | H3 WT | Initial CNS Tumor
PT_M9XXJ4GR | H3F3A K28M | Initial CNS Tumor
PT_KTRJ8TFY | H3F3A K28M | Initial CNS Tumor
PT_KTRJ8TFY | no report | Progressive   Disease Post-Mortem
PT_1E3E6GMF | H3F3A K28M | Initial CNS Tumor
PT_VPEMAQBN | HIST1H3B K28M | Initial CNS Tumor
PT_7P6J57H3 | H3F3A K28M | Initial CNS Tumor
PT_C5FKRB1P | H3 WT | Initial CNS Tumor
PT_CSKHQB16 | HIST1H3B K28M | Initial CNS Tumor
PT_1AAYYGGY | H3F3A K28M | Initial CNS Tumor
PT_NWYSD53S | H3F3A K28M | Initial CNS Tumor
PT_8P368R5B | No report | Initial CNS Tumor
PT_C9YDTZPA | H3F3A K28M | Initial CNS Tumor
PT_KAQMYFYB | H3F3A K28M | Initial CNS Tumor
PT_Y74CVASJ | H3F3A K28M | Initial CNS Tumor
PT_W5GP3F6B | H3F3A K28M | Initial CNS Tumor
PT_A06JR0E5 | H3F3A K28M | Initial CNS Tumor
PT_EN2RN5Y1 | H3F3A K28M | Initial CNS Tumor
PT_DMAF1J4A | H3F3A K28M | Initial CNS Tumor
PT_RYMG3M91 | H3F3A K28M | Initial CNS Tumor
PT_RE6AXQM1 | H3F3A K28M | Initial CNS Tumor
PT_G16VK7FR | HIST1H3B K28M | Initial CNS Tumor
PT_1YQH5NSH | H3F3A K28M | Initial CNS Tumor


#### Directories

```{r}
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")
results_dir <- "results"
```


#### Input

```{r}
# clinical pbta histology
clinical <- readr::read_tsv(file.path(data_dir,
                                "pbta-histologies-base.tsv"), guess_max = 10000) 

```


```{r}
library("tidyverse")

# clinical review file for PNOC003 
pnoc003_review <- readr::read_tsv(file.path("input","pnoc003-clinical-review.tsv")) %>%
  # removing these rows since we would have  
  # the subtyping from molecular-subtyping-HGG 
  # to over write this
  dplyr::filter(`H3 Status` != "no report")
  
# subtyping from molecular-subtype-HGG
subtype <- readr::read_tsv(file.path(results_dir,
                         "compiled_molecular_subtypes_and_report_info.tsv"),guess_max = 10000) 

```

#### Output file

```{r}
output_file <- file.path(results_dir, "compiled_molecular_subtypes_with_clinical_feedback_and_report_info.tsv")
```


Identify pnoc003 samples that have differences between clinical review subtype
and molecular subtype from histone variant mutation status

```{r}
diff_subtype <- subtype %>% left_join(pnoc003_review,by=c("Kids_First_Participant_ID","tumor_descriptor")) %>% 
  dplyr::filter(
    # checking for samples where clinically reviewed subtype is H3 WT
    # but molecular subtype is not HGG
    (grepl("H3 WT",.$`H3 Status`) & !grepl("HGG",molecular_subtype)) |
    # checking for samples where clinically reviewed subtype is H3 K28M
    # but molecular subtype is not DMG H3 K28
    (grepl("K28M",.$`H3 Status`) & !grepl("DMG, H3 K28",molecular_subtype)) | 
    # checking for only RNA-Seq HGG samples where clinically reviewed subtype is H3 K28M
    # but molecular subtype is HGG, To be classified
    (grepl("K28M",.$`H3 Status`) & grepl("HGG, To be classified",molecular_subtype)) |   
    # checking for samples were clinical review was not found  
    grepl("no report",.$`H3 Status`) )  %>%
  dplyr::select(Kids_First_Participant_ID,
                sample_id,
                `H3 Status`,
                molecular_subtype,
                tumor_descriptor) %>%
  distinct()

diff_subtype

```

Counts in each type of differences to identify how many samples already 
have subtyping updated from molecular-subtype-HGG module and how many
samples need molecular subtype updated from clinical review

```{r}

diff_subtype %>% 
  group_by(`H3 Status`,molecular_subtype) %>%
  tally()

```


Seems like for a subset of PNOC samples (WXS and matching RNA-Seq samples from PT_NK8A49X5, PT_QA9WJ679, PT_WGVEF96B)  Notes,molecular_subtype and integrated_diagnosis will need to be updated to capture this information.

```{r}
subtype_clinical_review_df <- subtype %>%
  # matching only on "Kids_First_Participant_ID","tumor_descriptor"
  left_join(pnoc003_review,by=c("Kids_First_Participant_ID","tumor_descriptor")) %>% 
  dplyr::mutate(
      # Since we have some additional values from IDH/BRAF alterations in 
      # molecular-subtping-HGG which are not part of the clinical review
      # I will be using `grepl` to check if the value from clinical review 
      # and molecular-subtyping-HGG values are different
    Notes =  case_when(
      # Don't think we would need to ever change from K28 to H3 wildtype
      # since there would be evidence for K28 from snv data for the sample
      # (grepl("H3 WT",.$`H3 Status`) & !grepl("HGG",molecular_subtype)) 
      # ~ paste(Notes,"Updated to HGG, H3 wildtype from clinical review",sep=","),
      #
      # IF we have K28M in clinical review and molecular subtype if not DMG K28M
      (grepl("K28M",.$`H3 Status`) & !grepl("DMG, H3 K28",molecular_subtype)) 
      # THEN the value in Notes will be updated to specify that the value is
      # updated because of clinical review
      ~ paste(Notes,"Updated to DMG, H3 K28 from clinical review", sep=","),
      # IF K28M in clinical review and molecular subtype is HGG, To be classified
      (grepl("K28M",.$`H3 Status`) & grepl("HGG, To be classified",molecular_subtype))
      # THEN the value in Notes which will be NA will need to be updated to 
      # specify that the it's DMG K28M from clinical review
      ~ paste("Updated to DMG, H3 K28 from clinical review", sep=","),
      # IF clinical review and molecular subtype match OR 
      # no clinical review is provided then keep values from Notes
      TRUE ~ Notes
    ),
    integrated_diagnosis =  case_when(
      # IF we have K28M in clinical review and molecular subtype if not DMG K28M
      (grepl("K28M",.$`H3 Status`) & !grepl("DMG, H3 K28",molecular_subtype)) 
      # THEN the value in integrated_diagnosis will be updated to specify that the value is
      # updated because of clinical review
      ~ "Diffuse midline glioma, H3 K28-mutant",
      # IF clinical review and molecular subtype match OR 
      # no clinical review is provided then keep values from Notes
      TRUE ~ integrated_diagnosis
    ),
    molecular_subtype =  case_when(
      # Don't think we would need to ever change from K28 to H3 wildtype
      # since there would be evidence for K28 from snv data for the sample
      #(grepl("H3 WT",.$`H3 Status`) & !grepl("HGG",molecular_subtype)) ~ 
      #  gsub("DMG, H3 K28","HGG, H3 wildtype",molecular_subtype),
      #
      # IF we have K28M in clinical review and molecular subtype is "HGG, H3 wildtype"
      (grepl("K28M",.$`H3 Status`) & grepl("HGG, H3 wildtype",molecular_subtype)) ~ 
        # THEN the value in molecular subtype will be updated 
        gsub("HGG, H3 wildtype","DMG, H3 K28",molecular_subtype),
      # IF we have K28M in clinical review and molecular subtype is "HGG, To be classified"
      (grepl("K28M",.$`H3 Status`) & grepl("HGG, To be classified",molecular_subtype)) ~ 
        # THEN the value in molecular subtype will be updated 
        gsub("HGG, To be classified","DMG, H3 K28",molecular_subtype),
      # IF clinical review and molecular subtype match OR
      # no clinical review is provided then keep values from molecular subtype
      TRUE ~ molecular_subtype
    )
  ) %>%
  # remove extra columns 
  dplyr::select(-`H3 Status`) %>%
  write_tsv(file.path(output_file))
  
```

