{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "# Importing modules \n",
    "import argparse\n",
    "import os\n",
    "import sys\n",
    "import numpy as np \n",
    "import pandas as pd\n",
    "from scipy.stats import iqr\n",
    "\n",
    "# Reading in table from results folder without the subgroup column\n",
    "EPN_final = pd.read_csv(\"results/EPN_all_data.tsv\", sep=\"\\t\")\n",
    "EPN_final[\"subgroup\"] = \"\"\n",
    "\n",
    "# Output to write \n",
    "outfile = open(\"results/EPN_all_data_withsubgroup.tsv\", \"w\")\n",
    "\n",
    "# This list will hold samples that have been assigned subgroups\n",
    "# that are of priority and cannot be used to assign multiple subgroup names\n",
    "samples_assigned = []\n",
    "\n",
    "# This function will take every row with a list of tuples of size two\n",
    "def prioritized_fusion(row, subgroupname, fusionlist, sample_list):\n",
    "    fusion_boolean = []\n",
    "    for fusionname in fusionlist:\n",
    "        #print(row[fusionname])\n",
    "        if(row[fusionname] > 0):\n",
    "            sample_list.append(row[\"sample_id\"])\n",
    "            fusion_boolean.append(1)\n",
    "        else:\n",
    "            fusion_boolean.append(0)\n",
    "    if 1 in fusion_boolean:\n",
    "        return(subgroupname)\n",
    "    else:\n",
    "        return(row[\"subgroup\"]) \n",
    "\n",
    "def prioritizing_PT_EPN(row, sample_list):\n",
    "    if((row[\"CXorf67_expr_zscore\"]>3 and row[\"1q_gain\"]>0) or\n",
    "        (row[\"TKTL1_expr_zscore\"]>3  and row[\"1q_gain\"]>0)):\n",
    "        return(\"PT_EPN_A\")\n",
    "        sample_list.append(row[\"sample_id\"])\n",
    "    elif((row[\"GPBP1_expr_zscore\"]>3 and row[\"6q_loss\"]>0) or\n",
    "          (row[\"GPBP1_expr_zscore\"]>3 and row[\"6p_loss\"]>0) or \n",
    "          (row[\"IFT46_expr_zscore\"]>3 and row[\"6q_loss\"]>0) or\n",
    "          (row[\"IFT46_expr_zscore\"]>3 and row[\"6p_loss\"]>0)):\n",
    "        return(\"PT_EPN_B\")\n",
    "        sample_list.append(row[\"sample_id\"])\n",
    "    else:\n",
    "        return(row[\"subgroup\"])\n",
    "\n",
    "# Certain fusion  are prioritized in RELA and YAP1 subgroups\n",
    "# The fusions are gievn as a list, every row here is given as input\n",
    "# If the samples have the fusions, the subgroup is assigned and sample_id \n",
    "# is added to samples_assigned  and they will omitted when assessing other categories\n",
    "st_epn_rela_fusions = [\"C11orf95--RELA\", \"LTBP3--RELA\"]\n",
    "EPN_final[\"subgroup\"] = EPN_final.apply(prioritized_fusion,\n",
    "                                        axis=1,\n",
    "                                        subgroupname=\"ST_EPN_RELA\",\n",
    "                                        fusionlist=st_epn_rela_fusions,\n",
    "                                        sample_list=samples_assigned)\n",
    "\n",
    "st_epn_yap1_fusions = [\"C11orf95--YAP1\", \"YAP1--MAMLD1\", \"YAP1--FAM118B\"]\n",
    "EPN_final[\"subgroup\"] = EPN_final.apply(prioritized_fusion,\n",
    "                                        axis=1,\n",
    "                                        subgroupname=\"ST_EPN_YAP1\",\n",
    "                                        fusionlist=st_epn_yap1_fusions,\n",
    "                                        sample_list=samples_assigned)\n",
    "\n",
    "# Same as the two. subrgoups above ET_EPN A and B are prioritized for certain gene expressions and \n",
    "# CNV changes, after these are assigned, samples are not considered for further categories\n",
    "EPN_final[\"subgroup\"] = EPN_final.apply(prioritizing_PT_EPN, axis=1, sample_list=samples_assigned)\n",
    "EPN_final[\"subgroup\"] = EPN_final.apply(prioritizing_PT_EPN, axis=1, sample_list=samples_assigned)\n",
    "\n",
    "# Things to note at this point of the script - 31 samples are categorized with priority subgrouping \n",
    "# and will not be considered for further. subgrouping\n",
    "\n",
    "\n",
    "\n",
    "def subgroup_func(row, subgroupname, column_values, sample_list):\n",
    "    current_subgroup = row[\"subgroup\"]\n",
    "    for columnname, value in column_values:\n",
    "        if row[columnname] > value and row[\"sample_id\"] not in sample_list:\n",
    "            if current_subgroup == '':\n",
    "                current_subgroup = subgroupname\n",
    "            elif subgroupname in current_subgroup.split(\",\"):\n",
    "                pass\n",
    "            else:\n",
    "                current_subgroup = current_subgroup + \",\" + subgroupname\n",
    "    return(current_subgroup)\n",
    "\n",
    "st_epn_rela_tests = [(\"PTEN--TAS2R1\",  0),\n",
    "                     (\"9p_loss\", 0),\n",
    "                     (\"9q_loss\", 0),\n",
    "                     (\"RELA_expr_zscore\", 3),\n",
    "                     (\"L1CAM_expr_zscore\",3)]\n",
    "# Calling function subgroup_func to  set  the values for last column \"subgroup\"\n",
    "EPN_final[\"subgroup\"] = EPN_final.apply(subgroup_func,\n",
    "                                        axis=1,\n",
    "                                        subgroupname=\"ST_EPN_RELA\",\n",
    "                                        column_values=st_epn_rela_tests,\n",
    "                                        sample_list=samples_assigned)\n",
    "\n",
    "#### Looking for ST_EPN_YAP1 sub-group samples\n",
    "st_epn_yap1_tests = [(\"C11orf95--MAML2\", 0),\n",
    "             (\"11q_loss\",  0),\n",
    "             (\"11q_gain\", 0),\n",
    "             (\"ARL4D_expr_zscore\", 3), \n",
    "             (\"CLDN1_expr_zscore\", 3)]  \n",
    "\n",
    "EPN_final[\"subgroup\"] = EPN_final.apply(subgroup_func,\n",
    "                                        axis=1,\n",
    "                                        subgroupname=\"ST_EPN_YAP1\",\n",
    "                                        column_values=st_epn_yap1_tests,\n",
    "                                        sample_list=samples_assigned)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Creating columns for SV and CNV breaks density\n",
    "EPN_final[\"SV instability\"] = \"\"\n",
    "EPN_final[\"CNV_instability\"] = \"\"\n",
    "sv_iqr = iqr(np.array(EPN_final[\"breaks_density-chromosomal_instability_SV\"].dropna()))\n",
    "sv_median = np.median(np.array(EPN_final[\"breaks_density-chromosomal_instability_SV\"].dropna()))\n",
    "cnv_iqr = iqr(np.array(EPN_final[\"breaks_density-chromosomal_instability_CNV\"].dropna()))\n",
    "cnv_median = np.median(np.array(EPN_final[\"breaks_density-chromosomal_instability_CNV\"].dropna()))\n",
    "EPN_final[\"SV instability\"] = EPN_final.apply(lambda x: (x[\"breaks_density-chromosomal_instability_SV\"]-sv_median)/sv_iqr, axis=1)\n",
    "EPN_final[\"CNV_instability\"] = EPN_final.apply(lambda x: (x[\"breaks_density-chromosomal_instability_CNV\"]-cnv_median)/cnv_iqr, axis=1)\n",
    "\n",
    "EPN_final.to_csv(outfile, sep=\"\\t\", header=True, index=False)\n",
    "outfile.close()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
