---
title: "Create consensus hotspot maf"
author: Krutika Gaonkar for D3b
output: html_notebook

---

In the previous step we gather all known cancer hotspots by scavenging calls by
- checking for overlap with amino acid positions in a curated and published cancer hotspot [database](https://www.cancerhotspots.org/files/hotspots_v2.xls)
- checking for overlap with non-coding region hotspots mutations , here we are using the in TERT promoter region

In this notebook, we will create consensus mafs from the filtered call `scratch/hotspot-detection/*RDS`. Since there are slight differences in read support from each caller, we will take a mean value to provide the information per site in the consensus maf file. All other vcf2maf columns will be added from strelka/mutect2  

## Setup
```{r}
library("tidyverse")
library("maftools")

root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")

results_dir <- file.path(root_dir,
                         "analyses",
                         "hotspots-detection",
                         "results")
# Make output folder
if (!dir.exists(results_dir )) {
  dir.create(results_dir , recursive = TRUE)
}

scratch_dir <- file.path(
  root_dir,
  "scratch",
  "hotspot-detection"
)

# Import function to split MNV to SNV
source(file.path(root_dir, "analyses", "snv-callers", "util", "split_mnv.R"))

maf_coltypes <- readRDS(file.path("input", "maf_coltypes.RDS"))
# maf_coltypes is a named vector with column types information for each column in a maf file
# More description about maf format is available
# [here](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/)

# combined hotspot calls
combined_hotspot_calls <- list.files(path = scratch_dir,pattern = ".RDS",full.names = TRUE) %>%
  map_dfr(readRDS) 

# MNVs are split into SNVs
combined_hotspot_calls_mnv_split <- split_mnv(combined_hotspot_calls)

combined_hotspot_calls <- combined_hotspot_calls %>%
  # removing MNVs here
  filter(!Variant_Type %in% c("DNP", "TNP", "ONP")) %>%
  # adding the MNV split SNV alleles here 
  union_all(combined_hotspot_calls_mnv_split)

# hotspot mutation maf file output
output_file <- file.path(results_dir,"pbta-snv-hotspots-plus-consensus.maf.tsv.gz")


# PBTA consensus maf 
consensus_maf_file <- read_tsv(file.path(data_dir,"pbta-snv-consensus-mutation.maf.tsv.gz"),
                                col_types = maf_coltypes)

```

## Gather consensus maf file for hotspots

The main tasks are:
 - filter `combined_hotspot_calls` to calls that are not found in consensus calls
 - calculate a mean of read supports for REF and ALT from multiple callers which can be slightly different for the consenus maf file
 
```{r warning=FALSE }

# These columns provide unique values 
# to check for exact in the 2 dataframes
join_cols <- c("Hugo_Symbol",
               "Entrez_Gene_Id",
               "NCBI_Build",
               "Chromosome",
               "Start_Position",
               "End_Position",
               "Strand",
               "Variant_Classification",
               "Variant_Type",
               "Reference_Allele",
               "Tumor_Seq_Allele1",
               "Tumor_Seq_Allele2",
               "Tumor_Sample_Barcode")

# Filter combined_hotspot_calls to calls that are only
# found from the hotspot scavenging analysis
combined_maf_hotspots_filt <- combined_hotspot_calls %>%
  anti_join(consensus_maf_file,
          by = join_cols) %>%
  unique()

# These columns are unique to each caller maf file.
# Read support and quality are intrinsic values gathered
# from the tools calling the variant so can vary in values.
unique_cols <- c("t_depth",
                 "n_depth",
                 "t_ref_count",
                 "n_ref_count",
                 "t_alt_count",
                 "n_alt_count",
                 "caller",
                 "vcf_qual")

# Gather mean values in `unique_cols` such that 
# calls from multiple callers are summarised as 1 unique row
combined_maf_hotspots_filt <- combined_maf_hotspots_filt %>%
  group_by_at(setdiff(names(combined_hotspot_calls), unique_cols)) %>%
  summarise(
            t_depth= mean(as.numeric(t_depth)),
            n_depth = mean(as.numeric(n_depth)),
            t_ref_count = mean(as.numeric(t_ref_count)),
            n_ref_count = mean(as.numeric(n_ref_count)),
            t_alt_count = mean(as.numeric(t_alt_count)),
            n_alt_count = mean(as.numeric(n_alt_count)),
            caller_count = n(),
            caller = toString(unique(caller)),
            vcf_qual = mean(as.numeric(vcf_qual))) %>% 
  ungroup() %>%
  # calculate VAF
  mutate(VAF = t_alt_count/(t_alt_count+t_ref_count))

```

## Mean read support for combination of callers

We will try to identify the read support for each combination of callers to visualize if any callers call hostspot sites with low read support calls which
```{r}

ggplot(combined_maf_hotspots_filt[,c("caller","caller_count","t_alt_count")],
       aes(x=as.factor(caller_count),y=t_alt_count))+
  geom_violin()+ 
  ggpubr::stat_compare_means()

```
Hotspot sites called by just 1 caller seem to have majority low read support and a few are found in 3 caller ( now captured in 3 callers because of the variant being present in vardict)

Lets look at these calls that are unique to 1 caller.

```{r}

unique_1caller_hotspots <- combined_maf_hotspots_filt %>%
  filter(caller_count==1)

unique_1caller_hotspots
```
There are 109 SNV or indel calls are only called by 1 caller.

What's the distrbution of the t_alt_count in calls only supported by 1 caller?
```{r}

summary.type <- unique_1caller_hotspots %>%  
  group_by(Variant_Type) %>%
  tally()


ggplot(unique_1caller_hotspots[,c("caller","t_alt_count","Variant_Type")],
       aes(x=as.factor(Variant_Type),y=t_alt_count))+
  geom_violin()+ 
  geom_jitter(height = 0, width = 0.1, aes(color=caller)) +
  # add number of calls at the top
  geom_text(data=summary.type ,aes(x = Variant_Type, y = 350, label=n), fontface =2, size = 4)+
  # draw a line where t_alt_count==5
  geom_hline(aes(yintercept=5),color="red")


```

Majority of the unique calls are deletions and dinucleotude mutations (these are 2 base polyphormism split into 2 SNVs in consensus maf df) followed by SNPs. Some of the unique per caller calls from lancet,mutect2 and vardict seem to be below the red line which denotes t_alt_count==5.


## Merge the combined hotspots with consensus 

We will now create a maf file with hotspots called uniquely from `combined_hotspot_calls` to `consensus_maf_file`
```{r}

combined_maf_hotspots_filt %>%
  bind_rows(consensus_maf_file) %>%
  write_tsv(gzfile(output_file),na = ".")



```

