---
title: "Combine individual caller maf"
output: html_notebook
params:
  db_file:
    label: "all callers db"
    value: scratch/snv_db.sqlite
    input: file
  cancer_hotspot_file:
    label: "MSKCC cancer hotspot file"
    value: input/hotspots_v2.xls
    input: file  
  genomic_site_hotspot_file:
    label: "Genomic locations to capture mutations as hotspots"
    value: input/tert_promoter_hotspots.tsv
    input: file
    
---

In this notebook we will combiine calls from each caller used in OpenPBTA snv

Filtered by:
- IMPACT == 'HIGH|MODERATE|MODIFIER' to remove any LOW mutations in the given amino acid position in hotspot database
- Hugo_Symbol  %in% c(hotspot_database_aa$Hugo_Symbol,hotspot_database_g$Hugo_Symbol)
- Amino acid hotspot overlap ( using a dataframe of hotspots per gene and amino acid position)
- TERT promoter region overlap (using a genomic region overlap filtering strategy )



## Setup
```{r}
library("tidyverse")
library("maftools")

root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")
mskcc_cancer_hotspot_file <- params$cancer_hotspot_file
genomic_region_file <- params$genomic_site_hotspot_file

results_dir <- file.path(root_dir,
                         "analyses",
                         "hotspots-detection",
                         "results")

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

```

## Hotspot database for filtering
```{r}
# Cancer database
hotspot_database_2017_snv <- readxl::read_xls(mskcc_cancer_hotspot_file,sheet = 1) 
hotspot_database_2017_indel <- readxl::read_xls(mskcc_cancer_hotspot_file,sheet = 2)
hotspot_database_aa <- bind_rows( hotspot_database_2017_snv, hotspot_database_2017_indel) %>%
  dplyr::select("Amino_Acid_Position","Hugo_Symbol") %>%
  dplyr::mutate(hotspot_database="MSKCC") %>%
  unique() %>%
  as.data.frame()

hotspot_database_g <- read_tsv(genomic_region_file)

```

## Read variant calls database 
```{r}

db_file <- file.path(root_dir, params$db_file)

# Start up connection
con <- DBI::dbConnect(RSQLite::SQLite(), db_file,path = ":memory:")

```

## Designate caller tables from SQL file
```{r}
strelka <- dplyr::tbl(con, "strelka")
mutect <- dplyr::tbl(con, "mutect")
vardict <- dplyr::tbl(con, "vardict")
lancet <- dplyr::tbl(con, "lancet") 	

```

## Subset maf per caller
```{r}
source ("utils/prepMaf.R")

gene_table <- data.frame("Hugo_Symbol"=c(hotspot_database_aa$Hugo_Symbol,hotspot_database_g$Hugo_Symbol))
strelka_subset<-prepMaf(strelka, gene_table = gene_table,
                        impact_values = "MODERATE|HIGH|MODIFIER",
                        hotspot_amino_acid_position_df = hotspot_database_aa, 
                        hotspot_genomic_site_df = hotspot_database_g) %>%
       mutate(caller="strelka")	
mutect_subset<-prepMaf(mutect, gene_table = gene_table,
                       impact_values = "MODERATE|HIGH|MODIFIER",
                       hotspot_amino_acid_position_df = hotspot_database_aa, 
                       hotspot_genomic_site_df = hotspot_database_g) %>%
	mutate(caller="mutect")
vardict_subset<-prepMaf(vardict, gene_table = gene_table,
                        impact_values = "MODERATE|HIGH|MODIFIER",
                        hotspot_amino_acid_position_df = hotspot_database_aa, 
                        hotspot_genomic_site_df = hotspot_database_g) %>%
	mutate(caller="vardict")
lancet_subset<-prepMaf(lancet, gene_table = gene_table,
                       impact_values = "MODERATE|HIGH|MODIFIER",
                       hotspot_amino_acid_position_df = hotspot_database_aa,
                       hotspot_genomic_site_df = hotspot_database_g) %>%
	mutate(caller="lancet")

# combine
combined_maf_subsets <-bind_rows(strelka_subset,
mutect_subset,
vardict_subset,
lancet_subset) 


# save for future runs
saveRDS(combined_maf_subsets, file.path("results","combined_maf_subsets.RDS"))
	      
```


### Get filtered numbers

Upset plots to visualize caller contributions

```{r}
source(file.path("utils","getUpset.R"))

```

### Upset graphs 

SNP: Single nucleotide polymorphism -- a substitution in one nucleotide
DNP: Double nucleotide polymorphism -- a substitution in two consecutive nucleotides
TNP: Triple nucleotide polymorphism -- a substitution in three consecutive nucleotides
ONP: Oligo-nucleotide polymorphism -- a substitution in more than three consecutive nucleotides


```{r}
getUpset(combined_maf_subsets,c("SNP","DNP","TNP","ONP"))
```
Maority of hotspots is called by all 4 callers, mutect2 individually called maximum hotspots!

INS: Insertion -- the addition of nucleotides
DEL: Deletion -- the removal of nucleotides
```{r}
getUpset(combined_maf_subsets,c("INS","DEL"))
```
Only 2 indels sites are captured with amino acids hotspot filtering which are called by all 4 callers.



```{r}
DBI::dbDisconnect(con)
```
