---
title: "Domain overlap hotspots"
output: html_notebook
params:
  cancer_hotspot_file :
    label: "MSKCC cancer hotspot file"
    value: input/hotspots_v2.xls
    input: file  
  gene_list :
    label: gene table with gene type annotation 
    value: ../fusion_filtering/references/genelistreference.txt 
    input: file  

---

In this notebook we will look at sites that overlap domains of interest in filtered snv/indels from step 1.

Domians of intests:
- Protein kinase domain ; pfam ID == "PF00069"


## Setup
```{r}
library("tidyverse")
library("biomaRt")
library("GenomicRanges")

root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
data_dir <- file.path(root_dir, "data")
mskcc_cancer_hotspot_file <- params$cancer_hotspot_file

results_dir <- file.path(root_dir,
                         "analyses",
                         "hotspots-detection",
                         "results")

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

```

## Cancer hotspot files
```{r}

# Cancer database
hotspot_database_2017_snv <- readxl::read_xls(mskcc_cancer_hotspot_file,sheet = 1) 
hotspot_database_2017_indel <- readxl::read_xls(mskcc_cancer_hotspot_file,sheet = 2)

hotspot_database <- bind_rows( hotspot_database_2017_snv, hotspot_database_2017_indel) %>%
  dplyr::select("Amino_Acid_Position","Hugo_Symbol") %>%
  dplyr::mutate(hotspot_database="MSKCC") %>%
  unique()

```

## Gene table for reccurence checks
```{r}
gene_table <- read_tsv("../fusion_filtering/references/genelistreference.txt")%>%
  dplyr::rename(Hugo_Symbol=Gene_Symbol) %>%
  dplyr::select(Hugo_Symbol,type)

brain_goi <- read_tsv("input/brain-goi-list-new.txt",col_names = "Hugo_Symbol") %>%
	dplyr::mutate(type="brain-goi")

gene_table <- bind_rows(gene_table,
			brain_goi) %>%
  group_by(Hugo_Symbol) %>%
  summarise(type=toString(unique(type)))

```

### Format pfam locations as granges
```{r}
ensembl <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",dataset = "hsapiens_gene_ensembl",host="ensembl.org")

source ("utils/get_pfam_granges.R")

# kinase domain 
kinase_gr <-get_pfam_granges("PF00069",ensembl = ensembl)

```

### Read filtered snv calls
```{r}
filtered_calls <- readRDS(file.path(results_dir,"combined_maf_subsets.RDS")) 
 
 filtered_calls_gr <- makeGRangesFromDataFrame(df = filtered_calls,
                                               keep.extra.columns = TRUE,
                                               seqnames.field = "Chromosome",
                                               start.field = "Start_Position",
                                               end.field = "End_Position",
                                               # To-do update
                                               ignore.strand = TRUE)

```


### Overlap with domain granges 
```{r}
overlaps_kinase_snv <- IRanges::subsetByOverlaps(filtered_calls_gr,
                                                 kinase_gr)

# save as the base maf format
kinase_overlap <- data.frame(Chromosome=seqnames(overlaps_kinase_snv),
           Start_Position=start(overlaps_kinase_snv),
           End_Position=end(overlaps_kinase_snv),
           Hugo_Symbol = overlaps_kinase_snv$Hugo_Symbol,
           Variant_Classification = overlaps_kinase_snv$Variant_Classification,
           IMPACT = overlaps_kinase_snv$IMPACT,
           Protein_position = overlaps_kinase_snv$Protein_position,
           Amino_Acid_Position = overlaps_kinase_snv$Amino_Acid_Position,
           Reference_Allele= overlaps_kinase_snv$Reference_Allele,
           Variant_Type = overlaps_kinase_snv$Variant_Type,
           Tumor_Seq_Allele2 = overlaps_kinase_snv$Tumor_Seq_Allele2,
           Tumor_Sample_Barcode = overlaps_kinase_snv$Tumor_Sample_Barcode,
           HGVSp_Short = overlaps_kinase_snv$HGVSp_Short,
           VAF = overlaps_kinase_snv$VAF,
           caller = overlaps_kinase_snv$caller) %>%
   unique() 

```

```{r}
# functions takes in a maf format dataframe and a variant_type to return an upset plot
source(file.path("utils","getUpset.R"))
```

```{r}
getUpset(kinase_overlap_novel,c("SNP","DNP","TNP","ONP"))
```

```{r}
getUpset(kinase_overlap,c("INS","DEL"))
```

### Gather maf for summary plots

```{r}

kinase_overlap %>%
  left_join(gene_table,by="Hugo_Symbol") %>%
  left_join(hotspot_database,by=c("Amino_Acid_Position","Hugo_Symbol")) %>%
  mutate(hotspot_database = case_when(
    grepl("Cosmic",type) & !is.na(hotspot_database) ~ paste0(hotspot_database,",Cosmic"),
    grepl("Cosmic",type) & is.na(hotspot_database) ~ "Cosmic",
    TRUE ~ hotspot_database
  )) %>%
  dplyr::filter(is.na(hotspot_database)) %>%
   dplyr::select(-caller) %>%
  group_by(Chromosome,Start_Position,End_Position,
           Amino_Acid_Position,Hugo_Symbol,
           Tumor_Sample_Barcode,
           Variant_Type,
           Reference_Allele,
           Tumor_Seq_Allele2,
           HGVSp_Short,
           Variant_Classification) %>%
  # doing a mean so that caller specific VAF 
  # calculations doesn't unique rows
  summarise(VAF = mean(VAF)) %>%
   write_tsv(file.path(results_dir,"kinase_snv_novel_sites.maf"))

kinasedomains_novel_sites <- maftools::read.maf("results/kinase_snv_novel_sites.maf")
domain.titv = titv(maf = kinasedomains_novel_sites, plot = FALSE, useSyn = TRUE)
#plot titv summary
plotTiTv(res = domain.titv)

```

```{r}
plotVaf(kinasedomains_novel_sites,vafCol = "VAF")

```
```{r}
pdf("results/kinasenovelhits.pdf",width = 15,height = 8)
lapply(c("TYRO3","PIM3","ACVRL1","GAK","GRK2","PKN1","CDKL4","ULK1","RPS6KA4"),function(x){
maftools::lollipopPlot(kinasedomainsmaf,gene = x ,AACol = "HGVSp_Short")})
dev.off()
```

