## Scripts exclusively for publication-ready plots

We list information about the relevant figures and modules for each script in this directory below for convenience.

| Individual script | Figure(s) | Linked analysis modules |
|--------|--------|------------------|
| [`fig1-sample-distribution.R`](fig1-sample-distribution.R) | Figure 1 & Supplementary Figure sample distribution figure | N/A |
| [`fig2-oncoprint-landscape.R`](fig2-oncoprint-landscape.R) | Figure 2 & Supplementary Figure 3 panel A | [`oncoprint-landscape`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/oncoprint-landscape) |
| [`fig3-chromothripsis-barplot.R`](fig3-chromothripsis-barplot.R) | Figure 3 | [`chromothripsis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/chromothripsis) |
| [`fig4-tp53-telomerase-panels.R`](fig4-tp53-telomerase-panels.R) | Figure 4 | [`tp53_nf1_score`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/tp53_nf1_score); [`telomerase-activity-prediction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/telomerase-activity-prediction/) |
| [`fig5-panels-gsva-umap.R`](fig5-panels-gsva-umap.R) | Figure 5 | [`transcriptomic-dimension-reduction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/transcriptomic-dimension-reduction) <br> [`collapse-rnaseq`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/collapse-rnaseq) <br> [`gene-set-enrichment-analysis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/gene-set-enrichment-analysis) |
| [`fig5-forest-plot.R`](fig5-forest-plot.R) | Figure 5 | [`survival-analysis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/survival-analysis) <br> [`immune-deconv`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/immune-deconv)  
| [`supp-subtype-umap.R`](supp-subtype-umap.R) | Supplementary Figure 6 panels A-D | [`transcriptomic-dimension-reduction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/transcriptomic-dimension-reduction)
| [`supp-snv-callers-panels.R`](supp-snv-callers-panels.R) | Supplementary Figure 2 panels A-G | [`snv-callers`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/snv-callers) (Note: requires 32 GB RAM)
| [`supp-tmb-compare-panels.R`](supp-tmb-compare-panels.R) | Supplementary Figure 2 panels H-I | [`tmb-compare`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/tmb-compare) 
| [`supp-S3-panels-BCD.R`](supp-S3-panels-BCD.R) | Supplementary Figure 3 panels B-D | [`cnv-chrom-plot`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/cnv-chrom-plot); [`chromothripsis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/chromothripsis); [`copy_number_consensus_call`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/copy_number_consensus_call) 
| [`supp-S5-panels-ABC.R`](supp-S5-panels-ABC.R) | Supplementary Figure 5 panels A-C | [`tp53_nf1_score`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/tp53_nf1_score) <br> [`telomerase-activity-prediction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/telomerase-activity-prediction/)
