## Scripts exclusively for publication-ready plots

We list information about the relevant figures and modules for each script in this directory below for convenience.

| Individual script | Figure(s) | Linked analysis modules |
|--------|--------|------------------|
| [`fig2-oncoprint-landscape.R`](fig2-oncoprint-landscape.R) | Figure 2 & Supplement | [`oncoprint-landscape`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/oncoprint-landscape) |
| [`fig3-chromothripsis-barplot.R`](fig3-chromothripsis-barplot.R) | Figure 3 | [`chromothripsis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/chromothripsis) |
| [`fig4-panels-gsva-umap.R`](fig4-panels-gsva-umap.R) | Figure 4 | [`transcriptomic-dimension-reduction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/transcriptomic-dimension-reduction) <br> [`collapse-rnaseq`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/collapse-rnaseq) <br> [`gene-set-enrichment-analysis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/gene-set-enrichment-analysis) |
| [`TelomeraseActivitites.R`](TelomeraseActivitites.R) |  Figure 4 & Supplement | [`telomerase-activity-prediction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/telomerase-activity-prediction/) |

