## Scripts exclusively for publication-ready plots

We list information about the relevant figures and modules for each script in this directory below for convenience.

| Individual script | Figure(s) | Linked analysis modules |
|--------|--------|------------------|
| [`fig1-sample-distribution.R`](fig1-sample-distribution.R) | Figure 1B | N/A |
| [`fig2_figS3-oncoprint-landscape.R`](fig2-oncoprint-landscape.R) | Figure 2 & Figure S3B | [`oncoprint-landscape`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/oncoprint-landscape) |
| [`fig3-chromothripsis-barplot.R`](fig3-chromothripsis-barplot.R) | Figure 3D | [`chromothripsis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/chromothripsis) |
| [`fig3_figS4-mutational-signatures-panels.R`](fig3_figS4-mutational-signatures-panels.R) | Figure 3E and Figure S4 | [`mutational-signatures`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/mutational-signatures) |
| [`fig4-tp53-telomerase-panels.R`](fig4-tp53-telomerase-panels.R) | Figure 4 panels A, B, C, D, F | [`tp53_nf1_score`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/tp53_nf1_score) <br> [`telomerase-activity-prediction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/telomerase-activity-prediction/) <br> [`survival-analysis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/survival-analysis)  |
| [`fig4-heatmap.R`](fig4-heatmap.R) | Figure 4E | [`mutational-signatures`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/mutational-signatures)  |
| [`fig4-hgg-kaplan-meier.R`](fig4-hgg-kaplan-meier.R) | Figure 4H | [`survival-analysis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/survival-analysis)  |
| [`fig4-hgg-subtype-forest-plot.R`](fig4-hgg-subtype-forest-plot.R) | Figure 4G | [`survival-analysis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/survival-analysis)  |
| [`fig5-panels-gsva-umap.R`](fig5-panels-gsva-umap.R) | Figure 5 panels A, B | [`transcriptomic-dimension-reduction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/transcriptomic-dimension-reduction) <br> [`collapse-rnaseq`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/collapse-rnaseq) <br> [`gene-set-enrichment-analysis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/gene-set-enrichment-analysis) |
| [`fig5-forest-plot.R`](fig5-forest-plot.R) | Figure 5D | [`survival-analysis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/survival-analysis) <br> [`immune-deconv`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/immune-deconv)
| [`fig5_figS6-immune-deconv-panels.R`](fig5_figS6-immune-deconv-panels.R) |  Figure 5 panels C, E and Figure S6 panels E, F | [`immune-deconv`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/immune-deconv)  |
| [`figS2-snv-callers-panels.R`](figS2-snv-callers-panels.R) | Figure S2 panels A-G | [`snv-callers`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/snv-callers) (:warning: _Requires 32 GB RAM_)
| [`figS2-tmb-compare-panels.R`](figS2-tmb-compare-panels.R) | Figure S2 panels H, I | [`tmb-compare`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/tmb-compare)
| [`figS3-panels-cn-chromothripsis.R`](figS3-panels-cn-chromothripsis.R) | Figure S3 panels C, D, E | [`cnv-chrom-plot`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/cnv-chrom-plot) <br> [`chromothripsis`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/chromothripsis) <br> [`copy_number_consensus_call`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/copy_number_consensus_call)
| [`figS5-all-panels.R`](figS5-all-panels.R) | Figure S5 A-C | [`tp53_nf1_score`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/tp53_nf1_score) <br> [`telomerase-activity-prediction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/telomerase-activity-prediction/)
| [`figS6-subtype-umap-panels.R`](figS6-subtype-umap-panels.R) | Figure S6 panels A-D | [`transcriptomic-dimension-reduction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/transcriptomic-dimension-reduction)
| [`figS7-seqcenter-barplot.R`](figS7-seqcenter-barplot.R) | Figure S7 panel A | None |
| [`figS7-UMAP-libraries.R`](figS7-UMAP-libraries.R) | Figure S7 panel B | None |
| [`figS7-tp53-telomerase-tumor-purity-threshold.R`](figS7-tp53-telomerase-tumor-purity-threshold.R) | Figure S7 panel G | [`transcriptomic-dimension-reduction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/transcriptomic-dimension-reduction) <br> [`telomerase-activity-prediction`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/telomerase-activity-prediction/) <br> [`tumor-purity-exploration`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/tumor-purity-exploration/)
