---
title: "Tumor Mutation Burden"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---


#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/tmb-calculations/01-tmb-calculations.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

```{r}
if (!("TCGAbiolinks" %in% installed.packages())) {
  install.packages("TCGAbiolinks")
}
# magrittr pipe
`%>%` <- dplyr::`%>%`


source(file.path("..", "snv-callers", "util", "wrangle_functions.R"))
```

Declare names of output directories.

```{r}
data_dir <- file.path("..", "..", "data")
scratch_dir <- file.path("..", "..", "scratch")
results_dir <- "results"
plots_dir <- "plots"
```

Create output directories in this analysis folder if they weren't made 

```{r}
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

Calculating the TMB for TCGA. 

```{r}
calc_tmb <- function(maf_df,
                     bed_granges = wgs_bed_granges, 
                     genome_size = wgs_genome_size,
                     bp_window = 0) {
  # Given a MAF formatted dta.frame and a BED regions data.frame; filter out
  # any variants of the MAF df that are not within the BED regions.
  #
  # Args:
  #   maf_df: A MAF formattted data.frame
  #   wgs_bed_granges: a file path to TSV file that has BED formatted columns with
  #                 chromosome, start, end positions in that order.
  #   bp_window: how many base pairs away can it be from the BED region to still
  #              be included? Default is 0 bp. This argument gets forwarded
  #              to GenomicRanges::findOverlaps's `maxgap` argument.
  #
  # Returns:
  # The same MAF formatted data.frame with the mutations that lie outside
  # the supplied BED regions filtered out.
  # Turn the MAF WXS sample mutations into a GRanges object
  maf_granges <- maf_to_granges(maf_df)

  # Find the overlap of the BED regions and the mutations This outputs a
  # special GenomicRanges object that contains indices of each of these
  # ranges that overlap
  overlap <- GenomicRanges::findOverlaps(
    maf_granges,
    bed_granges,
    maxgap = bp_window
  )

  # Calculate of ratio of variants in this BED using the @from slot which
  # indicates the indices of the ranges in `wxs_maf_ranges` that have overlaps
  # with `wxs_bed_ranges`
  ratio <- length(overlap@from) / nrow(maf_df)

  # What fraction of mutations are in these bed regions?
  cat(
    "Ratio of variants in this BED:", ratio, "\n",
    "Ratio of variants being filtered out:", 1 - ratio, "\n"
  )
  
  # Only keep those in the BED regions that overlap the `wxs_bed_granges`
  maf_df <- maf_df[unique(overlap@from), ] %>%
    dplyr::group_by(Tumor_Sample_Barcode) %>%
    
    # Count number of mutations for that sample
    dplyr::summarize(mutation_count = dplyr::n()) %>%
    
    # Calculate TMB
    dplyr::mutate(tmb = mutation_count / (genome_size / 1000000))
  
  return(maf_df)
}
```

## Read in and set up data

Download and read in the consensus MAF file. 

```{r}
# Declare file path for consensus file
consensus_file <- file.path(data_dir, "snv-consensus", "consensus_mutation_tmb.tsv")

# If this file hasn't been downloaded, download and unzip it.
if (!file.exists(consensus_file)) {
# Download the file
download.file("https://open-pbta.s3.amazonaws.com/data/snv-consensus/snv-consensus_1112019.zip", 
              destfile = file.path(data_dir, "snv-consensus_1112019.zip"))

# Get the files in this zip file
unzip(file.path(data_dir, "snv-consensus_1112019.zip"),
      exdir = data_dir)
}
```

Read in the consensus MAF file. 

```{r}
# Read in the file
tmb_pbta <- data.table::fread(consensus_file)
```

Set up the WGS data for applying to TCGA data for calculating TMB. 

```{r}
wgs_bed_file <- file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed")

# Set up BED region files for TMB calculations
wgs_bed <- readr::read_tsv(wgs_bed_file, col_names = FALSE) %>%
    dplyr::rename(Chromosome = X1, Start_Position = X2, End_Position = X3)

# Calculate size of genome surveyed
wgs_genome_size <- sum(wgs_bed[, 3] - wgs_bed[, 2])

# Turn the WXS bed regions into a GRanges object
wgs_bed_granges <- maf_to_granges(wgs_bed)
```

Set up the TCGA data. 

```{r}
maf <- list(
  TCGAbiolinks::GDCquery_Maf("LGG", pipelines = "mutect"),
  TCGAbiolinks::GDCquery_Maf("GBM", pipelines = "mutect")
)
```

Set up the TCGA metadata so we can get histology groups. 

```{r}
clinical <- list(
  TCGAbiolinks::GDCquery_clinic(project = "TCGA-LGG", type = "clinical"),
  TCGAbiolinks::GDCquery_clinic(project = "TCGA-GBM", type = "clinical")
)
# Bind the data together
clinical <- dplyr::bind_rows(clinical) 
```

Calculate TMB and set up the TCGA data with its histology. 

```{r}
# Calculate TMB
tmb_tcga <- lapply(maf, calc_tmb) 

# Get the patient barcode from the Tumor Sample Barcode category
tmb_tcga <- dplyr::bind_rows(tmb_tcga) %>% 
  dplyr::mutate(bcr_patient_barcode = substr(Tumor_Sample_Barcode, 0, 12)) %>% 
  # Add on the histology group 
  dplyr::left_join(dplyr::select(clinical, primary_diagnosis, bcr_patient_barcode), 
                   by =  "bcr_patient_barcode") %>% 
  # Rename so it matches PBTA data
  dplyr::rename(short_histology = primary_diagnosis)
```

## Combine the TCGA and PBTA data

```{r}
all_tmb <- dplyr::bind_rows(tmb_tcga, tmb_pbta, .id = "dataset") %>% 
  # Recode the dataset so we know which data is from which
  dplyr::mutate(dataset = dplyr::recode(dataset, `1` = "TCGA", `2` = "PBTA")) %>% 
  dplyr::filter(!is.na(short_histology)) 
```

## Plot the TMB as a jitterplot 

```{r}
jitter_plot <- function(data, colour = NULL, title) {
  data %>% 
    ggplot2::ggplot(ggplot2::aes(x = short_histology, y = tmb))+ 
    ggplot2::geom_jitter(color = colour, alpha = 0.5) + 
    ggplot2::stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean, 
                          geom = "crossbar", 
                          width = 0.6, size = 0.15) +
    ggplot2::theme_classic() + 
    ggplot2::ylab("Number of Mutations per Mb") +
    ggplot2::xlab("") +
    ggplot2::ylim(c(0, 30)) + 
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 60, hjust = 1), 
                 legend.position = "none") + 
    ggplot2::ggtitle(title)
}

# Make TCGA plot and get rid of y axis 
tcga_plot <- jitter_plot(tmb_tcga, colour = "darkred", title = "TCGA (Adult)") + 
  ggplot2::theme(axis.title.y  = ggplot2::element_blank(),
                 axis.text.y  = ggplot2::element_blank(),
                 axis.ticks.y = ggplot2::element_blank(), 
                 axis.line.y = ggplot2::element_blank()) 
  
# Make the plot for PBTA
pbta_plot <- jitter_plot(tmb_pbta, colour = "blue", title = "PBTA")

# Put the plots together
tmb_plot <- cowplot::plot_grid(pbta_plot, tcga_plot, align="h", axis = "left", 
                               rel_widths = c(2.75, 1),
                               label_size = 12)
# Print out tmb plot 
tmb_plot
```


```{r}
# Save the plot to a png
cowplot::save_plot(file.path(plots_dir, "tmb_tcga_and_pbta_plot.png"), plot = tmb_plot)
```

## Session Info

```{r}
sessionInfo()
```
