---
title: "Tumor Mutation Burden"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---


#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/tmb-calculations/02-tmb-calculations.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

```{r}
if (!("R.utils" %in% installed.packages())) {
  install.packages("R.utils")
}
if (!("annotatr" %in% installed.packages())) {
  BiocManager::install("annotatr")
}
```

`magrittr` pipe.

```{r}
`%>%` <- dplyr::`%>%`
```

Declare names of output directories.

```{r}
data_dir <- file.path("..", "..", "data")
scratch_dir <- file.path("..", "..", "scratch")
results_dir <- "results"
plots_dir <- "plots"
mutation_file <- file.path(scratch_dir, "strelka2.RDS")
metadata_dir <- file.path(scratch_dir, "metadata_filtered_maf_samples.tsv")
```

Create output directories in this analysis folder if they weren't made 

```{r}
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

Read in the vetted list of mutations. 

```{r}
# Read in the RDS file made previously
mutations <- readr::read_rds(mutation_file)

# Turn it into a data.frame
mut_df <- mutations@data %>% 
  # Get rid of any variables that have completely NAs.
  dplyr::select(-which(apply(is.na(.), 2, all)))
```

```{r}
if (file.exists(metadata_dir)) {
  # Read the ready-to-go files if these files exist
  metadata <- metadata <- readr::read_tsv(metadata_dir)
} else { # If any of the needed files don't exist, rerun this process:
  # Only import the sample names
  strelka2_samples <- data.table::fread(file.path(
    data_dir,
    "pbta-snv-strelka2.vep.maf.gz"
  ),
    select = "Tumor_Sample_Barcode",
    skip = 1,
    data.table = FALSE
  ) %>%
    dplyr::pull("Tumor_Sample_Barcode")
  
  # Isolate metadata to only the samples that are in the datasets
  metadata <- readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv")) %>%
    dplyr::filter(Kids_First_Biospecimen_ID %in% c(strelka2_samples, mutect2_samples)) %>%
    dplyr::distinct(Kids_First_Biospecimen_ID, .keep_all = TRUE) %>%
    dplyr::arrange() %>%
    dplyr::rename(Tumor_Sample_Barcode = Kids_First_Biospecimen_ID) %>%
    readr::write_tsv(file.path(scratch_dir, "metadata_filtered_maf_samples.tsv"))

  # Read in original strelka file with maftools
  strelka2 <- maftools::read.maf(file.path(data_dir, "pbta-snv-strelka2.vep.maf.gz"),
    clinicalData = metadata
  )

  # Save to RDS so we don't have to run this again
  saveRDS(strelka2, strelka2_dir)
}
```

Make our mutation data into a GRanges object. 

```{r}
mut_ranges <- GenomicRanges::GRanges(
  seqnames = mut_df$Chromosome,
  ranges = IRanges::IRanges(start = mut_df$Start_Position, 
                            end = mut_df$End_Position),
  strand = mut_df$Strand, 
  mcols = mut_df)
```

Set up annotation object if it has not been set up previously. 

```{r}
if (!file.exists(file.path(scratch_dir, "hg38_genomic_region_annotation.rds"))) {
  # Here we are specifying what types of annotation we would like
  annots <- c('hg38_basicgenes', 
              'hg38_genes_intergenic',
              'hg38_genes_intronexonboundaries',
              'hg38_genes_exonintronboundaries',
              'hg38_genes_5UTRs',
              'hg38_genes_exons',
              'hg38_genes_introns',
              'hg38_genes_3UTRs')

  # Build all the annotations into GRanges object
  annotations <- annotatr::build_annotations(genome = 'hg38', annotations = annots)

  # Write this object so we don't have to write it again
  readr::write_rds(annotations, 
                   file.path(scratch_dir, "hg38_genomic_region_annotation.rds"), 
                   compress = "gz")
} else {
  # If annotation object is already made, read it in!
  annotations <- readr::read_rds(file.path(scratch_dir, 
                                           "hg38_genomic_region_annotation.rds"))
}
```

Make a new data frame that contains the old information as well as the information
regarding what types of genomic region the mutation is in. 

```{r}
# Intersect the regions we read in with the annotations
mut_matches <- GenomicRanges::findOverlaps(annotations, mut_ranges)

# Extract the annotation from the regions that overlap the mut
mut_annot <- data.frame(
  annotations[mut_matches@from]@elementMetadata@listData, 
  mut_ranges[mut_matches@to]@elementMetadata@listData, 
  stringsAsFactors = FALSE) %>%
  dplyr::rename_at(dplyr::vars(dplyr::starts_with("mcols.")), substr, 7, 10000) %>%
  dplyr::mutate("type" = gsub("^hg38_genes_", "", type),
                # Calculate the variant allele frequency
                vaf = as.numeric(t_alt_count) / (as.numeric(t_ref_count) +
                                                   as.numeric(t_alt_count)),
                # Create a base_change variable
                base_change = paste0(Reference_Allele, ">", Allele))
```

Some sanity checks for how much these data agree.

```{r}
# Turn NA's into none so we can differentiate when both agree on there 
# being no gene name
mut_annot$Hugo_Symbol[which(is.na(mut_annot$Hugo_Symbol))] <- "none"
mut_annot$symbol[which(is.na(mut_annot$symbol))] <- "none"

# Let's match these
matching <- as.character(mut_annot$Hugo_Symbol) == as.character(mut_annot$symbol)

# Print out the ratio of them matching 
sum(matching)/length(matching)
```


```{r}
  # Calculate sum totals for each data group
mut_annot %>%
  dplyr::group_by(type, Variant_Classification) %>%
  dplyr::summarise(count = dplyr::n()) %>% 
  ggplot2::ggplot(ggplot2::aes(x = type, y = count, fill = Variant_Classification)) + 
  ggplot2::geom_bar(position = "dodge", stat = "identity") +
  ggplot2::theme_classic() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 1)) 
```

Calculate Tumor Mutation Burden and get the metadata. 

```{r}
tmb <- mut_df %>%
  dplyr::group_by(Tumor_Sample_Barcode) %>%
  dplyr::summarise(count = dplyr::n(), 
                   tmb_per_Mb = dplyr::n()/30) %>%
  #dplyr::filter(tmb_per_Mb < 5) %>%
  dplyr::inner_join(metadata, by = "Tumor_Sample_Barcode") %>% 
  dplyr::filter(experimental_strategy != "Panel") 

summary(as.factor(tmb$experimental_strategy))
```

Let's do some checks with the metadata 

```{r}
tmb %>% 
  ggplot2::ggplot(ggplot2::aes(x = experimental_strategy, y = count)) + 
  ggplot2::geom_violin(ggplot2::aes(color = experimental_strategy)) + 
  ggplot2::theme_classic()
```

```{r}
tmb %>% 
  ggplot2::ggplot(ggplot2::aes(x = short_histology, y = count)) + 
  ggplot2::geom_jitter(ggplot2::aes(color = short_histology)) + 
  ggplot2::theme_classic() + 
  ggplot2::ylab("Number of SNV Coding mutations") +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 1), 
                 legend.position = "none") 
```

## Session Info

```{r}
sessionInfo()
```
