---
title: "Tumor Mutation Burden Comparison with TCGA data"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---

#### Purpose

Plot the consensus tumor mutation burden statistics for PBTA in comparison to TCGA brain-related data. 

#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/tmb-compare-tcga/compare-tmb.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

```{r}
if (!("TCGAbiolinks" %in% installed.packages())) {
  install.packages("TCGAbiolinks")
}

# magrittr pipe
`%>%` <- dplyr::`%>%`

# Load in these functions so we can use `maf_to_granges`
source(file.path("..", "snv-callers", "util", "wrangle_functions.R"))
```

Declare names of input and output directories.

```{r}
data_dir <- file.path("..", "..", "data")
results_dir <- "results"
plots_dir <- "plots"
```

Create output directories for this analysis. 

```{r}
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

Custom function for plotting the TMB. 

```{r}
sina_plot <- function(data, colour = NULL, title) {
  # Given a data.frame with `short_histology` and `tmb` information, make a sina plot from it
  #
  # Args:
  #
  # data : a data.frame with `short_histology` and `tmb` as columns
  # colour: the color you would like the points to be
  # title: The title for the plot. To be passed to `ggtitle`
  #
  # Returns:
  #   A sina plot by histology group
  data %>%
    ggplot2::ggplot(ggplot2::aes(x = reorder(short_histology, tmb, mean), y = log10(tmb))) +
    ggforce::geom_sina(color = colour, alpha = 0.3) +
    ggplot2::stat_summary(
      fun.y = mean, fun.ymin = mean, fun.ymax = mean,
      geom = "crossbar",
      width = 0.95, size = 0.15
    ) +
    ggplot2::theme_classic() +
    ggplot2::ylab("log10 (Number of Mutations per Mb)") +
    ggplot2::xlab("") +
    ggplot2::theme(
      axis.text.x = ggplot2::element_text(angle = 60, hjust = 1),
      legend.position = "none"
    ) +
    ggplot2::ggtitle(title) + 
    ggplot2::theme(
    panel.grid.major.x = ggplot2::element_line(colour = c("gray93", "white"), size = 9)
)
}
```

## Set up consensus SNV PBTA data

Download consensus SNV files. 

```{r}
# Declare file path for consensus file
consensus_file <- file.path(data_dir, "snv-consensus", "consensus_mutation_tmb.tsv")

# If this file hasn't been downloaded, download and unzip it.
if (!file.exists(consensus_file)) {
  # Download the file
  download.file("https://open-pbta.s3.amazonaws.com/data/snv-consensus/snv-consensus_1112019.zip",
    destfile = file.path(data_dir, "snv-consensus_1112019.zip")
  )

  # Get the files in this zip file
  unzip(file.path(data_dir, "snv-consensus_1112019.zip"),
    exdir = data_dir
  )
}
```

Read in the consensus TMB file. 

```{r include=FALSE}
# Read in the file
tmb_pbta <- data.table::fread(consensus_file)
```

Set up the WGS BED region data for Strelka2 for applying to TCGA data for calculating TMB. 
Here we are using Strelka2 because the consensus mutation data uses it. 

```{r}
wgs_bed_file <- file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed")

# Set up BED region files for TMB calculations
wgs_bed <- readr::read_tsv(wgs_bed_file, col_names = FALSE) %>%
  dplyr::rename(Chromosome = X1, Start_Position = X2, End_Position = X3)

# Calculate size of genome surveyed
wgs_genome_size <- sum(wgs_bed[, 3] - wgs_bed[, 2])

# Turn the bed regions into a GRanges object
wgs_bed_granges <- maf_to_granges(wgs_bed)
```

## Set up the TCGA data. 

We will only use the brain related TCGA projects, so we are specifying the project abbreviations here. 

```{r}
tcga_datasets <- c("LGG", "GBM", "PCPG")
```

Download and read each of these TCGA project's MAF files using `TCGAbiolinks`. 

```{r}
maf <- lapply(tcga_datasets, function(dataset) {
  TCGAbiolinks::GDCquery_Maf(dataset, pipelines = "mutect") %>%
    dplyr::select(
      "Chromosome", "Start_Position", "End_Position",
      "Tumor_Sample_Barcode", "Strand"
    )
})
# Bind data together
maf_df <- dplyr::bind_rows(maf)
```

Calculate overlap with these BED ranges and the TCGA mutations. 

```{r}
maf_granges <- maf_to_granges(maf_df)

# Find the overlap of the BED regions and the mutations This outputs a
# special GenomicRanges object that contains indices of each of these
# ranges that overlap
overlap <- GenomicRanges::findOverlaps(
  maf_granges,
  wgs_bed_granges,
  maxgap = 0
)
```

Report the ratio of mutations that are within these BED windows. 

```{r}
# Calculate of ratio of TCGA variants in this Strelka2 BED using the @from slot which
# indicates the indices of the ranges
length(overlap@from) / nrow(maf_df)
```
All of these mutations are within the Strelka2 windows, so we don't need to filter the TCGA data. 

## Calculate TCGA tumor mutation burden

```{r}
# Calculate the TMB for TCGA samples
tmb_tcga <- maf_df %>%
  dplyr::group_by(Tumor_Sample_Barcode) %>%

  # Count number of mutations for that sample
  dplyr::summarize(mutation_count = dplyr::n()) %>%

  # Calculate TMB
  dplyr::mutate(tmb = mutation_count / (wgs_genome_size / 1000000))
```

Get the TCGA metadata for these projects and extract histology groups. 

```{r}
clinical <- lapply(tcga_datasets, function(dataset) {
  TCGAbiolinks::GDCquery_clinic(project = paste0("TCGA-", dataset), type = "clinical") %>%
    dplyr::select(primary_diagnosis, bcr_patient_barcode)
})
# Bind the data together
clinical <- dplyr::bind_rows(clinical)
```

Set up TCGA TMB data with its histology. 

```{r}
# Get the patient barcode from the Tumor Sample Barcode category
tmb_tcga <- tmb_tcga %>%
  dplyr::mutate(bcr_patient_barcode = substr(Tumor_Sample_Barcode, 0, 12)) %>%

  # Add on the histology group
  dplyr::left_join(clinical, by = "bcr_patient_barcode") %>%

  # Rename so it matches PBTA data
  dplyr::rename(short_histology = primary_diagnosis) %>%

  # If there is no information for this, for TCGA, we will get rid of these rows
  dplyr::filter(!is.na(short_histology)) %>%
  readr::write_tsv(file.path(results_dir, "brain_related_tcga_tmb.tsv"))
```

## Plot the TMB as a sina plot

Plot the TCGA data. 
```{r}
# Make TCGA plot and get rid of y axis
tcga_plot <- sina_plot(tmb_tcga, colour = "#3BC8A2", title = "TCGA (Adult)") +
  ggplot2::theme(
    axis.title.y = ggplot2::element_blank(),
    axis.text.y = ggplot2::element_blank(),
    axis.ticks.y = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(size = 7),
    panel.grid.major.x = ggplot2::element_line(colour = c("gray93", "white"), size = 9)
  )
```

Plot the PBTA data. 

```{r}
# Make the plot for PBTA
pbta_plot <- sina_plot(tmb_pbta, colour = "#630882", title = "PBTA")
```

Use cowplot to put them together. 

```{r}
# Put the plots together
tmb_plot <- cowplot::plot_grid(pbta_plot, tcga_plot,
  align = "v",
  axis = "left",
  rel_widths = c(2.5, 1),
  label_size = 12
)
# Print out tmb plot
tmb_plot
```

Save this final plot to a png file.

```{r}
# Save the plot to a png
cowplot::save_plot(file.path(plots_dir, "tmb_tcga_and_pbta_plot.png"), plot = tmb_plot)
```

## Session Info

```{r}
sessionInfo()
```
