---
title: "Subtyping chordoma"
output: html_notebook
author: Mateusz Koptyra 
date: 20191121
---

This notebook prepares _SMARCB1_ copy number and expression data for chordoma samples for the purpose of identifying poorly-differentiated chordoma samples, which are characterized by loss of _SMARCB1_.

## Set up

```{r}
library(dplyr)
library(readr)
library(ggplot2)
```

### Read in data

```{r}
# File path to subsetted files directory
input_dir <- "chordoma-subset"
```

Set up metadata

```{r}
# Read in non-subsetted metadata
histologies_df <- readr::read_tsv(file.path("..", "..", "data", "pbta-histologies-base.tsv"), guess_max = 10000)

# Subset metadata
subset_metadata <- histologies_df %>%
  dplyr::filter(short_histology == "Chordoma") %>%
  select(
    Kids_First_Biospecimen_ID, sample_id, Kids_First_Participant_ID,
    experimental_strategy,tumor_ploidy
  ) %>%
  rename(ploidy=tumor_ploidy)
```

```{r}
subset_focal_cn_df <-
  data.table::fread(file.path(input_dir, "chordoma-only_cn_autosomes.tsv.gz"))
```

```{r}
# we need to include the sample_id field from pbta-histologies-base.tsv in the final
# table (field will allow #us to map between RNA-seq (e.g., SMARCB1 expression
# values) and WGS data (e.g., SMARCB1 focal copy number status) from the same
# event for a given individual). To get the SMARCB1 jitter plot in the photo
# here #250 (comment), you will first need to read in the collapsed expression
# data
subset_expression_data <-
  read_rds(file.path(
    input_dir,
    "chordoma-only-gene-expression-rsem-fpkm-collapsed.stranded.rds"
  ))
```

### Output

```{r}
# scatterplot
plot_dir <- "plots"
if (!dir.exists(plot_dir)) {
  dir.create(plot_dir)
}
plot_file <- file.path(plot_dir, "smarcb1_expression_copy_status.png")

# tabular data
results_dir <- "results"
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
output_file <- file.path(results_dir, "chordoma_smarcb1_status.tsv")
```

## Prepare the data

Extracting the chordoma samples that have a loss of SMARCB1 from `subset_focal_cn_df`

```{r}
chordoma_loss <- subset_focal_cn_df %>%
  filter(
    gene_symbol == "SMARCB1",
    status == "loss"
  )
chordoma_loss
```

Distinguishing the chordoma samples with no copy number change chromosome 22
```{r}
copy_neutral_df <- subset_metadata %>%
  # the copy events can only be taken from WGS data not RNA-seq data
  # we also only want biospecimens where a loss was not recorded to avoid duplicates
  filter(
    experimental_strategy == "WGS",
    !(Kids_First_Biospecimen_ID %in% chordoma_loss$biospecimen_id)
  ) %>%
  # if there's no loss, let's assume status is copy neutral
  mutate(status = "neutral",
         copy_number = ploidy) %>%
  # let's get the columns to match chordoma_loss
  dplyr::rename(biospecimen_id = Kids_First_Biospecimen_ID) %>%
  select(biospecimen_id, status, copy_number,ploidy)
copy_neutral_df
```

Joining chordoma samples (both with chr 22 loss and no copy number change)
```{r}
chordoma_copy <- chordoma_loss %>%
  # join the losses with the neutrals to get a new data frame
  select(biospecimen_id, status,ploidy,copy_number) %>%
  bind_rows(copy_neutral_df)
chordoma_copy
```

Need to get the sample_id that corresponds to biospecimen_id into chordoma_copy so we can match WGS and RNA-seq biospecimens from the same event/sample:
```{r}
chordoma_copy <- chordoma_copy %>%
  # get only the Kids_First_Biospecimen_ID, sample_id columns from our identifier data.frame
  # then use biospecimen IDs to add the sample_id info
  inner_join(select(
    subset_metadata,
    Kids_First_Biospecimen_ID,
    sample_id
  ),
  by = c("biospecimen_id" = "Kids_First_Biospecimen_ID")
  )
chordoma_copy
```

Look at SMARCB1 expression values only in chordoma

```{r}
# get the row that contains the SMARCB1 values
# gene symbols are rownames
smarcb1_expression <- subset_expression_data[which(rownames(subset_expression_data) == "SMARCB1"), ]
```

```{r}
# now only the columns correspond to chordoma samples
smarcb1_expression <- smarcb1_expression[, which(colnames(subset_expression_data) %in% subset_metadata$Kids_First_Biospecimen_ID) ]

# Print out the expression for SMARCB1
smarcb1_expression
```

The `smarcb1_expression` is a not a friendly form ^^; Transposing needed: 

```{r}
# transpose such that samples are rows
smarcb1_expression <- t(smarcb1_expression) %>%
  # make a data.frame
  as.data.frame() %>%
  # we want the rownames that are biospecimen identifers as their own column called Kids_First_Biospecimen_ID
  tibble::rownames_to_column("Kids_First_Biospecimen_ID") %>%
  # give SMARCB1 column a slightly better column name
  dplyr::rename(SMARCB1_expression = SMARCB1)
smarcb1_expression
```

This also needs sample_id to add it in
```{r}
smarcb1_expression <- smarcb1_expression %>%
  inner_join(select(
    subset_metadata,
    Kids_First_Biospecimen_ID,
    sample_id
  ),
  by = "Kids_First_Biospecimen_ID"
  )
smarcb1_expression
```

Joining the copy number data with the expression data in this step
```{r}
chordoma_smarcb1_df <- smarcb1_expression %>%
  # any missing samples will get filled with NA when using a full join
  full_join(chordoma_copy, by = "sample_id") %>%
  dplyr::rename(
    Kids_First_Biospecimen_ID_DNA = biospecimen_id,
    Kids_First_Biospecimen_ID_RNA = Kids_First_Biospecimen_ID
  )

# this step adds in the participant identifier (sample_id to match between the two data.frame)
chordoma_smarcb1_df <- subset_metadata %>%
  select(sample_id, Kids_First_Participant_ID) %>%
  distinct() %>%
  inner_join(chordoma_smarcb1_df,
    by = "sample_id"
  )

chordoma_smarcb1_df
```

```{r}
chordoma_smarcb1_df <- chordoma_smarcb1_df %>%
  select(
    Kids_First_Participant_ID,
    Kids_First_Biospecimen_ID_DNA,
    Kids_First_Biospecimen_ID_RNA,
    sample_id,
    status,
    ploidy,
    copy_number,
    SMARCB1_expression
  ) %>%
  # 'status' is replaced a more descriptive name
  dplyr::rename(focal_SMARCB1_status = status,
                SMARCB1_copy_number = copy_number)
chordoma_smarcb1_df
```

### Plot _SMARCB1_ expression

copy loss vs. copy neutral 

```{r}
# this specifies that this is the data we want to plot
chordoma_smarcb1_df %>%
  # drop the sample that doesn't have WGS data
  tidyr::drop_na() %>%
  # this step specifies what should go on the x- and y-axes
  ggplot(aes(
    x = focal_SMARCB1_status,
    y = SMARCB1_expression
  )) +
  # we want a jitter plot where the points aren't too far
  # apart that's what width does
  geom_jitter(width = 0.1) +
  # this is plotting the median as a blue diamond
  stat_summary(
    fun.y = "median",
    geom = "point",
    size = 3,
    color = "blue",
    shape = 18
  ) +
  xlab("SMARCB1 status") +
  xlab("SMARCB1 status") +
  ylab("SMARCB1 expression")
```

```{r}
ggsave(filename = plot_file)
```

Write the table to file.

```{r}
chordoma_smarcb1_df %>%
  arrange(Kids_First_Participant_ID, sample_id) %>%
  write_tsv(output_file)
```

### Session Info

```{r}
sessionInfo()
```
