---
title: "Evaluate concordance between Mutect2 and Strelka2"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

Candace Savonen - CCDL for ALSF

This notebook analyzes the overlap and distinctions between between MuTect2 and 
Strelka2 results after having set the data up in 01-set-up.Rmd.
It addresses [issue \# 30 in OpenPBTA](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/30).

### Summary of Findings:

Due to some of the findings here with Mutect2 data, I suggest we should move 
forward with only Strelka2 data *OR* move forward with the variants that are 
detected by both algorithms. 

*Notable Findings:*  

- MuTect2 and Strelka2 detect [55,808 of the same variants](#venn-diagrams). 
This as defined by having the same Hugo gene symbol, base change, chromosomal 
start site, and sample of origin (see notebook [01-set-up.Rmd](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/mutect2-vs-strelka2/01-set-up.Rmd) 
for more details).
If moving forward we want only the most reliably called variants, this set of 
55,808 variants would give us plenty to work with. 

- MuTect2 and Strelka2 highly agree in their [Variant allele frequency (VAF) calculations](#variant-allele-frequency-scatterplot). 
This is good regardless of our choices moving forward. 

- Variants detected only by MuTect2 have a particularly low VAF compared to
variants detected only by Strelka2. (See [VAF Violin plots](#vaf-violin-plots)). 
These density plots suggest some of these MuTect2 calls may be noise. 
Although these low-VAF of MuTect2 could be identifying true variants, our further
analyses would probably benefit from a more robust, higher confidence set of 
variants. 

- MuTect2 also registers dinucleotide and larger variants where Strelka2 seems 
to break these variants into their single nucleotide changes.
In these analyses, these base changes have been grouped together and collectively
called [`long_changes`](#characterizing-the-variants-identified). 
The higher base resolution of Strelka2, and its ability to parse apart the SNVs
from each other, is more useful to us for this particular analyses, as the larger 
structural variants are better detected in the Manta or LUMPY analyses.

### Outline of analyses completed:

- [Venn diagram of variants detected](#venn-diagrams)  
- [VAF correlation hex plot](#variant-allele-frequency-scatterplot)   
- [Violin plot of VAF from MuTect2 or Strelka2 when they were detected by one or both algorithms](#vaf-violin-plots)  
- [Characterization of variants detected by one or both algorithms](#characterizing-the-variants-identified)  
  - [Number of base changes detected](#number-of-variants-categorized-by-their-base-change)  
  - [Percent of base changes detected](#percent-of-variants-by-their-base-change)  
  - [Percents of variants detected in coding versus non-coding regions](#percent-of-variants-that-are-in-coding-or-non-coding-regions)  
  - [Percent of variant type detected](#percent-of-variant-type)  

## Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/mutect2-vs-strelka2/02-analyze-concordance.Rmd', 
                              clean = TRUE)"
```
_This assumes you are in the top directory of the repository._
 
## Set Up

```{r}
# Will need hexbin for the hex plot
if (!("hexbin" %in% installed.packages())) {
  install.packages("hexbin")
}

# Will use colorblindr to make the plots colorblind-friendly
if (!("colorblindr" %in% installed.packages())) {
  devtools::install_github("clauswilke/colorblindr")
}

# Will need R.utils for zipping up the results file
if (!("R.utils" %in% installed.packages())) {
  install.packages("R.utils")
}

# Will need VennDiagrams
if (!("VennDiagrams" %in% installed.packages())) {
  install.packages("VennDiagrams")
}

# Get magrittr pipe
`%>%` <- dplyr::`%>%`
```

Declare names of output directories.

```{r}
results_dir <- "results"
plots_dir <- "plots"
```

Create output directories in this analysis folder if they weren't made 

```{r}
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

Make a function that turns a selected variable from the original MAF file into 
a long format that is ready for downstream purposes. 

```{r}
make_long_format <- function(df, variable_name = NULL, 
                             other_variables = "mutation_id") {
  # Reconfigure data.frame into long format with a column that indicates which 
  # algorithm or if both contain a particular variant
  # Args:
  #   df: vaf_df data.frame that contains the combined data for each algorithm
  #       with each algorithm having it's own column e.g. vaf.strelka2 and vaf.mutect2
  #   variable_name: a character string that indicates the base variable name
  #             from the original MAF file that you would like to make into
  #             long format
  #   other_variables: a character string or vector of characters that indicates
  #                    other columns you would like to be kept. Default is to 
  #                    keep the mutation_id.
  long_df <- df %>%
    dplyr::select(other_variables,
      paste0(variable_name, c(".strelka2", ".mutect2")),
      dataset
    ) %>%
    tidyr::gather(
      key = "data", value = "variable",
      paste0(variable_name, c(".strelka2", ".mutect2"))
    ) %>%
    dplyr::mutate("data" = gsub(paste0(variable_name, "."), "", data)) %>%
    # Filter out NAs that are only there because the dataset doesn't have 
    # that mutation. Keep NAs that are what the original label was.
    dplyr::filter(!grepl("strelka2_only mutect2|mutect2_only strelka2", 
                         paste(dataset, data)))
  return(long_df)
}
```

Barplot wrapper function.

```{r}
barplot_maf_var <- function(df, variable_name = NULL, filter_cutoff = 0,
                            as_percent = TRUE, omit.na = FALSE) {
  # Creates a ggplot barplot for the numbers of variants that fall into each
  # category of the variable selected.
  #
  # Args:
  #   df: combined data.frame that contains the MAF data for both algorithms. 
  #       (noted here as vaf_df)
  #   variable_name: a character string that indicates the base variable name
  #                  from the original MAF file that is indicated in "variable"
  #                  column. 
  #   filter_cutoff: variable categories with less than this cutoff will be removed
  #   as_percent: If true, will report numbers as percentages of the group they
  #               are from
  #   omit.na: If true, will remove NA's as a category
  #
  # Returns:
  #   ggplot2 grouped barplot with mutect_only, strelka2_only, and both as groups
  #   
  # Reconfigure data.frame into long format with a column that indicates which 
  # algorithm or if both contain a particular variant
  long_df <- make_long_format(df, variable_name = variable_name)
  
  # Summarize by the number of time each category appears by the data group
  # i.e. (`mutect2_only`, `strelka2_only`, or `both`).
  count_df <- long_df %>%
    dplyr::group_by(dataset, data, variable) %>%
    dplyr::summarise(count = dplyr::n())

  # Calculate sum totals for each data group
  totals <- count_df %>%
    dplyr::group_by(dataset) %>%
    dplyr::summarise(total = sum(count, na.rm = TRUE))

  # Calculate percentages based on totals and apply filter_cutoff
  count_df <- count_df %>%
    as.data.frame() %>%
    dplyr::mutate("percent" = count_df$count / totals$total[match(
      count_df$dataset,
      totals$dataset
    )]) %>%
    dplyr::mutate(report = ifelse(rep(as_percent, nrow(count_df)), percent, count)) %>%
    dplyr::filter(report > filter_cutoff)
  
  # Omit NA category if that was specified
  if (omit.na) {
    count_df <- dplyr::filter(count_df, !is.na(variable))
  }
  # Plot this!
  ggplot2::ggplot(count_df, ggplot2::aes(
    x = reorder(variable, -report),
    y = report, fill = dataset
  )) +
    ggplot2::geom_bar(position = "dodge", stat = "identity") +
    ggplot2::theme_classic() +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 1)) +
    colorblindr::scale_fill_OkabeIto() +
    ggplot2::xlab("") +
    ggplot2::ylab(ifelse(as_percent, "Percent of Data Group", "Count"))
}
```

Read in the combined dataset that was set up in `01-set-up.Rmd`, 
'combined_results.tsv'.
This file has been zipped so it can be shared on GitHub more easily. 
If no unzipped version exists, we will unzip this.

```{r}
if (!file.exists(file.path(results_dir, "combined_results.tsv"))) {
  unzip(file.path(results_dir, "combined_results.tsv.zip"))
}
```

Read in the combined dataset, 'combined_results.tsv.'

```{r}
vaf_df <- readr::read_tsv(file.path(results_dir, "combined_results.tsv"))
```

## Venn Diagrams

How many variants do both MuTect2 and Strelka2 agree on? Here we define a variant
as being identical in both datasets if its Hugo gene symbol, base change, 
chromosomal start site, and sample of origin. 
The variable `mutation_id` contains this information for each sample in a 
concatenated form, e.g. "ZMYND19_T>G_137590214_BS_5ZRZC3ZM".
It was previously set up in [01-set-up.Rmd](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/mutect2-vs-strelka2/01-set-up.Rmd).

```{r}
# Get the total count for variants found by each algorithm or by both.
count <- summary(as.factor(vaf_df$dataset))
```

Use the count object from above to make a Venn Diagram

```{r}
# Make the Venn diagram
grid::grid.newpage()
venn.plot <- VennDiagram::draw.pairwise.venn(
  area1 = count[3] + count[1],
  area2 = count[2] + count[1],
  cross.area = count[1],
  category = c("Strelka2", "MuTect2"),
  fill = c("blue", "yellow"),
  cex = 2,
  cat.cex = 1.5,
  cat.dist = c(-0.04, -0.031),
  ext.pos = 0,
  ext.dist = -0.01,
  ext.length = .8,
  ext.line.lwd = 2,
  ext.line.lty = "dashed"
)
grid::grid.draw(venn.plot) # Draw plot
```

Save the Venn diagram plot to a png. 

```{r}
# Make filename to save plot as
venn.plot.file <- file.path(
  plots_dir,
  "strelka2_mutect2_venn_diagram.png"
)
png(venn.plot.file)
grid::grid.draw(venn.plot)
dev.off()
```

## Variant allele frequency scatterplot

For variants identified in both datasets, are the VAF's similar? 
The VAFs for each mutation in each dataset were calculated by 
`t_alt_count) / (t_ref_count + t_alt_count) `, as following the [code used in 
`maftools`](https://github.com/PoisonAlien/maftools/blob/master/R/plot_vaf.R)
(See [01-set-up.Rmd](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/mutect2-vs-strelka2/01-set-up.Rmd)).

```{r}
# Calculate Pearson's correlation
pearson_r <- cor.test(vaf_df$vaf.strelka2, vaf_df$vaf.mutect2)

# Plot these data as a hex plot
vaf_df %>%
  ggplot2::ggplot(ggplot2::aes(x = vaf.strelka2, y = vaf.mutect2)) +
  ggplot2::geom_hex() +
  ggplot2::theme_classic() +
  ggplot2::xlab("VAF for each mutation for Strelka2") +
  ggplot2::ylab("VAF for each mutation for MuTect2") +
  ggplot2::ggtitle(
    label = "",
    subtitle = paste("R = ", round(pearson_r$estimate, 3))
  )

# Save this plot to a png
ggplot2::ggsave(file.path(plots_dir, "VAF_scatterplot.png"))
```

Overall, the VAF are highly correlated between the two algorithms.

## VAF Violin plots

How do the VAF distributions of MuTect2, Strelka2 look for variants found by
both algorithms, versus variants found only by one algorithm? 

```{r}
# Set up the VAF data for the density plot
vaf_df %>%
  tidyr::gather(key = "data", value = "vaf", vaf.strelka2, vaf.mutect2) %>%
  dplyr::mutate(data = gsub("^vaf.", "", data)) %>%
  dplyr::mutate(data.group = paste0(dataset, ":", data, " VAF")) %>%
  dplyr::filter(!is.na(vaf)) %>%
  # Plot it
  ggplot2::ggplot(ggplot2::aes(data.group, vaf)) +
  ggplot2::geom_violin(fill = "light blue") +
  ggplot2::theme_classic() +
  ggplot2::ylab("Density of VAF") +
  ggplot2::xlab(" ")

# Save this plot to a png
ggplot2::ggsave(file.path(plots_dir, "VAF_violin_plots.png"))
```

Variants that are only detected by MuTect2 appear to be particularly low VAF, 
even compared to variants only detected by Strelka2.
This may indicate it's picking up noise. 

## Characterizing the variants identified

What types of variants are are the most discrepant between the algorithms as 
compared to variants that are detected by both consistently?
We will make a series of grouped bar plots to try to get an overall picture of 
these variants.
I made the 

### Number of variants categorized by their base change

Note that `long_change` category is a combination of all base changes that are
more than a single nucleotide (See [01-set-up.Rmd](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/mutect2-vs-strelka2/01-set-up.Rmd)
for the code on how this was set up). 
Note that only MuTect2 reports these data as multiple base changes as opposed to 
parsing these changes a part as single base changes at a time.
Due to the lower base resolution lack of discernment of MuTect2 in this regard, 
this is another reason avoid using MuTect2 calls alone. 

```{r}
barplot_maf_var(
  df = vaf_df, variable_name = "change",
  as_percent = FALSE, omit.na = FALSE, filter_cutoff = 0
)
```

### Percent of variants by their base change

This is the same plot as above except normalized to be the percent of each data
group (`mutect2_only`, `strelka2_only`, or `both`).

MuTect2 shows a disproportional amount of deletions. 
It appears C>T and G>A changes are easiest for both algorithms to detect. 

```{r}
barplot_maf_var(
  df = vaf_df, variable_name = "change", filter_cutoff = 0,
  as_percent = TRUE, omit.na = FALSE
)
```

### Percent of variants that are in coding or non-coding regions

It's not surprising that these are mostly coding variants detected, however, 
there doesn't appear to be any strong biases between the groups. 

```{r}
barplot_maf_var(
  df = vaf_df, variable_name = "coding", filter_cutoff = 0,
  as_percent = TRUE, omit.na = FALSE
)
```

### Percent of variant type 

Only MuTect2 appears to record changes as DNP's or TNP's. 
Again we see MuTect2 shows more deletions. 

```{r}
barplot_maf_var(
  df = vaf_df, variable_name = "Variant_Type", filter_cutoff = 0,
  as_percent = TRUE, omit.na = FALSE
)
```

### Percent of estimated clinical effect

```{r}
barplot_maf_var(
  df = vaf_df, variable_name = "PolyPhen_category", filter_cutoff = 0,
  as_percent = TRUE, omit.na = FALSE
)
```

### Number of estimated clinical effect

```{r}
barplot_maf_var(
  df = vaf_df, variable_name = "PolyPhen_category", filter_cutoff = 0,
  as_percent = FALSE, omit.na = FALSE
)
```

## Write possibly/probably damaging variants to files for further analysis

```{r}
# Get a data.frame of only the damaging variants and which algorithms detected
# them
damaging_variants <- vaf_df %>%
  dplyr::filter(grepl("possibly_damaging|probably_damaging", 
                      paste0(PolyPhen_category.strelka2, PolyPhen_category.mutect2))) %>%
  make_long_format(variable_name = "PolyPhen_category") 
```

Make mutation id lists for each data group from the above data.frame. 

```{r}
strelka2_only_ids <- damaging_variants %>%
  dplyr::filter(dataset == "strelka2_only") %>%
  dplyr::pull(mutation_id)
  
mutect2_only_ids <- damaging_variants %>%
  dplyr::filter(dataset == "mutect2_only") %>%
  dplyr::pull(mutation_id)

both_detected_ids <- damaging_variants %>%
  dplyr::filter(dataset == "both") %>%
  dplyr::pull(mutation_id)
```

We will collect the sample ID, chromosomal coordinates and base change for each
mutation in we identified in the lists above. 

```{r}
# Here are the variables we will need
var_select <- c("Tumor_Sample_Barcode", "Chromosome", "Start_Position", 
                "End_Position", "base_change")
```

Get rid of dataset specific suffix (e.g. ".strelka2"), then write this to a TSV 
file. 

```{r}
strelka2_only <- vaf_df %>%
  # Keep only the mutation_id's identifed as only in strelka2
  dplyr::filter(mutation_id %in% strelka2_only_ids) %>%
  # We want the strelka2 versions of these variables
  dplyr::select(paste0(var_select, ".strelka2")) %>% 
  # But now we need to drop the suffix
  dplyr::rename_all(dplyr::funs(stringr::str_replace(., ".strelka2", ""))) %>%
  # Write to a TSV
  readr::write_tsv(file.path(results_dir, "variants_strelka2_only.tsv"))
```

Do same for MuTect2.

```{r}
mutect2_only <- vaf_df %>%
  # Keep only the mutation_id's identifed as only in mutect2
  dplyr::filter(mutation_id %in% mutect2_only_ids) %>%
  # We want the mutect2 versions of these variables
  dplyr::select(paste0(var_select, ".mutect2")) %>% 
  # But now we need to drop the suffix
  dplyr::rename_all(dplyr::funs(stringr::str_replace(., ".mutect2", ""))) %>%
  # Write to a TSV
  readr::write_tsv(file.path(results_dir, "variants_mutect2_only.tsv"))
```

Write the same information to a file for variants identified by both algorithms. 
Note that it doesn't matter whether we take Strelka2 or MuTect2's variables for
the coordinates and tumor sample ID; they are identical for these fields. 
To ensure this, I tested this by isolating using ".strelka2" and ".mutect2" 
variables and used `all.equal` and saw they were identical. 

```{r}
both_detected <- vaf_df %>%
  dplyr::filter(mutation_id %in% both_detected_ids) %>%
  # It doesn't matter whether we take strelka or mutect variables here
  dplyr::select(paste0(var_select, ".mutect2")) %>% 
  # But now we need to drop the suffix
  dplyr::rename_all(dplyr::funs(stringr::str_replace(., ".mutect2", ""))) %>%
  readr::write_tsv(file.path(results_dir, "variants_both_detected.tsv"))
```

Session Info: 

```{r}
sessionInfo()
```
