---
title: "Gather CNV changes to subtype LGAT biospecimens"
author: "K S Gaonkar (D3B)"
output: html_notebook

---

In this notebook, we will look for the following CNV changes that define subtypes of LGAT
 - LGG, FGFR
 harbors FGFR1 TKD (tyrosine kinase domain tandem duplication)
 
 - LGG, CDKN2A/B
 harbors focal CDKN2A and/or CDKN2B deletion

### Setup
```{r}
library("tidyverse")
```

### Input
```{r}
# Look for git root folder
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

# get subset folder
subset_dir <- file.path(root_dir, 
                        "analyses",
                        "molecular-subtyping-LGAT",
                        "lgat-subset")

# create if doesn't exist
if (!dir.exists(subset_dir)) {
  dir.create(subset_dir)
}

# manta calls
manta <- read_tsv(file.path(root_dir,
                            "data",
                            "pbta-sv-manta.tsv.gz")) 

# consensus seg calls
cnv_focal_anno <- read_tsv(file.path(root_dir, 
                                        "data",
                                        "pbta-cnv-consensus.seg.gz"))

# Gene location and domain overlap file 
bioMartDataPfamFGFR1 <- 
  readRDS(system.file("extdata", "pfamDataBioMart.RDS", package = "annoFuse")) %>%
  # to check if biospecimen harbors FGFR1 TKD (tyrosine kinase domain tandem duplication) 
  dplyr::filter(hgnc_symbol == "FGFR1",
                DESC == "Protein tyrosine kinase")

bioMartDataPfamCDKN <- 
  readRDS(system.file("extdata", "pfamDataBioMart.RDS", package = "annoFuse")) %>%
  # to check if biospecimen harbors FGFR1 TKD (tyrosine kinase domain tandem duplication) 
  dplyr::filter(hgnc_symbol %in% c("CDKN2A,CDKN2B"))

```
### Gather exon locations
Code below is adapted from focal-cn-file-preparation/04-prepare-cn-file.R

```{r}
#### Annotation file -----------------------------------------------------------

# this is the output of GenomicFeatures::makeTxDbFromGFF
# TODO: possibly update this when the GTF file gets included in the data
#       download; may also remove the --gtf_file option and hardcode it?
annotation_directory <- file.path(root_dir,
                                  "analyses",
                                  "focal-cn-file-preparation",
                                  "annotation_files")
annotation_file <- file.path(annotation_directory,
                             "txdb_from_gencode.v27.gtf.db")

if (!file.exists(annotation_file)) {
  # Define the annotations for the hg38 genome
  txdb <- GenomicFeatures::makeTxDbFromGFF(
    file = file.path(root_dir,"data","gencode.v27.primary_assembly.annotation.gtf.gz"),
    format = "gtf"
  )
  # can do this even if the directory exists
  dir.create(annotation_directory, showWarnings = FALSE)
  # write this to file to save time next time
  AnnotationDbi::saveDb(txdb, annotation_file)
} else {
  txdb <- AnnotationDbi::loadDb(annotation_file)
}

# extract the exons but include ensembl gene identifiers
tx_exons <- GenomicFeatures::exons(txdb, columns = "gene_id")

```

### Gather lgat subset

```{r}
# File from 00-LGAT-select-pathology-dx that is used for the pathology diagnosis
# was used to subset clinical file
lgat_specimens_df<-read_tsv(file.path(subset_dir, "lgat_metadata.tsv"))

# Filter to RNA-Seq samples
lgat_wgs_df <- lgat_specimens_df %>%
  dplyr::filter(experimental_strategy == "WGS") %>%
  dplyr::select(Kids_First_Biospecimen_ID)

# Filter CNV calls
manta_lgat <- manta %>%
  dplyr::filter(Kids.First.Biospecimen.ID.Tumor %in% lgat_wgs_df$Kids_First_Biospecimen_ID)

cnv_focal_anno_lgat <- cnv_focal_anno %>%
  dplyr::filter(biospecimen_id %in% lgat_wgs_df$Kids_First_Biospecimen_ID)

```

### Overlap tandem duplication in FGFR1 with exon

```{r}
# Genomic range for manta file 
manta_gr <- manta_lgat %>%
  # format
  dplyr::rename(chr = SV.chrom, start = SV.start, end = SV.end,
                ALT = ALT) %>%
  # SV overlaps only FGFR1 gene
  # ie. breakpoint is within FGFR1 gene
  dplyr::filter(grepl("^FGFR1$",manta_lgat$Gene.name),
                # filter for tandem duplication
                grepl("<DUP:TANDEM>",ALT)) %>%
  # format with chr to match to cnv
  dplyr::mutate(chr=paste0("chr",chr)) %>%
  # make genomic ranges
  GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = TRUE,
                                          starts.in.df.are.0based = FALSE)

# Genomic range for cnv seg file 
cnv_gr <- consensus_seg %>%
  # format
  dplyr::rename(chr = chrom, start = loc.start, end = loc.end,
                copy_number = copy.num) %>%
  dplyr::select(-num.mark, -seg.mean) %>%
  # make genomic ranges
  GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = TRUE,
                                          starts.in.df.are.0based = FALSE)

# Genomic range for gene location and domain overlap file 
domain_gr <- bioMartDataPfamFGFR1 %>%
  # keep only domain info
  dplyr::filter(!is.na(domain_start),!is.na(domain_end)) %>%
  # format
  dplyr::mutate(strand = if_else(strand=="-1","-","+"),
                chromosome_name = paste0("chr",chromosome_name)) %>%
  dplyr::rename(chr = chromosome_name, start = domain_start, end = domain_end) %>%
  # make genomic ranges
  GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = TRUE,
                                          starts.in.df.are.0based = FALSE)

# overlap manta and domain
overlaps_manta <- IRanges::mergeByOverlaps(manta_gr, domain_gr)

# overlap consensus seg with domain
overlaps_cnv <- IRanges::mergeByOverlaps(cnv_gr,domain_gr)

```

### Distribution of manta SV
CNV that overlap "Protein tyrosine kinase" domain in biospecimens

```{r}
overlaps_manta_df <- data.frame(
  biospecimen_id = overlaps_manta$Kids.First.Biospecimen.ID.Tumor,
  ALT = overlaps_manta$ALT,
  hgnc_symbol = overlaps_manta$hgnc_symbol,
  NAME = overlaps_manta$NAME,
  stringsAsFactors = FALSE
) %>%
  dplyr::distinct() 

overlaps_manta_df 

```
4 biospecimens have tandem duplication that overlap tyrosine kinase domain in FGFR1


### Distribution of copy number
CNV that overlap "Protein tyrosine kinase" domain in biospecimens

```{r}
overlaps_cnv_df <- data.frame(
  biospecimen_id = overlaps_cnv$ID,
  copy_number = overlaps_cnv$copy_number,
  hgnc_symbol = overlaps_cnv$hgnc_symbol,
  NAME = overlaps_cnv$NAME,
  stringsAsFactors = FALSE
) %>%
  dplyr::distinct() %>%
  dplyr::filter(copy_number >2) %>%
  

overlaps_cnv_df 

```
24 samples have copy_number changes > 2 that overlap with tyrosine kinase domain







