---
title: "Compile LGAT subtyping"
output: html_notebook
---

In this notebook, we will be compiling the subtype annotation for LGAT samples. Subtypes are described in [#790](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/790) which were gathered through the following scripts : 

`01-subset-files-for-LGAT.Rmd` gathers subtype associated SNV annotation 
`02-subset-fusion-files-LGAT.Rmd` gathers subttype associated Fusion annotation
`03-subset-cnv-files-LGAT.Rmd` gathers subtype associated CNV annotation

along with the above molecular data based annotation in this notebook we will include NF-1 cancer_predispositions from histology file as a criteria for LGG, NF1 as well

### Setup
```{r}
library("tidyverse")

root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

# get subset folder
subset_dir <- file.path(
  root_dir,
  "analyses",
  "molecular-subtyping-LGAT",
  "lgat-subset"
)

# result Dir
resultDir <- file.path(
  root_dir,
  "analyses",
  "molecular-subtyping-LGAT",
  "results"
)

```

### Input

```{r}
# These will only be samples that have gone through the inclusion/exclusion
# criteria that are enforced in 01-subset-files-for-LGAT.Rmd
lgat_metadata_file <- file.path("lgat-subset", "lgat_metadata.tsv")
bsid_to_include <- read_tsv(lgat_metadata_file) %>%
  pull(Kids_First_Biospecimen_ID)

# File from 00-LGAT-select-pathology-dx that is used for the pathology diagnosis
# that will used to recode subtype
path_dx_list <- jsonlite::fromJSON(
  file.path(subset_dir, 
            "lgat_subtyping_path_dx_strings.json")
)

# clinical file
lgat_manifest <- read_tsv(file.path(root_dir, 
                                    "data",
                                    "pbta-histologies-base.tsv"), 
                          guess_max = 10000) %>%
  filter(Kids_First_Biospecimen_ID %in% bsid_to_include) %>%
  # select columns for interest
  dplyr::select(Kids_First_Biospecimen_ID,
         Kids_First_Participant_ID,
         sample_id,
         experimental_strategy,
         cancer_predispositions,
         pathology_diagnosis,
         pathology_free_text_diagnosis)

lgat_dna <- lgat_manifest %>%
  filter(experimental_strategy != "RNA-Seq") %>%
  dplyr::select( -experimental_strategy, -cancer_predispositions)

lgat_rna <- lgat_manifest %>%
  filter(experimental_strategy == "RNA-Seq") %>%
  dplyr::select( -experimental_strategy, -cancer_predispositions)

all_lgat_matched <- full_join(lgat_dna, lgat_rna,
                         by = c("Kids_First_Participant_ID",
                                "sample_id",
                                "pathology_free_text_diagnosis",
                                "pathology_diagnosis"),
                         suffix = c("_DNA", "_RNA")) %>%
  dplyr::select(starts_with("Kids_First"), sample_id,
                pathology_free_text_diagnosis,
                pathology_diagnosis)

# SNV based annotation
lgat_snv <- read_tsv(file.path(subset_dir,"LGAT_snv_subset.tsv")) 

# CNV based annotation
lgat_cnv <- read_tsv(file.path(subset_dir,"LGAT_cnv_subset.tsv")) %>%
  # To-do add to CNV subset script
  dplyr::rename(Kids_First_Biospecimen_ID=lgat_wgs_biospecimen_ids)

# Fusion base annotation
lgat_fus <- read_tsv(file.path(subset_dir,"LGAT_fusion_subset.tsv")) 

```


### LGG, NF1
This subtype is characterized by germline NF1 variants. We do not have germline data within OpenPBTA, but we do have information about cancer predispositions. This group can be annotated if NF-1 is present within the cancer_predispositions field of the pbta-histologies.tsv file. We should also annotate somatic loss of NF1 via either missense, nonsense mutation or copy loss.
```{r}
nf1_snv <- lgat_snv %>%
  filter(NF1_mut=="Yes") %>%
  mutate(molecular_subtype = "NF1-somatic") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

nf1_germline <- lgat_manifest %>% 
  # to keep non-RNA files only since we will match the RNA 
  # as Kids_First_Biospecimen_ID_RNA column at a later step
  filter(experimental_strategy != "RNA-Seq") %>%
  # keeping "NF1 and NF2" in cancer_predisposition
  filter(cancer_predispositions %in% c("NF-1",
                                       "NF-1,Other inherited conditions NOS")) %>%
  mutate(molecular_subtype = "NF1-germline") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

nf1_biospecimen <- bind_rows(nf1_snv,
                             nf1_germline) %>%
  unique()
  
```





### LGG, KIAA1549-BRAF

This subtype contains KIAA1549-BRAF fusion
```{r}
kiaabraf_biospecimen <- lgat_fus %>%
  filter(KIAA_BRAF_fus=="Yes") %>%
  mutate(molecular_subtype = "KIAA1549-BRAF") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

```

### LGG, BRAF V600E

This subtype contains BRAF V600E or V599 SNV or non-canonical BRAF SNV/indel overlapping kinase domain
```{r}

brafV600_biospecimen <- lgat_snv %>%
  filter(BRAF_V600E_mut=="Yes") %>%
  mutate(molecular_subtype = "BRAF V600E") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

```

### LGG, other MAPK

This subtype contains non-canonical BRAF fusion other than KIAA1549-BRAF
OR contains RAF1 fusion
OR contains KRAS, NRAS, HRAS, MAP2K1, MAP2K2, MAP2K1, ARAF SNV or indel
```{r}

otherMAPK_fus_biospecimen <- lgat_fus %>%
  filter(MAPK_fus=="Yes") %>%
  mutate(molecular_subtype = "other MAPK") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype")

otherMAPK_snv_biospecimen <- lgat_snv %>%
  filter(MAPK_mut=="Yes") %>%
  mutate(molecular_subtype = "other MAPK") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 


```

### LGG, RTK

This subtype harbors a fusion in ALK, ROS1, NTRK1, NTRK2, or NTRK3 or
OR harbors a MET SNV **or
OR harbors a KIT SNV or
OR harbors a PDGFRA SNV or fusion
```{r}

rtk_fus_biospecimen <- lgat_fus %>%
  filter(RTK_fus=="Yes") %>%
  mutate(molecular_subtype = "RTK") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

rtk_snv_biospecimen <- lgat_snv %>%
  filter(RTK_mut=="Yes") %>%
  mutate(molecular_subtype = "RTK") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 



```

### LGG, FGFR

This subtype harbors FGFR1 p.N546K, p.K656E, p.N577, or p. K687 hotspot mutations or
OR harbors FGFR1 TKD (tyrosine kinase domain tandem duplication) or
OR harbors FGFR1-TACC1 fusion **or
OR harbors FGFR1 or FGFR2 fusions
```{r}
fgfr_fus_biospecimen <- lgat_fus %>%
  filter(FGFR_fus=="Yes") %>%
  mutate(molecular_subtype = "FGFR") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

fgfr_snv_biospecimen <- lgat_snv %>%
  filter(FGFR_mut=="Yes") %>%
  mutate(molecular_subtype = "FGFR") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

fgfr_cnv_biospecimen <- lgat_cnv %>%
  filter(FGFR_DUP_TANDEM=="Yes" |  FGFR_DUP == "Yes") %>%
  mutate(molecular_subtype = "FGFR") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

```

### LGG, IDH

This subtype harbors an IDH R132 mutation
```{r}
idh_biospecimen <- lgat_snv %>%
  filter(IDH_mut=="Yes") %>%
  mutate(molecular_subtype = "IDH") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

```

### LGG, H3

This subtype harbors an H3F3A K28M or G35R/V mutation
```{r}
h3_biospecimen <- lgat_snv %>%
  filter(H3.1_mut =="Yes"| H3.2_mut =="Yes" |H3.3_mut =="Yes") %>%
  mutate(molecular_subtype = "H3") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 

```
 
### LGG, MYB/MYBL1

This subtype harbors either a MYB-QKI fusion or other MYB or MYBL1 fusion
```{r}
myb_biospecimen <- lgat_fus %>%
  filter(MYB_fus=="Yes") %>%
  mutate(molecular_subtype = "MYB/MYBL1") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype")
```

### LGG, CDKN2A/B

This subtype harbors focal CDKN2A and/or CDKN2B deletion
This is a secondary co-occurring alteration with prognostic significance.
```{r}

cdkn_biospecimen <- lgat_cnv %>%
  filter(CDKN2A_DEL=="Yes"| CDKN2B_DEL == "Yes") %>%
  mutate(molecular_subtype = "CDKN2A/B") %>%
  dplyr::select("Kids_First_Biospecimen_ID","molecular_subtype") 
```

### Compile
```{r}

all_dna_subtype <- bind_rows(nf1_biospecimen,
                         brafV600_biospecimen,
                         otherMAPK_snv_biospecimen,
                         rtk_snv_biospecimen,
                         fgfr_cnv_biospecimen,
                         fgfr_snv_biospecimen,
                         idh_biospecimen,
                         h3_biospecimen,
                         cdkn_biospecimen
) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_DNA=Kids_First_Biospecimen_ID) %>%
  # add Kids_First_Biospecimen_ID_RNA to dna subtypes dataframe
  left_join(all_lgat_matched,by="Kids_First_Biospecimen_ID_DNA") 

all_rna_subtype <- bind_rows(kiaabraf_biospecimen,
                             otherMAPK_fus_biospecimen,
                             rtk_fus_biospecimen,
                             fgfr_fus_biospecimen,
                             myb_biospecimen) %>%
  dplyr::rename(Kids_First_Biospecimen_ID_RNA=Kids_First_Biospecimen_ID) %>%
  # add Kids_First_Biospecimen_ID_DNA to rna subtypes dataframe
  left_join(all_lgat_matched,by="Kids_First_Biospecimen_ID_RNA")

  
all_subtype <- bind_rows(all_rna_subtype,
                         all_dna_subtype) %>%
  full_join(all_lgat_matched, by=c("Kids_First_Biospecimen_ID_RNA","Kids_First_Biospecimen_ID_DNA",
                                   "sample_id",
                                   "pathology_free_text_diagnosis",
                                   "pathology_diagnosis",
                                   "Kids_First_Participant_ID")) %>%
  # lets group because there are some instances where LGAT DNA/RNA have multiple subtypes
  group_by(sample_id,Kids_First_Participant_ID,
           Kids_First_Biospecimen_ID_RNA,
           Kids_First_Biospecimen_ID_DNA,
           pathology_free_text_diagnosis,
           pathology_diagnosis) %>%
  # merging them here with summarise
  summarise(molecular_subtype= toString(molecular_subtype)) %>%
  ungroup() %>% 
  # formatting molecular_subtype columns
  dplyr::mutate(molecular_subtype = case_when(
    # molecular_subtype is NA and we didn't find any canonical
    # LGG variant in the DNA and no RNA abberations in RNA sample
    # subtype should be WT
    (molecular_subtype=="NA" & !is.na(Kids_First_Biospecimen_ID_DNA) &
       !is.na(Kids_First_Biospecimen_ID_RNA)) ~ "LGG, wildtype",
    # molecular_subtype is NA and we do not have a DNA sample
    # OR don't have RNA to check for LGG variant or fusion
    # so status will be To be classified
    molecular_subtype=="NA" & (is.na(Kids_First_Biospecimen_ID_DNA) |
                                 is.na(Kids_First_Biospecimen_ID_RNA))  ~ "LGG, To be classified",
    # when molecular_subtype is not NA 
    TRUE ~ stringr::str_c("LGG, ", molecular_subtype))) %>%
  # recode if glialneuronal tumors
  dplyr::mutate(molecular_subtype = case_when(
    str_detect(pathology_free_text_diagnosis, 
               paste(path_dx_list$recode_path_free_text,collapse = "|")) &
      pathology_diagnosis == "Low-grade glioma/astrocytoma (WHO grade I/II)" ~ 
      gsub("LGG,","GNT,",molecular_subtype),
    pathology_diagnosis == "Ganglioglioma" ~
      gsub("LGG,|GNT,","GNG,",molecular_subtype),
    TRUE ~ molecular_subtype)) %>%
  dplyr::select(-pathology_free_text_diagnosis,
                -pathology_diagnosis) %>%
  unique() %>%
  dplyr::arrange(sample_id) %>%
  write_tsv(file.path(resultDir,"lgat_subtyping.tsv"))

```

```{r}
sessionInfo()
```
