---
title: "Determine Variant Caller Consensus"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---

Purpose: After running the initial analyses of each variant caller, this notebook 
combines the output from those calculations, compares them, and outputs the consensus 
for the mutation callers. 

### Summary of Findings:

Due to some of the findings here, I suggest we should move 
forward with mutations called by Strelka2, Mutect2, and Lancet. And, due to it's 
oversensitivity that VarDict's data can be dropped entirely. 

#### Individual Caller's Reports 

- [Lancet](./plots/lancet/lancet_both_report.html)
- [MuTect2](./plots/mutect2/mutect2_both_report.html)
- [Strelka2](./plots/strelka2/strelka2_both_report.html)
- [Vardict](./plots/vardict/vardict_both_report.html)

*Notable Findings:*  

- [VarDict calls more mutations](#upset-graph) than any of the other callers. 
Many of these have [very low ](#vaf-distributions) suggesting that some may be sequencing errors. 

- Mutect2, Strelka2, and Lancet do have [plenty of mutations in common](#upset-graph). 

### Outline of analyses completed:

- [Upset graph](#upset-graph)  
- [VAF of combinations of callers](#vaf-for-combination-sets-of-callers)
- [Base changes by caller](#base-changes)
- [VAF distributions by caller](#vaf-distributions)
- [TMB distributions by caller](#tmb-distributions)
- [VAF correlations between callers](#vaf-correlations)
- [TMB correlations between callers](#tmb-correlations)

#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/snv-callers/consensus_snv_callers.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

Read in set up script.

```{r}
if (!("ggupset" %in% installed.packages())) {
  install.packages("ggupset", repos = "http://cran.us.r-project.org")
}
if (!("UpSetR" %in% installed.packages())) {
  install.packages("UpSetR", repos = "http://cran.us.r-project.org")
}
if (!("GGally" %in% installed.packages())) {
  install.packages("GGally", repos = "http://cran.us.r-project.org")
}
# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

Set up a function for calculating the percent off mutations of a caller for a 
given variable. 

```{r}
perc_mutation_by_caller <- function(df, var) {
  # Summarizes a variable and calculates the percent of mutations for a caller
  # for each category
  #
  # Args:
  #   df: a data.frame with `caller` as a column.
  #   var: a string identifying the column of the variable you need to summarize by.
  #
  # Returns:
  #   a data.frame with a summarized percent of caller mutations calculated in
  #   the `percent` column.
  # Summarize by the number of times each category shows up
  count_df <- df %>%
    dplyr::count(caller, eval(parse(text = var)), name = "count") %>%
    dplyr::add_count(caller, wt = count) %>%
    dplyr::mutate(percent = count / n) %>%
    # Rename the column as what it was before
    dplyr::rename(var = "eval(parse(text = var))")

  # Return this data.frame
  return(count_df)
}
```

Set up output directories. 

```{r}
base_results_dir <- "results"
base_plots_dir <- "plots"
```

Make new directories for the comparison analysis.

```{r}
consensus_results_dir <- file.path(base_results_dir, "consensus")
consensus_plots_dir <- file.path(base_plots_dir, "consensus")

# Make caller specific plots folder
if (!dir.exists(consensus_results_dir)) {
  dir.create(consensus_results_dir)
}
# Make caller specific plots folder
if (!dir.exists(consensus_plots_dir)) {
  dir.create(consensus_plots_dir)
}
```

## Import data 

Load in the merged data.frames from `03-merge_callers.R` output. 

```{r}
# Master VAF data.frame
vaf_df <- readr::read_rds(file.path(consensus_results_dir, "all_callers_vaf.rds"))

# Master TMB data.frame
tmb_df <- readr::read_rds(file.path(consensus_results_dir, "all_callers_tmb.rds"))

# Full mutation list
mutation_id_list <- readr::read_rds(file.path(consensus_results_dir, "mutation_id_list.rds"))

# Callers per mutation data.frame
callers_per_mutation <- readr::read_rds(file.path(consensus_results_dir, "callers_per_mutation.rds"))
```

## Upset graph

Make the upset plot. 

```{r}
# Make the plot
upset <- UpSetR::upset(UpSetR::fromList(mutation_id_list),
  order.by = "freq",
  text.scale = 2
)

# Save this plot
png(file.path(consensus_plots_dir, "upset_plot.png"), width = 900, height = 900)
upset
dev.off()

# Print this out
upset
```

## VAF for combination sets of callers

Graph mutations by combinations of callers and plot the VAF density.
Get list of callers per mutation and the median vaf for each. 
Graph the median VAF for each combination of callers.

```{r}
# Make this plot
callers_per_mutation %>%
  ggplot2::ggplot(ggplot2::aes(x = caller_combo, y = median_vaf)) +
  ggplot2::geom_violin() +
  ggplot2::theme_classic() +
  ggupset::scale_x_mergelist(sep = "-") +
  ggupset::axis_combmatrix(sep = "-") +
  ggplot2::xlab("") +
  ggplot2::ylab("Median VAF Across Callers")
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "upset_median_vaf_plot.png"))
```

## Base Changes 

Set up the summarized form of the base change information. 

```{r}
# Summarize by the number of times each base change shows up in each category.
perc_change_df <- vaf_df %>%
  dplyr::count(caller, change, name = "count") %>%
  dplyr::add_count(caller, wt = count) %>%
  dplyr::mutate(percent = count / n) %>%
  dplyr::filter(!grepl("G>G|A>A|N>A|N>T", change))

# Need to rearrange the order of the base change categories
perc_change_df <- perc_change_df %>%
  dplyr::mutate(
    change = as.factor(change),
    change = factor(change, c(grep("ins|del|long_change", levels(change), value = TRUE, invert = TRUE), "ins", "del", "long_change"))
  )
```

Make a barplot illustrating the percent of the mutations for each caller that 
that are each type of change. 

```{r}
perc_change_df %>%
  ggplot2::ggplot(ggplot2::aes(x = change, y = percent)) +
  ggplot2::theme_classic() +
  ggplot2::geom_bar(
    position = "dodge", stat = "identity",
    ggplot2::aes(fill = caller)
  ) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) +
  ggplot2::xlab("") +
  ggplot2::ylab("Percent of callers' mutations") +
  colorblindr::scale_fill_OkabeIto()
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "perc_base_change_plot.png"))
```

## VAF distributions

```{r}
vaf_df %>%
  ggplot2::ggplot(ggplot2::aes(x = caller, y = vaf, color = caller)) +
  ggplot2::geom_violin() +
  ggplot2::theme_classic() +
  colorblindr::scale_color_OkabeIto()
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "vaf_violin_plot.png"))
```

## TMB distributions

```{r}
tmb_df %>%
  ggplot2::ggplot(ggplot2::aes(x = caller, y = tmb, color = short_histology)) +
  ggplot2::geom_boxplot() +
  ggplot2::theme_classic()
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "tmb_boxplot_histology_plot.png"))
```

## VAF correlations

```{r}
vaf_long <- vaf_df %>%
  dplyr::group_by(mutation_id, caller) %>%
  dplyr::summarize(vaf = mean(vaf)) %>%
  dplyr::distinct(mutation_id, caller, vaf, .keep_all = TRUE) %>%
  tidyr::spread(caller, vaf) %>%
  tibble::column_to_rownames("mutation_id")
```

Make the plot with ggpairs. 

```{r}
GGally::ggpairs(vaf_long, mapping = ggplot2::aes(alpha = 0.05)) +
  ggplot2::theme_classic()
``` 

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "vaf_correlations_plot.png"))
```

## TMB correlations

```{r}
tmb_long <- tmb_df %>%
  dplyr::ungroup() %>%
  dplyr::select(-genome_size, -experimental_strategy) %>%
  dplyr::distinct(Tumor_Sample_Barcode, caller, tmb) %>%
  tidyr::spread(caller, tmb) %>%
  dplyr::select(-Tumor_Sample_Barcode)
```

Make the TMB correlation plot.

```{r}
GGally::ggpairs(tmb_long, mapping = ggplot2::aes(alpha = 0.05), cardinality_threshold = NULL) +
  ggplot2::theme_classic()
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "tmb_correlations_plot.png"))
```

## Session Info

```{r}
sessionInfo()
```
