---
title: "Plot Caller Comparisons"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---

Purpose: After creating a consensus file with `02-merge_callers.R`, this notebook 
plots these data in a series of plots to compare the consensus calls to the individual 
callers. 

### Summary of Findings:


#### Individual Caller's Reports 


*Notable Findings:*  

- [VarDict calls more mutations](#upset-graph) than any of the other callers. 
Many of these have [very low ](#vaf-distributions) suggesting that some may be sequencing errors. 

- Mutect2, Strelka2, and Lancet do have [plenty of mutations in common](#upset-graph). 

### Outline of analyses completed:

- [Upset graph](#upset-graph)  
- [VAF of combinations of callers](#vaf-for-combination-sets-of-callers)
- [Base changes by caller](#base-changes)
- [VAF distributions by caller](#vaf-distributions)
- [TMB distributions by caller](#tmb-distributions)
- [VAF correlations between callers](#vaf-correlations)
- [TMB correlations between callers](#tmb-correlations)

#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/snv-callers/compare_snv_callers_plots.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

Read in set up script.

```{r}
if (!("ggupset" %in% installed.packages())) {
  install.packages("ggupset", repos = "http://cran.us.r-project.org")
}
if (!("UpSetR" %in% installed.packages())) {
  install.packages("UpSetR", repos = "http://cran.us.r-project.org")
}
if (!("GGally" %in% installed.packages())) {
  install.packages("GGally", repos = "http://cran.us.r-project.org")
}
# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

Set up output directories. 

```{r}
scratch_dir <- file.path("..", "..", "scratch")
```

Make new directories for the comparison analysis.

```{r}
results_dir <- file.path("results", "comparison")
plots_dir <- file.path("plots", "comparison")

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

## Connect to database

Connect to SQLite database.

```{r}
# Start up connection
con <- DBI::dbConnect(RSQLite::SQLite(), 
                      file.path(scratch_dir, "testing_snv_db.sqlite"))
```

Note what columns we will join by.

```{r}
join_cols = c("Chromosome",
              "Start_Position",
              "Reference_Allele",
              "Allele",
              "Tumor_Sample_Barcode")
```

Set up tables from the database but only call the columns we need.

```{r}
strelka <- dplyr::tbl(con, "strelka") %>% 
  dplyr::select(join_cols, "VAF")

lancet <- dplyr::tbl(con, "lancet") %>% 
  dplyr::select(join_cols, "VAF")

mutect <- dplyr::tbl(con, "mutect") %>% 
  dplyr::select(join_cols, "VAF") 

vardict <- dplyr::tbl(con, "vardict") %>% 
  dplyr::select(join_cols, "VAF") 
```

```{r}
# Designate consensus table
consensus <- dplyr::tbl(con, "consensus")
```

```{r}
# Designate consensus table
metadata <- dplyr::tbl(con, "samples")
```

Because `DBI` does not support full_join, we had to use a series of `left_join` and
`union_all` calls in order to get a full join of all the callers. 

```{r}
# Join Strelka with Lancet data
# Union of two left joins
union_strelka_lancet <- dplyr::union_all(
 dplyr::left_join(lancet, strelka,
           by = join_cols,
           suffix = c("_lancet", "_strelka")) %>%
   # Want to keep the VAFs from each caller
   dplyr::select(join_cols, "VAF_lancet", "VAF_strelka"),
 dplyr::left_join(strelka, lancet,
           by = join_cols,
           suffix = c("_strelka", "_lancet")) %>%
   dplyr::select(join_cols, "VAF_lancet", "VAF_strelka") %>% 
   # VAFs for lancet will already be included in the previous left_join
   dplyr::filter(is.na(VAF_lancet))
) 

# Join Mutect with VarDict data
union_mutect_vardict <- dplyr::union_all(
 dplyr::left_join(vardict, mutect,
           by = join_cols,
           suffix = c("_vardict", "_mutect")) %>%
   dplyr::select(join_cols, "VAF_vardict", "VAF_mutect"),
 dplyr::left_join(mutect, vardict,
           by = join_cols,
           suffix = c("_mutect", "_vardict")) %>%
   dplyr::select(join_cols, "VAF_vardict", "VAF_mutect") %>% 
   dplyr::filter(is.na(VAF_vardict))
) 

# Join both combinations made above into one big data set
all_caller <- dplyr::union_all(
 dplyr::left_join(union_strelka_lancet, union_mutect_vardict,
           by = join_cols) %>%
   dplyr::select(join_cols, "VAF_lancet", "VAF_strelka", "VAF_vardict", "VAF_mutect"),
 dplyr::left_join(union_mutect_vardict, union_strelka_lancet,
           by = join_cols) %>%
   dplyr::select(join_cols, "VAF_lancet", "VAF_strelka", "VAF_vardict", "VAF_mutect") %>% 
   dplyr::filter(is.na(VAF_strelka) && is.na(VAF_lancet))
) 

# Check that this looks right
head(all_caller)
```

```{r}
vaf_df <- all_caller %>% 
  dplyr::select("VAF_lancet", "VAF_strelka", "VAF_vardict", "VAF_mutect") %>%
  as.data.frame() %>% 
  dplyr::mutate(index = factor(1:nrow(.))) %>% 
  tidyr::gather(key = "caller", value = "vaf", -index) %>% 
  dplyr::mutate(caller = gsub("VAF_", "", caller))
  
callers_per_mutation <- tapply(vaf_df$caller,
    vaf_df$index,
    paste0,
    collapse = "-"
  ) %>% 
  as.data.frame() %>%
  tibble::rownames_to_column("index")

vaf_med <- tapply(
    vaf_df$vaf,
    vaf_df$index,
    median
  ) %>% 
  as.data.frame() %>%
  tibble::rownames_to_column("index")

med_vaf <- dplyr::inner_join(callers_per_mutation, vaf_med)
```

## Upset graph

Make the upset plot. 

```{r}
# Make the plot
upset <- UpSetR::upset(callers_per_mutation,
  order.by = "freq",
  text.scale = 1.5
)

# Save this plot
png(file.path(plots_dir, "upset_plot.png"), width = 900, height = 900)
upset
dev.off()

# Print this out
upset
```

## VAF for combination sets of callers

Graph mutations by combinations of callers and plot the VAF density.
Get list of callers per mutation and the median vaf for each. 
Graph the median VAF for each combination of callers.

```{r}
# Make this plot
callers_per_mutation %>%
  ggplot2::ggplot(ggplot2::aes(x = caller_combo, y = median_vaf)) +
  ggplot2::geom_violin() +
  ggplot2::theme_classic() +
  ggupset::scale_x_mergelist(sep = "-") +
  ggupset::axis_combmatrix(sep = "-") +
  ggplot2::xlab("") +
  ggplot2::ylab("Median VAF Across Callers")
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "upset_median_vaf_plot.png"))
```

## Base Changes 

Set up the summarized form of the base change information. 

```{r}
# Summarize by the number of times each base change shows up in each category.
perc_change_df <- vaf_df %>%
  dplyr::count(caller, change, name = "count") %>%
  dplyr::add_count(caller, wt = count) %>%
  dplyr::mutate(percent = count / n) %>%
  dplyr::filter(!grepl("G>G|A>A|N>A|N>T", change))

# Need to rearrange the order of the base change categories
perc_change_df <- perc_change_df %>%
  dplyr::mutate(
    change = as.factor(change),
    change = factor(change, c(grep("ins|del|long_change", levels(change), value = TRUE, invert = TRUE), "ins", "del", "long_change"))
  )
```

Make a barplot illustrating the percent of the mutations for each caller that 
that are each type of change. 

```{r}
perc_change_df %>%
  ggplot2::ggplot(ggplot2::aes(x = change, y = percent)) +
  ggplot2::theme_classic() +
  ggplot2::geom_bar(
    position = "dodge", stat = "identity",
    ggplot2::aes(fill = caller)
  ) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) +
  ggplot2::xlab("") +
  ggplot2::ylab("Percent of callers' mutations") +
  colorblindr::scale_fill_OkabeIto()
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "perc_base_change_plot.png"))
```

## VAF distributions

Make a violin plot of each caller. 

```{r}
vaf_df %>%
  ggplot2::ggplot(ggplot2::aes(x = caller, y = vaf, color = caller)) +f
  ggplot2::geom_violin() +
  ggplot2::theme_classic() +
  colorblindr::scale_color_OkabeIto()
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "vaf_violin_plot.png"))
```

## TMB distributions

```{r}
tmb_df %>%
  ggplot2::ggplot(ggplot2::aes(x = caller, y = tmb, color = short_histology)) +
  ggplot2::geom_boxplot() +
  ggplot2::theme_classic()
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "tmb_boxplot_histology_plot.png"))
```

## VAF correlations

```{r}
vaf_long <- vaf_df %>%
  dplyr::group_by(mutation_id, caller) %>%
  dplyr::summarize(vaf = mean(vaf)) %>%
  dplyr::distinct(mutation_id, caller, vaf, .keep_all = TRUE) %>%
  tidyr::spread(caller, vaf) %>%
  tibble::column_to_rownames("mutation_id")
```

Make the plot with ggpairs. 

```{r}
GGally::ggpairs(vaf_long, mapping = ggplot2::aes(alpha = 0.05)) +
  ggplot2::geom_hex() + 
  ggplot2::theme_classic()
``` 

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "vaf_correlations_plot.png"))
```

## TMB correlations

```{r}
tmb_long <- tmb_df %>%
  dplyr::ungroup() %>%
  dplyr::select(-genome_size, -experimental_strategy) %>%
  dplyr::distinct(Tumor_Sample_Barcode, caller, tmb) %>%
  tidyr::spread(caller, tmb) %>%
  dplyr::select(-Tumor_Sample_Barcode)
```

Make the TMB correlation plot.

```{r}
GGally::ggpairs(tmb_long, mapping = ggplot2::aes(alpha = 0.05), cardinality_threshold = NULL) +
  ggplot2::theme_classic()
```

```{r}
# Save this plot
ggplot2::ggsave(file.path(consensus_plots_dir, "tmb_correlations_plot.png"))
```

## Session Info

```{r}
sessionInfo()
```
