---
title: "Compare Variant Callers"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---

Purpose: After running the initial analyses of each variant caller, this notebook 
combines the output from those calculations and compares them. 

### Summary of Findings:

Due to some of the findings here, I suggest we should move 
forward with mutations called by Strelka2, Mutect2, and Lancet. And, due to it's 
oversensitivity that VarDict's data can be dropped entirely. 

#### Individual Caller's Reports 

- [Lancet](./plots/lancet/lancet_both_report.html)
- [MuTect2](./plots/mutect2/mutect2_both_report.html)
- [Strelka2](./plots/strelka2/strelka2_both_report.html)
- [Vardict](./plots/vardict/vardict_both_report.html)

*Notable Findings:*  

- [VarDict calls more mutations](#upset-graph) than any of the other callers. 
Many of these have [very low ](#vaf-distributions) suggesting that some may be sequencing errors. 

- 

### Outline of analyses completed:

- [Upset graph](#upset-graph)  
- [Base changes by caller](#base-changes)
- [VAF distributions by caller](#vaf-distributions)
- [TMB distributions by caller](#tmb-distributions)
- [VAF correlations between callers](#vaf-correlations)
- [TMB correlations between callers](#tmb-correlations)
- [Genomic region of mutations from each caller](#genomic-regions)

#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/snv-callers/compare_snv_callers.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

Read in set up script.

```{r}
if (!("ggupset" %in% installed.packages())) {
  install.packages("ggupset", repos = "http://cran.us.r-project.org")
}
if (!("UpSetR" %in% installed.packages())) {
  install.packages("UpSetR", repos = "http://cran.us.r-project.org")
}
if (!("GGally" %in% installed.packages())) {
  install.packages("GGally", repos = "http://cran.us.r-project.org")
}
# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

Set up output directories. 

```{r}
base_results_dir <- "results"
base_plots_dir <- "plots"
```

Make new directories for the comparison analysis.

```{r}
compare_results_dir <- file.path(base_results_dir, "comparisons")
compare_plots_dir <- file.path(base_plots_dir, "comparisons")

# Make caller specific plots folder
if (!dir.exists(compare_results_dir)) {
  dir.create(compare_results_dir)
}
# Make caller specific plots folder
if (!dir.exists(compare_plots_dir)) {
  dir.create(compare_plots_dir)
}
```

## Import data 

Look at the existing folders within the results folder. 

```{r}
dir(base_results_dir)
```

Get the lists of each type of file. 

```{r}
# Get the lists of each type of file
vaf_files <- list.files(base_results_dir,
  pattern = "_vaf.tsv$", recursive = TRUE,
  full.names = TRUE
)
tmb_files <- list.files(base_results_dir,
  pattern = "_tmb.tsv$", recursive = TRUE,
  full.names = TRUE
)
region_annot_files <- list.files(base_results_dir,
  pattern = "_region.tsv$", recursive = TRUE,
  full.names = TRUE
)
```

Obtain the caller names from the file paths. 

```{r}
# Get the caller names
caller_names <- stringr::word(vaf_files, sep = "/", 2)
```

Make a big VAF data.frame

```{r include=FALSE}
vaf_list <- lapply(vaf_files, function(file) {
  readr::read_tsv(file) %>%
    dplyr::select("mutation_id", "vaf", "base_change", "change") %>%
    dplyr::distinct(mutation_id, .keep_all = TRUE)
})
names(vaf_list) <- caller_names

vaf_df <- dplyr::bind_rows(vaf_list, .id = "caller") %>%
  dplyr::mutate(caller = factor(caller))
```

Make a big tmb data.frame

```{r include=FALSE}
tmb_list <- lapply(tmb_files, readr::read_tsv)
names(tmb_list) <- caller_names
tmb_df <- dplyr::bind_rows(tmb_list, .id = "caller") %>%
  dplyr::mutate(caller = factor(caller))
```

Make a big genomic region data.frame

```{r include=FALSE}
region_annot <- lapply(region_annot_files, function(file) {
  readr::read_tsv(file) %>%
    dplyr::select("vaf", "base_change", "mutation_id", "change", "type")
})
names(region_annot) <- caller_names
region_annot_df <- dplyr::bind_rows(region_annot, .id = "caller") %>%
  dplyr::mutate(caller = factor(caller))
```

## Upset graph

```{r}
mutation_id_list <- lapply(vaf_list, function(caller) caller$mutation_id)
```

Make the upset plot. 

```{r}
upset <- UpSetR::upset(UpSetR::fromList(mutation_id_list), order.by = "freq")

# Print this out
upset
```

Graph mutations by combinations of callers and plot the VAF density.

Get list of callers per mutation and the median vaf for each. 

```{r}
callers_per_mutation <- tapply(vaf_df$caller,
  vaf_df$mutation_id,
  paste0,
  collapse = "-"
) %>%
  as.data.frame() %>%
  tibble::rownames_to_column("mutation_id")

vaf_med <- tapply(
  vaf_df$vaf,
  vaf_df$mutation_id,
  median
) %>%
  as.data.frame() %>%
  tibble::rownames_to_column("mutation_id")

callers_per_mutation <- callers_per_mutation %>%
  dplyr::inner_join(vaf_med, by = "mutation_id")

colnames(callers_per_mutation) <- c("mutation_id", "caller_combo", "median_vaf")
```

Graph the median VAF for each combination of callers.

```{r}
callers_per_mutation %>%
  ggplot2::ggplot(ggplot2::aes(x = caller_combo, y = median_vaf)) +
  ggforce::geom_sina(alpha = 0.05) +
  ggplot2::theme_classic() +
  ggupset::scale_x_mergelist(sep = "-") +
  ggupset::axis_combmatrix(sep = "-") +
  ggplot2::xlab("") +
  ggplot2::ylab("Median VAF Across Callers")
```

## Base Changes 

```{r}
region_annot_df %>%
  dplyr::filter(!(base_change %in% c("G>G", "A>A", "N>A", "N>T"))) %>%
  dplyr::group_by(change, caller) %>%
  dplyr::summarize(count = dplyr::n()) %>%
  ggplot2::ggplot(ggplot2::aes(x = reorder(change, -count), y = count)) +
  ggplot2::theme_classic() +
  ggplot2::geom_bar(
    position = "dodge", stat = "identity",
    ggplot2::aes(fill = caller)
  ) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) +
  ggplot2::xlab("") +
  ggplot2::ylab("Count") +
  colorblindr::scale_fill_OkabeIto()
```

## VAF distributions

```{r}
vaf_df %>%
  ggplot2::ggplot(ggplot2::aes(x = caller, y = vaf, color = caller)) +
  ggforce::geom_sina(alpha = 0.05) +
  ggplot2::theme_classic() +
  colorblindr::scale_color_OkabeIto()
```

## TMB distributions

```{r}
tmb_df %>%
  ggplot2::ggplot(ggplot2::aes(x = caller, y = tmb, color = short_histology)) +
  ggforce::geom_sina(alpha = 0.25) +
  ggplot2::theme_classic()
```

## VAF correlations

```{r}
vaf_long <- vaf_df %>%
  dplyr::group_by(mutation_id, caller) %>%
  dplyr::summarize(vaf = mean(vaf)) %>%
  dplyr::distinct(mutation_id, caller, vaf, .keep_all = TRUE) %>%
  tidyr::spread(caller, vaf) %>%
  tibble::column_to_rownames("mutation_id")
```

Make the plot with ggpairs. 

```{r}
GGally::ggpairs(vaf_long, mapping = ggplot2::aes(alpha = 0.05)) +
  ggplot2::theme_classic()
``` 

## TMB correlations

```{r}
tmb_long <- tmb_df %>%
  dplyr::distinct(Tumor_Sample_Barcode, caller, tmb) %>%
  tidyr::spread(caller, tmb) %>%
  tibble::column_to_rownames("Tumor_Sample_Barcode")
```

Make the TMB correlation plot.

```{r}
GGally::ggpairs(tmb_long, mapping = ggplot2::aes(alpha = 0.05)) +
  ggplot2::theme_classic()
```

## Genomic regions 

```{r}
region_annot_df %>%
  dplyr::group_by(type, caller) %>%
  dplyr::summarize(count = dplyr::n()) %>%
  ggplot2::ggplot(ggplot2::aes(x = reorder(type, -count), y = count)) +
  ggplot2::theme_classic() +
  ggplot2::geom_bar(
    position = "dodge", stat = "identity",
    ggplot2::aes(fill = caller)
  ) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) +
  ggplot2::xlab("") +
  ggplot2::ylab("Count") +
  colorblindr::scale_fill_OkabeIto()
```


```{r}
region_annot_df %>%
  ggplot2::ggplot(ggplot2::aes(x = caller, y = vaf)) +
  ggplot2::theme_classic() +
  ggforce::geom_sina(alpha = 0.1, ggplot2::aes(color = caller)) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) +
  ggplot2::xlab("") +
  ggplot2::ylab("VAF") +
  colorblindr::scale_color_OkabeIto() +
  ggplot2::facet_wrap(~type)
```

## Session Info

```{r}
sessionInfo()
```
