---
title: "TCGA Plot Caller Comparisons"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---

Purpose: Why are TCGA and PBTA data so different? 

#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/snv-callers/compare_snv_callers_plots-tcga.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

Read in set up script.

```{r}
# Magrittr pipe
`%>%` <- dplyr::`%>%`

source(file.path("..", "snv-callers", "util", "tmb_functions.R"))
```

```{r}
sina_plot <- function(data, colour = NULL, x_axis = "short_histology", title) {
  # Given a data.frame with metadata and `tmb` information, make a sina plot from it
  #
  # Args:
  #
  # data : a data.frame with `tmb` as columns
  # x_axis : A name of a column in data that you would like to make the x_axis from
  #           Given as a character string. 
  # colour: the color you would like the points to be
  # title: The title for the plot. To be passed to `ggtitle`
  #
  # Returns:
  #   A sina plot by histology group
  data %>%
    # eval parse bit will treat the variable as the text it has stored
    ggplot2::ggplot(ggplot2::aes(x = reorder(eval(parse(text = x_axis)), tmb, mean), 
                                 y = tmb)) +
    ggforce::geom_sina(color = colour, alpha = 0.3, maxwidth = 0.9) +
    ggplot2::stat_summary(
      fun.y = mean, fun.ymin = mean, fun.ymax = mean,
      geom = "crossbar",
      width = 0.95, size = 0.15
    ) +
    ggplot2::theme_classic() +
    ggplot2::ylab("log10 (Number of Mutations per Mb)") +
    ggplot2::xlab("") +
    ggplot2::theme(
      axis.text.x = ggplot2::element_text(angle = 60, hjust = 1),
      legend.position = "none"
    ) +
      ggplot2::scale_y_continuous(trans = "log1p", breaks = c(0, 1, 3, 10, 30)) +
    ggplot2::scale_x_continuous(limits = c(-0.2, 1.2), breaks = c()) +
    ggplot2::ggtitle(title) 
}
```

Set up directories. 

```{r}
scratch_dir <- file.path("..", "..", "scratch")
data_dir <- file.path("..", "..", "data")
```

```{r}
pbta_metadata <- readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv")) 
tcga_metadata <- readr::read_tsv(file.path(data_dir, "pbta-tcga-manifest.tsv")) %>% 
  dplyr::rename(short_histology = broad_histology)
```

## Connect to databases

Note what columns we will join by.

```{r}
join_cols = c("Chromosome",
              "Start_Position",
              "End_Position", 
              "Reference_Allele",
              "Allele",
              "Tumor_Sample_Barcode", 
              "Variant_Classification", 
              "t_depth",
              "t_ref_count", 
              "t_alt_count", 
              "n_depth", 
              "n_ref_count", 
              "n_alt_count", 
              "VAF")
```

Set up tables from the database but only call the columns we need.

```{r}
set_up_tcga <- function(data.name = "strelka") {
    # Start up connection
  con <- DBI::dbConnect(RSQLite::SQLite(), file.path(scratch_dir, "tcga_snv_db.sqlite"))
  
  # Connect to SQL
  df <- dplyr::tbl(con, data.name) %>% 
    dplyr::select(join_cols) %>% 
    as.data.frame() %>% 
    # Shorten the Tumor_Sample_Barcode so it matches
    dplyr::mutate(Tumor_Sample_Barcode = substr(Tumor_Sample_Barcode, 0, 12)) %>% 
    dplyr::inner_join(tcga_metadata %>% dplyr::select(Tumor_Sample_Barcode, broad_histology))
    dplyr::mutate(experimental_strategy = "WXS", short_histology = broad_histology)
    
  return(df)
}

tcga_lancet <- set_up_tcga("lancet")
tcga_mutect <- set_up_tcga("mutect")
tcga_strelka <- set_up_tcga("strelka")
```

```{r}
set_up_pbta <- function(data.name = "strelka") {
  # Start up connection
  con <- DBI::dbConnect(RSQLite::SQLite(), 
                        file.path(scratch_dir, "snv_db.sqlite"))
  
  # Connect to SQL
  df <- dplyr::tbl(con, data.name) %>% 
    dplyr::select(join_cols, "VAF") %>% 
    as.data.frame() %>% 
    dplyr::inner_join(
    pbta_metadata %>% 
      dplyr::select(Tumor_Sample_Barcode = Kids_First_Biospecimen_ID,
                    experimental_strategy, short_histology))
    
  return(df)
}

pbta_lancet <- set_up_pbta("lancet")
pbta_mutect <- set_up_pbta("mutect")
pbta_strelka <- set_up_pbta("strelka")
```

```{r}
lancet <- dplyr::bind_rows(list(tcga = tcga_lancet,  
                                pbta = pbta_lancet), .id = "dataset") %>% 
  dplyr::mutate(dataset = as.factor(dataset))

mutect <- dplyr::bind_rows(list(tcga = tcga_mutect, 
                                pbta = pbta_mutect), .id = "dataset") %>% 
  dplyr::mutate(dataset = as.factor(dataset)) 

strelka <- dplyr::bind_rows(list(tcga = tcga_strelka, 
                                pbta = pbta_strelka), .id = "dataset") %>% 
  dplyr::mutate(dataset = as.factor(dataset)) 
```

## Is this a tumor sequencing depth problem for certain callers? 

```{r}
lancet %>% 
  ggplot2::ggplot(ggplot2::aes(x = log10(t_depth), color = dataset)) + 
  ggplot2::geom_density() 
```

```{r}
mutect %>% 
  ggplot2::ggplot(ggplot2::aes(x = log10(t_depth), color = dataset)) + 
  ggplot2::geom_density() 
```

```{r}
strelka %>% 
  ggplot2::ggplot(ggplot2::aes(x = VAF, color = dataset)) + 
  ggplot2::geom_density() 
```

```{r}
lancet %>% 
  dplyr::group_by(Tumor_Sample_Barcode, dataset) %>% 
  dplyr::summarize(n_mut = dplyr::n(), 
                   med_depth = median(n_depth)) %>% 
  ggplot2::ggplot(ggplot2::aes(x = med_depth, y = n_mut, color = dataset)) + 
  ggplot2::geom_point() + 
  ggplot2::facet_wrap(~dataset)
```

```{r}
mutect %>% 
  dplyr::group_by(Tumor_Sample_Barcode, dataset) %>% 
  dplyr::summarize(n_mut = dplyr::n(), 
                   med_depth = median(n_depth)) %>% 
  ggplot2::ggplot(ggplot2::aes(x = med_depth, y = n_mut, color = dataset)) + 
  ggplot2::geom_point() + 
  ggplot2::facet_wrap(~dataset)
```

```{r}
strelka  %>% 
  dplyr::group_by(Tumor_Sample_Barcode, dataset) %>% 
  dplyr::summarize(n_mut = dplyr::n(), 
                   med_depth = median(n_depth)) %>% 
  ggplot2::ggplot(ggplot2::aes(x = med_depth, y = n_mut, color = dataset)) + 
  ggplot2::geom_point() + 
  ggplot2::facet_wrap(~dataset)
```

## Is this a caller specific problem?? 

```{r}
tmb_combo <- function(pbta_df, tcga_df, data.name) {
# Calculate coding only TMBs and write to file
  pbta_coding_df <- calculate_tmb(pbta_df,
    bed_wgs = file.path(data_dir, "intersect_cds_lancet_strelka_mutect_WGS.bed"),
    bed_wxs = file.path(scratch_dir, "intersect_cds_WXS.bed")
    )

  # Calculate coding only TMBs and write to file
  tcga_coding_df <- calculate_tmb(tcga_df,
    bed_wgs = file.path(scratch_dir, "intersect_cds_lancet_strelka_mutect_WGS.bed"),
    bed_wxs = file.path(scratch_dir, "intersect_cds_WXS.bed")
    )
  
  # Make the plot for PBTA
  pbta_plot <- sina_plot(pbta_coding_df, colour = "#630882", title = "PBTA")
  
  # Make TCGA plot and get rid of y axis
  tcga_plot <- sina_plot(tcga_coding_df, colour = "#3BC8A2", title = "TCGA (Adult)") +
    ggplot2::theme(
      axis.title.y = ggplot2::element_blank(),
      axis.text.y = ggplot2::element_blank(),
      axis.ticks.y = ggplot2::element_blank(),
      axis.text.x = ggplot2::element_text(size = 7),
      panel.grid.major.x = ggplot2::element_line(colour = c("gray93", "white"), size = 9)
    )

  # Put the plots together
  tmb_plot <- cowplot::plot_grid(pbta_plot, tcga_plot,
    align = "v",
    axis = "left",
    rel_widths = c(2.5, 1),
    label_size = 12
  )
  
  return(tmb_plot)
}
```

```{r}
tmb_combo(pbta_lancet, tcga_lancet)
```

```{r}
tmb_combo(pbta_mutect, tcga_mutect)
```

```{r}
tmb_combo(pbta_strelka, tcga_strelka)
```

## Let's find the overlap of the target regions of both datasets

```{r}
"https://api.gdc.cancer.gov/data/7f0d3ab9-8bef-4e3b-928a-6090caae885b"


```
## Session Info

```{r}
sessionInfo()
```