---
title: "TCGA Plot Caller Comparisons"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---

Purpose: Why are TCGA and PBTA data so different? 

#### Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/snv-callers/compare_snv_callers_plots-tcga.Rmd', 
                              clean = TRUE)"
```

_This assumes you are in the top directory of the repository._

## Setup

#### Packages and functions

Read in set up script.

```{r}
# Magrittr pipe
`%>%` <- dplyr::`%>%`

source(file.path("..", "snv-callers", "util", "tmb_functions.R"))
```

Set up directories. 

```{r}
scratch_dir <- file.path("..", "..", "scratch")
data_dir <- file.path("..", "..", "data")
```

```{r}
pbta_metadata <- readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv")) 
tcga_metadata <- readr::read_tsv(file.path(data_dir, "pbta-tcga-manifest.tsv")) %>% 
  dplyr::rename(short_histology = broad_histology)
```

## Connect to databases

Note what columns we will join by.

```{r}
join_cols = c("Chromosome",
              "Start_Position",
              "End_Position", 
              "Reference_Allele",
              "Allele",
              "Tumor_Sample_Barcode", 
              "Variant_Classification", 
              "t_depth",
              "t_ref_count", 
              "t_alt_count", 
              "n_depth", 
              "n_ref_count", 
              "n_alt_count", 
              "VAF")
```

Set up tables from the database but only call the columns we need.

```{r}
set_up_tcga <- function(data.name = "strelka") {
    # Start up connection
  con <- DBI::dbConnect(RSQLite::SQLite(), file.path(scratch_dir, "tcga_snv_db.sqlite"))
  
  # Connect to SQL
  df <- dplyr::tbl(con, data.name) %>% 
    dplyr::select(join_cols) %>% 
    as.data.frame() %>% 
    # Shorten the Tumor_Sample_Barcode so it matches
    dplyr::mutate(Tumor_Sample_Barcode = substr(Tumor_Sample_Barcode, 0, 12)) %>% 
    dplyr::inner_join(tcga_metadata %>% dplyr::select(Tumor_Sample_Barcode, broad_histology))
    dplyr::mutate(experimental_strategy = "WXS", short_histology = broad_histology)
    
  return(df)
}

tcga_lancet <- set_up_tcga("lancet")
tcga_mutect <- set_up_tcga("mutect")
tcga_strelka <- set_up_tcga("strelka")
```

```{r}
set_up_pbta <- function(data.name = "strelka") {
  # Start up connection
  con <- DBI::dbConnect(RSQLite::SQLite(), 
                        file.path(scratch_dir, "snv_db.sqlite"))
  
  # Connect to SQL
  df <- dplyr::tbl(con, data.name) %>% 
    dplyr::select(join_cols, "VAF") %>% 
    as.data.frame() %>% 
    dplyr::inner_join(
    pbta_metadata %>% 
      dplyr::select(Tumor_Sample_Barcode = Kids_First_Biospecimen_ID,
                    experimental_strategy, short_histology))
    
  return(df)
}

pbta_lancet <- set_up_pbta("lancet")
pbta_mutect <- set_up_pbta("mutect")
pbta_strelka <- set_up_pbta("strelka")
```

```{r}
lancet <- dplyr::bind_rows(list(tcga = tcga_lancet,  
                                pbta = pbta_lancet), .id = "dataset") %>% 
  dplyr::mutate(dataset = as.factor(dataset))

mutect <- dplyr::bind_rows(list(tcga = tcga_mutect, 
                                pbta = pbta_mutect), .id = "dataset") %>% 
  dplyr::mutate(dataset = as.factor(dataset)) 

strelka <- dplyr::bind_rows(list(tcga = tcga_strelka, 
                                pbta = pbta_strelka), .id = "dataset") %>% 
  dplyr::mutate(dataset = as.factor(dataset)) 
```

## Is this a tumor sequencing depth problem for certain callers? 

```{r}
lancet %>% 
  ggplot2::ggplot(ggplot2::aes(x = log10(t_depth), color = dataset)) + 
  ggplot2::geom_density() 
```

```{r}
mutect %>% 
  ggplot2::ggplot(ggplot2::aes(x = log10(t_depth), color = dataset)) + 
  ggplot2::geom_density() 
```

```{r}
strelka %>% 
  ggplot2::ggplot(ggplot2::aes(x = VAF, color = dataset)) + 
  ggplot2::geom_density() 
```

```{r}
lancet %>% 
  dplyr::group_by(Tumor_Sample_Barcode, dataset) %>% 
  dplyr::summarize(n_mut = dplyr::n(), 
                   med_depth = median(n_depth)) %>% 
  ggplot2::ggplot(ggplot2::aes(x = med_depth, y = n_mut, color = dataset)) + 
  ggplot2::geom_point() + 
  ggplot2::facet_wrap(~dataset)
```

```{r}
mutect %>% 
  dplyr::group_by(Tumor_Sample_Barcode, dataset) %>% 
  dplyr::summarize(n_mut = dplyr::n(), 
                   med_depth = median(n_depth)) %>% 
  ggplot2::ggplot(ggplot2::aes(x = med_depth, y = n_mut, color = dataset)) + 
  ggplot2::geom_point() + 
  ggplot2::facet_wrap(~dataset)
```

```{r}
strelka  %>% 
  dplyr::group_by(Tumor_Sample_Barcode, dataset) %>% 
  dplyr::summarize(n_mut = dplyr::n(), 
                   med_depth = median(n_depth)) %>% 
  ggplot2::ggplot(ggplot2::aes(x = med_depth, y = n_mut, color = dataset)) + 
  ggplot2::geom_point() + 
  ggplot2::facet_wrap(~dataset)
```

## Is this a caller specific problem?? 

```{r}
tmb_combo <- function(pbta_df, tcga_df) {
# Calculate coding only TMBs and write to file
  pbta_coding_df <- calculate_tmb(pbta_df,
    bed_wgs = file.path(data_dir, "intersect_cds_lancet_strelka_mutect_WGS.bed"),
    bed_wxs = file.path(scratch_dir, "intersect_cds_WXS.bed")
    )

  # Calculate coding only TMBs and write to file
  tcga_coding_df <- calculate_tmb(tcga_df,
    bed_wgs = file.path(scratch_dir, "intersect_cds_lancet_strelka_mutect_WGS.bed"),
    bed_wxs = file.path(scratch_dir, "intersect_cds_WXS.bed")
    )

  tmb_df <- dplyr::bind_rows(list(pbta = pbta_df, 
                                  tcga = tcga_df), .id = "dataset") %>% 
    ggplot2::ggplot(ggplot2::aes(x = short_histology, y = log10(tmb), color = dataset)) + 
    ggforce::geom_sina() + 
    ggplot2::theme(
        axis.text.x = ggplot2::element_text(angle = 55, hjust = 1)) 
}
```

```{r}
tmb_combo(pbta_lancet, tcga_lancet)
```

```{r}
tmb_combo(pbta_mutect, tcga_mutect)
```

```{r}
tmb_combo(pbta_strelka, tcga_strelka)
```

## Session Info

```{r}
sessionInfo()
```