---
title: "Padded vs unpadded"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: C. Savonen for ALSF CCDL
date: 2019
---


Declare names of input and output directories.

```{r}
# magrittr pipe
`%>%` <- dplyr::`%>%`

data_dir <- file.path("..", "..", "data")
scratch_dir <- file.path("..", "..", "scratch")
results_dir <- "results"
plots_dir <- "plots"

# Load in these functions so we can use `maf_to_granges`
source(file.path("..", "snv-callers", "util", "tmb_functions.R"))
```


```{r}
metadata <- readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv")) %>% 
  dplyr::select(sample_id, experimental_strategy) %>% 
  tibble::rowid_to_column() %>% 
  tidyr::spread(experimental_strategy, sample_id) %>% 
  dplyr::filter(!is.na(WGS) & !is.na(WXS))

metadata
```
```{r}
wxs_padded <- readr::read_tsv(file.path(data_dir, "WXS.hg38.lancet.400bp_padded.bed"), 
                              col_names = FALSE)

# Turn the bed regions df into a GRanges object
padded_ranges <- GenomicRanges::GRanges(
  seqnames = wxs_padded$X1,
  ranges = IRanges::IRanges(
    start = wxs_padded$X2,
    end = wxs_padded$X3
))
    

# Turn the bed regions df into a GRanges object
unpadded_ranges <- GenomicRanges::GRanges(
  seqnames = wxs_padded$X1,
  ranges = IRanges::IRanges(
    start = wxs_padded$X2 + 400,
    end = wxs_padded$X3 - 400
))  
```

Connect to SQLite database.

```{r}
# Start up connection
con <- DBI::dbConnect(RSQLite::SQLite(), 
                      file.path(scratch_dir, "v11_snv_db.sqlite"))
```

Note what columns we will join by.

```{r}
join_cols = c("Chromosome",
              "Start_Position",
              "End_Position", 
              "Reference_Allele",
              "Allele",
              "Tumor_Sample_Barcode", 
              "Variant_Classification")
```

```{r}
lancet <- dplyr::tbl(con, "lancet") %>% 
  dplyr::select(join_cols, "VAF") %>% 
    dplyr::inner_join(
    dplyr::tbl(con, "samples") %>% 
      dplyr::select(Tumor_Sample_Barcode = Kids_First_Biospecimen_ID,
                    experimental_strategy, short_histology)) %>% 
    dplyr::filter(experimental_strategy == "WXS") %>% 
  as.data.frame()

lancet_wgs <- dplyr::tbl(con, "lancet") %>% 
  dplyr::select(join_cols, "VAF") %>% 
    dplyr::inner_join(
    dplyr::tbl(con, "samples") %>% 
      dplyr::select(Tumor_Sample_Barcode = Kids_First_Biospecimen_ID,
                    experimental_strategy, short_histology)) %>% 
    dplyr::filter(experimental_strategy == "WGS") %>% 
  as.data.frame()
```

```{r}
# Turn the MAF sample mutations into a GRanges object
lancet_granges <- maf_to_granges(lancet)

overlap <- GenomicRanges::findOverlaps(
  lancet_granges,
  unpadded_ranges
)
  
unpadded_lancet <- lancet[unique(overlap@from), ]

overlap <- GenomicRanges::findOverlaps(
  lancet_granges,
  padded_ranges
)
  
# Only keep those in the BED regions that overlap the `wxs_bed_granges`
padded_lancet <- lancet[unique(overlap@from), ]

lancet_combo <- dplyr::bind_rows(list(padded = padded_lancet, 
                                      unpadded = unpadded_lancet, 
                                      wgs = lancet_wgs), .id = "pad")

ggplot2::ggplot(lancet_combo, ggplot2::aes(x = pad, y = VAF, fill = pad)) + 
  ggplot2::geom_violin() + 
  ggplot2::theme_classic() + 
  ggplot2::ggtitle("Lancet VAF WXS samples only")
```
