---
title: "High-Grade Glioma Molecular Subtyping - Data Prep"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: Chante Bethell for ALSF CCDL
date: 2019
---

This notebook addresses the data wrangling for the purpose of molecular
subtyping High-grade glioma samples.

# Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

`Rscript -e "rmarkdown::render('analyses/molecular-subtyping-HGG/03-HGG-molecular-subtyping-data-prep.Rmd', clean = TRUE)"`

# Set Up

```{r}
# Install `DT` and `htmlwidgets` packages for displaying tables
if (!("DT" %in% installed.packages())) {
  install.packages("DT")
}
if (!("htmlwidgets" %in% installed.packages())) {
  install.packages("htmlwidgets")
}

library(tidyverse)
```

```{r}
# We can use functions from the `snv-callers` module of the OpenPBTA project
# TODO: if a common util folder is established, use that instead
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
# source(file.path(root_dir, "analyses", "snv-callers", "util",
#                  "tmb_functions.R"))
```

## Directories and Files

```{r}
# File path to results directory
input_dir <-
  file.path(root_dir, "analyses", "molecular-subtyping-HGG", "hgg-subset")

# File path to results directory
results_dir <-
  file.path(root_dir, "analyses", "molecular-subtyping-HGG", "results")

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# Read in output tsv with defining lesions as variables generated in 
# `01-HGG-molecular-subtyping-defining-lesions.Rmd`
hgg_df <- read_tsv(file.path(results_dir, "HGG_defining_lesions.tsv"))

# Read in metadata
metadata <- read_tsv(file.path(root_dir, "data", "pbta-histologies.tsv"))

# Select wanted columns in metadata for merging and assign to a new object
select_metadata <- metadata %>%
  select(sample_id,
         Kids_First_Biospecimen_ID)

# Read in HGG subset z-scored RNA expression data
stranded_expression <-
  read_rds(
    file.path(
      input_dir,
      "hgg_zscored_expression.RDS"
    )
  )

# Read in HGG subset focal CN data
cn_df <- read_tsv(
  file.path(
    input_dir,
    "hgg_focal_cn.tsv.gz"
  )
)

# Read in HGG subset fusion data
fusion_df <-
  read_tsv(file.path(input_dir,
                     "hgg_fusion.tsv"))

# Read in HGG subset GISTIC broad value data 
gistic_df <-
  read_tsv(file.path(input_dir, "hgg_gistic_broad_values.tsv"))

# Read in HGG subset snv maf data 
snv_df <-
  read_tsv(file.path(input_dir, "hgg_snv_maf.tsv"))
```

## Custom Function

```{r}
# Custom datatable function
# Function code adapted from: https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/49acc98f5ffd86853fc70f220623311e13e3ca9f/analyses/collapse-rnaseq/02-analyze-drops.Rmd#L23
viewDataTable <- function(data) {
  DT::datatable(
    data,
    rownames = FALSE,
    filter = "bottom",
    class = 'cell-border stripe',
    options = list(
      pageLength = 5,
      searchHighlight = TRUE,
      scrollX = TRUE,
      dom = 'tpi',
      initComplete = htmlwidgets::JS(
        "function(settings, json) {",
        "$(this.api().table().header()).css({'background-color':
                                            '#004467', 'color': '#fff'});",
        "}"
      )
    )
  )
}
```

# Prepare Data

## Metadata

```{r}
# Filter `HGG_defining_lesions.tsv` for samples classified and reclassified as
# HGG
hgg_df <- hgg_df %>%
  filter(
    disease_type_new == "High-grade glioma" |
      grepl("High-grade glioma", disease_type_reclassified)
  )

# Are there any duplicated sample IDs?
any(duplicated(hgg_df$sample_id))
```

```{r}
hgg_df %>% 
  group_by(sample_id) %>%
  filter(n() > 1) %>%
  arrange(sample_id)
```

## DNA-seq data

We will first join all DNA-seq assays using the `Kids_First_Biospecimen_ID` 
identifier.

### CN data

```{r}
# Filter focal CN data for PDGFRA, PTEN and MYCN status
# All of this is 'necessary' because there are some cases where a gene
# symbol has two calls (e.g., gain AND loss) in different rows
# This is not to say that there is not a more elegant way to do this...
genes_cn_df <- cn_df %>%
  # Drop the cytoband column
  select(-cytoband) %>%
  # Look at the genes of interest
  filter(gene_symbol %in% c("PDGFRA", "PTEN", "MYCN")) %>%
  # Spread such that the gene symbols become columns
  rowid_to_column("id") %>%
  group_by(biospecimen_id) %>%
  spread(gene_symbol, status) %>%
  # However each biospecimen may have had multiple rows in the original 
  # data frame and are now duplicated
  group_by(biospecimen_id) %>%
  # If something is NA, it won't get included in the paste step and instead 
  # becomes "" rather than NA -- this is more difficult to work with but we can
  # detect the absence of gain, loss, or amplification
  mutate(
    PDGFRA_focal_status = paste(sort(unique(PDGFRA)), collapse = ", "),
    PTEN_focal_status = paste(sort(unique(PTEN)), collapse = ", "),
    MYCN_focal_status = paste(sort(unique(MYCN)), collapse = ", ")
  ) %>%
  # Drop the columns that are no longer relevant and remove duplicates
  select(-c("id", "PDGFRA", "PTEN", "MYCN")) %>%
  distinct() %>%
  # Here if the value for a focal status column doesn't contain loss, gain,
  # or amplification -- call it neutral
  mutate_at(vars(ends_with("focal_status")),
            function(x) {
              case_when(
                !grepl("gain|loss|amplification", x) ~ "neutral",
                TRUE ~ x
              )
            })

# Add in the samples that did not have any copy number changes for either
# of these genes
missing_cn_bsid <-
  setdiff(unique(cn_df$biospecimen_id), genes_cn_df$biospecimen_id)
genes_cn_df <- metadata %>%
  filter(Kids_First_Biospecimen_ID %in% missing_cn_bsid) %>%
  select(sample_id,
         Kids_First_Participant_ID,
         Kids_First_Biospecimen_ID) %>%
  rename(biospecimen_id = Kids_First_Biospecimen_ID) %>%
  bind_rows(genes_cn_df) %>%
  # `bind_rows` fills with NAs -- if something in the focal status variables is
  # NA, there is no evidence for a copy number alteration in that gene and we
  # will call it neutral
  replace(is.na(.), "neutral")

# Display `genes_cn_df`
genes_cn_df %>%
  head(n = 15)
```

### GISTIC data

```{r}
# Mutate columns denoting whether or not each Kids_First_Biospecimen_ID has a 
# chromosomal loss for each of the selected chromosomes
gistic_df <- gistic_df %>%
  group_by(Kids_First_Biospecimen_ID) %>%
  mutate(
    chr_1p_loss = paste(sort(unique(
      case_when(`1p` < 0 ~ "Yes", TRUE ~ "No")
    )), collapse = ", "),
    chr_19q_loss = paste(sort(unique(
      case_when(`19q` < 0 ~ "Yes", TRUE ~ "No")
    )), collapse = ", "),
    chr_7_loss = paste(sort(unique(
      case_when(`7p` < 0 |
                  `7q` < 0 ~ "Yes", TRUE ~ "No")
    )), collapse = ", "),
    chr_10_loss = paste(sort(unique(
      case_when(`10p` < 0 |
                  `10q` < 0 ~ "Yes", TRUE ~ "No")
    )), collapse = ", ")
  ) %>%
  select(-c("1p", "19q", "7p", "7q", "10p", "10q"))

# Display `gistic_df`
gistic_df %>%
  head(n = 15)
```

```{r}
other_goi <- c("ACVR1", "TP53", "ATRX", "SETD2", "IDH1", "H3F3A", "IDH1")
```

```{r}
# Filter metadata for `High-grade glioma` and samples that should be classified
# as High-grade glioma based on defining lesions and are stranded RNA-seq
# samples
hgg_metadata <- metadata %>%
  filter(sample_id %in% hgg_df$sample_id) %>%
  mutate(age_at_diagnosis_years = as.numeric(age_at_diagnosis_days) /
           365) %>%
  select(
    sample_id,
    Kids_First_Participant_ID,
    Kids_First_Biospecimen_ID,
    age_at_diagnosis_years,
    glioma_brain_region
  ) %>%
  group_by(sample_id,
           Kids_First_Participant_ID) %>%
  summarize_all(function(x)
    paste(sort(unique(x)), collapse = ", "))

# Display metadata subsetted for HGG samples
hgg_metadata %>%
  head(n = 15)

# Merge output from `01-HGG-molecular-subtyping-defining-legions.Rmd`
final_df <- hgg_df %>%
  select(
    -Kids_First_Participant_ID,
    -Kids_First_Biospecimen_ID,
    -disease_type_new,
    -disease_type_reclassified
  ) %>%
  group_by(sample_id) %>%
  mutate(
    H3F3A.K28M = paste(sort(unique(H3F3A.K28M)), collapse = ", "),
    HIST1H3B.K28M = paste(sort(unique(HIST1H3B.K28M)), collapse = ", "),
    H3F3A.G35R.V = paste(sort(unique(
      H3F3A.G35R, H3F3A.G35V
    )), collapse = ", ")
  ) %>%
  select(-H3F3A.G35R,-H3F3A.G35V) %>%
  distinct() %>%
  right_join(metadata, by = "sample_id")
```

## Filter and Join RNA expression, CN, Fusion and GISTIC data 

### RNA expression data 

```{r}
# Filter to only the genes of interest and merge metadata with expression data
filtered_expression_metadata <- stranded_expression %>%
  as.data.frame() %>%
  select(FOXG1_expression_zscore = FOXG1,
         OLIG2_expression_zscore = OLIG2) %>%
  tibble::rownames_to_column("Kids_First_Biospecimen_ID") %>%
  left_join(select_metadata, by = "Kids_First_Biospecimen_ID") %>%
  select(-Kids_First_Biospecimen_ID) %>%
  distinct()

# Display `filtered_expression_metadata`
filtered_expression_metadata %>%
  head(n = 15)

# Join expression data with running final data.frame
final_df <- final_df %>%
  left_join(filtered_expression_metadata,
            by = "sample_id")

# Remove data we no longer need
rm(filtered_expression_metadata)
```



### SV data

```{r}
# Define target lesions
H3_K28 <-
  c("TP53",
    "H3F3A",
    "HIST1H3B",
    "ACVR1",
    "ATRX",
    "FGFR1",
    "IDH1",
    "BRAF")

H3_G35 <- c("H3F3A", "SETD2", "IDH1", "ATRX", "DAXX")

# Filter CN data for the target SV lesions and join this data.frame with
# the selected variables of the metadata
sv_df_filtered <- snv_df %>%
  filter(
    stringr::str_detect(Hugo_Symbol, H3_K28) |
      stringr::str_detect(Hugo_Symbol, H3_G35) |
      stringr::str_detect(Hugo_Symbol, c("P73-AS1", "CIC", "FUBP1"))
  ) %>%
  select(sample_id, Hugo_Symbol) %>%
  group_by(sample_id) %>%
  mutate(
    ACVR1_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "ACVR1") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    TP53_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "TP53") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    ATRX_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "ATRX") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    FGFR1_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "FGFR1") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    SETD2_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "SETD2") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    TERT_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "TERT") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    IDH1_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "IDH1") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    CIC_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "CIC") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    FUBP1_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "FUBP1") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    DAXX_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "DAXX") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", "),
    P73_AS1_mutated = paste(sort(unique(
      case_when(stringr::str_detect(Hugo_Symbol, "P73-AS1") ~ "Yes",
                TRUE ~ "No")
    )), collapse = ", ")
  ) %>%
  select(-Hugo_Symbol) %>%
  distinct()

# Display `sv_df_filtered`
sv_df_filtered %>%
  head(n = 15)

# Join SV data with the running final data.frame
final_df <- final_df %>%
  left_join(sv_df_filtered, by = "sample_id")

# Remove data we don't need
rm(sv_df_filtered)
```

### Fusion data 

```{r}
# Filter the fusion data to only `NTRK` fusions
fusion_df_filtered <- fusion_df %>%
  mutate(
    NTRK_fused = case_when(stringr::str_detect(FusionName, "NTRK") ~ "Yes",
                           TRUE ~ "No"),
    FGFR1_fused = case_when(stringr::str_detect(FusionName, "FGFR1") ~ "Yes",
                            TRUE ~ "No")
  ) %>%
  select(
    -c(
      "FusionName",
      "age_at_diagnosis_days",
      "glioma_brain_region",
      "Kids_First_Participant_ID",
      "Sample"
    )
  )

# Display `fusion_df_filtered`
fusion_df_filtered %>%
  head(n = 15)

# Join fusion data with the running final data.frame
final_df <- final_df %>%
  left_join(fusion_df_filtered, by = "sample_id") %>%
  distinct()

# Remove data we don't need
rm(fusion_df_filtered)
```



### SNV data

```{r}
snv_df <- snv_df %>%
  mutate(
    IDH1.R132H = case_when(Hugo_Symbol == "IDH1" &
                             HGVSp_Short == "p.R132H" ~ "Yes",
                           TRUE ~ "No")
  ) %>%
  select(sample_id, IDH1.R132H)

# Display `snv_df`
snv_df %>%
  head(n = 15)

# Join `IDH1-R132H` data with running final data.frame
final_df <- final_df %>%
  left_join(snv_df, by = "sample_id") %>%
  distinct()

# Remove data we no longer need
rm(snv_df)
```

# Save final table of results

```{r}
# Merge the manta data with the existing data.frame
final_df <- final_df %>%
  select(
    Kids_First_Participant_ID,
    sample_id,
    Kids_First_Biospecimen_ID,
    age_at_diagnosis_years,
    glioma_brain_region,
    H3F3A.K28M,
    HIST1H3B.K28M,
    ACVR1_mutated,
    TP53_mutated,
    ATRX_mutated,
    FGFR1_mutated,
    FGFR1_fused,
    PDGFRA_focal_status,
    PTEN_focal_status,
    IDH1_mutated,
    H3F3A.G35R.V,
    DAXX_mutated,
    SETD2_mutated,
    NTRK_fused,
    IDH1.R132H,
    everything()
  )
# NOTE: There are still duplicate rows for some `sample_id`s due to various
# `OLIG2` and `FOXG1` expression z-scores for multiple
# `Kids_First_Biospecimen_ID`s.

# Save final data.frame to file
write_tsv(final_df,
          file.path(results_dir, "HGG_molecular_subtypes.tsv"))

# Display `final_df`
viewDataTable(final_df)
```

# Session Info

```{r}
# Print the session information
sessionInfo()
```

