---
title: "High-Grade Glioma Molecular Subtyping - Gene Expression"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: Chante Bethell and Jaclyn Taroni for ALSF CCDL
date: 2020
---

This notebook prepares gene expression data for the purposes of subtyping HGG
samples ([`AlexsLemonade/OpenPBTA-analysis#249`](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/249)).

## Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/molecular-subtyping-HGG/06-HGG-molecular-subtyping-gene-expression.Rmd', clean = TRUE)"
```

## Set up

### Libraries and Functions

```{r}
library(tidyverse)
```

### Directories

```{r}
# Detect the ".git" folder -- this will in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no
# matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

# File path to results directory
input_dir <-
  file.path(root_dir, "analyses", "molecular-subtyping-HGG", "hgg-subset")

# File path to results directory
results_dir <-
  file.path(root_dir, "analyses", "molecular-subtyping-HGG", "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

### Read in Files

The gene expression files were generated by filtering z-scored expression data
(handling polyA selection and stranded expression files separately) in the `02-HGG-molecular-subtyping-subset-files.R` script found in this module, for the
purpose of subtyping HGG samples.

```{r}
# Read in subsetted polyA expression data
polya_expression_matrix <-
  read_rds(file.path(input_dir, "hgg_zscored_expression.polya.RDS"))

# Read in subsetted stranded expression data
stranded_expression_matrix <-
  read_rds(file.path(input_dir, "hgg_zscored_expression.stranded.RDS"))
```

### Output Files

```{r}
# File path to cleaned polyA expression output
polya_output_file <- file.path(results_dir, "HGG_cleaned_expression.polya.tsv")

# File path to cleaned stranded expression output
stranded_output_file <-
  file.path(results_dir, "HGG_cleaned_expression.stranded.tsv")
```

## Which genes' expression values are we interested in?

There are three genes of interest according to [`AlexsLemonade/OpenPBTA-analysis#249`](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/249).
For the IDH mutant HGG subtype:

* _OLIG2_ and _FOXG1_ should be highly expressed.
* _TP73-AS1_ methylation and downregulation cooccur with _TP53_ mutations.

## Clean and Wrangle Expression Data

```{r}
clean_wrangle_expression <-
  function(filtered_expression, output_file_path) {
    # Given an expression matrix filtered to include only samples that meet
    # the HGG classification criteria, select the genes of interest. It displays
    # and writes the cleaned and filtered expression table to the given output
    # file path.
    
    cleaned_expression <- filtered_expression %>%
      as.data.frame() %>%
      rownames_to_column("Kids_First_Biospecimen_ID") %>%
      dplyr::select(Kids_First_Biospecimen_ID,
             FOXG1_expression_zscore = FOXG1,
             OLIG2_expression_zscore = OLIG2,
             TP73_AS1_expression_zscore = `TP73-AS1`) %>%
      distinct()
    
    # Save expression data to file
    write_tsv(cleaned_expression, output_file_path)
    
    # Display `cleaned_expression`
    cleaned_expression
  }

# Run `clean_wrangle_expression` on poly-A filtered expression data
polya_cleaned_expression <-
  clean_wrangle_expression(polya_expression_matrix, polya_output_file)

# Run `clean_wrangle_expression` on stranded filtered expression data
stranded_cleaned_expression <-
  clean_wrangle_expression(stranded_expression_matrix, stranded_output_file)
```

## Density Plots

```{r}
plot_expression_density <- function(expression_df) {
  # Given a cleaned expression data.frame from the `clean_wrangle_expression`
  # function, generate a density plot for each of the three genes of interest.
  # This function returns three density plots (one per gene).
  
  # Plot FOXG1 expression data and save plot
  FOXG1_expression_density_plot <-
    ggplot(expression_df,
           aes(FOXG1_expression_zscore)) +
    geom_density(alpha = 0.1) +
    theme_classic() +
    theme(text = element_text(size = 8))

  # Plot OLIG2 expression data and save plot
  OLIG2_expression_density_plot <-
    ggplot(expression_df,
           aes(OLIG2_expression_zscore)) +
    geom_density(alpha = 0.1) +
    theme_classic() +
    theme(text = element_text(size = 8))

  # Plot TP73-AS1 expression data and save plot
  TP73_AS1_expression_density_plot <-
    ggplot(expression_df,
           aes(TP73_AS1_expression_zscore)) +
    geom_density(alpha = 0.1) +
    theme_classic() +
    theme(text = element_text(size = 8))
  
  # Use `cowplot::plot_grid` function to combine and display plots
  cowplot::plot_grid(
    FOXG1_expression_density_plot,
    OLIG2_expression_density_plot,
    TP73_AS1_expression_density_plot,
    labels = c("FOXG1", "OLIG2", "TP73-AS1"),
    label_size = 8,
    scale = 0.8,
    ncol = 3,
    nrow = 1
  )
}

# Run `plot_expression_density` on cleaned polyA data
plot_expression_density(polya_cleaned_expression)

# Run `plot_expression_density` on cleaned stranded data
plot_expression_density(stranded_cleaned_expression)
```

## Session Info

```{r}
# Print session information
sessionInfo()
```

