---
title: "High-Grade Glioma Molecular Subtyping - Combine DNA Assays"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: Chante Bethell and Jaclyn Taroni for ALSF CCDL
date: 2020
---

This notebook joins copy number alteration, mutation, fusion and gene expression data for the
purposes of subtyping HGG samples
([`AlexsLemonade/OpenPBTA-analysis#249`](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/249)).

All data are cleaned in notebooks upstream:

* Copy number data in [`03-HGG-molecular-subtyping-cnv.Rmd`](./03-HGG-molecular-subtyping-cnv.Rmd)
* Mutation data in [`04-HGG-molecular-subtyping-mutation.Rmd`](./04-HGG-molecular-subtyping-mutation.Rmd); specific defining lesions are prepared in [`01-HGG-molecular-subtyping-defining-lesions.Rmd`](./01-HGG-molecular-subtyping-defining-lesions.Rmd)
* Fusion data in [`05-HGG-molecular-subtyping-fusion.Rmd`](./05-HGG-molecular-subtyping-fusion.Rmd)
* Gene expression data in [`06-HGG-molecular-subtyping-gene-expression.Rmd`](./06-HGG-molecular-subtyping-gene-expression.Rmd)

## Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/molecular-subtyping-HGG/07-HGG-molecular-subtyping-combine-dna.Rmd', clean = TRUE)"
```

## Set up

### Libraries and Functions

```{r}
library(tidyverse)
```

### Directories

```{r}
# Detect the ".git" folder -- this will in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no
# matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))

# File path to results directory -- this contains the cleaned data
results_dir <-
  file.path(root_dir, "analyses", "molecular-subtyping-HGG", "results")

data_dir <- file.path(root_dir, "data")
```

### Read in Files

```{r message=FALSE}
cn_gistic_df <- read_tsv(file.path(results_dir, "HGG_cleaned_cnv.tsv"))
mutation_df <- read_tsv(file.path(results_dir, "HGG_cleaned_mutation.tsv"))
histologies_df <- read_tsv(file.path(data_dir, "pbta-histologies-base.tsv"), guess_max = 10000)
fusion_df <- read_tsv(file.path(results_dir, "HGG_cleaned_fusion.tsv"))
polya_exp <- read_tsv(file.path(results_dir,
                                "HGG_cleaned_expression.polya.tsv"))
stranded_exp <- read_tsv(file.path(results_dir,
                                   "HGG_cleaned_expression.stranded.tsv"))

# the cell line composition file comes from #610 in OpenPBTA_analysis repo
cell_line_composition <- read_tsv(file.path("input","cell-line-composition.tsv"),
                                  col_types =
                                    readr::cols( aliquot_id  = readr::col_character()))


```

### Output Files

```{r}
# full table
output_all <- file.path(results_dir, "HGG_cleaned_all_table.tsv")
# molecular subtype table
output_subtype <- file.path(results_dir, "HGG_molecular_subtype.tsv")
```

## Join Together

### DNA-seq Data (CNV, Mutation)

The CNV and SV data, and therefore the GISTIC data that is _downstream_ of the copy number consensus calls, only includes WGS samples.
Let's look at what biospecimen identifiers are present in the mutation data but not in the copy number data.

```{r}
histologies_df %>%
  filter(Kids_First_Biospecimen_ID %in%
           setdiff(mutation_df$Tumor_Sample_Barcode,
                   cn_gistic_df$Kids_First_Biospecimen_ID)) %>%
  count(experimental_strategy)
```

```{r}
dna_df <- mutation_df %>%
  rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>%
  left_join(cn_gistic_df)
```


```{r}
any(duplicated(dna_df$sample_id))
```
What are they?

If we only group by sample_id what does composition,aliquot_id and cell_line_composition look like:

```{r}
histologies_df %>%
  left_join(cell_line_composition) %>%
  filter(sample_id %in% dna_df$sample_id[duplicated(dna_df$sample_id)],
         experimental_strategy %in% c("WGS"),
         sample_type == "Tumor") %>%
  group_by(sample_id) %>%
  arrange(sample_id) %>%
  dplyr::select(Kids_First_Biospecimen_ID,
                sample_id,
                composition,
                cell_line_composition,
                aliquot_id) %>%
  summarize(
            composition =  paste(composition, collapse = ", "),
            cell_line_composition = paste(cell_line_composition, collapse = ", "),
            aliquot_id = paste(aliquot_id, collapse = ", "))

```
The duplicates are mostly from multiple Solid tumor samples per sample_id or there are non-tumor ( derived cell-line ) samples for sample_id .



```{r}
rm(mutation_df, cn_gistic_df)
```

### Join RNA-seq Data (Fusion, Gene Expression)

We have two cleaned gene expression data files corresponding to the stranded and poly-A datasets.
The (z-scored) distributions for the 3 genes of interest look pretty similar: https://alexslemonade.github.io/OpenPBTA-analysis/analyses/molecular-subtyping-HGG/06-HGG-molecular-subtyping-gene-expression.nb.html#density_plots
Therefore, we'll put all of this in the same table.

```{r}
exp_df <- bind_rows(polya_exp, stranded_exp)
rm(polya_exp, stranded_exp)
```

#### Digging into Fusion Data

There are some biospecimen identifiers that are in the expression files, but not the fusion files.
What are they?

```{r}
missing_bsids <- setdiff(exp_df$Kids_First_Biospecimen_ID,
                         fusion_df$Sample)

histologies_df %>%
  filter(Kids_First_Biospecimen_ID %in% missing_bsids)
```

No obvious pattern -- the `sample_type` and `composition` columns contain values we'd expect based on the retention of these identifiers.
Are these samples present in the files prior to subsetting in [`02-HGG-molecular-subtyping-subset-files.R`](./02-HGG-molecular-subtyping-subset-files.R)?

```{r}
# read in full file
putative_oncogenic_df <-
  read_tsv(file.path(root_dir,
    "analyses",
    "fusion_filtering",
    "results",
    "pbta-fusion-putative-oncogenic.tsv"))
```

```{r}
all(missing_bsids %in% putative_oncogenic_df$Sample)
```

```{r}
sum(!(missing_bsids %in% putative_oncogenic_df$Sample))
```

No, none of these identifiers are in the original putative oncogenic list.
What about the unfiltered fusion files?

STARFusion first.

```{r}
starfusion_df <- read_tsv(file.path(data_dir, "pbta-fusion-starfusion.tsv.gz"))
```

```{r}
all(missing_bsids %in% starfusion_df$tumor_id)
```

```{r}
arriba_df <- read_tsv(file.path(data_dir, "pbta-fusion-arriba.tsv.gz"))
```

```{r}
all(missing_bsids %in% arriba_df$tumor_id)
```

So all samples that are "missing" from the putative oncogenic fusion file are present in the Arriba file, but some are not present in the STARFusion file.
Putative oncogenic fusions are retained if they are detected by _either_ caller.
We will consider biospecimen identifiers that were present in the Arriba file to have no evidence for these fusions, rather than consider them missing data.

```{r}
fusion_df <- fusion_df %>%
  # Add in the instances where we have no evidence
  bind_rows(data.frame(Sample = missing_bsids)) %>%
  replace(is.na(.), "None")
```

```{r}
rm(arriba_df, putative_oncogenic_df, starfusion_df, missing_bsids)
```

### Join the Expression and Fusion Data

```{r}
rna_df <- left_join(exp_df, fusion_df,
                    by = c("Kids_First_Biospecimen_ID" = "Sample"))
head(rna_df, n = 10)
```

```{r}
rna_df <- histologies_df %>%
  select(Kids_First_Participant_ID,
         sample_id,
         Kids_First_Biospecimen_ID) %>%
  inner_join(rna_df)
```

```{r}
rm(fusion_df, exp_df)
```

Are there duplicated `sample_id` in the RNA data frame?

```{r}
any(duplicated(rna_df$sample_id))
```
What are they?

Checking for duplicate RNA sample_ids. If we only group by sample_id what does composition and cell_line_composition look like:

```{r}
histologies_df %>%
  left_join(cell_line_composition) %>%
  filter(sample_id %in% rna_df$sample_id[which(duplicated(rna_df$sample_id))],
         experimental_strategy == "RNA-Seq") %>%
  group_by(sample_id) %>%
  summarize(RNA_library = paste(RNA_library, collapse = ", "),
            composition =  paste(composition, collapse = ", "),
            cell_line_compositio = paste(cell_line_composition, collapse = ", "),
            aliquot_id = paste(aliquot_id, collapse = ", "))
```

The duplicates are mostly from multiple Solid tumor samples per sample_id or there are non-tumor ( derived cell-line ) samples for sample_id.


We are only trying to find a solution for multiple sample_ids per Kids_First_Participant_ID becuase of additional non-tumor `Derived Cell Line` samples .

Checking : if we group by `sample_id` and `aliquot_id` do we get unique matches between DNA and RNA?

```{r}

histologies_df %>%
  dplyr::filter(composition=="Derived Cell Line") %>%
  # grouping by sample_id and aliquot_id
  group_by(sample_id,aliquot_id) %>%
  # here we are checking if can aggregate DNA and RNA samples using
  # grouping by sample_id,aliquot_id
  summarise(experimental_strategy=paste(experimental_strategy,collapse = ", "),count=n()) %>%
  # if we can successfully aggregate `count` woulbe be >1
  dplyr::filter(count >1)

```
Don't find any aggregated DNA and RNA samples by grouping by `sample_id,aliquot_id`


If we group by `sample_id` and `cell_line_composition` do we get unique matches between DNA and RNA?
```{r}
histologies_df %>%
  left_join(cell_line_composition) %>%
  dplyr::filter(composition=="Derived Cell Line") %>%
  # grouping by sample_id and cell_line_composition
  group_by(sample_id,cell_line_composition) %>%
  # here we are checking if can aggregate DNA and RNA samples using
  # sample_id and cell_line_composition
  summarise(experimental_strategy=paste(experimental_strategy,collapse = ", "),count=n()) %>%
  # if we can successfully aggregate`count` woulbe be >1
  dplyr::filter(count >1)

```
We find 16 pairs of WGS and RNA-Seq samples grouping by sample_id and cell_line_composition. Let's add `cell_line_composition`column to dna and rna files to match `Derived Cell Line` samples appropriately in `All Data` section of notebook.

```{r}

dna_df <- dna_df %>%
  left_join(cell_line_composition,by = c("Kids_First_Participant_ID", "sample_id", "Kids_First_Biospecimen_ID"))

rna_df <- rna_df %>%
  left_join(cell_line_composition,by = c("Kids_First_Participant_ID", "sample_id", "Kids_First_Biospecimen_ID"))

```


### All Data

```{r}
all_data_df <- full_join(dna_df, rna_df,
                         by = c("Kids_First_Participant_ID",
                                "sample_id","cell_line_composition"),
                         suffix = c("_DNA", "_RNA")) %>%
  select(starts_with("Kids_First"), sample_id, everything())

head(all_data_df, n = 10)
```

Add in brain region and age.

```{r}
relevant_clinical_df <- histologies_df %>%
  filter(sample_id %in% all_data_df$sample_id,
         sample_type == "Tumor") %>%
  mutate(age_at_diagnosis_yr =
           floor(as.integer(age_at_diagnosis_days)/365)) %>%
  group_by(Kids_First_Participant_ID, sample_id) %>%
  summarize(CNS_region = paste(sort(unique(CNS_region)),
                               collapse = ", "),
            age_at_diagnosis_yr = paste(sort(unique(age_at_diagnosis_yr)),
                                        collapse = ", "))
```

```{r}
all_data_df <- inner_join(relevant_clinical_df,
                          all_data_df) %>%
  dplyr::arrange(Kids_First_Participant_ID, sample_id)
```

Write to file.

```{r}
write_tsv(all_data_df, output_all)
```

## Tables by Molecular Subtype

In this section, we'll divide up the table based on the molecular subtypes described in [`AlexsLemonade/OpenPBTA-analysis#249`](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/249).

### H3 K28 mutant

> * These tumors contain H3F3A K28M or HIST1H3B K28M mutations
> * Co-occurring lesions include: ACVR1, TP53, ATRX mutations; PDGFRA amplification; PTEN loss
> * Mutually-exclusive lesions: FGFR1 mutations/fusions (thalamic); IDH1 mutations; BRAF V600E (low-grade gliomas)
> * Average age of 9 years
> * Majority should be midline localized

```{r}
h3_k28_df <- all_data_df %>%
  # Only rows with H3 K28M mutations
  filter(H3F3A.K28M == "Yes" | HIST1H3B.K28M == "Yes" |
           HIST1H3C.K28M == "Yes" | HIST2H3C.K28M == "Yes") %>%
  # Only the relevant columns
  select(Kids_First_Participant_ID,
         sample_id,
         H3F3A.K28M,
         HIST1H3B.K28M,
         HIST1H3C.K28M,
         HIST2H3C.K28M,
         H3F3A.G35R,
         H3F3A.G35V,
         age_at_diagnosis_yr,
         CNS_region,
         relevant_coding_mutations,
         PTEN_focal_status,
         PDGFRA_focal_status,
         BRAF_V600E,
         FGFR1_fusions) %>%
  distinct()

h3_k28_df
```

Mutually exclusive with the H3 G35 mutations?

```{r}
any(h3_k28_df$H3F3A.G35R == "Yes")
```

```{r}
any(h3_k28_df$H3F3A.G35V == "Yes")
```

Are the majority midline localized?

```{r}
table(h3_k28_df$CNS_region)
```

Yes.

```{r}
h3_k28_df %>%
  group_by(relevant_coding_mutations) %>%
  count()
```

The coding mutation list looks like what we would expect based on the description, with the exception of some BRAF mutations.
There are no _IDH1_ mutations.

```{r}
h3_k28_df %>%
  group_by(PTEN_focal_status, PDGFRA_focal_status) %>%
  count()
```

The copy number information looks a bit different than what is described, but these are not lesions that must occur.

```{r}
summary(as.integer(h3_k28_df$age_at_diagnosis_yr))
```

Age is similar to what is described.

```{r}
table(h3_k28_df$FGFR1_fusions)
```

No _FGFR1_ fusions.

The histologies file that gets distributed as part of the project and contains molecular subtypes contains one row per biospecimen.
So we'll create a table that contains the biospecimen IDs and the calls.

```{r}
h3_k28_df_biospecimen <- all_data_df %>%
  filter(H3F3A.K28M == "Yes" | HIST1H3B.K28M == "Yes" |
           HIST1H3C.K28M == "Yes" | HIST2H3C.K28M == "Yes") %>%
  select(Kids_First_Participant_ID,
         sample_id,
         Kids_First_Biospecimen_ID_DNA,
         Kids_First_Biospecimen_ID_RNA,
         BRAF_V600E) %>%
  mutate(molecular_subtype = case_when(
    BRAF_V600E == "p.V600E" ~ "DMG, H3 K28, BRAF V600E",
    TRUE ~ "DMG, H3 K28"
  )) %>%
  select(-BRAF_V600E)
```

### H3 G35 mutant

> * These tumors contain H3F3A G35R/V mutations
> * Co-occurring lesions include: ATRX/DAXX, TP53, SETD2 mutations, NTRK fusions
> * Mutually exclusive lesions: IDH1 mutations
> * Average age of 20 years

```{r}
h3_g35_df <- all_data_df %>%
  # Only rows with H3 G35 mutations
  filter(H3F3A.G35R == "Yes" | H3F3A.G35V == "Yes") %>%
  # Only the relevant columns
  select(Kids_First_Participant_ID,
         sample_id,
         H3F3A.K28M,
         HIST1H3B.K28M,
         H3F3A.G35R,
         H3F3A.G35V,
         age_at_diagnosis_yr,
         relevant_coding_mutations,
         NTRK_fusions) %>%
  distinct()

h3_g35_df
```

Make a biospecimen-centric table.

```{r}
h3_g35_df_biospecimen <- all_data_df %>%
  # Only rows with H3 G35 mutations
  filter(H3F3A.G35R == "Yes" | H3F3A.G35V == "Yes") %>%
  select(Kids_First_Participant_ID,
         sample_id,
         Kids_First_Biospecimen_ID_DNA,
         Kids_First_Biospecimen_ID_RNA) %>%
  mutate(molecular_subtype = "HGG, H3 G35")
```

### IDH mutant

> * These tumors contain IDH1 R132H mutations
> * Co-occurring lesions include: TP53 mutations; P73-AS1 promoter methylation and downregulation
> * High expression of FOXG1 and OLIG2
> * Mutually exclusive lesions: chr7 gain and chr10 loss

```{r}
idh_df <- all_data_df %>%
  filter(IDH1_mutation == "p.R132H")
idh_df
```

In these instances, the broad copy data look consistent with the description, all tumors have _TP53_ mutations, and the _TP73-AS1_ expression is consistent with downregulation.

Make a biospecimen-centric table.

```{r}
idh_df_biospecimen <- all_data_df %>%
  filter(IDH1_mutation == "p.R132H") %>%
  select(Kids_First_Participant_ID,
         sample_id,
         Kids_First_Biospecimen_ID_DNA,
         Kids_First_Biospecimen_ID_RNA) %>%
  mutate(molecular_subtype = "HGG, IDH")
```


### H3.3 and IDH wildtype

> * High-grade gliomas absent of H3F3A and IDH mutations
> * Defining lesions: MYCN, PDGFRA amplification, TP53 and TERT mutations

```{r}
wildtype_df <- all_data_df %>%
  filter(H3F3A.K28M == "No",
         HIST1H3B.K28M == "No",
         HIST1H3C.K28M == "No",
         HIST2H3C.K28M == "No",
         H3F3A.G35R == "No",
         H3F3A.G35V == "No",
         !grepl("IDH1", relevant_coding_mutations)) %>%
  select(Kids_First_Participant_ID,
         sample_id,
         relevant_coding_mutations,
         TERT_variant_classification,
         MYCN_focal_status,
         PDGFRA_focal_status) %>%
  distinct()

wildtype_df %>% arrange(MYCN_focal_status, PDGFRA_focal_status)
```

We will label samples without H3 mutations as `HGG, H3 wildtype` below.

### 1p/19q co-deleted oligodendrogliomas

> * Co-deletion of chr 1p and 19q (LOH, loss of heterozygosity of both) results in translocation t(1p;19q)
> * Nearly all have a co-occurring IDH1 mutation (R132 or R172)
> * Other co-occurring lesions include: TERT promoter, CIC, and FUBP1 mutations
> * Mutually exclusive of TP53 and ATRX mutation
> * Typically occurs in adult tumors

Let's first filter by the broad arm data and see what we get.

```{r}
all_data_df %>%
  filter(`1p` == "loss" | `19q` == "loss")
```

We don't see the _IDH1_ mutations we would expect and we do see an H3 K28M mutation, so we will leave this off as well for the moment and in a subsequent notebook we'll address [`AlexsLemonade/OpenPBTA-analysis#435 (comment)`](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/435#issuecomment-576898275):

> We expect very few, if any of these (maybe 1 when we looked by CNV), so I think maybe we can approach this cohort by starting to look for 1p/19q codeletions in all samples + IDH mutations, and if we see any, we may manually want to check the CNV plots just to confirm.

### Samples without histone mutations that have _BRAF_ V600 mutations

From [`AlexsLemonade/OpenPBTA-analysis#435 (comment)`](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/435#issuecomment-576889305):

> A few patients ... initially had dx of HGG, have no histone mutations, but have only BRAF V600E mutations. These patients' biospecimens can possibly be LGGs. I think for these, we should note that they are exclusively `BRAF_V600E_mutant`

```{r}
braf_df_biospecimen <- all_data_df %>%
  filter(H3F3A.K28M == "No",
         HIST1H3B.K28M == "No",
         HIST1H3C.K28M == "No",
         HIST2H3C.K28M == "No",
         H3F3A.G35R == "No",
         H3F3A.G35V == "No",
         BRAF_V600E == "p.V600E") %>%
  select(Kids_First_Participant_ID,
         sample_id,
         Kids_First_Biospecimen_ID_DNA,
         Kids_First_Biospecimen_ID_RNA) %>%
  mutate(molecular_subtype = "BRAF V600E")
```

## Biospecimen Table

Join all the tables that contain the biospecimen ids and `molecular_subtype` column together.

```{r}
molecular_subtype_table <- bind_rows(h3_g35_df_biospecimen,
                                     h3_k28_df_biospecimen,
                                     idh_df_biospecimen,
                                     braf_df_biospecimen)

# We'll add HGG, H3 wildtype values in cases that do not yet have labels
molecular_subtype_table <- all_data_df %>%
  select(Kids_First_Participant_ID, sample_id,
         Kids_First_Biospecimen_ID_DNA, Kids_First_Biospecimen_ID_RNA) %>%
  left_join(molecular_subtype_table) %>%
  distinct() %>%
  dplyr::mutate(molecular_subtype = case_when(
    # molecular_subtype is NA and we didn't find any canonical histone
    # variant in the DNA sample so the status will be H3 wildtype
    is.na(molecular_subtype) & !is.na(Kids_First_Biospecimen_ID_DNA) ~ "HGG, H3 wildtype",
    # molecular_subtype is NA and we do not have a DNA sample
    # to check for histone variant so status will be
    # To be classified
    is.na(molecular_subtype) & is.na(Kids_First_Biospecimen_ID_DNA) ~ "HGG, To be classified",
    # when molecular_subtype is not NA (we know the H3 variant subtype from DNA )
    TRUE ~ molecular_subtype))
```

Write to file.

```{r}
write_tsv(molecular_subtype_table, output_subtype)
```

## Session Info

```{r}
sessionInfo()
```
