---
title: "HGAT samples without histone mutations that have `BRAF V600E` mutations"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
params:
  plot_ci: yes
  is_ci: FALSE
author: Chante Bethell for ALSF CCDL
date: 2020
---

This notebook will look at HGAT samples without histone mutations that have `BRAF V600E` mutations.

The purpose is to identify samples that may need to be reclassified as LGAT instead of HGAT using t-SNE and UMAP clustering.

## Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/molecular-subtyping-HGG/09-HGG-with-braf-clustering.Rmd', clean = TRUE, params=list(is_ci = ${IS_CI}))"
```

## Set up

### Libraries and functions

```{r}
# Load in tidyverse functions
library(tidyverse)
```

### CI
```{r}
# Assigning params$is_ci to running_in_ci avoids a locked binding error
running_in_ci <- params$is_ci
# Are we testing? In case of a non 0/1 number, we recast as logical, and then ensure logical.
if (running_in_ci %in% c(0,1)) running_in_ci <- as.logical(running_in_ci)
if (!(is.logical(running_in_ci)))
{
  stop("\n\nERROR: The parameter `is_ci` should be FALSE/TRUE (or 0/1).")
}
```

### Directories

```{r}
# File path to scratch directory -- this contains the cleaned data
scratch_dir <- file.path("..", "..", "scratch")

# File path to data directory
data_dir <- file.path("..", "..", "data")

# File path to plots directory -- this contains the multipanel output plots
plots_dir <- file.path("plots")

if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

```

### Functions

```{r}
# Source the functions for dimension reduction
source(
  file.path(
    "..",
    "transcriptomic-dimension-reduction",
    "util",
    "dimension-reduction-functions.R"
  )
)

# Alter the function to plot
plot_dimension_reduction <- function(expression_df,
                                     metadata_df,
                                     library,
                                     perplexity_param,
                                     score1 = 1,
                                     score2 = 2) {
  # Given a data.frame that contains the expression scores and the information
  # we want from the metadata, make a scatterplot.
  #
  # Args:
  #   expression_df: data.frame containing dimension reduction scores,
  #                           and relative information from the metadata
  #   metadata_df: data.frame containing the relevant metadata
  #   library: a string indicated the selection strategy used to produce the
  #            expression data
  #   score1: the column number of the first dimension reduction score that we
  #           want to plot, 1 by default
  #   score2: the column number of the second dimension reduction score that
  #           we want to plot, 2 by default
  #
  # Returns:
  #   dimension_reduction_plot: the plot representing the dimension reduction
  #                             scores in the given data.frame, using the
  #                             values in `broad_histology` as symbols to shape
  #                             and `molecular_subtype` to color the points
  
  
  # The first column will be either transcript or column ids
  feature_identifier <- colnames(expression_df)[1]
  
  # drop any columns that contain other identifers
  # expression_data <- expression_data %>%
  #   dplyr::select(!!rlang::sym(feature_identifier), dplyr::starts_with("BS_")) %>%
  #   dplyr::filter(complete.cases(.)) %>%
  #   tibble::column_to_rownames(var = feature_identifier)
  
  # Filter out low count genes
  genes_to_keep <-
    rowSums(expression_df) >= gene_count_threshold
  expression_df <- expression_df[genes_to_keep,]
  
  expression_df <- log2(expression_df + 1)
  
  # Transpose the data
  transposed_exp_data <- t(expression_df)
  
  # UMAP
  umap_file <- paste0("HGG_umap_", library)
  umap_aligned_scores <- dimension_reduction_wrapper(
    transposed_expression_matrix = transposed_exp_data,
    method = "UMAP",
    seed = 2020,
    metadata_df = metadata_df,
    filename_lead = umap_file,
    output_directory = scratch_dir,
    neighbors_parameter = 15
  )
  
  # t-SNE
  tsne_file <- paste0("HGG_tsne_", library)
  tsne_aligned_scores <- dimension_reduction_wrapper(
    transposed_expression_matrix = transposed_exp_data,
    method = "t-SNE",
    seed = 2020,
    metadata_df = metadata_df,
    filename_lead = tsne_file,
    output_directory = scratch_dir,
    perplexity_parameter = perplexity_param
  )
  
  # Plot
  # Empty lists that will hold plots
  plot_list <- list()
  
  # Plot using UMAP scores
  plot_list[["UMAP"]] <- ggplot2::ggplot(umap_aligned_scores,
                                         ggplot2::aes(
                                           x = dplyr::pull(umap_aligned_scores, score1),
                                           y = dplyr::pull(umap_aligned_scores, score2)
                                         )) +
    ggplot2::geom_point(alpha = 0.6,
                        ggplot2::aes(color = molecular_subtype,
                                     shape = broad_histology)) +
    ggplot2::scale_color_manual(values = (
      c(
        "#2b3fff", "#ec102f", "#235e31",
        "#1a6587", "#11e38c", "#a22f80",
        "#fe5900", "#1945c5", "#51f310",
        "#8b20d3", "#799d10", "#881c23",
        "#3fc6f8", "#fe5cde", "#0a7fb2",
        "#f2945a", "#6b4472", "#f4d403"
      )
    ), na.value = "#a6b6f9") +
    ggplot2::labs(x = "UMAP1", y = "UMAP2") +
    ggplot2::theme_bw()
  
  # Plot using t-SNE scores
  plot_list[["t-SNE"]] <- ggplot2::ggplot(tsne_aligned_scores,
                                          ggplot2::aes(
                                            x = dplyr::pull(tsne_aligned_scores, score1),
                                            y = dplyr::pull(tsne_aligned_scores, score2)
                                          )) +
    ggplot2::geom_point(alpha = 0.6,
                        ggplot2::aes(color = molecular_subtype,
                                     shape = broad_histology)) +
    ggplot2::scale_color_manual(values = (
      c(
        "#2b3fff", "#ec102f", "#235e31",
        "#1a6587", "#11e38c", "#a22f80",
        "#fe5900", "#1945c5", "#51f310",
        "#8b20d3", "#799d10", "#881c23",
        "#3fc6f8", "#fe5cde", "#0a7fb2",
        "#f2945a", "#6b4472", "#f4d403"
      )
    ), na.value = "#a6b6f9") +
    ggplot2::labs(x = "t-SNE1", y = "t-SNE2") +
    ggplot2::theme_bw()
  
  # Define output name of plot list
  output_filename <- paste0("HGG_", library, "_plot_list.RDS")
  
  # Save plot list
  readr::write_rds(plot_list,
                   path = file.path(scratch_dir, output_filename))
}

# Extract the individual plots from the list produced in
# `plot_dimension_reduction` function above
generate_multipanel_plot <- function(rds_filename) {
  # Given a data.frame that contains the scores of a dimension reduction
  # analysis and the information we want from the metadata, make a scatterplot.
  #
  # Args:
  #   rds_filename: the list of plots representing the dimension reduction scores
  #
  # Returns:
  #   final_plot: multipanel plot with visualizations of t-SNE and UMAP scores
  
  
  # First, we're going to infer some things about the plot from the name of
  # the input file -- specifically, the plot title and the output file name
  
  # Change file extension
  output_file <- sub("_plot_list.RDS", ".pdf", rds_filename)
  
  # Putting the title together
  plot_title <-
    toupper(paste(sub("_plot_list.RDS", "", rds_filename)))
  
  # We need to read in the list generated via get-plot-list.R
  plot_list <- readr::read_rds(file.path(scratch_dir, rds_filename))
  
  # Extract legend from the plot list + adjust text size
  plot_legend <- cowplot::get_legend(
    plot_list[[1]] +
      guides(color = guide_legend(ncol = 1)) +
      theme(
        text = element_text(size = 10),
        legend.box.margin = margin(15, 15, 15, 15)
      )
  )
  
  # We'll remove the legends from all the plots for arranging in a grid
  plot_list <- lapply(plot_list,
                      function(p)
                        p + theme(legend.position = "none"))
  
  # Multipanel plot of the dimension reduction scatter plots with panel labels
  multipanel_plot <-
    cowplot::plot_grid(
      plotlist = plot_list,
      nrow = 1,
      align = "h",
      labels = "AUTO"
    )
  
  # This is the plot title
  plot_title_panel <- cowplot::ggdraw() +
    cowplot::draw_label(plot_title, fontface = "bold", size = 15)
  
  # Add the title such that it is centered over the scatter plots
  multipanel_with_title <-
    cowplot::plot_grid(plot_title_panel,
                       multipanel_plot,
                       ncol = 1,
                       rel_heights = c(0.1, 1))
  
  # Pad the right side of the multipanel plot that now has a title
  multipanel_spacer <- cowplot::plot_grid(multipanel_with_title,
                                          NULL,
                                          nrow = 1,
                                          rel_widths = c(1, 0.05))
  
  # Add in the legend to the right of the multipanel plot
  final_plot <- cowplot::plot_grid(
    multipanel_with_title,
    plot_legend,
    nrow = 1,
    rel_widths = c(1, 0.25)
  ) +
    theme(plot.margin = unit(c(1, 1, 1, 1), "cm"))
  
  # Save to file
  cowplot::save_plot(
    file.path(plots_dir, output_file),
    final_plot,
    base_width = 24,
    base_height = 8
  )
  
  # Display final plot
  final_plot
}

```

### Read in data

```{r}
# Read in HGG molecular subtype tsv file -- output of `07-HGG-molecular-subtyping-combine-table.Rmd`
hgg_molecular_subtype_df <- readr::read_tsv(file.path("results",
                                                      "HGG_molecular_subtype.tsv"))

# Read in metadata
metadata_df <-
  readr::read_tsv(
    file.path(data_dir, "pbta-histologies.tsv"),
    col_types = cols(molecular_subtype = col_character())
  )

# Read in expression files
stranded_expression <- readr::read_rds(
  file.path(
    data_dir,
    "release-v13-20200116",
    "pbta-gene-expression-rsem-fpkm-collapsed.stranded.rds"
  )
)

polya_expression <- readr::read_rds(
  file.path(
    data_dir,
    "release-v13-20200116",
    "pbta-gene-expression-rsem-fpkm-collapsed.polya.rds"
  )
)

# Define gene count countoff 
gene_count_threshold <- 100
```

## Join HGG molecular subtype column with metadata

```{r}
# Join HGG molecular_subtype column with metadata and filter
metadata_df <- metadata_df %>%
  dplyr::left_join(hgg_molecular_subtype_df, by = "sample_id") %>%
  dplyr::group_by(sample_id) %>%
  dplyr::mutate(molecular_subtype = paste(sort(
    unique(molecular_subtype.x, molecular_subtype.y)
  ), collapse = ", ")) %>%
  dplyr::select(-molecular_subtype.x,-molecular_subtype.y) %>%
  dplyr::distinct() %>%
  dplyr::filter(
    broad_histology %in% c(
      "Low-grade astrocytic tumor",
      "Diffuse astrocytic and oligodendroglial tumor"
    )
  )

# Fill in blanks with NA's
metadata_df$molecular_subtype[metadata_df$molecular_subtype == ""] <- NA

```


## Dimension Reduction and Plotting

```{r fig.width = 22, fig.height = 10}
# Define the perplexity parameter for the t-SNE dimension reduction step
if (!(running_in_ci)) {
  perplexity_param = 10
} else {
  perplexity_param = 2
}

# Dimension reduction and plot list generation
stranded_plot_list <-
  plot_dimension_reduction(
    expression_df = stranded_expression,
    metadata_df = metadata_df,
    library = "stranded",
    perplexity_param = perplexity_param
  )

polya_plot_list <-
  plot_dimension_reduction(
    expression_df = polya_expression,
    metadata_df = metadata_df,
    library = "polya",
    perplexity_param = perplexity_param
  )

# Generate multipanel plots with custom function 
generate_multipanel_plot("HGG_stranded_plot_list.RDS")

generate_multipanel_plot("HGG_polya_plot_list.RDS")

```

## Session Info

```{r}
sessionInfo()
```

