---
title: "Tables output for manuscript"
author: "Run Jin, Stephanie J. Spielman, and Jo Lynne Rokita"
date: "2021-2022"
output:
  html_document: default
params:
  release: release-v21-20210820
---

```{r setup, include=FALSE}
library(tidyverse)
library(openxlsx)
```

## Output Tables for OpenPBTA Manuscript

This Rmarkdown generates tables used in the manuscript, including both main text and supplementary material. 

```{r define directories and result files}
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
working_dir <- file.path(root_dir, "tables")
analyses_dir <- file.path(root_dir, "analyses")

# Define data_dir based on ci. The `testing` param covers if we are in CI where 1 is CI.
if (params$release == "testing") {
  data_dir <- file.path(root_dir, "data")
} else {
  data_dir <- file.path(root_dir, "data", params$release)
}

results_dir <- file.path(working_dir, "results")
if(!dir.exists(results_dir)){
  dir.create(results_dir, recursive=TRUE)
}


# Define input files
histology_file <- file.path(data_dir, "pbta-histologies.tsv")
tmb_all_file <- file.path(data_dir, "pbta-snv-consensus-mutation-tmb-all.tsv")
tmb_coding_file <- file.path(data_dir, "pbta-snv-consensus-mutation-tmb-coding.tsv")
mutsigs_file <- file.path(analyses_dir, "mutational-signatures", "results", "deconstructsigs_exposures_merged.tsv")
chromo_file <- file.path(analyses_dir, "chromothripsis", "results", "chromothripsis_summary_per_sample.txt")
tp53_file <- file.path(analyses_dir, "tp53_nf1_score", "results", "tp53_altered_status.tsv")
quantiseq_file <- file.path(analyses_dir, "immune-deconv", "results", "quantiseq_deconv-output.rds")
telomerase_counts_polya_file <- file.path(analyses_dir, "telomerase-activity-prediction", "results", "TelomeraseScores_PTBAPolya_counts.txt")
telomerase_counts_stranded_file <- file.path(analyses_dir, "telomerase-activity-prediction", "results", "TelomeraseScores_PTBAStranded_counts.txt")
telomerase_fpkm_polya_file <- file.path(analyses_dir, "telomerase-activity-prediction", "results", "TelomeraseScores_PTBAPolya_FPKM.txt")
telomerase_fpkm_stranded_file <- file.path(analyses_dir, "telomerase-activity-prediction", "results", "TelomeraseScores_PTBAStranded_FPKM.txt")
CNS_match_json <- file.path(working_dir, "input", "CNS_primary_site_match.json")

# define output files and sheet names, when appropriate
table_1_file <- file.path(results_dir, "Table1-molecular-subtypes.tsv")
table_s1_file <- file.path(results_dir, "TableS1-histologies.xlsx")
table_s2_file <- file.path(results_dir, "TableS2-DNA-results-table.xlsx")
table_s3_file <- file.path(results_dir, "TableS3-RNA-results-table.xlsx")
table_s4_file <- file.path(results_dir, "TableS4-Key-Resources-table.xlsx")
```

## Table 1: Molecular subtypes determined for this project
```{r table 1}
histology_df <- read_tsv(histology_file, guess_max =10000)

histology_df %>%
  # remove rows to not include
  filter(!is.na(pathology_diagnosis),
         !is.na(molecular_subtype),
         !grepl("To be classified", molecular_subtype)) %>%
  # Count broad histology and subtype combinations
  count(broad_histology, molecular_subtype) %>%
  # export
  write_tsv(table_1_file)

```

## Table S1: V21 histologies table

```{r v21 histology as excel}
# Sheet 1: README tab

readme <- tribble(
~`Histology column`,~Definition,~`Possible values`,
"age_at_diagnosis_days","Patient age at diagnosis in days","numeric",
"age_last_update_days","Patient age at the last clinical event/update in days","numeric",
"aliquot_id","External aliquot identifier","alphanumeric",
"broad_histology","Broad WHO 2016 classification of cancer type",paste(unique(histology_df$broad_histology), collapse = "; "),
"cancer_group","Harmonized cancer groupings for plots",paste(unique(histology_df$cancer_group), collapse = "; "),
"cancer_predispositions","Reported cancer predisposition syndromes",paste(unique(histology_df$cancer_predispositions), collapse = "; "),
"CNS_region","Harmonized brain region based on `primary_site`",paste(unique(histology_df$CNS_region), collapse = "; "),
"cohort","Scientific cohort",paste(unique(histology_df$cohort), collapse = "; "),
"cohort_participant_id","Scientific cohort participant ID","C#####-C######",
"composition","Sample composition",paste(unique(histology_df$composition), collapse = "; "),
"ethnicity","Patient reported ethnicity",paste(unique(histology_df$ethnicity), collapse = "; "),
"experimental_strategy","Sequencing strategy",paste(unique(histology_df$experimental_strategy), collapse = "; "),
# leaving this non-programmatic because of the duplicates that would come up (eg two selections in one patient, needing data cleanup)
"extent_of_tumor_resection","Amount of tumor resected at time of surgical event","Biopsy only;Partial resection;Gross/Near total resection;Not Reported;Unavailable",
"germline_sex_estimate","Predicted sex of patient based on germline X and Y ratio calculation (described in methods)",paste(unique(histology_df$germline_sex_estimate), collapse = "; "),
"harmonized_diagnosis","`integrated_diagnosis` if exists or updated and harmonized diagnosis using pathology_free_text_diagnosis information","text",
"integrated_diagnosis","2016 WHO diagnosis integrated from pathology diagnosis and molecular subtyping","text",
"Kids_First_Biospecimen_ID","KidsFirst biospecimen identifier","BS_########",
"Kids_First_Participant_ID","KidsFirst patient identifier","PT_########",
"molecular_subtype","Molecular subtype defined by WHO 2016 guidelines","text",
"normal_fraction","Theta2 normal DNA fraction estimate","numeric",
"Notes","Free text field describing changes from `pathology_diagnosis` to `integrated_diagnosis` or manner in which molecular_subtype was determined","text",
"OS_days","Overall survival in days","numeric",
"OS_status","Overall survival status",paste(unique(histology_df$OS_status), collapse = "; "),
"parent_aliquot_id","External identifier combining sample_id; sample_type; aliquot_id; sequencing_strategy for some samples","alphanumeric",
"pathology_diagnosis","Reported and/or harmonized patient diagnosis from pathology reports","text",
"pathology_free_text_diagnosis","Free text patient diagnosis from pathology reports","text",
"PFS_days","Progression-free survival in days","numeric",
"primary_site","Bodily site(s) from which specimen was derived","text",
"race","Patient reported race",paste(unique(histology_df$race), collapse = "; "),
"reported_gender","Patient reported gender",paste(unique(histology_df$reported_gender), collapse = "; "),
"RNA_library","Type of RNA-Sequencing library preparation",paste(unique(histology_df$RNA_library), collapse = "; "),
"sample_id","External biospecimen identifier","alphanumeric",
"sample_type","Broad sample type",paste(unique(histology_df$sample_type), collapse = "; "),
"seq_center","Sequencing center",paste(unique(histology_df$seq_center), collapse = "; "),
"short_histology","Abbreviated `integrated_diagnosis` or `broad_histology` for plotting purposes",paste(unique(histology_df$short_histology), collapse = "; "),
"tumor_descriptor","Phase of therapy from which tumor was derived",paste(unique(histology_df$tumor_descriptor), collapse = "; "),
"tumor_fraction","Theta2 tumor DNA fraction estimate","numeric",
"tumor_ploidy","Control-FREEC ploidy","numeric"
)

# Sheet 2: Define array of samples that need to be classified from [ticket](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1206) which are not in v21, but set to be changed in final release [v22](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/1207)
to_be_classified_samples <- c("BS_3J4T2YYW", "BS_7F07M7JG", "BS_AG0BZN6F", "BS_HZNKSQ17", "BS_V6HKXX1Z", "BS_WP9J88EB", "BS_YYAPSA5P")

histologies_table <- histology_df %>% 
  # Update definitions of molecular subtype for TBD samples
  mutate(molecular_subtype = ifelse(
    Kids_First_Biospecimen_ID %in% to_be_classified_samples,
    "MB, To be classified",
    molecular_subtype
  ))

# Sheet 3: CNS region definition based on definitions from [Cassie Kline](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/tables/input/CNS_primary_site_match.json) with additional manual review of [HGG primary_site](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/1025). 
# This is integrated upstream of molecular subtyping

# CNS_region ~ primary_site matches
cns_regions_df <- purrr::imap_dfr(jsonlite::fromJSON(CNS_match_json), 
                                     function(x, name) { tibble::tibble(CNS_region = name, primary_site = paste0(x, collapse = ';')) })


# Combine and output
list_s1_table <- list(README = readme,
                      histologies_file = histologies_table,
                      CNS_region_definitions = cns_regions_df
                      )
write.xlsx(list_s1_table, 
           table_s1_file, 
           overwrite=TRUE, 
           keepNA=TRUE)

```

## Table S2: DNA results table
### TMB
```{r s2 DNA results part 1}

# read in tmb all file, select and rename columns
tmb_all <- read_tsv(tmb_all_file) %>%
  select(Tumor_Sample_Barcode, tmb) %>%
  rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>% 
  rename(Tmb_all = tmb) 

# read in tmb coding file, select and rename columns
tmb_coding <- read_tsv(tmb_coding_file) %>%
  select(Tumor_Sample_Barcode, tmb) %>%
  rename(Kids_First_Biospecimen_ID = Tumor_Sample_Barcode) %>% 
  rename(Tmb_coding = tmb)

# combine files
tmb_combined <- full_join(tmb_all, tmb_coding) 
```

### CNS mutational signatures with `deconstructSigs`
```{r s2 DNA results part 2}
# use the deconstructSigs table, not the sigfit table, for consistency with main text
# for convenience we can load the file deconstructsigs_exposures_merged.tsv which already has full sample information "merged" in _and_ uses the correct Signal RefSig names
cns_mut_list <- read_tsv(mutsigs_file)

cns_signatures <- cns_mut_list %>%
  select(Kids_First_Biospecimen_ID, exposure, signature) %>%
  # label signatures more clearly as RefSig or other
  mutate(signature = ifelse(
    signature == "Other", 
    "Other signatures",
    glue::glue("RefSig_{signature}")
  )) %>%
  spread(signature, exposure) %>%
  # order columns so "Other" is at the end
  select(Kids_First_Biospecimen_ID, contains("RefSig"), everything())

```

### Chromothripsis regions per sample
```{r s2 DNA results part 3}
chromothripsis_region_df <- readr::read_tsv(chromo_file)  

```

### combine S2 table
```{r s2 DNA results combined}
list_s2_table <- list(tumor_mutation_burden = tmb_combined,
                      cns_mutational_signatures = cns_signatures,
                      chromothripsis_events = chromothripsis_region_df
                     )
write.xlsx(list_s2_table, 
           table_s2_file,
           overwrite=TRUE, 
           # NA as NA and not as blank cells
           keepNA=TRUE)
```


## Table S3: RNA results table

### TP553 scores
```{r s3 RNA results table p1}
# get tp53 scores 
tp53_scores <- read_tsv(tp53_file) 
```

### Telomerase scores
```{r s3 RNA results table p2}
# Function to process the telomerase files
read_process_telomerase_file <- function(file, new_column_name, library_type) {
  read_tsv(file) %>%
    select(SampleID, NormEXTENDScores) %>%
    rename(Kids_First_Biospecimen_ID_RNA = SampleID,
           !!new_column_name := NormEXTENDScores) %>%
    mutate(library = library_type)
  
}
  
# Read in and process all telomerase scores
telomerase_scores_polya_count    <- read_process_telomerase_file(telomerase_counts_polya_file, "NormEXTENDScores_counts", "polyA")
telomerase_scores_stranded_count <- read_process_telomerase_file(telomerase_counts_stranded_file, "NormEXTENDScores_counts", "stranded")
telomerase_scores_polya_fpkm     <- read_process_telomerase_file(telomerase_fpkm_polya_file, "NormEXTENDScores_fpkm", "polyA")
telomerase_scores_stranded_fpkm  <- read_process_telomerase_file(telomerase_fpkm_stranded_file, "NormEXTENDScores_fpkm", "stranded")

# Combine all scores
telomerase_scores_combined <- 
  full_join(telomerase_scores_stranded_count,
          telomerase_scores_stranded_fpkm) %>%
  bind_rows(
    full_join(telomerase_scores_polya_count,
          telomerase_scores_polya_fpkm)
  ) %>%
  # order columns
  select(Kids_First_Biospecimen_ID_RNA,
         contains("Norm"), 
         library)


```

### Immune deconvolution with `quanTIseq` scores
```{r s3 RNA results table p3}
# get quantiseq immune scores file
quantiseq_immune <- read_rds(quantiseq_file) %>%
  select(sample, score, cell_type) %>% 
  rename(Kids_First_Biospecimen_ID_RNA = sample) %>%
  spread(cell_type, score)
```


### combine and output file
```{r s3 table combine and output}
list_s3_table <- list(tp53_scores = tp53_scores,
                      telomerase_scores = telomerase_scores_combined,
                      quantiseq_fractions = quantiseq_immune
                      )
write.xlsx(list_s3_table, 
           table_s3_file, 
           overwrite=TRUE, 
           keepNA=TRUE)

```

### Create `Key Resources` table S4

```{r table_s4}
# Sheet 1: R packages in Docker image
r_packages <- data.frame(installed.packages()[, c(1,3)]) %>%
  rename(`R Package` = "Package")

# Sheet 2: Python libraries in Docker image

# libraries not included in `requirements.txt`
python_libraries <- tribble(
  ~Software,~Version,
  "setuptools","46.3.0",
  "six","1.14.0",
  "wheel ","0.34.2"
)

# parse `requirements.txt`
python_requirements <- read_tsv(
  file.path(root_dir, "requirements.txt"), 
  col_names = FALSE) %>%
  separate(X1, into = c("Software", "Version"), sep = "==")

# merge `python_requirements` into `python_libraries`
python_libraries <- bind_rows(python_libraries, python_requirements)


# Sheet 3: Command line software in Docker image
cli_tools <- tribble(
~Software,~Version,~Source,
"bedtools","2.28.0","https://github.com/arq5x/bedtools2/",
"bedops","2.4.37","https://github.com/bedops/",
"HTSlib","1.9","https://github.com/samtools/htslib/",
"MCR","R2014a","https://www.mathworks.com/supportfiles/downloads/R2014a/",
"GISTIC2","2.0.23","ftp://ftp.broadinstitute.org/pub/GISTIC2.0/"
)



# Sheet 4: Tools used in workflows
workflow_tools <- tribble(
  ~"Software", ~"Version", ~"Associated Workflow(s) or Script",
  "NGSCheckMate","1.3","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/bcf_call.cwl",
  "ControlFreeC","11.6","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_combined_somatic_wgs_cnv_wf.cwl",
  "CNVKit","0.9.3","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_combined_somatic_wgs_cnv_wf.cwl",
  "strelka2","2.9.3","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_strelka2_mutect2_manta_workflow.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-mutect2_strelka2-wf.cwl",
  "mutect2 (GATK)","4.1.1.0","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_strelka2_mutect2_manta_workflow.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-mutect2_strelka2-wf.cwl",
  "manta","1.4.0","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_strelka2_mutect2_manta_workflow.cwl",
  "STAR aligner","2.6.1d","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_RNAseq_workflow.cwl",
  "rsem","1.3.1","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_RNAseq_workflow.cwl",
  "STAR-Fusion","1.5.0","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_RNAseq_workflow.cwl",
  "Arriba","1.1.0","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_RNAseq_workflow.cwl",
  "bwa","0.7.17","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-cram-only-wf.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-fq-input.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-fq-input-cram-only-wf.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-wf.cwl",
  "sambamba","0.6.3","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-cram-only-wf.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-fq-input.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-fq-input-cram-only-wf.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-wf.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-mendqc-wf.cwl",
  "GATK","4.0.3.0","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-cram-only-wf.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-fq-input.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-fq-input-cram-only-wf.cwl; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-alignment-wf.cwl",
  "lancet","1.0.7","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-lancet-wf.cwl",
  "UCSC-Treehouse MendQC","1.1.1","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-mendqc-wf.cwl",
  "GATK","4.0.12.0","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-mutect2_strelka2-wf.cwl",
  "VarDict Java v1.5.8","1.5.8","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc-vardict-wf.cwl",
  "VEP","r93","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_annot_vcf_sub_wf.cwl",
  "bcftools","1.7","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/cwl/kfdrc_annot_vcf_sub_wf.cwl",
  "GISTIC","2.0.23","https://github.com/d3b-center/OpenPBTA-workflows/blob/master/bash/run-gistic.sh ; https://github.com/d3b-center/OpenPBTA-workflows/blob/master/bash/run_gistic_consensus.sh"
)


# Combine and output
list_s4_table <- list(r_packages = r_packages,
                      python_libraries = python_libraries,
                      other_command_line_tools = cli_tools,
                      workflow_repository_tools = workflow_tools
                      )
write.xlsx(list_s4_table, 
           table_s4_file, 
           overwrite=TRUE, 
           keepNA=TRUE)

```


### print out session info
```{r}
sessionInfo()
```

