---
title: "Examining CNS signature fitting"
author: "S. Spielman"
date: "2021"
output:
  html_document:
    toc: yes
    df_print: paged
editor_options:
  chunk_output_type: console
---

<br><br><br><br>

#### Packages and paths


```{r}
`%>%` <- dplyr::`%>%`

plot_path <- "plots/"

maf_file <- file.path("..", "..", "scratch", "mutational-signatures", "pbta-snv-consensus-wgs.tsv.gz")
wgs_bed_file <- file.path("..", "..", "data", "WGS.hg38.strelka2.unpadded.bed")
histology_color_file <- file.path("..", "..", "figures", "palettes", "histology_label_color_table.tsv")

fitted_file <- file.path("results", "fitted_cns_signature_exposures.RDS")
```

```{r read_prep_data}
maf <- data.table::fread(maf_file, data.table = FALSE)

histology_label_mapping <- readr::read_tsv(histology_color_file) %>% 
  # Select just the columns we will need for plotting
  dplyr::select(display_group, display_order, hex_codes, Kids_First_Biospecimen_ID) %>% 
  # Reorder display_group based on display_order
  dplyr::mutate(display_group = forcats::fct_reorder(display_group, display_order)) 

# WGS information for calculating mutation per Mb
wgs_bed <- readr::read_tsv(wgs_bed_file, col_names=FALSE)
wgs_size <- sum(wgs_bed[, 3] - wgs_bed[, 2]) 


# Exposures into a tibble
raw_exposures <- readr::read_rds(fitted_file)
raw_exposures$mean %>%
  tibble::as_tibble(rownames = "Kids_First_Biospecimen_ID") %>%
  tidyr::gather(-Kids_First_Biospecimen_ID, 
                key = "signature", 
                value = "exposure") -> exposures


# Tabulate mutation counts
sigs_input <- deconstructSigs::mut.to.sigs.input(
  mut.ref = maf,
  sample.id = "Tumor_Sample_Barcode",
  chr = "Chromosome",
  pos = "Start_Position",
  ref = "Reference_Allele",
  alt = "Allele",
  bsg = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
)
```

```{r get_counts_exposures}
# Sample mutation counts per WGS Mb, considering only samples with >=50 mutations, since sigfit does not consider samples with <50
rowSums(sigs_input) %>%
  # throws a warning - ok for us since our package behavior is not going to change in Docker.
  tibble::as_tibble(rownames = "Kids_First_Biospecimen_ID") %>%
  dplyr::rename(mutation_count = "value") %>%
  dplyr::filter(mutation_count >= 50) %>%
  dplyr::mutate(sample_count_per_mb = mutation_count / (wgs_size / 10^6)) -> counts_per_mb


# Find sample counts for each signature
dplyr::left_join(counts_per_mb, exposures) %>%
  dplyr::mutate(signature_count_per_mb = sample_count_per_mb * exposure) %>%
  dplyr::select(-exposure) -> signature_counts
signature_counts
```
<br><br>

```{r barplot_mean_median_counts, fig.width = 8, fig.height = 4}

signature_counts %>%
  dplyr::select(-mutation_count, -sample_count_per_mb) %>%
  # Need to do calculations per grouping
  dplyr::inner_join(histology_label_mapping) %>% 
  dplyr::group_by(display_group, signature) %>% 
  dplyr::summarize(
    # proportion of samples within a group that actually contain the signature. If 1, then all samples have non-zero signature counts. This is expected and correct - we FITTED not EXTRACTED.
    prop_tumors_with_sig  = sum(signature_count_per_mb > 0) / dplyr::n(), # --> all of these values are 1, not worth plotting
    # Only calc for samples that have nonzero count. Do median and mean for comparison purposes
    median_count = median(signature_count_per_mb[signature_count_per_mb != 0]), 
    mean_count   = mean(signature_count_per_mb[signature_count_per_mb != 0])) %>%
  dplyr::ungroup() %>%
  # add back in groupings that summarize borks
  dplyr::inner_join(histology_label_mapping) %>%
  tidyr::gather("count_type",
                "count", 
                median_count:mean_count) -> cns_exposure_counts
cns_exposure_counts

cns_exposure_counts %>%
  #prettier labeling
  dplyr::mutate(signature_label = stringr::str_replace(signature, "refsig_", ""),
                count_type_label = stringr::str_replace(count_type, "_count", ""),
                count_type_label = stringr::str_to_title(count_type_label),
                count_type_label = paste( count_type_label, "number of mutations\nacross samples" )) %>%
  ggplot2::ggplot() + 
  ggplot2::aes(x = forcats::fct_reorder(signature_label, count, .desc=TRUE), y = count, fill = hex_codes) +
  ggplot2::geom_col() +
  ggplot2::scale_fill_identity() + 
  ggplot2::facet_wrap(~count_type_label, scales="free_y") +
  ggplot2::labs(x = "CNS RefSig Signature",
       y = "Mutation count per Mb",
       title = "Exposures across histology groups") -> mean_median_exposure_barplot

mean_median_exposure_barplot

ggplot2::ggsave( file.path(plot_path, "mean_median_exposure_barplot.png"), 
                 mean_median_exposure_barplot, 
                 width = 8,
                 height = 4)
```




