---
title: "CNV GISTIC Plots"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
---

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

```
Rscript -e "rmarkdown::render('analyses/cnv-chrom-plot/gistic_plot.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}
# Magrittr pipe
`%>%` <- dplyr::`%>%`
```

### Directories and Files

```{r}
# Path to input directory
input_dir <- file.path("..", "..", "data")
scratch_dir <- file.path("..", "..", "scratch")

# Path to output directory
plots_dir <- "plots"

# Create the plots_dir if it does not exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir, recursive = TRUE)
}
```

Extract file we need from GISTIC zip. 

```{r}
gistic_zip <- file.path(input_dir, "pbta-cnv-consensus-gistic.zip")
gistic_dir <- file.path(input_dir, "pbta-cnv-consensus-gistic")

if (!dir.exists(gistic_dir)) {
  unzip(gistic_zip,
    exdir = input_dir,
    files = file.path("pbta-cnv-consensus-gistic", "broad_values_by_arm.txt")
  )
}
```

### Read in data 

```{r}
# Read in metadata
metadata <-
  readr::read_tsv(file.path(input_dir, "pbta-histologies.tsv"))
```


```{r}
# Read in the segment copy number data
seg_df <- data.table::fread(file.path(
  "..",
  "focal-cn-file-preparation", 
  "results", 
  "consensus_seg_annotated_cn_autosomes.tsv.gz"), 
  data.table = FALSE) %>% 
  dplyr::filter(!is.na(cytoband))
```

```{r}
status_df <- seg_df %>% 
  dplyr::select(biospecimen_id, status, gene_symbol) %>% 
  dplyr::distinct(biospecimen_id, gene_symbol, .keep_all = TRUE) %>%
  tidyr::spread(gene_symbol, status) %>%
  tibble::column_to_rownames("biospecimen_id")

head(status_df)
```

## Set up heatmap annotation objects

Extract chromosome labels and make an alternating color key for them. 

```{r}
# Make a key for cytoband to gene
gene_cytoband_key <- seg_df$cytoband
names(gene_cytoband_key) <- seg_df$gene_symbol

# Make a key for assigning alternating colors to the cytobands
cytoband_colors <- rep(c("grey", "lightblue"), 
                  length.out = length(unique(gene_cytoband_key)))
names(cytoband_colors) <- unique(gene_cytoband_key)

# Gene color key for coloring alternating by cytoband
gene_color_key <- dplyr::recode(colnames(status_df), !!!cytoband_colors)
names(gene_color_key) <- colnames(status_df)
```

Make color key. 

```{r}
color_key <- structure(c("#f4a582", "#0571b0", "#ca0020"), 
                       names = unique(seg_df$status)) 
```

### Set up for making heatmaps of the breakpoints

Given the `GenomicRanges` objects for each sample, create a combined plot for 
each. 

Make histology labeling `HeatmapAnnotation` object.

```{r}
# Get the histologies for the samples in this set and order them by histology
histologies <-
  data.frame(Kids_First_Biospecimen_ID = rownames(status_df)) %>%
  dplyr::inner_join(metadata %>%
    dplyr::select(Kids_First_Biospecimen_ID, short_histology)) %>%
  dplyr::mutate(
    short_histology = tools::toTitleCase(short_histology),
    short_histology = as.factor(short_histology)
  ) %>%
  dplyr::filter(!is.na(short_histology)) %>%
  dplyr::arrange(short_histology) %>%
  tibble::column_to_rownames("Kids_First_Biospecimen_ID")

# TODO: Better colors
# Get values that can be used to make colors equi distant hues away for the
# number of histology groups we have
col_val <- seq(
  from = 0, to = 1,
  length.out = length(unique(histologies$short_histology))
)

# Translate into colors
histologies_colors <- hsv(h = col_val, s = col_val, v = 1)

# Make this named based on histology
names(histologies_colors) <- unique(histologies$short_histology)
```

```{r}
# Create the Heatmap annotation object
arm_annot <- ComplexHeatmap::HeatmapAnnotation(
  df = data.frame(gene_color_key),
  col = list(cytoband = gene_color_key),
  name = "",
  show_legend = FALSE,
  show_annotation_name = FALSE
)
```

```{r}
# Create the Heatmap annotation object
hist_annot <- ComplexHeatmap::HeatmapAnnotation(
  df = data.frame(histologies),
  col = list(short_histology = histologies_colors),
  which = "row",
  show_annotation_name = FALSE
)
```

```{r}
# Plot on a heatmap
heatmap <- ComplexHeatmap::Heatmap(as.matrix(status_df),
  heatmap_legend_param = list(title = "CN status"),
  col = color_key,
  cluster_columns = FALSE,
  cluster_rows = FALSE,
  show_column_names = FALSE,
  show_row_names = FALSE,
  bottom_annotation = arm_annot,
  left_annotation = hist_annot,
  na_col = "#f7f7f7"
)
```

## Print out heatmap

```{r}
heatmap
```


```{r}
# Save plot as PDF
pdf(file.path(plots_dir, "cn_status_heatmap.pdf"))
heatmap
dev.off()
```

# Session Info

```{r}
sessionInfo()
```

