---
title: "Summary of collapsed symbols"
output: html_notebook
---

```{r}
# load libraries
library(tidyverse)
library(reshape2)
library(refGenome)

# actual code
print("Generating input matrix...!")
expr <- readRDS(input.dat)
  
# reduce dataframe
expr <- expr[which(rowSums(expr[,2:ncol(expr)]) > 0),] # remove all genes with no expression
expr <- expr[grep('_PAR_', expr$gene_id, invert = T),] # discard PAR_* chromosomes from analysis

# collapse to matrix of HUGO symbols x Sample identifiers
# take mean per row and use the max value for duplicated gene symbols
expr.collapsed <- expr %>% 
  separate(gene_id, c("gene_id", "gene_symbol"), sep = "\\_", extra = "merge") %>%
  pivot_longer(-c(gene_id, gene_symbol), 
               names_to = "sample_name", values_to = "fpkm") %>% 
  group_by(gene_id) %>%
  mutate(means = mean(fpkm)) %>% 
  group_by(gene_symbol) %>%
  filter(means == max(means)) %>% 
  select(-gene_id) %>% unique()

# matrix of HUGO symbols x Sample identifiers
expr.input <- acast(expr.collapsed, gene_symbol~sample_name, value.var = 'fpkm')
print(dim(expr.input))

# save matrix
saveRDS(object = expr.input, file = output.mat)
print("Matrix generated. Done!!")

```

