---
title: "GISTIC entire PBTA cohort versus individual histology comparison of results"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: Chante Bethell for ALSF CCDL
date: 2020
---

This notebook will look at the files in the `run-gistic/results` directory and compare those for the entire cohort to those of three individual histologies, namely, LGAT, HGAT and medulloblastoma.

The purpose is to identify, if any, disagreement between GISTIC results for the entire cohort versus the three individual histologies.

## Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/gistic-cohort-vs-histology-comparison/gistic-cohort-vs-histology-comparison.Rmd', clean = TRUE)"
```

## Set up

```{r}
library(tidyverse)
```

### Files and Directories

```{r}
# Path to input directory
data_dir <- file.path("..", "run-gistic", "results")

# Path to output directory
plots_dir <- "plots"

# Create the plots_dir if it does not exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir, recursive = TRUE)
}
```

### Functions

```{r}
# Source `generate-multipanel-plot-functions.R` script
source(file.path("..",
                 "..",
                 "analyses",
                 "transcriptomic-dimension-reduction",
                 "util",
                 "generate-multipanel-plot-functions.R"))
```

## Compare "scores.gistic" files

### Custom function

```{r}
# Code adapted from `analyses/cnv-chrom-plot/gistic_plot.Rmd`
plot_gistic_scores <- function(gistic_scores_file) {
  # Read in and format gistic scores data
  gistic_scores <- data.table::fread(gistic_scores_file,
                                     data.table = FALSE) %>%
    dplyr::rename("gscore" = "G-score") %>%
    # Recode 23 and 24 as X and Y.
    dplyr::mutate(
      Chromosome = as.character(Chromosome),
      Chromosome = dplyr::recode(Chromosome,
                                 "23" = "X",
                                 "24" = "Y"),
      # Turn `Del` scores into negative `G-scores`
      # This is how GISTIC shows the scores.
      gscore = dplyr::case_when(Type == "Del" ~ -gscore,
                                TRUE ~ gscore)
    )
  
  # Make GISTIC data into GRanges object
  gistic_ranges <- GenomicRanges::GRanges(
    seqnames = gistic_scores$Chromosome,
    ranges = IRanges::IRanges(start = gistic_scores$Start,
                              end = gistic_scores$End),
    score = gistic_scores$gscore,
    mcols = gistic_scores
  )
  
  # Plot the GISTIC scores
  gistic_plot <-
    ggbio::autoplot(
      gistic_ranges,
      ggplot2::aes(y = score, fill = mcols.Type),
      geom = "bar",
      scales = "free_x",
      space = "free_x"
    ) +
    ggplot2::theme_classic() +
    ggplot2::theme(axis.text.x = ggplot2::element_text(
      size = 3,
      angle = 45,
      hjust = 1
    )) +
    colorblindr::scale_fill_OkabeIto(name = "Type") +
    ggplot2::ylab("G-scores") +
    ggplot2::scale_y_continuous(limits = c(-1, 1.2), breaks = seq(-1, 1.2, 0.2))
  
  # Return plot
  return(gistic_plot@ggplot)
  
}
```

### Unzip GISTIC results folders

```{r}
# Unzip and set up GISTIC results folders and files.
cohort_gistic_dir <- file.path(data_dir, "pbta-cnv-consensus-gistic")
cohort_gistic_zip <- file.path(data_dir, "pbta-cnv-consensus-gistic.zip")

if (!dir.exists(cohort_gistic_dir)) {
  unzip(cohort_gistic_zip, 
        exdir = data_dir)
}

lgat_gistic_dir <- file.path(data_dir, "pbta-cnv-consensus-lgat-gistic")
lgat_gistic_zip <- file.path(data_dir, "pbta-cnv-consensus-lgat-gistic.zip")

if (!dir.exists(lgat_gistic_dir)) {
  unzip(lgat_gistic_zip, 
        exdir = data_dir)
}

hgat_gistic_dir <- file.path(data_dir, "pbta-cnv-consensus-hgat-gistic")
hgat_gistic_zip <- file.path(data_dir, "pbta-cnv-consensus-hgat-gistic.zip")

if (!dir.exists(hgat_gistic_dir)) {
  unzip(hgat_gistic_zip, 
        exdir = data_dir)
}

medulloblastoma_gistic_dir <- file.path(data_dir, "pbta-cnv-consensus-medulloblastoma-gistic")
medulloblastoma_gistic_zip <- file.path(data_dir, "pbta-cnv-consensus-medulloblastoma-gistic.zip")

if (!dir.exists(medulloblastoma_gistic_dir)) {
  unzip(medulloblastoma_gistic_zip, 
        exdir = data_dir)
}

```

### Run `plot_gistic_scores` for each file

```{r}
# Define file paths to `scores.gistic` files and run `plot_gistic_scores`
# function for each file
cohort_scores <-
  file.path(cohort_gistic_dir, "scores.gistic")
cohort_scores_plot <- plot_gistic_scores(cohort_scores)

lgat_scores <-
  file.path(lgat_gistic_dir, "scores.gistic")
lgat_scores_plot <- plot_gistic_scores(lgat_scores)

hgat_scores <-
  file.path(hgat_gistic_dir, "scores.gistic")
hgat_scores_plot <- plot_gistic_scores(hgat_scores)

medulloblastoma_scores <-
  file.path(medulloblastoma_gistic_dir, "scores.gistic")
medulloblastoma_scores_plot <-
  plot_gistic_scores(medulloblastoma_scores)
```

### Generate multipanel plots 

```{r fig.height = 10, fig.width = 24}
lgat_plot_list <- list(lgat_scores_plot, cohort_scores_plot)
generate_multipanel_plot(plot_list = lgat_plot_list,
                         plot_title = "LGAT scores versus entire cohort",
                         output_directory = plots_dir,
                         output_filename = "lgat_gistic_scores_multipanel_plot.png")

hgat_plot_list <- list(hgat_scores_plot, cohort_scores_plot)
generate_multipanel_plot(plot_list = hgat_plot_list,
                         plot_title = "HGAT scores versus entire cohort",
                         output_directory = plots_dir,
                         output_filename = "hgat_gistic_scores_multipanel_plot.png")

medulloblastoma_plot_list <- list(medulloblastoma_scores_plot, cohort_scores_plot)
generate_multipanel_plot(plot_list = medulloblastoma_plot_list,
                         plot_title = "Medulloblastoma scores versus entire cohort",
                         output_directory = plots_dir,
                         output_filename = "medulloblastoma_gistic_scores_multipanel_plot.png")
```

## Compare `amp_genes.conf_90.txt` and `del_genes.conf_90.txt` files

### Custom function
```{r}
plot_amp_del_genes <-
  function(cohort_genes_file,
           histology_genes_file) {
    # Given the GISTIC result `amp_genes_conf_90.txt` or `del_genes_conf_90.txt`
    # files for the entire cohort and a specific histology, plot a Venn Diagram
    # showing the counts of rows that overlap/do not overlap between the two
    # files.
    #
    # Args:
    #   cohort_genes_file: the file path to the amp/del genes file for
    #                           the entire cohort
    #   histology_genes_file: the file path to the amp/del genes file
    #                              for an individual histology
    #
    # Returns:
    #  This function displays a Venn Diagram that represents the data that
    #  overlaps/does not overlap between the two given files
    
    # Read in files
    cohort_genes <-
      data.table::fread(cohort_genes_file,
                        data.table = FALSE)
    
    histology_genes <-
      data.table::fread(histology_genes_file,
                        data.table = FALSE)
    
    # Wrnagle data in cohort file for plotting
    cohort_genes <- t(cohort_genes)
    
    cohort_genes <- cohort_genes[-1,] %>%
      as.data.frame() %>%
      dplyr::mutate(summary_info = paste0(V1, ", ", V2, ", ", V3, ", ", V4)) %>%
      dplyr::select(-c("V1", "V2", "V3", "V4")) %>%
      dplyr::select(summary_info, dplyr::everything()) %>%
      tidyr::gather(V5, V6,-summary_info) %>%
      dplyr::group_by(summary_info) %>%
      dplyr::mutate(genes = paste0(sort(unique(V6)), collapse = ", ")) %>%
      dplyr::select(-V5,-V6) %>%
      dplyr::distinct() %>%
      dplyr::mutate(summary = paste0(summary_info, genes)) %>%
      dplyr::filter(!(summary == "NA, NA, NA, NA")) %>%
      dplyr::arrange()
    
    # Wrangle data in individual histology file for plotting
    histology_genes <- t(histology_genes)
    
    histology_genes <- histology_genes[-1,] %>%
      as.data.frame() %>%
      dplyr::mutate(summary_info = paste0(V1, ", ", V2, ", ", V3, ", ", V4)) %>%
      dplyr::select(-c("V1", "V2", "V3", "V4")) %>%
      dplyr::select(summary_info, dplyr::everything()) %>%
      tidyr::gather(V5, V6,-summary_info) %>%
      dplyr::group_by(summary_info) %>%
      dplyr::mutate(genes = paste0(sort(unique(V6)), collapse = ", ")) %>%
      dplyr::select(-V5,-V6) %>%
      dplyr::distinct() %>%
      dplyr::mutate(summary = paste0(summary_info, genes)) %>%
      dplyr::filter(!(summary == "NA, NA, NA, NA")) %>%
      dplyr::arrange()
    
    # Define vectors for Venn Diagram plotting
    cohort <- as.data.frame(cohort_genes$summary) %>%
      dplyr::rename(summary = `cohort_genes$summary`)
    
    individual_histology <- as.data.frame(histology_genes$summary) %>%
      dplyr::rename(summary = `histology_genes$summary`)
    
    
    # Define list for input in `venn` function
    input <-
      list(cohort = cohort,
           individual_histology = individual_histology)
    
    # Display Venn Diagram
    gplots::venn(input)
    
  }
```

### Run `plot_amp_del_genes` on each `amp_genes.conf_90.txt` file

```{r warning = FALSE, message = FALSE}
# Run comparison using LGAT file
cohort_amp_genes <- file.path(cohort_gistic_dir, "amp_genes.conf_90.txt")
lgat_amp_genes <- file.path(lgat_gistic_dir, "amp_genes.conf_90.txt")
plot_amp_del_genes(cohort_amp_genes, lgat_amp_genes)

# Run comparison using HGAT file
hgat_amp_genes <- file.path(hgat_gistic_dir, "amp_genes.conf_90.txt")
plot_amp_del_genes(cohort_amp_genes, hgat_amp_genes)

# Run comparison using medulloblastoma file
medulloblastoma_amp_genes <- file.path(medulloblastoma_gistic_dir, "amp_genes.conf_90.txt")
plot_amp_del_genes(cohort_amp_genes, medulloblastoma_amp_genes)
```

### Run `plot_amp_del_genes` on each `del_genes.conf_90.txt` file

```{r warning = FALSE, message = FALSE}
# Run comparison using LGAT file
cohort_del_genes <- file.path(cohort_gistic_dir, "del_genes.conf_90.txt")
lgat_del_genes <- file.path(lgat_gistic_dir, "del_genes.conf_90.txt")
plot_amp_del_genes(cohort_del_genes, lgat_del_genes)

# Run comparison using HGAT file
hgat_del_genes <- file.path(hgat_gistic_dir, "del_genes.conf_90.txt")
plot_amp_del_genes(cohort_del_genes, hgat_del_genes)

# Run comparison using medulloblastoma file
medulloblastoma_del_genes <- file.path(medulloblastoma_gistic_dir, "del_genes.conf_90.txt")
plot_amp_del_genes(cohort_del_genes, medulloblastoma_del_genes)
```

#### The above plots may not be the best visualizations for these particular files. I summarized the data for each row in the data.frames and used this summary info to compare the data.frames. I am not sure this was the best decision here.

#### The following plots are a series of `amp_qplot.png` files generated using GISTIC:

![Entire cohort amp plot](../run-gistic/results/pbta-cnv-consensus-gistic/amp_qplot.png)

![LGAT amp plot](../run-gistic/results/pbta-cnv-consensus-lgat-gistic/amp_qplot.png)

![HGAT amp plot](../run-gistic/results/pbta-cnv-consensus-hgat-gistic/amp_qplot.png)

![Medulloblastoma amp plot](../run-gistic/results/pbta-cnv-consensus-medulloblastoma-gistic/amp_qplot.png)

#### The following plots are a series of `amp_qplot.png` files generated using GISTIC:

![Entire cohort del plot](../run-gistic/results/pbta-cnv-consensus-gistic/del_qplot.png)

![LGAT del plot](../run-gistic/results/pbta-cnv-consensus-lgat-gistic/del_qplot.png)

![HGAT del plot](../run-gistic/results/pbta-cnv-consensus-hgat-gistic/del_qplot.png)

![Medulloblastoma del plot](../run-gistic/results/pbta-cnv-consensus-medulloblastoma-gistic/del_qplot.png)

## Session Info

```{r}
sessionInfo()
```

