---
title: "GISTIC specific histology versus entire PBTA cohort comparison of results"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: Chante Bethell for ALSF CCDL
date: 2020
---

This notebook will look at the files in the `run-gistic/results` directory and compare those for the entire cohort to those of three individual histologies, namely, LGAT, HGAT and medulloblastoma.

The purpose is to identify, if any, disagreement between GISTIC results for the entire cohort versus the three individual histologies.

## Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/gistic-cohort-vs-histology-comparison/gistic-cohort-vs-histology-comparison.Rmd', clean = TRUE)"
```

## Set up

```{r}
# Get the magrittr pipe
`%>%` <- dplyr::`%>%`
```

### Files and Directories

```{r}
# Path to input directory
data_dir <- file.path("..", "run-gistic", "results")

# Path to output directory
plots_dir <- "plots"

# Create the plots_dir if it does not exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir, recursive = TRUE)
}
```

### Functions

```{r}
# Source `generate-multipanel-plot-functions.R` script
source(file.path("..",
                 "..",
                 "analyses",
                 "transcriptomic-dimension-reduction",
                 "util",
                 "generate-multipanel-plot-functions.R"))
```

## Compare "scores.gistic" files

### Custom function

```{r}
# Code adapted from `analyses/cnv-chrom-plot/gistic_plot.Rmd`
plot_gistic_scores <- function(gistic_scores_file) {
  # Read in and format gistic scores data
  gistic_scores <- data.table::fread(gistic_scores_file,
                                     data.table = FALSE) %>%
    dplyr::rename("gscore" = "G-score") %>%
    # Recode 23 and 24 as X and Y.
    dplyr::mutate(
      Chromosome = as.character(Chromosome),
      Chromosome = dplyr::recode(Chromosome,
                                 "23" = "X",
                                 "24" = "Y"),
      # Turn `Del` scores into negative `G-scores`
      # This is how GISTIC shows the scores.
      gscore = dplyr::case_when(Type == "Del" ~ -gscore,
                                TRUE ~ gscore)
    )
  
  # Make GISTIC data into GRanges object
  gistic_ranges <- GenomicRanges::GRanges(
    seqnames = gistic_scores$Chromosome,
    ranges = IRanges::IRanges(start = gistic_scores$Start,
                              end = gistic_scores$End),
    score = gistic_scores$gscore,
    mcols = gistic_scores
  )
  
  # Plot the GISTIC scores
  gistic_plot <-
    ggbio::autoplot(
      gistic_ranges,
      ggplot2::aes(y = score, fill = mcols.Type),
      geom = "bar",
      scales = "free_x",
      space = "free_x"
    ) +
    ggplot2::theme_classic() +
    ggplot2::theme(axis.text.x = ggplot2::element_text(
      size = 3,
      angle = 45,
      hjust = 1
    )) +
    colorblindr::scale_fill_OkabeIto(name = "Type") +
    ggplot2::ylab("G-scores") +
    ggplot2::scale_y_continuous(limits = c(-1, 1.2), breaks = seq(-1, 1.2, 0.2))
  
  # Return plot
  return(gistic_plot@ggplot)
  
}
```

### Run `plot_gistic_scores` for each file

```{r}
cohort_scores <-
  file.path(data_dir, "pbta-cnv-consensus-gistic", "scores.gistic")
cohort_scores_plot <- plot_gistic_scores(cohort_scores)

lgat_scores <-
  file.path(data_dir, "pbta-cnv-consensus-lgat-gistic", "scores.gistic")
lgat_scores_plot <- plot_gistic_scores(lgat_scores)

hgat_scores <-
  file.path(data_dir, "pbta-cnv-consensus-hgat-gistic", "scores.gistic")
hgat_scores_plot <- plot_gistic_scores(hgat_scores)

medulloblastoma_scores <-
  file.path(data_dir,
            "pbta-cnv-consensus-medulloblastoma-gistic",
            "scores.gistic")
medulloblastoma_scores_plot <- plot_gistic_scores(medulloblastoma_scores)
```

### Generate multipanel plots 

```{r fig.height = 10, fig.width = 24}
lgat_plot_list <- list(lgat_scores_plot, cohort_scores_plot)
generate_multipanel_plot(plot_list = lgat_plot_list,
                         plot_title = "LGAT scores versus entire cohort",
                         output_directory = plots_dir,
                         output_filename = "lgat_gistic_scores_multipanel_plot.png")

hgat_plot_list <- list(hgat_scores_plot, cohort_scores_plot)
generate_multipanel_plot(plot_list = hgat_plot_list,
                         plot_title = "HGAT scores versus entire cohort",
                         output_directory = plots_dir,
                         output_filename = "hgat_gistic_scores_multipanel_plot.png")

medulloblastoma_plot_list <- list(medulloblastoma_scores_plot, cohort_scores_plot)
generate_multipanel_plot(plot_list = medulloblastoma_plot_list,
                         plot_title = "Medulloblastoma scores versus entire cohort",
                         output_directory = plots_dir,
                         output_filename = "medulloblastoma_gistic_scores_multipanel_plot.png")
```

## Compare "sample_seg_counts.txt" files

### Custom function
```{r}
plot_sample_seg_counts <-
  function(cohort_sample_seg_file,
           histology_sample_seg_file) {
    # Given the GISTIC result `sample_seg_counts.txt` files for the entire cohort
    # and a specific histology, plot a Venn Diagram showing the counts of rows that
    # overlap/do not overlap between the two files.
    #
    # Args:
    #   cohort_sample_seg_file: the file path to the sample seg counts file for
    #                           the entire cohort
    #   histology_sample_seg_file: the file path to the sample seg counts file
    #                              for an individual histology
    #
    # Returns:
    #  This function displays a Venn Diagram that represents the data that
    #  overlaps/does not overlap between the two given files
    
    # Read in files
    cohort_sample_seg_counts <-
      data.table::fread(cohort_sample_seg_file,
                        data.table = FALSE)
    
    histology_sample_seg_counts <-
      data.table::fread(histology_sample_seg_file,
                        data.table = FALSE)
    
    # Create a column that contains the information across columns for each row
    cohort_sample_seg_counts <- cohort_sample_seg_counts %>%
      dplyr::mutate(sample_seg_counts = paste0(sample, ", ", segment_count, ", ", included))
    
    histology_sample_seg_counts <- histology_sample_seg_counts %>%
      dplyr::mutate(sample_seg_counts = paste0(sample, ", ", segment_count, ", ", included))
    
    cohort_samples <-
      cohort_sample_seg_counts[!is.na(cohort_sample_seg_counts$sample_seg_counts), "sample_seg_counts"]
    histology_samples <-
      histology_sample_seg_counts[!is.na(histology_sample_seg_counts$sample_seg_counts), "sample_seg_counts"]
    
    # Define list for input in `venn` function
    input <-
      list(cohort_samples = cohort_samples, individual_histology_samples = histology_samples)
    
    # Plot Venn Diagram
    gplots::venn(input)
    
  }
```

### Run `plot_sample_seg_counts` on each file

```{r}
# Run comparison using LGAT file
cohort_sample_seg_counts <- file.path(data_dir, "pbta-cnv-consensus-gistic", "sample_seg_counts.txt")
lgat_sample_seg_counts <- file.path(data_dir, "pbta-cnv-consensus-lgat-gistic", "sample_seg_counts.txt")
plot_sample_seg_counts(cohort_sample_seg_counts, lgat_sample_seg_counts)

# Run comparison using HGAT file
hgat_sample_seg_counts <- file.path(data_dir, "pbta-cnv-consensus-hgat-gistic", "sample_seg_counts.txt")
plot_sample_seg_counts(cohort_sample_seg_counts, hgat_sample_seg_counts)

# Run comparison using medulloblastoma file
medulloblastoma_sample_seg_counts <- file.path(data_dir, "pbta-cnv-consensus-medulloblastoma-gistic", "sample_seg_counts.txt")
plot_sample_seg_counts(cohort_sample_seg_counts, medulloblastoma_sample_seg_counts)
```

## Session Info

```{r}
sessionInfo()
```

