---
title: "Molecularly Subtype Craniopharyngiomas into Adamantinomatous or Papillary"
author: "Daniel Miller <millerd15@email.chop.edu> and Jo Lynne Rokita <rokita@chop.edu> for D3B"
date: "October 2020"
output:
  html_notebook:
    toc: TRUE
    toc_float: TRUE
---

# Background

This notebook looks at the defining lesions for all samples for the issue of
molecular subtyping craniopharyngiomas into adamantinomatous or papillary samples
in the OpenPBTA dataset.

Defining Lesions:

CRANIO, ADAM
- Contains SNV in exon 3 of CTNNB1
- Tumor occurs mostly in childhood or young adolescence (0-39 years), but can be seen in adults

CRANIO, PAP
- Contains BRAF V600E mutation
- Tumor occurs exclusively in adults (40+ years)

# Usage

This notebook is intended to be run via the command line from the top directory
of the repository as follows:

```
Rscript -e "rmarkdown::render('analyses/molecular-subtyping-CRANIO/00-craniopharyngiomas-molecular-subtype.Rmd', clean = TRUE)"
```

# Set up

```{r load-library}
library(tidyverse)
```

## Directories and File Inputs/Outputs

```{r set-dir-and-file-names}
# Detect the ".git" folder -- this will in the project root directory.
# Use this as the root directory to ensure proper sourcing of functions no
# matter where this is called from
root_dir <- rprojroot::find_root(rprojroot::has_dir(".git"))
# File path to results directory
results_dir <-
  file.path(root_dir, "analyses", "molecular-subtyping-CRANIO", "results")
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# Outputs
output_lesions <- file.path(results_dir, "CRANIO_defining_lesions.tsv")
output_subtype <- file.path(results_dir, "CRANIO_molecular_subtype.tsv")
```

## SNV for subtyping 
 - We will use the `pbta-snv-consensus-mutation.maf.tsv.gz` from `snv-callers` module which gathers calls that are present in all 3 callers (strelka2,mutect2 and lancet)
 - In addition, we will also use `pbta-snv-scavenged-hotspots.maf.tsv.gz` from `hotspot-detection` module to gather calls that overlap MSKCC hotspots found in any caller (except if only vardict calls the site as variant, we remove these calls since we have a lot of calls unique to vardict which we consider as false positive as discussed [here](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/snv-callers#snv-caller-comparison-analysis))

```{r load-inputs-please-wait}
# Inputs
## Read in metadata
metadata <-
  read_tsv(file.path(root_dir, "data", "pbta-histologies-base.tsv")) %>%
  filter(pathology_diagnosis == "Craniopharyngioma")

snv_consensus_maf <- data.table::fread(
  file.path(root_dir, "data" , "pbta-snv-consensus-mutation.maf.tsv.gz"),
                                   select = c("Chromosome",
                                              "Start_Position",
                                              "End_Position",
                                              "Strand",
                                              "Variant_Classification",
                                              "Tumor_Sample_Barcode",
                                              "Hugo_Symbol",
                                              "HGVSp_Short",
                                              "Exon_Number"),
                                   data.table = FALSE) 

## Read in snv hotspot mutation data
snv_hotspot_maf <- data.table::fread(
  file.path(root_dir, "analyses" , "hotspots-detection" , "results" , "pbta-snv-scavenged-hotspots.maf.tsv.gz"),
                                   select = c("Chromosome",
                                              "Start_Position",
                                              "End_Position",
                                              "Strand",
                                              "Variant_Classification",
                                              "Tumor_Sample_Barcode",
                                              "Hugo_Symbol",
                                              "HGVSp_Short",
                                              "Exon_Number"),
                                   data.table = FALSE) %>%
  select(colnames(snv_consensus_maf))

snv_consensus_hotspot_maf <- snv_consensus_maf %>%
  bind_rows(snv_hotspot_maf) %>%
  unique()

```

# Process Data

## Collecting Tumor Sample Barcodes with Variants of Interest

Collect Tumor_Sample_Barcodes with BRAF V600E mutations.

```{r braf-tsb}
braf_v600e_tsb <-
  snv_consensus_hotspot_maf %>%
  filter(Hugo_Symbol == "BRAF" & HGVSp_Short == "p.V600E") %>%
  pull(Tumor_Sample_Barcode)
```

Collect Tumor_Sample_Barcodes with CTNNB1 mutation on exon 3.

```{r ctnnb1-tsb}
ctnnb1_e3_tsb <-
  snv_consensus_hotspot_maf %>%
  filter(Hugo_Symbol == "CTNNB1" & Exon_Number == "3/15") %>%
  pull(Tumor_Sample_Barcode)
```

## Defining Lesion Table
First, make a frame with DNA and RNA samples per patient as well as age at diagnosis.

```{r allmeta-table}
dnameta <-
  metadata %>%
  filter(experimental_strategy %in% c("WGS","WXS")) %>%
  select(Kids_First_Participant_ID,sample_id,Kids_First_Biospecimen_ID,age_at_diagnosis_days)

rnameta <-
  metadata %>%
  filter(experimental_strategy == "RNA-Seq") %>%
  select(Kids_First_Participant_ID,sample_id,Kids_First_Biospecimen_ID,age_at_diagnosis_days)

allmeta <-
  full_join(dnameta,
            rnameta,
            by=c("Kids_First_Participant_ID","sample_id", "age_at_diagnosis_days"),
            suffix = c("_DNA", "_RNA"))
```

Annotate the metadata table where the Kids_First_Biospecimen_ID_DNA matches the Tumor_Sample_Barcodes (TSBs) collected above.
Select the columns from the table that the PIs have requested.

```{r snv-lesion-table}
snv_lesion_df <-
  allmeta %>%
  mutate(CTNNB1.Exon3 = Kids_First_Biospecimen_ID_DNA %in% ctnnb1_e3_tsb) %>%
  mutate(BRAF.V600E = Kids_First_Biospecimen_ID_DNA %in% braf_v600e_tsb) %>%
  select(Kids_First_Participant_ID,
         sample_id,
         Kids_First_Biospecimen_ID_DNA,
         Kids_First_Biospecimen_ID_RNA,
         age_at_diagnosis_days,
         CTNNB1.Exon3,
         BRAF.V600E) %>%
  dplyr::arrange(Kids_First_Participant_ID, sample_id)
```

## Defining Molecular Subtype Table

Append molecular subtypes defined as follows:
1. `CRANIO, PAP`: patient has BRAF.V600E mutation and is older than the age_of_adulthood (40 years in this case)
1. `CRANIO, ADAM`:patient has CTNNB1.Exon3 mutation and is younger than the age_of_adulthood (40 years in this case)
1. `CRANIO, ADAM;CRANIO, PAP` Both of the above (this is actually impossible given the restriction on age)
1. `CRANIO, To be classified` None of the above or DNA sample does not exist
Select the columns from the table that the PIs have requested

```{r meta-subtype-table}
# Dispute this all you like
age_of_adulthood <- 40 * 365.25

snv_subtype <-
  snv_lesion_df %>%
  mutate(fourty_or_older = as.numeric(age_at_diagnosis_days) >= age_of_adulthood) %>%
  mutate(molecular_subtype =
           case_when(
             (BRAF.V600E & fourty_or_older) & !(CTNNB1.Exon3 & !fourty_or_older) ~ "CRANIO, PAP",
             !(BRAF.V600E & fourty_or_older) & (CTNNB1.Exon3 & !fourty_or_older) ~ "CRANIO, ADAM",
             (BRAF.V600E & fourty_or_older) & (CTNNB1.Exon3 & !fourty_or_older) ~ "CRANIO, ADAM;CRANIO, PAP",
             !(BRAF.V600E & fourty_or_older) & !(CTNNB1.Exon3 & !fourty_or_older) ~ "CRANIO, To be classified",
            is.na(Kids_First_Biospecimen_ID_DNA) ~ "CRANIO, To be classified"
            )) %>%
  select(Kids_First_Participant_ID,sample_id,molecular_subtype)

meta_subtype <-
  full_join(allmeta,
            snv_subtype,
            by=c("Kids_First_Participant_ID","sample_id")) %>%
  dplyr::arrange(Kids_First_Participant_ID, sample_id)
```

# Print Out Tables

```{r print-tables}
write_tsv(meta_subtype, output_subtype)
write_tsv(snv_lesion_df, output_lesions)
```

# SessionInfo

```{r session-info}
sessionInfo()
```
