---
title: "Unsupervised Analysis of Transcriptomic Differences"
author: Chante Bethell for CCDL 2019
output:
  html_notebook:
    toc: true
    toc_float: true
---

This notebook demonstrates samples in the PBTA cluster by experimental strategy
and cancer type using dimensionality reduction techniques, namely, 
[Principal Component Analysis](https://www.rdocumentation.org/packages/stats/versions/3.6.0/topics/prcomp) (PCA), 
[t-Distributed Stochastic Neighbor Embedding](https://cran.r-project.org/web/packages/Rtsne/Rtsne.pdf) (t-SNE), 
and [Uniform Manifold Approximation and Projection](https://cran.r-project.org/web/packages/umap/vignettes/umap.html) (UMAP).

It addresses [issue #9](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/9)
in the Open-PBTA analysis repository. 

## Output files:
- `analyses/transcriptomic-dimension-reduction/plots/meta_grid_kallisto_strategy.pdf`
- `analyses/transcriptomic-dimension-reduction/plots/meta_grid_kallisto_type.pdf`
- `analyses/transcriptomic-dimension-reduction/plots/meta_grid_rsem_strategy.pdf`
- `analyses/transcriptomic-dimension-reduction/plots/meta_grid_rsem_type.pdf`

# Usage

This script is intended to be run via the command line from the top directory of the repository as follows: 
```
Rscript -e "rmarkdown::render('analyses/transcriptomic-dimension-reduction/01-transcriptomic-analysis.Rmd', clean = TRUE)"
```

# Set Up

Set seed for reproducibility.
```{r}
set.seed(2019)
```

Install packages.
```{r, warning = FALSE, message = FALSE}
# This is needed for running the t-SNE analysis
if (!("Rtsne" %in% installed.packages())) {
  install.packages("Rtsne")
}
# This is needed for running the umap analysis
if (!("umap" %in% installed.packages())) {
  install.packages("umap")
}
```

Assign functions. 
```{r}
# magrittr pipe
`%>%` <- dplyr::`%>%`

align_metadata <- function(name, id) {
  # Given a data.frame that contains scores from a dimensionality reduction
  # technique, align the metadata to the data.frame in preparation for plotting
  #
  # Note: The rownames provided in the id argument are synonymous with the
  # metadata's `Kids_First_Biospecimen_ID`
  #
  # Args:
  #   name: Name of the data.frame containing dimension reduction scores
  #   id: Vector containing Biospecimen IDs extracted from the transposed
  #       expression data
  #
  # Returns:
  #   name: A data.frame containing the dimension reduction scores along with
  #         the variables from `metadata_df`
  
  # Assign the relevant rownames(Kids_First_Biospecimen_ID) to column named `Biospecimen_ID`
  name$Biospecimen_ID <- id
  # Align the metadata with the `Biospecimen_ID` variable in the data.frame
  metadata <- metadata_df %>%
    dplyr::filter(Kids_First_Biospecimen_ID %in% name$Biospecimen_ID) %>%
    dplyr::filter(!duplicated(Kids_First_Biospecimen_ID))
  # Filter the data.frame to contain only the `Biospecimen_ID` values common
  # to the metadata
  name <- name %>%
    dplyr::filter(Biospecimen_ID %in% metadata$Kids_First_Biospecimen_ID)
  # Assign Kids_First_Participant_ID to column `type` of the data.frame
  name$Kids_First_Participant_ID <-
    as.factor(metadata$Kids_First_Participant_ID)
  # Assign cancer type to column `type` of the data.frame
  name$disease_type_new <- as.factor(metadata$disease_type_new)
  # Filter data.frame for unique participant ID
  name <- name %>%
    dplyr::filter(!duplicated(Kids_First_Participant_ID))
  # Filter for the participant ids found in `name` data.frame
  metadata <- metadata_df %>%
    dplyr::select(Kids_First_Participant_ID, experimental_strategy) %>%
    dplyr::filter(Kids_First_Participant_ID %in% name$Kids_First_Participant_ID)
  # Filter the metadata for unique cases of `Kids_First_Participant_ID`
  metadata <- metadata %>%
    dplyr::filter(!duplicated(metadata$Kids_First_Participant_ID))
  # Assign the `experimental_strategy` variable from the metadata to 
  # `experimental_strategy` column in `name` data.frame
  name$experimental_strategy <- as.factor(metadata$experimental_strategy)
  return(name)
}

plot_dimension_reduction <- function(scores_df, point_color) {
  # Given a data.frame that contains the scores of a dimension reduction
  # analysis and the information we want from the metadata, make a scatterplot.
  #
  # Args:
  #   scores_df: data.frame containing dimension reduction scores, and
  #              relative information from the metadata
  #   point_color: the variable whose information will be used to color
  #                the points on the plot
  #
  # Returns:
  #   dimension_reduction_plot: the plot representing the dimension reduction
  #                             scores in the given data.frame, using the values
  #                             in `point_color` as symbols to color the points
  
  # transform the strings in `point_color` into symbols for plotting
  color_sym <- rlang::sym(point_color)
  
  dimension_reduction_plot <- ggplot2::ggplot(scores_df,
                                              ggplot2::aes(
                                                x = scores_df[, 1],
                                                y = scores_df[, 2],
                                                color = !!color_sym
                                              )) +
    ggplot2::geom_point() +
    ggplot2::theme(legend.position = "none")
}
```

Assign names of the output directories.
```{r}
# Assign name of output directory
data_dir <- file.path("..", "..", "data")
plots_dir <- "plots"

# Create directory to hold the output plots.
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

Read the metadata into a data.frame named `metadata_df`.
Read the expression data into data.frames named `exp_kallisto` and `exp_rsem`, 
respectively.
```{r}
# Read in dataset
metadata_df <-
  data.frame(readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv")))

# Read in kallisto expression data
exp_kallisto <- data.frame(readr::read_rds(file.path(
  data_dir, "pbta-gene-expression-kallisto.rds"
)))

# Read in RSEM expression data
exp_rsem <- data.frame(readr::read_rds(file.path(
  data_dir, "pbta-gene-expression-rsem.fpkm.rds"
)))
```

# Prep Data
Make the expression data.frame all numeric by transforming the non-numeric
`gene_id` column into rownames. 
Filter out the 
Transpose the data and save the rownames of the transposed data.frames for 
alignment to the metadata. 

```{r}
# Transform the non-numeric "gene_id" column into rownames
exp_kallisto <- exp_kallisto[, -1] %>%
  dplyr::filter(!duplicated(gene_id)) %>%
  tibble::column_to_rownames("gene_id") %>%
  na.omit()

exp_rsem <- exp_rsem %>%
  tibble::column_to_rownames("gene_id") %>%
  na.omit()

# Filter out low count genes
kallisto_genes_to_keep <- rowSums(exp_kallisto) >= 100
exp_kallisto <- exp_kallisto[kallisto_genes_to_keep, ]

rsem_genes_to_keep <- rowSums(exp_rsem) >= 100
exp_rsem  <- exp_rsem[rsem_genes_to_keep, ]

# Transpose the data
transposed_rsem_data <- t(exp_rsem)
transposed_kallisto_data <- t(exp_kallisto)

# Save rownames as a vector
rsem_ID <- rownames(transposed_rsem_data)
kallisto_ID <- rownames(transposed_kallisto_data)
```

# RSEM 

## Run PCA

```{r}
# Run PCA on RSEM
rsem_pca <- prcomp(transposed_rsem_data)
# Make a data.frame with PCA scores
rsem_pca_data <- data.frame(rsem_pca$x[, 1:2])
# Run the align_metadata function
rsem_pca_data <- align_metadata(rsem_pca_data, rsem_ID)
# Run the plot_dimension_reduction function and color by cancer type
rsem_pca_plot <- plot_dimension_reduction(rsem_pca_data, "disease_type_new")
# Run the plot_dimension_reduction function and color by experimental strategy
rsem_pca_plot_strategy <-
  plot_dimension_reduction(rsem_pca_data, "experimental_strategy")
```

## Run t-SNE

```{r}
# Run t-SNE on RSEM
rsem_tsne <- Rtsne::Rtsne(transposed_rsem_data)
# Make a data.frame with t-SNE scores
rsem_tsne_data <- data.frame(rsem_tsne$Y)
# Run the align_metadata function
rsem_tsne_data <- align_metadata(rsem_tsne_data, rsem_ID)
# Run the plot_dimension_reduction function and color by cancer type
rsem_tsne_plot <- plot_dimension_reduction(rsem_tsne_data, "disease_type_new")
# Run the plot_dimension_reduction function and color by experimental strategy
rsem_tsne_plot_strategy <-
  plot_dimension_reduction(rsem_tsne_data, "experimental_strategy")
```

## Run UMAP

```{r}
# Run UMAP on RSEM
rsem_umap <- umap::umap(transposed_rsem_data)
# Make a data.frame with umap scores
rsem_umap_data <- data.frame(rsem_umap$layout)
# Run the align_metadata function
rsem_umap_data <- align_metadata(rsem_umap_data, rsem_ID)
# Run the plot_dimension_reduction function and color by cancer type
rsem_umap_plot <- plot_dimension_reduction(rsem_umap_data, "disease_type_new") +
  ggplot2::theme(legend.position = "bottom")
# Run the plot_dimension_reduction function and color by experimental strategy
rsem_umap_plot_strategy <-
  plot_dimension_reduction(rsem_umap_data, "experimental_strategy") +
  ggplot2::theme(legend.position = "bottom")

# Extract the plots' legends
legend_a <- cowplot::get_legend(
  rsem_umap_plot +
    ggplot2::guides(color = ggplot2::guide_legend(nrow = 1)) +
    ggplot2::theme(legend.box.margin = ggplot2::margin(0, 0, 0, 30))
)
legend_b <- cowplot::get_legend(
  rsem_umap_plot_strategy +
    ggplot2::guides(color = ggplot2::guide_legend(nrow = 5)) +
    ggplot2::theme(legend.box.margin = ggplot2::margin(0, 0, 0, 18))
)
```

## RSEM Plots Grid 
The grid of plots below shows the dimension reductions scores for the RSEM data,
colored by experimental sequencing strategy and tumor type, respectively.  
These plots seem to suggest the presence of batch effects, which is likely due 
to the need for normalization across the two batches of RNA-Seq preparation 
(poly-A versus stranded).

```{r, fig.width = 17, fig.height = 12}
# Plot grid with RSEM data colored by experimental strategy
rsem_grid_strategy <- cowplot::plot_grid(
  rsem_pca_plot_strategy,
  rsem_tsne_plot_strategy,
  rsem_umap_plot_strategy + ggplot2::theme(legend.position = "none"),
  align = 'vh',
  labels = "AUTO",
  nrow = 1
)
# Add the extracted legend to the plot grid
rsem_grid_strategy <- cowplot::plot_grid(rsem_grid_strategy, legend_b, rel_widths = c(3, .4))
# Plot grid with RSEM data colored by cancer type
rsem_grid <- cowplot::plot_grid(
  rsem_pca_plot,
  rsem_tsne_plot,
  rsem_umap_plot + ggplot2::theme(legend.position = "none"),
  align = 'vh',
  labels = "AUTO",
  nrow = 1
)
# Add the extracted legend to the plot grid
rsem_grid <- cowplot::plot_grid(rsem_grid, legend_a, rel_widths = c(4, .2))
# Save grids
ggplot2::ggsave(
  file.path(plots_dir, "meta_grid_rsem_strategy.pdf"),
  rsem_grid_strategy,
  width = 18,
  height = 12
)
ggplot2::ggsave(
  file.path(plots_dir, "meta_grid_rsem_type.pdf"),
  rsem_grid,
  width = 18,
  height = 12
)
rsem_grid_strategy
rsem_grid
```


# Kallisto

## Run PCA

```{r}
# Run PCA on kallisto
kallisto_pca <- prcomp(transposed_kallisto_data)
# Make a data.frame with PCA scores
kallisto_pca_data <- data.frame(kallisto_pca$x[, 1:2])
# Run the align_metadata function
kallisto_pca_data <- align_metadata(kallisto_pca_data, kallisto_ID)
# Run the plot_dimension_reduction function and color by cancer type
kallisto_pca_plot <-
  plot_dimension_reduction(kallisto_pca_data, "disease_type_new")
# Run the plot_dimension_reduction function and color by experimental strategy
kallisto_pca_plot_strategy <-
  plot_dimension_reduction(kallisto_pca_data, "experimental_strategy")
```

## Run t-SNE 

```{r}
# Run t-SNE on kallisto
kallisto_tsne <- Rtsne::Rtsne(transposed_kallisto_data)
# Make a data.frame with t-SNE scores
kallisto_tsne_data <- data.frame(kallisto_tsne$Y)
# Run the align_metadata function
kallisto_tsne_data <- align_metadata(kallisto_tsne_data, kallisto_ID)
# Run the plot_dimension_reduction function and color by cancer type
kallisto_tsne_plot <-
  plot_dimension_reduction(kallisto_tsne_data, "disease_type_new")
# Run the plot_dimension_reduction function and color by experimental strategy
kallisto_tsne_plot_strategy <-
  plot_dimension_reduction(kallisto_tsne_data, "experimental_strategy")
```

## Run UMAP 

```{r}
# Run UMAP on kallisto
kallisto_umap <- umap::umap(transposed_kallisto_data)
# Make a data.frame with umap scores
kallisto_umap_data <- data.frame(kallisto_umap$layout)
# Run the align_metadata function
kallisto_umap_data <- align_metadata(kallisto_umap_data, kallisto_ID)
# Run the plot_dimension_reduction function and color by cancer type
kallisto_umap_plot <-
  plot_dimension_reduction(kallisto_umap_data, "disease_type_new") +
  ggplot2::theme(legend.position = "bottom")
# Run the plot_dimension_reduction function and color by experimental strategy
kallisto_umap_plot_strategy <-
  plot_dimension_reduction(kallisto_umap_data, "experimental_strategy") +
  ggplot2::theme(legend.position = "bottom")

# Extract the plots' legends
legend_a <- cowplot::get_legend(
  kallisto_umap_plot +
    ggplot2::guides(color = ggplot2::guide_legend(nrow = 1)) +
    ggplot2::theme(legend.box.margin = ggplot2::margin(0, 0, 0, 30))
)
legend_b <- cowplot::get_legend(
  kallisto_umap_plot_strategy +
    ggplot2::guides(color = ggplot2::guide_legend(nrow = 5)) +
    ggplot2::theme(legend.box.margin = ggplot2::margin(0, 0, 0, 12))
)
```

## Kallisto Plots Grid
The grid of plots below shows the dimension reductions scores for the kallisto 
data, colored by experimental sequencing strategy and tumor type, respectively. 
These plots seem to suggest the presence of batch effects, which is likely due 
to the need for normalization across the two batches of RNA-Seq preparation 
(poly-A versus stranded).

```{r, fig.width = 17, fig.height = 12}
# Plot grid with kallisto data colored by experimental strategy 
kallisto_grid_strategy <- cowplot::plot_grid(
  kallisto_pca_plot_strategy,
  kallisto_tsne_plot_strategy,
  kallisto_umap_plot_strategy + ggplot2::theme(legend.position="none"),
  align = 'vh',
  labels = "AUTO",
  nrow = 1
)
# Add the extracted legend to the plot grid
kallisto_grid_strategy <- cowplot::plot_grid(kallisto_grid_strategy, legend_b, rel_widths = c(3, .4))
# Plot grid with kallisto data colored by cancer type 
kallisto_grid <- cowplot::plot_grid(
  rsem_pca_plot,
  rsem_tsne_plot,
  rsem_umap_plot + ggplot2::theme(legend.position="none"),
  align = 'vh',
  labels = "AUTO",
  nrow = 1
)
# Add the extracted legend to the plot grid
kallisto_grid <- cowplot::plot_grid(kallisto_grid, legend_a, rel_widths = c(4, .2))
# Save grids
ggplot2::ggsave(file.path(plots_dir, "meta_grid_kallisto_strategy.pdf"),
                kallisto_grid_strategy,
                width = 18,
                height = 12
)
ggplot2::ggsave(file.path(plots_dir, "meta_grid_kallisto_type.pdf"),
                kallisto_grid,
                width = 18,
                height = 12
)
kallisto_grid_strategy
kallisto_grid
```

# Session Info

```{r}
sessionInfo()
```

