---
title: "Unsupervised Analysis of Transcriptomic Differences - Data Prep"
author: Chante Bethell for CCDL 2019
output:
  html_notebook:
    toc: true
    toc_float: true
---

This notebook is the first of two notebooks that demonstrate samples in the PBTA cluster by cancer type using dimensionality reduction techniques, namely, 
[Principal Component Analysis](https://www.rdocumentation.org/packages/stats/versions/3.6.0/topics/prcomp) (PCA), 
[t-Distributed Stochastic Neighbor Embedding](https://cran.r-project.org/web/packages/Rtsne/Rtsne.pdf) (t-SNE), 
and [Uniform Manifold Approximation and Projection](https://cran.r-project.org/web/packages/umap/vignettes/umap.html) (UMAP).

This notebook focuses on the preparation of the expression data and the execution of the dimension reduction techniques. 

It addresses [issue #9](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/9)
in the Open-PBTA analysis repository. 

## Output files:
- `analysis/transcriptomic-dimension-reduction/results/15.17_PCA_scores.tsv`
- `analysis/transcriptomic-dimension-reduction/results/15.17_t-SNE_scores.tsv`
- `analysis/transcriptomic-dimension-reduction/results/15.17_UMAP_scores.tsv`
- `analysis/transcriptomic-dimension-reduction/results/109.299_PCA_scores.tsv`
- `analysis/transcriptomic-dimension-reduction/results/109.299_t-SNE_scores.tsv`
- `analysis/transcriptomic-dimension-reduction/results/109.299_UMAP_scores.tsv`
- `analysis/transcriptomic-dimension-reduction/results/rsem_pca_model.RDS`
- `analysis/transcriptomic-dimension-reduction/results/rsem_tsne_model.RDS`
- `analysis/transcriptomic-dimension-reduction/results/rsem_umap_model.RDS`
- `analysis/transcriptomic-dimension-reduction/results/kallisto_pca_model.RDS`
- `analysis/transcriptomic-dimension-reduction/results/kallisto_tsne_model.RDS`
- `analysis/transcriptomic-dimension-reduction/results/kallisto_umap_model.RDS`

_Note that the tsv files holding the dimension reduction scores represent the scores for the RSEM (denoted by the output tsv files beginning with 15.17) and Kallisto expression data (denoted by the output tsv files beginning with 109.299)._

# Usage

This script is intended to be run via the command line from the top directory of the repository as follows: 
```
Rscript -e "rmarkdown::render('analyses/transcriptomic-dimension-reduction/01-transcriptomic-analysis-prep.Rmd', clean = TRUE)"
```

# Set Up

Set seed for reproducibility.
```{r}
set.seed(2019)
```

Install packages.
```{r, warning = FALSE, message = FALSE}
# This is needed for running the t-SNE analysis
if (!("Rtsne" %in% installed.packages())) {
  install.packages("Rtsne")
}
# This is needed for running the umap analysis
if (!("umap" %in% installed.packages())) {
  install.packages("umap")
}
```

Assign functions. 
```{r}
# magrittr pipe
`%>%` <- dplyr::`%>%`

perform_dimension_reduction <- function(transposed_expression_matrix, method, seed = 2019) {
  # Given a data.frame that containes the transposed expression matrix and the
  # name of a dimension reduction method, perform the dimension reduction 
  # technique on the transposed matrix
  # Args:
  #   transposed_expression_matrix: transposed `RSEM` or `Kallisto` expression
  #                                 matrix
  #   method: dimension reduction method to be performed
  #   seed: seed set to a default of 2019
  #
  # Returns:
  #   dimension_reduction_df: data.frame containing the resulting scores of the
  #                           dimension reduction technique, with a column that
  #                           includes sample identifiers
  
  # Set seed for reproducibility
  set.seed(seed)
  
  # Save rownames as a vector
  ID <- rownames(transposed_expression_matrix)
  # Perform dimension reduction 
  if (method == "PCA" ){
  prcomp_results <- prcomp(transposed_expression_matrix)
  dimension_reduction_df <- data.frame(prcomp_results$x[, 1:2])
  } else if (method == "t-SNE" ) {
  tsne_results <- Rtsne::Rtsne(transposed_expression_matrix, perplexity = 15)
  dimension_reduction_df <- data.frame(tsne_results$Y)
  } else if (method == "UMAP" ) {
  umap_results <- umap::umap(transposed_expression_matrix)
  dimension_reduction_df <- data.frame(umap_results$layout)
  } else {
    print ("Error")
  }
  
  # To set the tsv filename 
  matrix_name <- as.name(transposed_expression_matrix)
  # Assign the relevant rownames(Kids_First_Biospecimen_ID) to a column
  dimension_reduction_df$Kids_First_Biospecimen_ID <- ID
  # Write the resulting data.frame to a tsv file
  readr::write_tsv(dimension_reduction_df, 
                   file.path(results_dir, file = paste(matrix_name, method, "scores.tsv", sep = "_")))
  
  return(dimension_reduction_df)
}
align_metadata <- function(dimension_reduction_df, metadata_df, filename) {
  # Given a data.frame that contains scores from a dimensionality reduction
  # technique, align the metadata to the data.frame in preparation for plotting
  #
  # Note: The rownames provided in the id argument are synonymous with the
  # metadata's `Kids_First_Biospecimen_ID`
  #
  # Args:
  #   dimension_reduction_df: Name of the data.frame containing dimension reduction scores
  #   metadata_df : Name of the data.frame containing the relevant metadata
  #
  # Returns:
  #   aligned_scores_df: A data.frame containing the dimension reduction scores along with
  #         the variables from `metadata_df`
  
  # Join the dimension reductions scores data.frame and the metadata data.frame 
  aligned_scores_df <- dimension_reduction_df %>%
    dplyr::inner_join(metadata_df, by = c("Kids_First_Biospecimen_ID")) 
  
  # Write the resulting metadata aligned data.frame to a RDS file
  readr::write_rds(aligned_scores_df, 
                   file.path(results_dir, file = filename))
  
  return(aligned_scores_df)
}

dimension_reduction_wrapper <- function(transposed_expression_matrix, method, seed = 2019, metadata_df, filename){
  # Given a transposed expression matrix, the metadata data.frame, the name of 
  # the dimension reduction method to be performed and a seed with a default 
  # value, perform the dimension reduction, and align the metadata with the 
  # resulting scores
  # 
  # Args:
  #   transposed_expression_matrix: transposed `RSEM` or `Kallisto` expression
  #                                 matrix
  #   method: dimension reduction method to be performed
  #   seed: seed set to default 2019
  #   metadata_df: Name of the data.frame containing the relevant metadata
  #   filename: Filename for the RDS output file
  #
  # Returns:
  #   aligned_scores_df: data.frame containing dimension reduction scores and
  #                      aligned metadata variables
  #                              
  
  # Run `perform_dimension_reduction` function 
  dimension_reduction_df <- perform_dimension_reduction(transposed_expression_matrix, method, seed = 2019)
  # Run `align_metadata` function
  aligned_scores_df <- align_metadata(dimension_reduction_df, metadata_df, filename)
  
  return(aligned_scores_df)
}

```

Assign names of the output directories.
```{r}
# Assign name of output directory
data_dir <- file.path("..", "..", "data")
plots_dir <- "plots"
results_dir <- "results"

# Create directory to hold the output.
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

Read the metadata into a data.frame named `metadata_df`.
Read the expression data into data.frames named `exp_kallisto` and `exp_rsem`, 
respectively.
```{r}
# Read in dataset
metadata_df <-
  data.frame(readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv")))

# Read in kallisto expression data
exp_kallisto <- data.frame(readr::read_rds(file.path(
  data_dir, "pbta-gene-expression-kallisto.rds"
)))

# Read in RSEM expression data
exp_rsem <- data.frame(readr::read_rds(file.path(
  data_dir, "pbta-gene-expression-rsem.fpkm.rds"
)))
```

# Prep Data
Make the expression data.frame all numeric by transforming the non-numeric
`gene_id` column into rownames. 
Filter out the low count genes.
Transpose the data and save the rownames of the transposed data.frames for 
alignment to the metadata. 

```{r}
# Transform the non-numeric "gene_id" column into rownames
exp_kallisto <- exp_kallisto[, -1] %>%
  dplyr::filter(!duplicated(gene_id)) %>%
  dplyr::filter(complete.cases(.)) %>%
  tibble::column_to_rownames("gene_id")

exp_rsem <- exp_rsem %>%
  dplyr::filter(complete.cases(.)) %>%
  tibble::column_to_rownames("gene_id")

# Filter out low count genes
gene_count_threshold <- 100

kallisto_genes_to_keep <- rowSums(exp_kallisto) >= gene_count_threshold
exp_kallisto <- exp_kallisto[kallisto_genes_to_keep, ]

rsem_genes_to_keep <- rowSums(exp_rsem) >= gene_count_threshold
exp_rsem  <- exp_rsem[rsem_genes_to_keep, ]

# Transpose the data
transposed_rsem_data <- t(exp_rsem)
transposed_kallisto_data <- t(exp_kallisto)

```

# RSEM 

## Run PCA

```{r}
# Run `dimension_reduction_wrapper` function 
rsem_pca_model <-
  dimension_reduction_wrapper(transposed_rsem_data,
                              "PCA",
                              seed = 2019,
                              metadata_df,
                              "rsem_pca_model.RDS")
```

## Run t-SNE

```{r}
# Run `dimension_reduction_wrapper` function 
rsem_tsne_model <-
  dimension_reduction_wrapper(transposed_rsem_data,
                              "t-SNE",
                              seed = 2019,
                              metadata_df,
                              "rsem_tsne_model.RDS")
```

## Run UMAP

```{r}
# Run `dimension_reduction_wrapper` function 
rsem_umap_model <-
  dimension_reduction_wrapper(transposed_rsem_data,
                              "UMAP",
                              seed = 2019,
                              metadata_df,
                              "rsem_umap_model.RDS")
```


# Kallisto

## Run PCA

```{r}
# Run `dimension_reduction_wrapper` function 
kallisto_pca_model <-
  dimension_reduction_wrapper(transposed_kallisto_data,
                              "PCA",
                              seed = 2019,
                              metadata_df,
                              "kallisto_pca_model.RDS")
```

## Run t-SNE 

```{r}
# Run `dimension_reduction_wrapper` function 
kallisto_tsne_model <-
  dimension_reduction_wrapper(transposed_kallisto_data,
                              "t-SNE",
                              seed = 2019,
                              metadata_df,
                              "kallisto_tsne_model.RDS")
```

## Run UMAP 

```{r}
# Run `dimension_reduction_wrapper` function 
kallisto_umap_model <-
  dimension_reduction_wrapper(transposed_kallisto_data,
                              "UMAP",
                              seed = 2019,
                              metadata_df,
                              "kallisto_umap_model.RDS")
```

# Session Info

```{r}
sessionInfo()
```

